
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Fast primer screening with the pydna.primer_screen module &#8212; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=b530ca2f" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eea9577"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/Example_primer_screen';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pydna 6.0.0a24.post17+b7b559bd66 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/pydna-group/pydna" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Pydna
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../getting_started.html">Getting started</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="Dseq.html">Representing sequences in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="Dseq_Features.html">Working with Features using the Dseqrecord class</a></li>
<li class="toctree-l2"><a class="reference internal" href="Importing_Seqs.html">Importing and viewing sequence files in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="Restrict_Ligate_Cloning.html">Restriction and Ligation</a></li>
<li class="toctree-l2"><a class="reference internal" href="PCR.html">PCR</a></li>

<li class="toctree-l2"><a class="reference internal" href="primer_design.html">Primer design</a></li>
<li class="toctree-l2"><a class="reference internal" href="Gibson.html">Gibson Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="CRISPR.html">CRISPR-Cas9</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html">Cloning history / Cloning strategy</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../example_gallery.html">Example gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="Example_Restriction.html">Plasmid Restriction/Ligation Cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="Example_Gibson.html">Gibson Assembly from the original publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="Example_CRISPR.html">Designing a CRISPR-Cas9 toolbox for genome engineering.</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../how_pydna_works.html">How pydna works</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="new_assemblies.html">How assemblies work</a></li>


</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reference/pydna.html">pydna package</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/pydna-group/pydna/blob/main/docs/notebooks/Example_primer_screen.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on Colab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="Colab logo" src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/pydna-group/pydna" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/pydna-group/pydna/edit/main/docs/notebooks/Example_primer_screen.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/Example_primer_screen.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Fast primer screening with the pydna.primer_screen module</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#primer-list">Primer list</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-data-files">Read data files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#most-important-functions-provided-by-pydna-primer-screen">Most important functions provided by <code class="docutils literal notranslate"><span class="pre">pydna.primer_screen</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-forward-primers">Function <code class="docutils literal notranslate"><span class="pre">forward_primers</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-reverse-primers">Function <code class="docutils literal notranslate"><span class="pre">reverse_primers</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-primer-pairs">Function <code class="docutils literal notranslate"><span class="pre">primer_pairs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-flanking-primer-pairs">Function <code class="docutils literal notranslate"><span class="pre">flanking_primer_pairs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-diff-primer-pairs">Function <code class="docutils literal notranslate"><span class="pre">diff_primer_pairs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-diff-primer-triplets">Function <code class="docutils literal notranslate"><span class="pre">diff_primer_triplets</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#automaton-for-a-list-of-primers">Automaton for a list of primers.</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="fast-primer-screening-with-the-pydna-primer-screen-module">
<h1>Fast primer screening with the pydna.primer_screen module<a class="headerlink" href="#fast-primer-screening-with-the-pydna-primer-screen-module" title="Link to this heading">#</a></h1>
<blockquote>
<div><p>Visit the full library documentation <a class="reference external" href="https://pydna-group.github.io/pydna/">here</a></p>
</div></blockquote>
<p>This notebook showcases how to use the <code class="docutils literal notranslate"><span class="pre">pydna.primer_screen</span></code> module to screen your primer collection. I routinely use it to find the annealing positions of about two thousands primers in a fraction of a second.</p>
<p>A provision is to have the sequences primer sequences in a format that pydna can read using the <code class="docutils literal notranslate"><span class="pre">pydna.parsers.parse_primers</span></code>.</p>
<p>The primer list can consist of <code class="docutils literal notranslate"><span class="pre">Primer</span></code> objects returned by :func:<code class="docutils literal notranslate"><span class="pre">pydna.parsers.parse_primers</span></code> or any objects with a <code class="docutils literal notranslate"><span class="pre">seq</span></code> attribute, such as <code class="docutils literal notranslate"><span class="pre">pydna.seqrecord.SeqRecord</span></code> or <code class="docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>.</p>
<p>This module uses the <a class="reference external" href="https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm">Aho–Corasick algorithm</a>.</p>
<p>If the same primer list is used repeatedly, creating an automaton with <code class="docutils literal notranslate"><span class="pre">pydna.make_automaton</span></code> greatly speeds up repeated searches. See at the end of this notebook for how to create, save, and use the automaton.</p>
<a target="_blank" href="https://colab.research.google.com/github/pydna-group/pydna/blob/master/docs/notebooks/Example_primer_screen.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="c1"># Install pydna and download data files from the pydna Github repository (only when running on Colab)</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="k">if</span> <span class="s1">&#39;google.colab&#39;</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">modules</span><span class="p">:</span>
    <span class="o">%</span><span class="n">pip</span> <span class="n">install</span> <span class="n">pydna</span><span class="p">[</span><span class="n">clipboard</span><span class="p">,</span><span class="n">download</span><span class="p">,</span><span class="n">express</span><span class="p">,</span><span class="n">gel</span><span class="p">]</span>
    <span class="err">!</span><span class="n">curl</span> <span class="o">-</span><span class="n">LO</span> <span class="s2">&quot;https://github.com/pydna-group/pydna/raw/master/docs/notebooks/FAS2_S288C_wild-type_locus.gb&quot;</span>
    <span class="err">!</span><span class="n">curl</span> <span class="o">-</span><span class="n">LO</span> <span class="s2">&quot;https://github.com/pydna-group/pydna/raw/master/docs/notebooks/fas2__NatMX4_locus.gb&quot;</span>
    <span class="err">!</span><span class="n">curl</span> <span class="o">-</span><span class="n">LO</span> <span class="s2">&quot;https://github.com/pydna-group/pydna/raw/master/docs/notebooks/fas2__KanMX4_locus.gb&quot;</span>
    <span class="err">!</span><span class="n">curl</span> <span class="o">-</span><span class="n">LO</span> <span class="s2">&quot;https://github.com/pydna-group/pydna/raw/master/docs/notebooks/pIL68.gb&quot;</span>
    <span class="err">!</span><span class="n">curl</span> <span class="o">-</span><span class="n">LO</span> <span class="s2">&quot;https://github.com/pydna-group/pydna/raw/master/docs/notebooks/pIL75.gb&quot;</span>
</pre></div>
</div>
</div>
</div>
<section id="primer-list">
<h2>Primer list<a class="headerlink" href="#primer-list" title="Link to this heading">#</a></h2>
<p>The code cell below establishes a <strong>primer list</strong>. This list is a sublist of actual oligonucleotides that sit in a freezer in a lab in the north of Portugal. That freezer contain over two thousand primers, all of which are named “N_primername” where N is an integer.
The number in the first part of the primer name indicate the index of that primer in the original list.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.parsers</span><span class="w"> </span><span class="kn">import</span> <span class="n">parse_primers</span>

<span class="c1"># Eleven primers in FASTA format</span>
<span class="n">primers</span> <span class="o">=</span> <span class="n">parse_primers</span> <span class="p">(</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">&gt;51_TefTermFwd A.gos 20-mer</span>
<span class="s2">CAGATGCGAAGTTAAGTGCG</span>
<span class="s2">&gt;82_MSW_fwd this primer sits inside the loxP of pUG6</span>
<span class="s2">TCCTTGACAGTCTTGACG</span>
<span class="s2">&gt;149_MX4rev (25-mer) Primer in the Ashbya gossypi TEF terminator in the reverse direction. NEW (2021-01-05)</span>
<span class="s2">ACAAATGACAAGTTCTTGAAAACAA</span>
<span class="s2">&gt;255_kanC (22-mer)</span>
<span class="s2">TGATTTTGATGACGAGCGTAAT</span>
<span class="s2">&gt;594_pRS306_rev 24-mer</span>
<span class="s2">tcgctttcttcccttcctttctcg</span>
<span class="s2">&gt;595_pRS306_fwd</span>
<span class="s2">taaagggagcccccgatttagagc</span>
<span class="s2">&gt;607_XDHcasrv (28-mer)</span>
<span class="s2">aaacagctatgaccatgattacgccaag</span>
<span class="s2">&gt;700_sc_fas2-B1: (25-mer)</span>
<span class="s2">ATTTCTCTATGTAAAGACAGAGCAG</span>
<span class="s2">&gt;701_sc_fas2-A1: (25-mer)</span>
<span class="s2">CTATATTTCTATTCTATCCGAACTC</span>
<span class="s2">&gt;1215_URA3r 27-mer</span>
<span class="s2">CGGTTTCTTTGAAATTTTTTTGATTCG</span>
<span class="s2">&gt;1564_KANMX_rev</span>
<span class="s2">CACTCGCATCAACCAAACC</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>

<span class="c1"># A list of None is created</span>
<span class="n">pl</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1565</span><span class="p">)]</span>

<span class="k">for</span> <span class="n">primer</span> <span class="ow">in</span> <span class="n">primers</span><span class="p">:</span>
    <span class="c1"># Extract the number from the name</span>
    <span class="n">number</span><span class="p">,</span> <span class="n">rest</span> <span class="o">=</span> <span class="n">primer</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">maxsplit</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># put the primer in the correct location in the list</span>
    <span class="n">pl</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">number</span><span class="p">)]</span> <span class="o">=</span> <span class="n">primer</span>
</pre></div>
</div>
</div>
</div>
<p>The primer list is just a list where most elements are empty (None)
so that the primer index in the name matches the index of the primer.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">pl</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1565</span>
</pre></div>
</div>
</div>
</div>
<p>Eleven elements should have a value that is not None</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">assert</span> <span class="nb">len</span><span class="p">([</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pl</span> <span class="k">if</span> <span class="n">p</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">])</span> <span class="o">==</span> <span class="mi">11</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="read-data-files">
<h2>Read data files<a class="headerlink" href="#read-data-files" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.readers</span><span class="w"> </span><span class="kn">import</span> <span class="n">read</span>
</pre></div>
</div>
</div>
</div>
<p>The following three files are wild-type (wt) and two mutant alleles
where the coding sequence of a gene has been replaced by two different
marker genes. These were made in the lab by one of my students.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">wt</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;FAS2_S288C_wild-type_locus.gb&quot;</span><span class="p">)</span>
<span class="n">nat</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;fas2__NatMX4_locus.gb&quot;</span><span class="p">)</span>
<span class="n">kan</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;fas2__KanMX4_locus.gb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The following two files are plasmids from Liachko I, Dunham MJ. (2014). An autonomously replicating sequence for use in a wide range of budding
yeasts. FEMS Yeast Res. 14(2):364–367 (<a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/24205893/">pubmed</a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pIL68</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;pIL68.gb&quot;</span><span class="p">)</span>
<span class="n">pIL75</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;pIL75.gb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="most-important-functions-provided-by-pydna-primer-screen">
<h2>Most important functions provided by <code class="docutils literal notranslate"><span class="pre">pydna.primer_screen</span></code><a class="headerlink" href="#most-important-functions-provided-by-pydna-primer-screen" title="Link to this heading">#</a></h2>
<p>The six main functions are imported below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">forward_primers</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">reverse_primers</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">primer_pairs</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">flanking_primer_pairs</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">diff_primer_pairs</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">diff_primer_triplets</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/pydna/pydna/src/pydna/primer_screen.py:63: FutureWarning: The primer_screen module is experimental and not yet extensively tested. api may change in future versions.
  warnings.warn(
</pre></div>
</div>
</div>
</div>
</section>
<section id="function-forward-primers">
<h2>Function <code class="docutils literal notranslate"><span class="pre">forward_primers</span></code><a class="headerlink" href="#function-forward-primers" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">kan</span></code> sequence is the <em>Saccharomyces cerevisiae</em> <a class="reference external" href="https://www.yeastgenome.org/locus/S000006152">fas2</a> locus where the FAS2 open reading frame (orf) and 1000 bp upstream and downstream. The orf has been replaced by the <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/8692690/">KanMX4</a> marker gene from the pUG6 plasmid.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kan</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Dseqrecord(-3613)
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">forward_primers</span></code> function returns a dictionary <code class="docutils literal notranslate"><span class="pre">{int:</span> <span class="pre">list}</span></code> representing the <strong>index</strong> and <strong>position(s)</strong> of forward primers from the <code class="docutils literal notranslate"><span class="pre">pl</span></code> list annealing to the <code class="docutils literal notranslate"><span class="pre">kan</span></code> sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">forward_primers</span><span class="p">(</span><span class="n">kan</span><span class="p">,</span> <span class="n">pl</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{701: [534], 82: [1168], 255: [1979], 51: [2434]}
</pre></div>
</div>
</div>
</div>
<p>The returned dictionary has the form:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>{ primer_A_index : [primer_A_location1, primer_A_location2, ...]
  primer_B_index : [primer_B_location1, primer_B_location2, ...] }
</pre></div>
</div>
<p>The location is the position where the 3’ part of the primer anneals. The forward primer in the figure below anneals at position 14 on the 20 bp template.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> 5-gtcatgatctagtcgatgtta-3
   |||||||||||||||||||||

         5&#39;-tagtcg-3&#39;      &lt;=== forward primer, location = 14
            ||||||
   |||||||||||||||||||||
 3-cagtactagatcagctacaat-5
                 ^
   0-------------1-----2 position
                 4     0
</pre></div>
</div>
</section>
<section id="function-reverse-primers">
<h2>Function <code class="docutils literal notranslate"><span class="pre">reverse_primers</span></code><a class="headerlink" href="#function-reverse-primers" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">reverse_primers</span></code> function returns a dictionary <code class="docutils literal notranslate"><span class="pre">{int:</span> <span class="pre">list}</span></code> representing the <em>index</em> and <em>position(s)</em> of reverse primers from <code class="docutils literal notranslate"><span class="pre">pl</span></code> annealing to the <code class="docutils literal notranslate"><span class="pre">kan</span></code> sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reverse_primers</span><span class="p">(</span><span class="n">kan</span><span class="p">,</span> <span class="n">pl</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{149: [2306], 1564: [1940]}
</pre></div>
</div>
</div>
</div>
<p>The resulting dict has the same form as in the previous example describing the <code class="docutils literal notranslate"><span class="pre">forward_primers</span></code> function. In the example below, the reverse primeranneals at position 9 on thje 20 bp template.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>5-gtcatgatctagtcgatgtta-3
  |||||||||||||||||||||
           ||||||
         3-atcagc-5       &lt;=== reverse primer, location = 9

  |||||||||||||||||||||
3-cagtactagatcagctacaat-5
           ^
  0--------9----------2 position
                      0
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">primer_pairs</span><span class="p">(</span><span class="n">kan</span><span class="p">,</span> <span class="n">pl</span><span class="p">,</span> <span class="n">short</span><span class="o">=</span><span class="mi">350</span><span class="p">,</span> <span class="n">long</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[amplicon_tuple(fp=701, rp=149, fposition=534, rposition=2306, size=1822),
 amplicon_tuple(fp=701, rp=1564, fposition=534, rposition=1940, size=1450),
 amplicon_tuple(fp=82, rp=149, fposition=1168, rposition=2306, size=1181),
 amplicon_tuple(fp=82, rp=1564, fposition=1168, rposition=1940, size=809),
 amplicon_tuple(fp=255, rp=149, fposition=1979, rposition=2306, size=374)]
</pre></div>
</div>
</div>
</div>
</section>
<section id="function-primer-pairs">
<h2>Function <code class="docutils literal notranslate"><span class="pre">primer_pairs</span></code><a class="headerlink" href="#function-primer-pairs" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">primer_pairs</span></code> function return a list of named tuples. They each indicate a forward and a reverse primer, their positions and the size of the potential PCR product. PCR products with a size between optional keyword arguments <code class="docutils literal notranslate"><span class="pre">short</span></code> (500 bp by default) and <code class="docutils literal notranslate"><span class="pre">long</span></code> (2000 bp by default) are returned.</p>
</section>
<section id="function-flanking-primer-pairs">
<h2>Function <code class="docutils literal notranslate"><span class="pre">flanking_primer_pairs</span></code><a class="headerlink" href="#function-flanking-primer-pairs" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">flanking_primer_pairs</span></code> function returns primers pairs that flank a target, in the example above from position 550 to 1200.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">flanking_primer_pairs</span><span class="p">(</span><span class="n">kan</span><span class="p">,</span> <span class="n">pl</span><span class="p">,</span> <span class="n">target</span><span class="o">=</span><span class="p">(</span><span class="mi">550</span><span class="p">,</span> <span class="mi">1200</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[amplicon_tuple(fp=82, rp=1564, fposition=1168, rposition=1940, size=809),
 amplicon_tuple(fp=82, rp=149, fposition=1168, rposition=2306, size=1181)]
</pre></div>
</div>
</div>
</div>
</section>
<section id="function-diff-primer-pairs">
<h2>Function <code class="docutils literal notranslate"><span class="pre">diff_primer_pairs</span></code><a class="headerlink" href="#function-diff-primer-pairs" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">diff_primer_pairs</span></code> function takes a list or tuple with Dseqrecord objects and returns another kind of named tuple <code class="docutils literal notranslate"><span class="pre">primer_tuple</span></code>the same kind of named tuples as the previous two functions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">diff_primer_pairs</span><span class="p">((</span><span class="n">nat</span><span class="p">,</span> <span class="n">kan</span><span class="p">),</span> <span class="n">pl</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[(primer_tuple(seq=Dseqrecord(-3308), fp=82, rp=149, size=944),
  primer_tuple(seq=Dseqrecord(-3613), fp=82, rp=149, size=1181))]
</pre></div>
</div>
</div>
</div>
<p>Each tuple contain potential PCR products from a primer pair that results in different PCR product sizes. This can be used to find primers that can differentiate between two or more DNA sequences, such as before and after some genetic modification.</p>
<p>For example, Primers 1 and 2 both form PCR products from sequenceA and B below, but of different sizes. Primers 1 and 2 could be used to verify genetic modifications such as cloning an insert into a plasmid vector.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> 1&gt;              &lt;2
-------NNNNNNNNN----  sequenceA


 1&gt;          &lt;2
-------XXXXX----      sequenceB
</pre></div>
</div>
<p>Using primers 82 and 149 results in a 944 bp PCR product from the <code class="docutils literal notranslate"><span class="pre">kan</span></code> sequence and and 1181 bp product from the <code class="docutils literal notranslate"><span class="pre">nat</span></code> sequence.</p>
<p>In this example, the nat and kan represents the same kind of deletion in the S. cerevisiae fas2 locus, but one was wade with the KanMX4 and the other with the NatMX4 from the pAG25 plasmid <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/10514571/">pubmed</a>. They differ in size by about 300 bp:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nat</span><span class="p">,</span> <span class="n">kan</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(Dseqrecord(-3308), Dseqrecord(-3613))
</pre></div>
</div>
</div>
</div>
</section>
<section id="function-diff-primer-triplets">
<h2>Function <code class="docutils literal notranslate"><span class="pre">diff_primer_triplets</span></code><a class="headerlink" href="#function-diff-primer-triplets" title="Link to this heading">#</a></h2>
<p>The two plasmids pIL68 and pIL75 are closely related and there is no single primer pair in our list that would result in PCR products of different sizes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">diff_primer_pairs</span><span class="p">((</span><span class="n">pIL68</span><span class="p">,</span> <span class="n">pIL75</span><span class="p">),</span> <span class="n">pl</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pIL68</span><span class="p">,</span> <span class="n">pIL75</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(Dseqrecord(o4852), Dseqrecord(o5213))
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">diff_primer_triplets</span></code> function looks for primer triplets that can be used to distinguish two sequences.</p>
<p>For example, Primers 1, 2 and 3 form PCR products from sequenceA and B below, but of different sizes. Primer 1 binds both sequences while primers 2 and 3 bind one sequence each. This primer triplet could be used to verify genetic modifications by using a PCR reaction with three primers.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span> 1&gt;        &lt;2
-------NNNNNNNNN----  sequenceA

 1&gt;     &lt;3
-------XXXXX--------  sequenceB
</pre></div>
</div>
<p>The fuction returns a list of tuples. Each tuple contain potential PCR products from a primer pair that results in different PCR product sizes. This can be used to find primers that can differentiate between two or more DNA sequences, such as before and after some genetic modification.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">diff_primer_triplets</span><span class="p">((</span><span class="n">pIL68</span><span class="p">,</span> <span class="n">pIL75</span><span class="p">),</span> <span class="n">pl</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[(primer_tuple(seq=Dseqrecord(o4852), fp=1215, rp=594, size=1474),
  primer_tuple(seq=Dseqrecord(o5213), fp=51, rp=594, size=548)),
 (primer_tuple(seq=Dseqrecord(o4852), fp=1215, rp=594, size=1474),
  primer_tuple(seq=Dseqrecord(o5213), fp=255, rp=594, size=1005))]
</pre></div>
</div>
</div>
</div>
<p>In this example, primers 1215, 51 and 594 could be used to amplify a 1474 bp product from the pIL68 (4852 bp) and a 548 bp product from the pIL75 (5213 bp). Alternatively, the primer 51 could be substiuted for primer 255 to give two products slightly closer in size.</p>
</section>
<section id="automaton-for-a-list-of-primers">
<h2>Automaton for a list of primers.<a class="headerlink" href="#automaton-for-a-list-of-primers" title="Link to this heading">#</a></h2>
<p>An <a class="reference external" href="https://github.com/WojciechMula/pyahocorasick">automaton</a> can
be made prior to primer screening for a list of Primer
objects for faster primer search.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer_screen</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_automaton</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># build automaton</span>
<span class="n">atm</span> <span class="o">=</span> <span class="n">make_automaton</span><span class="p">(</span><span class="n">pl</span><span class="p">,</span> <span class="n">limit</span> <span class="o">=</span> <span class="mi">16</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># save automaton</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span> <span class="c1"># The standard lib pickle </span>
<span class="n">atm</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;atm.automaton&quot;</span><span class="p">,</span> <span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load automaton</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">ahocorasick</span>
<span class="n">atm</span> <span class="o">=</span> <span class="n">ahocorasick</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;atm.automaton&quot;</span><span class="p">,</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># use automaton</span>
<span class="n">forward_primers</span><span class="p">(</span><span class="n">kan</span><span class="p">,</span> <span class="n">pl</span><span class="p">,</span> <span class="n">automaton</span><span class="o">=</span><span class="n">atm</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{701: [534], 82: [1168], 255: [1979], 51: [2434]}
</pre></div>
</div>
</div>
</div>
<p>This automaton can be reused as an optional argument across calls to</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">forward_primers</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">reverse_primers</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">primer_pairs</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">flanking_primer_pairs</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">diff_primer_pairs</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">diff_primer_triplets</span></code></p></li>
</ul>
<p>The primer list can contain None, this can be used to remove primers
from the primer_list for the automaton, while keeping the original index
for each primer.</p>
<p>The limit is the part of the primer used to find annealing positions.
The automaton processes the uppercase 3’ part of each primer up to <code class="docutils literal notranslate"><span class="pre">limit</span></code>.
It has to be rebuilt if a different limit is needed.</p>
<p>The primers can contain ambiguous bases from the extended IUPAC DNA alphabet:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Code</p></th>
<th class="head"><p>Bases</p></th>
<th class="head"><p>Meaning</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>A</p></td>
<td><p>Adenine</p></td>
</tr>
<tr class="row-odd"><td><p>C</p></td>
<td><p>C</p></td>
<td><p>Cytosine</p></td>
</tr>
<tr class="row-even"><td><p>G</p></td>
<td><p>G</p></td>
<td><p>Guanine</p></td>
</tr>
<tr class="row-odd"><td><p>T</p></td>
<td><p>T</p></td>
<td><p>Thymine</p></td>
</tr>
<tr class="row-even"><td><p>U</p></td>
<td><p>T</p></td>
<td><p>Uracil (RNA)</p></td>
</tr>
<tr class="row-odd"><td><p>R</p></td>
<td><p>AG</p></td>
<td><p>puRine</p></td>
</tr>
<tr class="row-even"><td><p>Y</p></td>
<td><p>CT</p></td>
<td><p>pYrimidine</p></td>
</tr>
<tr class="row-odd"><td><p>S</p></td>
<td><p>GC</p></td>
<td><p>Strong (3 H-bonds)</p></td>
</tr>
<tr class="row-even"><td><p>W</p></td>
<td><p>AT</p></td>
<td><p>Weak (2 H-bonds)</p></td>
</tr>
<tr class="row-odd"><td><p>K</p></td>
<td><p>GT</p></td>
<td><p>Keto</p></td>
</tr>
<tr class="row-even"><td><p>M</p></td>
<td><p>AC</p></td>
<td><p>aMino</p></td>
</tr>
<tr class="row-odd"><td><p>B</p></td>
<td><p>CGT</p></td>
<td><p>not A</p></td>
</tr>
<tr class="row-even"><td><p>D</p></td>
<td><p>AGT</p></td>
<td><p>not C</p></td>
</tr>
<tr class="row-odd"><td><p>H</p></td>
<td><p>ACT</p></td>
<td><p>not G</p></td>
</tr>
<tr class="row-even"><td><p>V</p></td>
<td><p>ACG</p></td>
<td><p>not T</p></td>
</tr>
<tr class="row-odd"><td><p>N</p></td>
<td><p>ACGT</p></td>
<td><p>any</p></td>
</tr>
<tr class="row-even"><td><p>X</p></td>
<td><p>-</p></td>
<td><p>unknown</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#primer-list">Primer list</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-data-files">Read data files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#most-important-functions-provided-by-pydna-primer-screen">Most important functions provided by <code class="docutils literal notranslate"><span class="pre">pydna.primer_screen</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-forward-primers">Function <code class="docutils literal notranslate"><span class="pre">forward_primers</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-reverse-primers">Function <code class="docutils literal notranslate"><span class="pre">reverse_primers</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-primer-pairs">Function <code class="docutils literal notranslate"><span class="pre">primer_pairs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-flanking-primer-pairs">Function <code class="docutils literal notranslate"><span class="pre">flanking_primer_pairs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-diff-primer-pairs">Function <code class="docutils literal notranslate"><span class="pre">diff_primer_pairs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#function-diff-primer-triplets">Function <code class="docutils literal notranslate"><span class="pre">diff_primer_triplets</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#automaton-for-a-list-of-primers">Automaton for a list of primers.</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Björn F. Johansson
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Björn F. Johansson.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>