
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>pydna package &#8212; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=b530ca2f" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eea9577"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'reference/pydna';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="How assemblies work" href="../notebooks/new_assemblies.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pydna 6.0.0a24.post17+b7b559bd66 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/pydna-group/pydna" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Pydna
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../getting_started.html">Getting started</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Dseq.html">Representing sequences in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Dseq_Features.html">Working with Features using the Dseqrecord class</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Importing_Seqs.html">Importing and viewing sequence files in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Restrict_Ligate_Cloning.html">Restriction and Ligation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/PCR.html">PCR</a></li>

<li class="toctree-l2"><a class="reference internal" href="../notebooks/primer_design.html">Primer design</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Gibson.html">Gibson Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/CRISPR.html">CRISPR-Cas9</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/history.html">Cloning history / Cloning strategy</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../example_gallery.html">Example gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Example_Restriction.html">Plasmid Restriction/Ligation Cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Example_Gibson.html">Gibson Assembly from the original publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/Example_CRISPR.html">Designing a CRISPR-Cas9 toolbox for genome engineering.</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../how_pydna_works.html">How pydna works</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../notebooks/new_assemblies.html">How assemblies work</a></li>


</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">pydna package</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/pydna-group/pydna" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/pydna-group/pydna/edit/main/docs/reference/pydna.rst" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/reference/pydna.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>pydna package</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna">pydna</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna-package-layout">pydna package layout</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-use-the-documentation">How to use the documentation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna-source-code">pydna source code</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-get-more-help">How to get more help</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#examples-of-pydna-in-use">Examples of pydna in use</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.get_env"><code class="docutils literal notranslate"><span class="pre">get_env()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.logo"><code class="docutils literal notranslate"><span class="pre">logo()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#submodules">Submodules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.all">pydna.all module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal"><code class="docutils literal notranslate"><span class="pre">Anneal</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.forward_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.forward_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.reverse_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.reverse_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.template"><code class="docutils literal notranslate"><span class="pre">Anneal.template</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.limit"><code class="docutils literal notranslate"><span class="pre">Anneal.limit</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.products"><code class="docutils literal notranslate"><span class="pre">Anneal.products</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.report"><code class="docutils literal notranslate"><span class="pre">Anneal.report()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.pcr"><code class="docutils literal notranslate"><span class="pre">pcr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.genbank"><code class="docutils literal notranslate"><span class="pre">genbank()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Genbank"><code class="docutils literal notranslate"><span class="pre">Genbank</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Genbank.nucleotide"><code class="docutils literal notranslate"><span class="pre">Genbank.nucleotide()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.download_text"><code class="docutils literal notranslate"><span class="pre">download_text()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.add_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.add_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.cas9"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.circular"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.circular</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.copy_fasta_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_fasta_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.copy_gb_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_gb_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.extract_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.figure"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.find"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.find_aa"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aa()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.find_aminoacids"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aminoacids()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.format"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.format()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.from_string"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.history"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.history()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.linearize"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.linearize()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.looped"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.lower"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.m"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.m()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.map_trace_files"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.map_trace_files()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.number_of_cuts"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.number_of_cuts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.orfs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.orfs_to_features"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs_to_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.rc"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.seguid"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.shifted"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.source"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.source</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.synced"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.synced()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.upper"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.write"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.write()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq"><code class="docutils literal notranslate"><span class="pre">Dseq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.T4"><code class="docutils literal notranslate"><span class="pre">Dseq.T4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseq.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cas9"><code class="docutils literal notranslate"><span class="pre">Dseq.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cut"><code class="docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cutsite_is_valid"><code class="docutils literal notranslate"><span class="pre">Dseq.cutsite_is_valid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.exo1_end"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.exo1_front"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_front()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.fill_in"><code class="docutils literal notranslate"><span class="pre">Dseq.fill_in()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.find"><code class="docutils literal notranslate"><span class="pre">Dseq.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.five_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.five_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.from_full_sequence_and_overhangs"><code class="docutils literal notranslate"><span class="pre">Dseq.from_full_sequence_and_overhangs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.from_representation"><code class="docutils literal notranslate"><span class="pre">Dseq.from_representation()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.from_string"><code class="docutils literal notranslate"><span class="pre">Dseq.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.get_cut_parameters"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cut_parameters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.get_cutsite_pairs"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsite_pairs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.get_cutsites"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsites()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.isblunt"><code class="docutils literal notranslate"><span class="pre">Dseq.isblunt()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.left_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.left_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.looped"><code class="docutils literal notranslate"><span class="pre">Dseq.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.lower"><code class="docutils literal notranslate"><span class="pre">Dseq.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.mung"><code class="docutils literal notranslate"><span class="pre">Dseq.mung()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.mw"><code class="docutils literal notranslate"><span class="pre">Dseq.mw()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.quick"><code class="docutils literal notranslate"><span class="pre">Dseq.quick()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.rc"><code class="docutils literal notranslate"><span class="pre">Dseq.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.right_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.right_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.seguid"><code class="docutils literal notranslate"><span class="pre">Dseq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.shifted"><code class="docutils literal notranslate"><span class="pre">Dseq.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.t4"><code class="docutils literal notranslate"><span class="pre">Dseq.t4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseq.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.three_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.three_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseq.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.transcribe"><code class="docutils literal notranslate"><span class="pre">Dseq.transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.translate"><code class="docutils literal notranslate"><span class="pre">Dseq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.trunc"><code class="docutils literal notranslate"><span class="pre">Dseq.trunc</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.upper"><code class="docutils literal notranslate"><span class="pre">Dseq.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.watson_ovhg"><code class="docutils literal notranslate"><span class="pre">Dseq.watson_ovhg()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.read"><code class="docutils literal notranslate"><span class="pre">read()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.read_primer"><code class="docutils literal notranslate"><span class="pre">read_primer()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.parse"><code class="docutils literal notranslate"><span class="pre">parse()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.parse_primers"><code class="docutils literal notranslate"><span class="pre">parse_primers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.primer_design"><code class="docutils literal notranslate"><span class="pre">primer_design()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.assembly_fragments"><code class="docutils literal notranslate"><span class="pre">assembly_fragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.eq"><code class="docutils literal notranslate"><span class="pre">eq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.gbtext_clean"><code class="docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.amplicon">pydna.amplicon module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon"><code class="docutils literal notranslate"><span class="pre">Amplicon</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Amplicon.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Amplicon.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.rc"><code class="docutils literal notranslate"><span class="pre">Amplicon.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.figure"><code class="docutils literal notranslate"><span class="pre">Amplicon.figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.set_forward_primer_footprint"><code class="docutils literal notranslate"><span class="pre">Amplicon.set_forward_primer_footprint()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.set_reverse_primer_footprint"><code class="docutils literal notranslate"><span class="pre">Amplicon.set_reverse_primer_footprint()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.program"><code class="docutils literal notranslate"><span class="pre">Amplicon.program()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.dbd_program"><code class="docutils literal notranslate"><span class="pre">Amplicon.dbd_program()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.primers"><code class="docutils literal notranslate"><span class="pre">Amplicon.primers()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.amplify">pydna.amplify module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal"><code class="docutils literal notranslate"><span class="pre">Anneal</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.forward_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.forward_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.reverse_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.reverse_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.template"><code class="docutils literal notranslate"><span class="pre">Anneal.template</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.limit"><code class="docutils literal notranslate"><span class="pre">Anneal.limit</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.products"><code class="docutils literal notranslate"><span class="pre">Anneal.products</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.report"><code class="docutils literal notranslate"><span class="pre">Anneal.report()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.pcr"><code class="docutils literal notranslate"><span class="pre">pcr()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.assembly">pydna.assembly module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.assembly2">pydna.assembly2 module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gather_overlapping_locations"><code class="docutils literal notranslate"><span class="pre">gather_overlapping_locations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.ends_from_cutsite"><code class="docutils literal notranslate"><span class="pre">ends_from_cutsite()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.restriction_ligation_overlap"><code class="docutils literal notranslate"><span class="pre">restriction_ligation_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.combine_algorithms"><code class="docutils literal notranslate"><span class="pre">combine_algorithms()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.blunt_overlap"><code class="docutils literal notranslate"><span class="pre">blunt_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_sub_strings"><code class="docutils literal notranslate"><span class="pre">common_sub_strings()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gibson_overlap"><code class="docutils literal notranslate"><span class="pre">gibson_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.sticky_end_sub_strings"><code class="docutils literal notranslate"><span class="pre">sticky_end_sub_strings()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.zip_match_leftwards"><code class="docutils literal notranslate"><span class="pre">zip_match_leftwards()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.zip_match_rightwards"><code class="docutils literal notranslate"><span class="pre">zip_match_rightwards()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.seqrecord2_uppercase_DNA_string"><code class="docutils literal notranslate"><span class="pre">seqrecord2_uppercase_DNA_string()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.primer_template_overlap"><code class="docutils literal notranslate"><span class="pre">primer_template_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fill_left"><code class="docutils literal notranslate"><span class="pre">fill_left()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fill_right"><code class="docutils literal notranslate"><span class="pre">fill_right()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fill_dseq"><code class="docutils literal notranslate"><span class="pre">fill_dseq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.reverse_complement_assembly"><code class="docutils literal notranslate"><span class="pre">reverse_complement_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.filter_linear_subassemblies"><code class="docutils literal notranslate"><span class="pre">filter_linear_subassemblies()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.remove_subassemblies"><code class="docutils literal notranslate"><span class="pre">remove_subassemblies()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly2str"><code class="docutils literal notranslate"><span class="pre">assembly2str()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly2str_tuple"><code class="docutils literal notranslate"><span class="pre">assembly2str_tuple()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly_has_mismatches"><code class="docutils literal notranslate"><span class="pre">assembly_has_mismatches()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly_is_circular"><code class="docutils literal notranslate"><span class="pre">assembly_is_circular()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assemble"><code class="docutils literal notranslate"><span class="pre">assemble()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.annotate_primer_binding_sites"><code class="docutils literal notranslate"><span class="pre">annotate_primer_binding_sites()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.edge_representation2subfragment_representation"><code class="docutils literal notranslate"><span class="pre">edge_representation2subfragment_representation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.subfragment_representation2edge_representation"><code class="docutils literal notranslate"><span class="pre">subfragment_representation2edge_representation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.get_assembly_subfragments"><code class="docutils literal notranslate"><span class="pre">get_assembly_subfragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.extract_subfragment"><code class="docutils literal notranslate"><span class="pre">extract_subfragment()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.is_sublist"><code class="docutils literal notranslate"><span class="pre">is_sublist()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.circular_permutation_min_abs"><code class="docutils literal notranslate"><span class="pre">circular_permutation_min_abs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assembly_is_valid"><code class="docutils literal notranslate"><span class="pre">Assembly.assembly_is_valid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.add_edges_from_match"><code class="docutils literal notranslate"><span class="pre">Assembly.add_edges_from_match()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.format_assembly_edge"><code class="docutils literal notranslate"><span class="pre">Assembly.format_assembly_edge()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_linear_assemblies"><code class="docutils literal notranslate"><span class="pre">Assembly.get_linear_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.node_path2assembly_list"><code class="docutils literal notranslate"><span class="pre">Assembly.node_path2assembly_list()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_unique_linear_paths"><code class="docutils literal notranslate"><span class="pre">Assembly.get_unique_linear_paths()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_possible_assembly_number"><code class="docutils literal notranslate"><span class="pre">Assembly.get_possible_assembly_number()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_circular_assemblies"><code class="docutils literal notranslate"><span class="pre">Assembly.get_circular_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.format_insertion_assembly"><code class="docutils literal notranslate"><span class="pre">Assembly.format_insertion_assembly()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.format_insertion_assembly_edge_case"><code class="docutils literal notranslate"><span class="pre">Assembly.format_insertion_assembly_edge_case()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_insertion_assemblies"><code class="docutils literal notranslate"><span class="pre">Assembly.get_insertion_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assemble_insertion"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_insertion()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_locations_on_fragments"><code class="docutils literal notranslate"><span class="pre">Assembly.get_locations_on_fragments()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assembly_uses_only_adjacent_edges"><code class="docutils literal notranslate"><span class="pre">Assembly.assembly_uses_only_adjacent_edges()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly"><code class="docutils literal notranslate"><span class="pre">PCRAssembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly.get_linear_assemblies"><code class="docutils literal notranslate"><span class="pre">PCRAssembly.get_linear_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly.get_circular_assemblies"><code class="docutils literal notranslate"><span class="pre">PCRAssembly.get_circular_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly.get_insertion_assemblies"><code class="docutils literal notranslate"><span class="pre">PCRAssembly.get_insertion_assemblies()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly.get_circular_assemblies"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly.get_circular_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly.get_insertion_assemblies"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly.get_insertion_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly.get_linear_assemblies"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly.get_linear_assemblies()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_function_assembly_products"><code class="docutils literal notranslate"><span class="pre">common_function_assembly_products()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gibson_assembly"><code class="docutils literal notranslate"><span class="pre">gibson_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.in_fusion_assembly"><code class="docutils literal notranslate"><span class="pre">in_fusion_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fusion_pcr_assembly"><code class="docutils literal notranslate"><span class="pre">fusion_pcr_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.in_vivo_assembly"><code class="docutils literal notranslate"><span class="pre">in_vivo_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.restriction_ligation_assembly"><code class="docutils literal notranslate"><span class="pre">restriction_ligation_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.golden_gate_assembly"><code class="docutils literal notranslate"><span class="pre">golden_gate_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.ligation_assembly"><code class="docutils literal notranslate"><span class="pre">ligation_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly_is_multi_site"><code class="docutils literal notranslate"><span class="pre">assembly_is_multi_site()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gateway_assembly"><code class="docutils literal notranslate"><span class="pre">gateway_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_function_integration_products"><code class="docutils literal notranslate"><span class="pre">common_function_integration_products()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_handle_insertion_fragments"><code class="docutils literal notranslate"><span class="pre">common_handle_insertion_fragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_function_excision_products"><code class="docutils literal notranslate"><span class="pre">common_function_excision_products()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.homologous_recombination_integration"><code class="docutils literal notranslate"><span class="pre">homologous_recombination_integration()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.homologous_recombination_excision"><code class="docutils literal notranslate"><span class="pre">homologous_recombination_excision()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.cre_lox_integration"><code class="docutils literal notranslate"><span class="pre">cre_lox_integration()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.cre_lox_excision"><code class="docutils literal notranslate"><span class="pre">cre_lox_excision()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.crispr_integration"><code class="docutils literal notranslate"><span class="pre">crispr_integration()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.pcr_assembly"><code class="docutils literal notranslate"><span class="pre">pcr_assembly()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.codon">pydna.codon module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.common_sub_strings">pydna.common_sub_strings module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.common_sub_strings.common_sub_strings"><code class="docutils literal notranslate"><span class="pre">common_sub_strings()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.common_sub_strings.terminal_overlap"><code class="docutils literal notranslate"><span class="pre">terminal_overlap()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.conftest">pydna.conftest module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.contig">pydna.contig module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig"><code class="docutils literal notranslate"><span class="pre">Contig</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.from_string"><code class="docutils literal notranslate"><span class="pre">Contig.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Contig.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Contig.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.rc"><code class="docutils literal notranslate"><span class="pre">Contig.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.detailed_figure"><code class="docutils literal notranslate"><span class="pre">Contig.detailed_figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.figure"><code class="docutils literal notranslate"><span class="pre">Contig.figure()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.cre_lox">pydna.cre_lox module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.cre_loxP_overlap"><code class="docutils literal notranslate"><span class="pre">cre_loxP_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.get_regex_dict"><code class="docutils literal notranslate"><span class="pre">get_regex_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.find_loxP_sites"><code class="docutils literal notranslate"><span class="pre">find_loxP_sites()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.annotate_loxP_sites"><code class="docutils literal notranslate"><span class="pre">annotate_loxP_sites()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.crispr">pydna.crispr module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9"><code class="docutils literal notranslate"><span class="pre">cas9</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.scaffold"><code class="docutils literal notranslate"><span class="pre">cas9.scaffold</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.pam"><code class="docutils literal notranslate"><span class="pre">cas9.pam</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.size"><code class="docutils literal notranslate"><span class="pre">cas9.size</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.fst5"><code class="docutils literal notranslate"><span class="pre">cas9.fst5</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.fst3"><code class="docutils literal notranslate"><span class="pre">cas9.fst3</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.ovhg"><code class="docutils literal notranslate"><span class="pre">cas9.ovhg</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.search"><code class="docutils literal notranslate"><span class="pre">cas9.search()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.protospacer"><code class="docutils literal notranslate"><span class="pre">protospacer()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.design">pydna.design module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.design.primer_design"><code class="docutils literal notranslate"><span class="pre">primer_design()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.design.assembly_fragments"><code class="docutils literal notranslate"><span class="pre">assembly_fragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.design.circular_assembly_fragments"><code class="docutils literal notranslate"><span class="pre">circular_assembly_fragments()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.download">pydna.download module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.download.download_text"><code class="docutils literal notranslate"><span class="pre">download_text()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.dseq">pydna.dseq module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq"><code class="docutils literal notranslate"><span class="pre">Dseq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.trunc"><code class="docutils literal notranslate"><span class="pre">Dseq.trunc</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.quick"><code class="docutils literal notranslate"><span class="pre">Dseq.quick()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.from_string"><code class="docutils literal notranslate"><span class="pre">Dseq.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.from_representation"><code class="docutils literal notranslate"><span class="pre">Dseq.from_representation()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.from_full_sequence_and_overhangs"><code class="docutils literal notranslate"><span class="pre">Dseq.from_full_sequence_and_overhangs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.mw"><code class="docutils literal notranslate"><span class="pre">Dseq.mw()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.upper"><code class="docutils literal notranslate"><span class="pre">Dseq.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.lower"><code class="docutils literal notranslate"><span class="pre">Dseq.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.find"><code class="docutils literal notranslate"><span class="pre">Dseq.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.rc"><code class="docutils literal notranslate"><span class="pre">Dseq.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.shifted"><code class="docutils literal notranslate"><span class="pre">Dseq.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.looped"><code class="docutils literal notranslate"><span class="pre">Dseq.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseq.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.five_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.five_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.three_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.three_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.watson_ovhg"><code class="docutils literal notranslate"><span class="pre">Dseq.watson_ovhg()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.fill_in"><code class="docutils literal notranslate"><span class="pre">Dseq.fill_in()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.transcribe"><code class="docutils literal notranslate"><span class="pre">Dseq.transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.translate"><code class="docutils literal notranslate"><span class="pre">Dseq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.mung"><code class="docutils literal notranslate"><span class="pre">Dseq.mung()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.T4"><code class="docutils literal notranslate"><span class="pre">Dseq.T4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.t4"><code class="docutils literal notranslate"><span class="pre">Dseq.t4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.exo1_front"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_front()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.exo1_end"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.seguid"><code class="docutils literal notranslate"><span class="pre">Dseq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.isblunt"><code class="docutils literal notranslate"><span class="pre">Dseq.isblunt()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cas9"><code class="docutils literal notranslate"><span class="pre">Dseq.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseq.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cut"><code class="docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cutsite_is_valid"><code class="docutils literal notranslate"><span class="pre">Dseq.cutsite_is_valid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.get_cutsites"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsites()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.left_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.left_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.right_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.right_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.get_cut_parameters"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cut_parameters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseq.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.get_cutsite_pairs"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsite_pairs()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.dseqrecord">pydna.dseqrecord module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.source"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.source</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.from_string"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.circular"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.circular</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.m"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.m()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.extract_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.add_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.add_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.seguid"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.looped"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.format"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.format()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.write"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.write()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.find"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.find_aminoacids"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aminoacids()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.find_aa"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aa()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.map_trace_files"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.map_trace_files()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.linearize"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.linearize()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.number_of_cuts"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.number_of_cuts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.cas9"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.rc"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.synced"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.synced()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.upper"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.lower"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.orfs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.orfs_to_features"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs_to_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_gb_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_fasta_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.figure"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.shifted"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.annotations"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.annotations</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.dbxrefs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.dbxrefs</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.history"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.history()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.fakeseq">pydna.fakeseq module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fakeseq.FakeSeq"><code class="docutils literal notranslate"><span class="pre">FakeSeq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fakeseq.FakeSeq.m"><code class="docutils literal notranslate"><span class="pre">FakeSeq.m()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fakeseq.FakeSeq.M"><code class="docutils literal notranslate"><span class="pre">FakeSeq.M()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.fusionpcr">pydna.fusionpcr module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fusionpcr.fuse_by_pcr"><code class="docutils literal notranslate"><span class="pre">fuse_by_pcr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fusionpcr.list_parts"><code class="docutils literal notranslate"><span class="pre">list_parts()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.gateway">pydna.gateway module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gateway.gateway_overlap"><code class="docutils literal notranslate"><span class="pre">gateway_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gateway.find_gateway_sites"><code class="docutils literal notranslate"><span class="pre">find_gateway_sites()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gateway.annotate_gateway_sites"><code class="docutils literal notranslate"><span class="pre">annotate_gateway_sites()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.gel">pydna.gel module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gel.interpolator"><code class="docutils literal notranslate"><span class="pre">interpolator()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gel.gel"><code class="docutils literal notranslate"><span class="pre">gel()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbank">pydna.genbank module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbank.Genbank"><code class="docutils literal notranslate"><span class="pre">Genbank</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbank.Genbank.nucleotide"><code class="docutils literal notranslate"><span class="pre">Genbank.nucleotide()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbank.genbank"><code class="docutils literal notranslate"><span class="pre">genbank()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbankfile">pydna.genbankfile module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile"><code class="docutils literal notranslate"><span class="pre">GenbankFile</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">GenbankFile.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile.reverse_complement"><code class="docutils literal notranslate"><span class="pre">GenbankFile.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile.rc"><code class="docutils literal notranslate"><span class="pre">GenbankFile.rc()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbankfixer">pydna.genbankfixer module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.parseGBLoc"><code class="docutils literal notranslate"><span class="pre">parseGBLoc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.strip_multiline"><code class="docutils literal notranslate"><span class="pre">strip_multiline()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.toInt"><code class="docutils literal notranslate"><span class="pre">toInt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.strip_indent"><code class="docutils literal notranslate"><span class="pre">strip_indent()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.concat_dict"><code class="docutils literal notranslate"><span class="pre">concat_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.toJSON"><code class="docutils literal notranslate"><span class="pre">toJSON()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.wrapstring"><code class="docutils literal notranslate"><span class="pre">wrapstring()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.locstr"><code class="docutils literal notranslate"><span class="pre">locstr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.originstr"><code class="docutils literal notranslate"><span class="pre">originstr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.toGB"><code class="docutils literal notranslate"><span class="pre">toGB()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.gbtext_clean"><code class="docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbankrecord">pydna.genbankrecord module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord"><code class="docutils literal notranslate"><span class="pre">GenbankRecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.from_string"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.rc"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.pydna_code"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.pydna_code()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.biopython_code"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.biopython_code()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.goldengate">pydna.goldengate module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.ladders">pydna.ladders module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.ligate">pydna.ligate module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.ligate.ligate"><code class="docutils literal notranslate"><span class="pre">ligate()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.opencloning_models">pydna.opencloning_models module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.id_mode"><code class="docutils literal notranslate"><span class="pre">id_mode()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.get_id"><code class="docutils literal notranslate"><span class="pre">get_id()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.from_biopython_location"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.from_biopython_location()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.to_biopython_location"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.to_biopython_location()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.field_validator"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.field_validator()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.from_start_and_end"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.from_start_and_end()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.ConfiguredBaseModel"><code class="docutils literal notranslate"><span class="pre">ConfiguredBaseModel</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.ConfiguredBaseModel.model_config"><code class="docutils literal notranslate"><span class="pre">ConfiguredBaseModel.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.TextFileSequence"><code class="docutils literal notranslate"><span class="pre">TextFileSequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.TextFileSequence.from_dseqrecord"><code class="docutils literal notranslate"><span class="pre">TextFileSequence.from_dseqrecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.TextFileSequence.model_config"><code class="docutils literal notranslate"><span class="pre">TextFileSequence.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PrimerModel"><code class="docutils literal notranslate"><span class="pre">PrimerModel</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PrimerModel.from_primer"><code class="docutils literal notranslate"><span class="pre">PrimerModel.from_primer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PrimerModel.model_config"><code class="docutils literal notranslate"><span class="pre">PrimerModel.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput"><code class="docutils literal notranslate"><span class="pre">SourceInput</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput.sequence"><code class="docutils literal notranslate"><span class="pre">SourceInput.sequence</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">SourceInput.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput.model_config"><code class="docutils literal notranslate"><span class="pre">SourceInput.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.left_location"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.left_location</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.right_location"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.right_location</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.reverse_complemented"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.reverse_complemented</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.from_biopython_location"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.from_biopython_location()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.model_config"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source"><code class="docutils literal notranslate"><span class="pre">Source</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.input"><code class="docutils literal notranslate"><span class="pre">Source.input</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">Source.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.input_models"><code class="docutils literal notranslate"><span class="pre">Source.input_models()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">Source.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.add_to_history_graph"><code class="docutils literal notranslate"><span class="pre">Source.add_to_history_graph()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.history_string"><code class="docutils literal notranslate"><span class="pre">Source.history_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.model_config"><code class="docutils literal notranslate"><span class="pre">Source.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource"><code class="docutils literal notranslate"><span class="pre">AssemblySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.circular"><code class="docutils literal notranslate"><span class="pre">AssemblySource.circular</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">AssemblySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">AssemblySource.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.from_subfragment_representation"><code class="docutils literal notranslate"><span class="pre">AssemblySource.from_subfragment_representation()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.model_config"><code class="docutils literal notranslate"><span class="pre">AssemblySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource.restriction_enzymes"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource.restriction_enzymes</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource.model_config"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GibsonAssemblySource"><code class="docutils literal notranslate"><span class="pre">GibsonAssemblySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GibsonAssemblySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">GibsonAssemblySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GibsonAssemblySource.model_config"><code class="docutils literal notranslate"><span class="pre">GibsonAssemblySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InFusionSource"><code class="docutils literal notranslate"><span class="pre">InFusionSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InFusionSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">InFusionSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InFusionSource.model_config"><code class="docutils literal notranslate"><span class="pre">InFusionSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource"><code class="docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource.model_config"><code class="docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InVivoAssemblySource"><code class="docutils literal notranslate"><span class="pre">InVivoAssemblySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InVivoAssemblySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">InVivoAssemblySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InVivoAssemblySource.model_config"><code class="docutils literal notranslate"><span class="pre">InVivoAssemblySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.LigationSource"><code class="docutils literal notranslate"><span class="pre">LigationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.LigationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">LigationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.LigationSource.model_config"><code class="docutils literal notranslate"><span class="pre">LigationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource"><code class="docutils literal notranslate"><span class="pre">GatewaySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">GatewaySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.reaction_type"><code class="docutils literal notranslate"><span class="pre">GatewaySource.reaction_type</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.greedy"><code class="docutils literal notranslate"><span class="pre">GatewaySource.greedy</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.model_config"><code class="docutils literal notranslate"><span class="pre">GatewaySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.HomologousRecombinationSource"><code class="docutils literal notranslate"><span class="pre">HomologousRecombinationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.HomologousRecombinationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">HomologousRecombinationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.HomologousRecombinationSource.model_config"><code class="docutils literal notranslate"><span class="pre">HomologousRecombinationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CRISPRSource"><code class="docutils literal notranslate"><span class="pre">CRISPRSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CRISPRSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">CRISPRSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CRISPRSource.model_config"><code class="docutils literal notranslate"><span class="pre">CRISPRSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CreLoxRecombinationSource"><code class="docutils literal notranslate"><span class="pre">CreLoxRecombinationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CreLoxRecombinationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">CreLoxRecombinationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CreLoxRecombinationSource.model_config"><code class="docutils literal notranslate"><span class="pre">CreLoxRecombinationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource"><code class="docutils literal notranslate"><span class="pre">PCRSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">PCRSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource.add_primer_features"><code class="docutils literal notranslate"><span class="pre">PCRSource.add_primer_features</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource.model_config"><code class="docutils literal notranslate"><span class="pre">PCRSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.left_edge"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.left_edge</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.right_edge"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.right_edge</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.BASE_MODEL"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.BASE_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.ENZYME_MODEL"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.ENZYME_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.from_parent"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.from_parent()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.model_config"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy"><code class="docutils literal notranslate"><span class="pre">CloningStrategy</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.primers"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.add_primer"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.add_primer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.add_dseqrecord"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.add_dseqrecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.reassign_ids"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.reassign_ids()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.from_dseqrecords"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.from_dseqrecords()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.model_dump_json"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.model_dump_json()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.model_dump"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.model_dump()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.model_config"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.model_config</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.parsers">pydna.parsers module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.extract_from_text"><code class="docutils literal notranslate"><span class="pre">extract_from_text()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.embl_gb_fasta"><code class="docutils literal notranslate"><span class="pre">embl_gb_fasta()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.parse"><code class="docutils literal notranslate"><span class="pre">parse()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.parse_primers"><code class="docutils literal notranslate"><span class="pre">parse_primers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.parse_snapgene"><code class="docutils literal notranslate"><span class="pre">parse_snapgene()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.primer">pydna.primer module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer"><code class="docutils literal notranslate"><span class="pre">Primer</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer.footprint"><code class="docutils literal notranslate"><span class="pre">Primer.footprint</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer.tail"><code class="docutils literal notranslate"><span class="pre">Primer.tail</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Primer.reverse_complement()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.readers">pydna.readers module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.readers.read"><code class="docutils literal notranslate"><span class="pre">read()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.readers.read_primer"><code class="docutils literal notranslate"><span class="pre">read_primer()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.seq">pydna.seq module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq"><code class="docutils literal notranslate"><span class="pre">Seq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.translate"><code class="docutils literal notranslate"><span class="pre">Seq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.gc"><code class="docutils literal notranslate"><span class="pre">Seq.gc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.cai"><code class="docutils literal notranslate"><span class="pre">Seq.cai()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.rarecodons"><code class="docutils literal notranslate"><span class="pre">Seq.rarecodons()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.startcodon"><code class="docutils literal notranslate"><span class="pre">Seq.startcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.stopcodon"><code class="docutils literal notranslate"><span class="pre">Seq.stopcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.express"><code class="docutils literal notranslate"><span class="pre">Seq.express()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.orfs2"><code class="docutils literal notranslate"><span class="pre">Seq.orfs2()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.orfs"><code class="docutils literal notranslate"><span class="pre">Seq.orfs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.seguid"><code class="docutils literal notranslate"><span class="pre">Seq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Seq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.rc"><code class="docutils literal notranslate"><span class="pre">Seq.rc()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq"><code class="docutils literal notranslate"><span class="pre">ProteinSeq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.translate"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.complement_rna"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.complement_rna()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.rc"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.reverse_complement_rna"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.reverse_complement_rna()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.transcribe"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.back_transcribe"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.back_transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.seguid"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.molecular_weight"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.molecular_weight()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.pI"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.pI()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.instability_index"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.instability_index()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.seqrecord">pydna.seqrecord module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord.from_Bio_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.locus"><code class="docutils literal notranslate"><span class="pre">SeqRecord.locus</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.accession"><code class="docutils literal notranslate"><span class="pre">SeqRecord.accession</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.definition"><code class="docutils literal notranslate"><span class="pre">SeqRecord.definition</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">SeqRecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">SeqRecord.isorf()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.translate"><code class="docutils literal notranslate"><span class="pre">SeqRecord.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_colors_to_features_for_ape()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.add_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.list_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.list_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.extract_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.sorted_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.sorted_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.seguid"><code class="docutils literal notranslate"><span class="pre">SeqRecord.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.comment"><code class="docutils literal notranslate"><span class="pre">SeqRecord.comment()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.datefunction"><code class="docutils literal notranslate"><span class="pre">SeqRecord.datefunction()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.stamp"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stamp()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.lcs"><code class="docutils literal notranslate"><span class="pre">SeqRecord.lcs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.gc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">SeqRecord.cai()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rarecodons()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.startcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stopcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.express"><code class="docutils literal notranslate"><span class="pre">SeqRecord.express()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.copy"><code class="docutils literal notranslate"><span class="pre">SeqRecord.copy()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.dump"><code class="docutils literal notranslate"><span class="pre">SeqRecord.dump()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.isorf()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.gc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.cai()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rarecodons()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.startcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.stopcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.express"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.express()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.sequence_picker">pydna.sequence_picker module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.sequence_picker.genbank_accession"><code class="docutils literal notranslate"><span class="pre">genbank_accession()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.sequence_regex">pydna.sequence_regex module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.sequence_regex.compute_regex_site"><code class="docutils literal notranslate"><span class="pre">compute_regex_site()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.sequence_regex.dseqrecord_finditer"><code class="docutils literal notranslate"><span class="pre">dseqrecord_finditer()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.threading_timer_decorator_exit">pydna.threading_timer_decorator_exit module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.cdquit"><code class="docutils literal notranslate"><span class="pre">cdquit()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.exit_after"><code class="docutils literal notranslate"><span class="pre">exit_after()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.a"><code class="docutils literal notranslate"><span class="pre">a()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.b"><code class="docutils literal notranslate"><span class="pre">b()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.c"><code class="docutils literal notranslate"><span class="pre">c()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.d"><code class="docutils literal notranslate"><span class="pre">d()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.countdown"><code class="docutils literal notranslate"><span class="pre">countdown()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.main"><code class="docutils literal notranslate"><span class="pre">main()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.tm">pydna.tm module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_default"><code class="docutils literal notranslate"><span class="pre">tm_default()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_dbd"><code class="docutils literal notranslate"><span class="pre">tm_dbd()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_product"><code class="docutils literal notranslate"><span class="pre">tm_product()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.ta_default"><code class="docutils literal notranslate"><span class="pre">ta_default()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.ta_dbd"><code class="docutils literal notranslate"><span class="pre">ta_dbd()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.program"><code class="docutils literal notranslate"><span class="pre">program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.taq_program"><code class="docutils literal notranslate"><span class="pre">taq_program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.dbd_program"><code class="docutils literal notranslate"><span class="pre">dbd_program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.pfu_sso7d_program"><code class="docutils literal notranslate"><span class="pre">pfu_sso7d_program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.Q5"><code class="docutils literal notranslate"><span class="pre">Q5()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tmbresluc"><code class="docutils literal notranslate"><span class="pre">tmbresluc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_neb"><code class="docutils literal notranslate"><span class="pre">tm_neb()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.types">pydna.types module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.user_cloning">pydna.user_cloning module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.utils">pydna.utils module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.three_frame_orfs"><code class="docutils literal notranslate"><span class="pre">three_frame_orfs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.shift_location"><code class="docutils literal notranslate"><span class="pre">shift_location()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.shift_feature"><code class="docutils literal notranslate"><span class="pre">shift_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.smallest_rotation"><code class="docutils literal notranslate"><span class="pre">smallest_rotation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.cai"><code class="docutils literal notranslate"><span class="pre">cai()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.rarecodons"><code class="docutils literal notranslate"><span class="pre">rarecodons()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.express"><code class="docutils literal notranslate"><span class="pre">express()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.open_folder"><code class="docutils literal notranslate"><span class="pre">open_folder()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.rc"><code class="docutils literal notranslate"><span class="pre">rc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.complement"><code class="docutils literal notranslate"><span class="pre">complement()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.identifier_from_string"><code class="docutils literal notranslate"><span class="pre">identifier_from_string()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.flatten"><code class="docutils literal notranslate"><span class="pre">flatten()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.seq31"><code class="docutils literal notranslate"><span class="pre">seq31()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomRNA"><code class="docutils literal notranslate"><span class="pre">randomRNA()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomDNA"><code class="docutils literal notranslate"><span class="pre">randomDNA()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomORF"><code class="docutils literal notranslate"><span class="pre">randomORF()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomprot"><code class="docutils literal notranslate"><span class="pre">randomprot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.eq"><code class="docutils literal notranslate"><span class="pre">eq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.cuts_overlap"><code class="docutils literal notranslate"><span class="pre">cuts_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.location_boundaries"><code class="docutils literal notranslate"><span class="pre">location_boundaries()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.locations_overlap"><code class="docutils literal notranslate"><span class="pre">locations_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.sum_is_sticky"><code class="docutils literal notranslate"><span class="pre">sum_is_sticky()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.limit_iterator"><code class="docutils literal notranslate"><span class="pre">limit_iterator()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.create_location"><code class="docutils literal notranslate"><span class="pre">create_location()</span></code></a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="module-pydna">
<span id="pydna-package"></span><h1>pydna package<a class="headerlink" href="#module-pydna" title="Link to this heading">#</a></h1>
<dl class="field-list simple">
<dt class="field-odd">copyright<span class="colon">:</span></dt>
<dd class="field-odd"><p>Copyright 2013-2023 by Björn Johansson. All rights reserved.</p>
</dd>
<dt class="field-even">license<span class="colon">:</span></dt>
<dd class="field-even"><p>This code is part of the pydna package, governed by the
license in LICENSE.txt that should be included as part
of this package.</p>
</dd>
</dl>
<section id="pydna">
<h2>pydna<a class="headerlink" href="#pydna" title="Link to this heading">#</a></h2>
<p>Pydna is a python package providing code for simulation of the creation of
recombinant DNA molecules using
<a class="reference external" href="https://en.wikipedia.org/wiki/Molecular_biology">molecular biology</a>
techniques. Development of pydna happens in this Github <a class="reference external" href="https://github.com/pydna-group/pydna">repository</a>.</p>
<dl class="simple">
<dt>Provided:</dt><dd><ol class="arabic simple">
<li><p>PCR simulation</p></li>
<li><p>Assembly simulation based on shared identical sequences</p></li>
<li><p>Primer design for amplification of a given sequence</p></li>
<li><p>Automatic design of primer tails for Gibson assembly
or homologous recombination.</p></li>
<li><p>Restriction digestion and cut&amp;paste cloning</p></li>
<li><p>Agarose gel simulation</p></li>
<li><p>Download sequences from Genbank</p></li>
<li><p>Parsing various sequence formats including the capacity to
handle broken Genbank format</p></li>
</ol>
</dd>
</dl>
<section id="pydna-package-layout">
<h3>pydna package layout<a class="headerlink" href="#pydna-package-layout" title="Link to this heading">#</a></h3>
<p>The most important modules and how to import functions or classes from
them are listed below. Class names starts with a capital letter,
functions with a lowercase letter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>from pydna.module import function
from pydna.module import Class

Example: from pydna.gel import Gel

pydna
   ├── amplify
   │         ├── Anneal
   │         └── pcr
   ├── assembly
   │          └── Assembly
   ├── design
   │        ├── assembly_fragments
   │        └── primer_design
   ├── download
   │          └── download_text
   ├── dseqrecord
   │            └── Dseqrecord
   ├── gel
   │     └── Gel
   ├── genbank
   │         ├── genbank
   │         └── Genbank
   ├── parsers
   │         ├── parse
   │         └── parse_primers
   └── readers
             ├── read
             └── read_primers
</pre></div>
</div>
</section>
<section id="how-to-use-the-documentation">
<h3>How to use the documentation<a class="headerlink" href="#how-to-use-the-documentation" title="Link to this heading">#</a></h3>
<p>Documentation is available as docstrings provided in the source code for
each module.
These docstrings can be inspected by reading the source code directly.
See further below on how to obtain the code for pydna.</p>
<p>In the python shell, use the built-in <code class="docutils literal notranslate"><span class="pre">help</span></code> function to view a
function’s docstring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna</span><span class="w"> </span><span class="kn">import</span> <span class="n">readers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">readers</span><span class="o">.</span><span class="n">read</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>The doctrings are also used to provide an automaticly generated reference
manual available online at
<a class="reference external" href="https://pydna-group.github.io/pydna">read the docs</a>.</p>
<p>Docstrings can be explored using <a class="reference external" href="http://ipython.org/">IPython</a>, an
advanced Python shell with
TAB-completion and introspection capabilities. To see which functions
are available in <cite>pydna</cite>,
type <cite>pydna.&lt;TAB&gt;</cite> (where <cite>&lt;TAB&gt;</cite> refers to the TAB key).
Use <cite>pydna.open_config_folder?&lt;ENTER&gt;`to view the docstring or
`pydna.open_config_folder??&lt;ENTER&gt;</cite> to view the source code.</p>
<p>In the <a class="reference external" href="https://github.com/spyder-ide/spyder">Spyder IDE</a> it is possible
to place the cursor immediately before the name of a module,class or
function and press ctrl+i to bring up docstrings in a separate window in Spyder</p>
<p>Code snippets are indicated by three greater-than signs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="mi">41</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">42</span>
</pre></div>
</div>
</section>
<section id="pydna-source-code">
<h3>pydna source code<a class="headerlink" href="#pydna-source-code" title="Link to this heading">#</a></h3>
<p>The pydna source code is
<a class="reference external" href="https://github.com/pydna-group/pydna">available on Github</a>.</p>
</section>
<section id="how-to-get-more-help">
<h3>How to get more help<a class="headerlink" href="#how-to-get-more-help" title="Link to this heading">#</a></h3>
<p>Please join the
<a class="reference external" href="https://groups.google.com/forum/#!forum/pydna">Google group</a>
for pydna, this is the preferred location for help. If you find bugs
in pydna itself, open an issue at the
<a class="reference external" href="https://github.com/pydna-group/pydna/issues">Github repository</a>.</p>
</section>
<section id="examples-of-pydna-in-use">
<h3>Examples of pydna in use<a class="headerlink" href="#examples-of-pydna-in-use" title="Link to this heading">#</a></h3>
<dl class="simple">
<dt>See this repository for a collection of</dt><dd><p><a class="reference external" href="https://github.com/MetabolicEngineeringGroupCBMA/pydna-examples?tab=readme-ov-file#pydna-examples">examples</a>.</p>
</dd>
</dl>
</section>
</section>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.get_env">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">get_env</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna.html#get_env"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.get_env" title="Link to this definition">#</a></dt>
<dd><p>Print a an ascii table containing all environmental variables.</p>
<p>Pydna related variables have names that starts with <cite>pydna_</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.logo">
<span class="sig-prename descclassname"><span class="pre">pydna.</span></span><span class="sig-name descname"><span class="pre">logo</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna.html#logo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.logo" title="Link to this definition">#</a></dt>
<dd><p>Ascii-art logotype of pydna.</p>
</dd></dl>

<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">#</a></h2>
</section>
<section id="module-pydna.all">
<span id="pydna-all-module"></span><h2>pydna.all module<a class="headerlink" href="#module-pydna.all" title="Link to this heading">#</a></h2>
<p>This module provide most pydna functionality in the local namespace.</p>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.all</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.all</span><span class="w"> </span><span class="kn">import</span> <span class="n">__all__</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">__all__</span>
<span class="go">[&#39;Anneal&#39;, &#39;pcr&#39;, &#39;Assembly&#39;, &#39;genbank&#39;, &#39;Genbank&#39;, &#39;download_text&#39;, &#39;Dseqrecord&#39;,</span>
<span class="go">&#39;Dseq&#39;, &#39;read&#39;, &#39;read_primer&#39;, &#39;parse&#39;, &#39;parse_primers&#39;, &#39;primer_design&#39;, &#39;assembly_fragments&#39;, &#39;eq&#39;, &#39;gbtext_clean&#39;]</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.all.Anneal">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">Anneal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplify.html#Anneal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Anneal" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The Anneal class has the following important attributes:</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.all.Anneal.forward_primers">
<span class="sig-name descname"><span class="pre">forward_primers</span></span><a class="headerlink" href="#pydna.all.Anneal.forward_primers" title="Link to this definition">#</a></dt>
<dd><p>Description of <cite>forward_primers</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.all.Anneal.reverse_primers">
<span class="sig-name descname"><span class="pre">reverse_primers</span></span><a class="headerlink" href="#pydna.all.Anneal.reverse_primers" title="Link to this definition">#</a></dt>
<dd><p>Description of <cite>reverse_primers</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.all.Anneal.template">
<span class="sig-name descname"><span class="pre">template</span></span><a class="headerlink" href="#pydna.all.Anneal.template" title="Link to this definition">#</a></dt>
<dd><p>A copy of the template argument. Primers annealing sites has been
added as features that can be visualized in a seqence editor such as
ApE.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.all.Anneal.limit">
<span class="sig-name descname"><span class="pre">limit</span></span><a class="headerlink" href="#pydna.all.Anneal.limit" title="Link to this definition">#</a></dt>
<dd><p>The limit of PCR primer annealing, default is 13 bp.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.all.Anneal.products">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">products</span></span><a class="headerlink" href="#pydna.all.Anneal.products" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Anneal.report">
<span class="sig-name descname"><span class="pre">report</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.all.Anneal.report" title="Link to this definition">#</a></dt>
<dd><p>returns a short report describing if or where primer
anneal on the template.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.pcr">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">pcr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><span class="pre">Amplicon</span></a></span></span><a class="reference internal" href="../_modules/pydna/amplify.html#pcr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.pcr" title="Link to this definition">#</a></dt>
<dd><p>pcr is a convenience function for the Anneal class to simplify its
usage, especially from the command line. If more than one or no PCR
product is formed, a ValueError is raised.</p>
<p>args is any iterable of Dseqrecords or an iterable of iterables of
Dseqrecords. args will be greedily flattened.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>iterable containing sequence objects</em>) – Several arguments are also accepted.</p></li>
<li><p><strong>limit</strong> (<em>int = 13</em><em>, </em><em>optional</em>) – limit length of the annealing part of the primers.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>sequences in args could be of type:</p>
<ul class="simple">
<li><p>string</p></li>
<li><p>Seq</p></li>
<li><p>SeqRecord (or subclass)</p></li>
<li><p>Dseqrecord (or sublcass)</p></li>
</ul>
<p>The last sequence will be assumed to be the template while
all preceeding sequences will be assumed to be primers.</p>
<p>This is a powerful function, use with care!</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>product</strong> – An <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> object representing the PCR
product. The direction of the PCR product will be the same as for
the template sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon">Amplicon</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.readers</span><span class="w"> </span><span class="kn">import</span> <span class="n">read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.amplify</span><span class="w"> </span><span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;cgactgtatcatctgatagcac&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">([</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.all.Assembly">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">Assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">common_sub_strings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly.html#Assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Assembly" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Assembly of a list of linear DNA fragments into linear or circular
constructs. The Assembly is meant to replace the Assembly method as it
is easier to use. Accepts a list of Dseqrecords (source fragments) to
initiate an Assembly object. Several methods are available for analysis
of overlapping sequences, graph construction and assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fragments</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – a list of Dseqrecord objects.</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – The shortest shared homology to be considered</p></li>
<li><p><strong>algorithm</strong> (<em>function</em><em>, </em><em>optional</em>) – The algorithm used to determine the shared sequences.</p></li>
<li><p><strong>max_nodes</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The maximum number of nodes in the graph. This can be tweaked to
manage sequences with a high number of shared sub sequences.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly</span><span class="w"> </span><span class="kn">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;acgatgctatactgCCCCCtgtgctgtgctcta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtgctgtgctctaTTTTTtattctggctgtatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tattctggctgtatcGGGGGtacgatgctatactg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">((</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">),</span> <span class="n">limit</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 33bp 34bp 35bp</span>
<span class="go">limit(bp)....: 14</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()</span>
<span class="go">[Contig(o59), Contig(o59)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">watson</span>
<span class="go">&#39;acgatgctatactgCCCCCtgtgctgtgctctaTTTTTtattctggctgtatcGGGGGt&#39;</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Assembly.assemble_circular">
<span class="sig-name descname"><span class="pre">assemble_circular</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.all.Assembly.assemble_circular" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Assembly.assemble_linear">
<span class="sig-name descname"><span class="pre">assemble_linear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.all.Assembly.assemble_linear" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.genbank">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">genbank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">accession</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CS570233.1'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord" title="pydna.genbankrecord.GenbankRecord"><span class="pre">GenbankRecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/genbank.html#genbank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.genbank" title="Link to this definition">#</a></dt>
<dd><p>Download a genbank nuclotide record.</p>
<p>This function takes the same paramenters as the
:func:pydna.genbank.Genbank.nucleotide method. The email address stored
in the <cite>pydna_email</cite> environment variable is used. The easiest way set
this permanantly is to edit the <cite>pydna.ini</cite> file.
See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<p>if no accession is given, a <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/CS570233.1">very short Genbank
entry</a>
is used as an example (see below). This can be useful for testing the
connection to Genbank.</p>
<p>Please note that this result is also cached by default by settings in
the pydna.ini file.
See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">LOCUS</span>       <span class="n">CS570233</span>                  <span class="mi">14</span> <span class="n">bp</span>    <span class="n">DNA</span>     <span class="n">linear</span>   <span class="n">PAT</span> <span class="mi">18</span><span class="o">-</span><span class="n">MAY</span><span class="o">-</span><span class="mi">2007</span>
<span class="n">DEFINITION</span>  <span class="n">Sequence</span> <span class="mi">6</span> <span class="kn">from</span><span class="w"> </span><span class="nn">Patent</span> <span class="n">WO2007025016</span><span class="o">.</span>
<span class="n">ACCESSION</span>   <span class="n">CS570233</span>
<span class="n">VERSION</span>     <span class="n">CS570233</span><span class="mf">.1</span>
<span class="n">KEYWORDS</span>    <span class="o">.</span>
<span class="n">SOURCE</span>      <span class="n">synthetic</span> <span class="n">construct</span>
  <span class="n">ORGANISM</span>  <span class="n">synthetic</span> <span class="n">construct</span>
            <span class="n">other</span> <span class="n">sequences</span><span class="p">;</span> <span class="n">artificial</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">REFERENCE</span>   <span class="mi">1</span>
  <span class="n">AUTHORS</span>   <span class="n">Shaw</span><span class="p">,</span><span class="n">R</span><span class="o">.</span><span class="n">W</span><span class="o">.</span> <span class="ow">and</span> <span class="n">Cottenoir</span><span class="p">,</span><span class="n">M</span><span class="o">.</span>
  <span class="n">TITLE</span>     <span class="n">Inhibition</span> <span class="n">of</span> <span class="n">metallo</span><span class="o">-</span><span class="n">beta</span><span class="o">-</span><span class="n">lactamase</span> <span class="n">by</span> <span class="n">double</span><span class="o">-</span><span class="n">stranded</span> <span class="n">dna</span>
  <span class="n">JOURNAL</span>   <span class="n">Patent</span><span class="p">:</span> <span class="n">WO</span> <span class="mi">2007025016</span><span class="o">-</span><span class="n">A1</span> <span class="mi">6</span> <span class="mi">01</span><span class="o">-</span><span class="n">MAR</span><span class="o">-</span><span class="mi">2007</span><span class="p">;</span>
            <span class="n">Texas</span> <span class="n">Tech</span> <span class="n">University</span> <span class="n">System</span> <span class="p">(</span><span class="n">US</span><span class="p">)</span>
<span class="n">FEATURES</span>             <span class="n">Location</span><span class="o">/</span><span class="n">Qualifiers</span>
     <span class="n">source</span>          <span class="mf">1..14</span>
                     <span class="o">/</span><span class="n">organism</span><span class="o">=</span><span class="s2">&quot;synthetic construct&quot;</span>
                     <span class="o">/</span><span class="n">mol_type</span><span class="o">=</span><span class="s2">&quot;unassigned DNA&quot;</span>
                     <span class="o">/</span><span class="n">db_xref</span><span class="o">=</span><span class="s2">&quot;taxon:32630&quot;</span>
                     <span class="o">/</span><span class="n">note</span><span class="o">=</span><span class="s2">&quot;This is a 14bp aptamer inhibitor.&quot;</span>
<span class="n">ORIGIN</span>
        <span class="mi">1</span> <span class="n">atgttcctac</span> <span class="n">atga</span>
<span class="o">//</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.all.Genbank">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">Genbank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">users_email</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">tool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pydna'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbank.html#Genbank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Genbank" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Class to facilitate download from genbank. It is easier and
quicker to use the <a class="reference internal" href="#pydna.genbank.genbank" title="pydna.genbank.genbank"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.genbank()</span></code></a> function directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>users_email</strong> (<em>string</em>) – Has to be a valid email address. You should always tell
Genbanks who you are, so that they can contact you.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.genbank</span><span class="w"> </span><span class="kn">import</span> <span class="n">Genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gb</span><span class="o">=</span><span class="n">Genbank</span><span class="p">(</span><span class="s2">&quot;bjornjobb@gmail.com&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span> <span class="o">=</span> <span class="n">gb</span><span class="o">.</span><span class="n">nucleotide</span><span class="p">(</span><span class="s2">&quot;LP002422.1&quot;</span><span class="p">)</span>   <span class="c1"># &lt;- entry from genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Genbank.nucleotide">
<span class="sig-name descname"><span class="pre">nucleotide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Literal" title="(in Python v3.8)"><span class="pre">Literal</span></a><span class="p"><span class="pre">[</span></span><span class="m"><span class="pre">1</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="m"><span class="pre">2</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord" title="pydna.genbankrecord.GenbankRecord"><span class="pre">GenbankRecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/genbank.html#Genbank.nucleotide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Genbank.nucleotide" title="Link to this definition">#</a></dt>
<dd><p>This method downloads a genbank nuclotide record from genbank. This method is
cached by default. This can be controlled by editing the <strong>pydna_cached_funcs</strong> environment
variable. The best way to do this permanently is to edit the edit the
<cite>pydna.ini</cite> file. See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<p>Item is a string containing one genbank accession number
for a nucleotide file. Genbank nucleotide accession numbers have this format:</p>
<div class="line-block">
<div class="line">A12345   = 1 letter  + 5 numerals</div>
<div class="line">AB123456 = 2 letters + 6 numerals</div>
</div>
<p>The accession number is sometimes followed by a point and version number</p>
<div class="line-block">
<div class="line">BK006936.2</div>
</div>
<p>Item can also contain optional interval information in the following formats:</p>
<div class="line-block">
<div class="line">BK006936.2 REGION: complement(613900..615202)</div>
<div class="line">NM_005546 REGION: 1..100</div>
<div class="line">NM_005546 REGION: complement(1..100)</div>
<div class="line">21614549:1-100</div>
<div class="line">21614549:c100-1</div>
<div class="line">21614549 1-100</div>
<div class="line">21614549 c100-1</div>
</div>
<p>It is useful to set an interval for large genbank records to limit the download time.
The items above containing interval information and can be obtained directly by
looking up an entry in Genbank and setting the <cite>Change region shown</cite> on the
upper right side of the page. The <cite>ACCESSION</cite> line of the displayed Genbank
file will have the formatting shown.</p>
<p>Alternatively, seq_start and seq_stop can be set explicitly to the sequence intervals to be
downloaded.</p>
<p>If strand is 2. “c”, “C”, “crick”, “Crick”, “antisense”,”Antisense”,
“2”, 2, “-” or “-1”, the antisense (Crick) strand is returned, otherwise
the sense (Watson) strand is returned.</p>
<p>Result is returned as a <a class="reference internal" href="#pydna.genbankrecord.GenbankRecord" title="pydna.genbankrecord.GenbankRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.genbankrecord.GenbankRecord</span></code></a> object.</p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id1" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">1</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm">http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm</a></p>
</aside>
<aside class="footnote brackets" id="id2" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">2</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch</a></p>
</aside>
</aside>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.download_text">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">download_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/download.html#download_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.download_text" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">Dseqrecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<p>Dseqrecord is a double stranded version of the Biopython SeqRecord <a class="footnote-reference brackets" href="#id1" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> class.
The Dseqrecord object holds a Dseq object describing the sequence.
Additionally, Dseqrecord hold meta information about the sequence in the
from of a list of SeqFeatures, in the same way as the SeqRecord does.</p>
<p>The Dseqrecord can be initialized with a string, Seq, Dseq, SeqRecord
or another Dseqrecord. The sequence information will be stored in a
Dseq object in all cases.</p>
<p>Dseqrecord objects can be read or parsed from sequences in FASTA, EMBL or Genbank formats.
See the <a class="reference internal" href="#module-pydna.readers" title="pydna.readers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.readers</span></code></a> and <a class="reference internal" href="#module-pydna.parsers" title="pydna.parsers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.parsers</span></code></a> modules for further information.</p>
<p>There is a short representation associated with the Dseqrecord.
<code class="docutils literal notranslate"><span class="pre">Dseqrecord(-3)</span></code> represents a linear sequence of length 2
while <code class="docutils literal notranslate"><span class="pre">Dseqrecord(o7)</span></code>
represents a circular sequence of length 7.</p>
<p>Dseqrecord and Dseq share the same concept of length. This length can be larger
than each strand alone if they are staggered as in the example below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<em>string</em><em>, </em><a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><em>Seq</em></a><em>, </em><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>, </em><a class="reference internal" href="#pydna.all.Dseq" title="pydna.all.Dseq"><em>Dseq</em></a><em> or </em><em>other Dseqrecord object</em>) – This data will be used to form the seq property</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True or False reflecting the shape of the DNA molecule</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True or False reflecting the shape of the DNA molecule</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">3</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</aside>
</aside>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'misc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.add_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.add_feature" title="Link to this definition">#</a></dt>
<dd><p>Add a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates start of the feature</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates end of the feature</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4), strand=1), type=&#39;misc&#39;, qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.apply_cut">
<span class="sig-name descname"><span class="pre">apply_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.apply_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.apply_cut" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.cas9">
<span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RNA</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.cas9" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.circular">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">circular</span></span><a class="headerlink" href="#pydna.all.Dseqrecord.circular" title="Link to this definition">#</a></dt>
<dd><p>The circular property can not be set directly.
Use <a class="reference internal" href="#pydna.all.Dseqrecord.looped" title="pydna.all.Dseqrecord.looped"><code class="xref py py-meth docutils literal notranslate"><span class="pre">looped()</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.copy_fasta_to_clipboard">
<span class="sig-name descname"><span class="pre">copy_fasta_to_clipboard</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.copy_fasta_to_clipboard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.copy_fasta_to_clipboard" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.copy_gb_to_clipboard">
<span class="sig-name descname"><span class="pre">copy_gb_to_clipboard</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.copy_gb_to_clipboard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.copy_gb_to_clipboard" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.cut">
<span class="sig-name descname"><span class="pre">cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.cut" title="Link to this definition">#</a></dt>
<dd><p>Digest a Dseqrecord object with one or more restriction enzymes.</p>
<p>returns a list of linear Dseqrecords. If there are no cuts, an empty
list is returned.</p>
<p>See also <a class="reference internal" href="#pydna.all.Dseq.cut" title="pydna.all.Dseq.cut"><code class="xref py py-func docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></a>
:param enzymes: A Bio.Restriction.XXX restriction object or iterable of such.
:type enzymes: enzyme object or iterable of such objects</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>Dseqrecord_frags</strong> – list of Dseqrecord objects formed by the digestion</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">BamHI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="go">(Dseqrecord(-5), Dseqrecord(-5))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="p">,</span> <span class="n">frag2</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag2</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">gatcc</span>
<span class="go">    g</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.cutters">
<span class="sig-name descname"><span class="pre">cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.extract_feature">
<span class="sig-name descname"><span class="pre">extract_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.extract_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.extract_feature" title="Link to this definition">#</a></dt>
<dd><p>Extracts a feature and creates a new Dseqrecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates the feature to extract</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-2)</span>
<span class="go">gt</span>
<span class="go">ca</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\x1b[48;5;11m'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\x1b[0m'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.figure" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.find" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.find_aa">
<span class="sig-name descname"><span class="pre">find_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.all.Dseqrecord.find_aa" title="Link to this definition">#</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.find_aminoacids">
<span class="sig-name descname"><span class="pre">find_aminoacids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.find_aminoacids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.find_aminoacids" title="Link to this definition">#</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.format">
<span class="sig-name descname"><span class="pre">format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gb'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.format" title="Link to this definition">#</a></dt>
<dd><p>Returns the sequence as a string using a format supported by Biopython
SeqIO <a class="footnote-reference brackets" href="#id2" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. Default is “gb” which is short for Genbank.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;gb&quot;</span><span class="p">))</span>
<span class="go">LOCUS       name                       3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  description.</span>
<span class="go">ACCESSION   id</span>
<span class="go">VERSION     id</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id6" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">4</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.from_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><span class="pre">SeqRecord</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.from_SeqRecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.from_string">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.from_string" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.history">
<span class="sig-name descname"><span class="pre">history</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.history"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.history" title="Link to this definition">#</a></dt>
<dd><p>Returns a string representation of the cloning history of the sequence.
Returns an empty string if the sequence has no source.</p>
<p>Check the documentation notebooks for extensive examples.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>str</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>A string representation of the cloning history of the sequence.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">gibson_assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fragments</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>   <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;TTTTacgatAAtgctccCCCC&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment1&quot;</span><span class="p">),</span>
<span class="gp">... </span>   <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;CCCCtcatGGGG&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment2&quot;</span><span class="p">),</span>
<span class="gp">... </span>   <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;GGGGatataTTTT&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment3&quot;</span><span class="p">),</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">product</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">gibson_assembly</span><span class="p">(</span><span class="n">fragments</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">product</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;product_name&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">product</span><span class="o">.</span><span class="n">history</span><span class="p">())</span>
<span class="go">╙── product_name (Dseqrecord(o34))</span>
<span class="go">    └─╼ GibsonAssemblySource</span>
<span class="go">        ├─╼ fragment1 (Dseqrecord(-21))</span>
<span class="go">        ├─╼ fragment2 (Dseqrecord(-12))</span>
<span class="go">        └─╼ fragment3 (Dseqrecord(-13))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.linearize">
<span class="sig-name descname"><span class="pre">linearize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.linearize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.linearize" title="Link to this definition">#</a></dt>
<dd><p>Similar to <cite>:func:cut</cite>.</p>
<p>Throws an exception if there is not excactly one cut
i.e. none or more than one digestion products.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.looped">
<span class="sig-name descname"><span class="pre">looped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.looped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.looped" title="Link to this definition">#</a></dt>
<dd><p>Circular version of the Dseqrecord object.</p>
<p>The underlying linear Dseq object has to have compatible ends.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.looped" title="pydna.dseq.Dseq.looped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.looped</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.lower">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.lower" title="Link to this definition">#</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">upper</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">upper</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">AAA</span>
<span class="go">TTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lower</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lower</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseqrecord object in lowercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.upper" title="pydna.dseqrecord.Dseqrecord.upper"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.upper</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.m">
<span class="sig-name descname"><span class="pre">m</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.m"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.m" title="Link to this definition">#</a></dt>
<dd><p>This method returns the mass of the DNA molecule in grams. This is
calculated as the product between the molecular weight of the Dseq object
and the</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.map_trace_files">
<span class="sig-name descname"><span class="pre">map_trace_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.map_trace_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.map_trace_files" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.n_cutters">
<span class="sig-name descname"><span class="pre">n_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.n_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.n_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.no_cutters">
<span class="sig-name descname"><span class="pre">no_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.no_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.no_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.number_of_cuts">
<span class="sig-name descname"><span class="pre">number_of_cuts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.number_of_cuts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.number_of_cuts" title="Link to this definition">#</a></dt>
<dd><p>The number of cuts by digestion with the Restriction enzymes
contained in the iterable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.once_cutters">
<span class="sig-name descname"><span class="pre">once_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.once_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.once_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.orfs">
<span class="sig-name descname"><span class="pre">orfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.orfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.orfs" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.orfs_to_features">
<span class="sig-name descname"><span class="pre">orfs_to_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.orfs_to_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.orfs_to_features" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.all.Dseqrecord.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.seguid" title="Link to this definition">#</a></dt>
<dd><p>Url safe SEGUID for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be part of a URL.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;ldseguid=TEwydy0ugvGXh3VJnVwgtxoyDQA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.shifted">
<span class="sig-name descname"><span class="pre">shifted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.shifted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.shifted" title="Link to this definition">#</a></dt>
<dd><p>Circular Dseqrecord with a new origin &lt;shift&gt;.</p>
<p>This only works on circular Dseqrecords. If we consider the following
circular sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">GAAAT</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">watson</span> <span class="pre">strand</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">CTTTA</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">crick</span> <span class="pre">strand</span></code></div>
</div>
<p>The T and the G on the watson strand are linked together as well
as the A and the C of the of the crick strand.</p>
<p>if <code class="docutils literal notranslate"><span class="pre">shift</span></code> is 1, this indicates a new origin at position 1:</p>
<div class="line-block">
<div class="line-block">
<div class="line">new origin at the | symbol:</div>
<div class="line"><br /></div>
</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">G|AAAT</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">C|TTTA</span></code></div>
</div>
<p>new sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">AAATG</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">TTTAC</span></code></div>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aaat</span>
<span class="go">ttta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">shifted</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aata</span>
<span class="go">ttat</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.source">
<span class="sig-name descname"><span class="pre">source</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#pydna.opencloning_models.Source" title="pydna.opencloning_models.Source"><span class="pre">Source</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#pydna.all.Dseqrecord.source" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.synced">
<span class="sig-name descname"><span class="pre">synced</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.synced"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.synced" title="Link to this definition">#</a></dt>
<dd><p>This method returns a new circular sequence (Dseqrecord object), which has been rotated
in such a way that there is maximum overlap between the sequence and
ref, which may be a string, Biopython Seq, SeqRecord object or
another Dseqrecord object.</p>
<p>The reason for using this could be to rotate a new recombinant plasmid so
that it starts at the same position after cloning. See the example below:</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;gaat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">gaat</span>
<span class="go">ctta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">a</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-4)</span>
<span class="go">atga</span>
<span class="go">tact</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;CCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span> <span class="o">=</span> <span class="p">(</span><span class="n">d</span><span class="o">+</span><span class="n">insert</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">atgaCCC</span>
<span class="go">tactGGG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">synced</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">gaCCCat</span>
<span class="go">ctGGGta</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.terminal_transferase">
<span class="sig-name descname"><span class="pre">terminal_transferase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.terminal_transferase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.terminal_transferase" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.tolinear">
<span class="sig-name descname"><span class="pre">tolinear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.tolinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.tolinear" title="Link to this definition">#</a></dt>
<dd><p>Returns a linear, blunt copy of a circular Dseqrecord object. The
underlying Dseq object has to be circular.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="p">[:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.twice_cutters">
<span class="sig-name descname"><span class="pre">twice_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.twice_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.twice_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.unique_cutters">
<span class="sig-name descname"><span class="pre">unique_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.unique_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.unique_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.upper">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.upper" title="Link to this definition">#</a></dt>
<dd><p>Returns an uppercase copy.
&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord
&gt;&gt;&gt; my_seq = Dseqrecord(“aAa”)
&gt;&gt;&gt; my_seq.seq
Dseq(-3)
aAa
tTt
&gt;&gt;&gt; upper = my_seq.upper()
&gt;&gt;&gt; upper.seq
Dseq(-3)
AAA
TTT
&gt;&gt;&gt;</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseqrecord object in uppercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.lower" title="pydna.dseqrecord.Dseqrecord.lower"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.lower</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseqrecord.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gb'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseqrecord.write" title="Link to this definition">#</a></dt>
<dd><p>Writes the Dseqrecord to a file using the format f, which must
be a format supported by Biopython SeqIO for writing <a class="footnote-reference brackets" href="#id4" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a>. Default
is “gb” which is short for Genbank. Note that Biopython SeqIO reads
more formats than it writes.</p>
<p>Filename is the path to the file where the sequece is to be
written. The filename is optional, if it is not given, the
description property (string) is used together with the format.</p>
<p>If obj is the Dseqrecord object, the default file name will be:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;.&lt;f&gt;</span></code></p>
<p>Where &lt;f&gt; is “gb” by default. If the filename already exists and
AND the sequence it contains is different, a new file name will be
used so that the old file is not lost:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;_NEW.&lt;f&gt;</span></code></p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id8" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">5</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</aside>
</aside>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.all.Dseq">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">Dseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">watson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ovhg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><code class="xref py py-class docutils literal notranslate"><span class="pre">Seq</span></code></a></p>
<p>Dseq holds information for a double stranded DNA fragment.</p>
<p>Dseq also holds information describing the topology of
the DNA fragment (linear or circular).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>watson</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a string representing the watson (sense) DNA strand.</p></li>
<li><p><strong>crick</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><em>optional</em>) – a string representing the crick (antisense) DNA strand.</p></li>
<li><p><strong>ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – A positive or negative number to describe the stagger between the
watson and crick strands.
see below for a detailed explanation.</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True indicates that sequence is linear, False that it is circular.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True indicates that sequence is circular, False that it is linear.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Dseq is a subclass of the Biopython Seq object. It stores two
strings representing the watson (sense) and crick(antisense) strands.
two properties called linear and circular, and a numeric value ovhg
(overhang) describing the stagger for the watson and crick strand
in the 5’ end of the fragment.</p>
<p>The most common usage is probably to create a Dseq object as a
part of a Dseqrecord object (see <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a>).</p>
<p>There are three ways of creating a Dseq object directly listed below, but you can also
use the function Dseq.from_full_sequence_and_overhangs() to create a Dseq:</p>
<p>Only one argument (string):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>The given string will be interpreted as the watson strand of a
blunt, linear double stranded sequence object. The crick strand
is created automatically from the watson strand.</p>
<p>Two arguments (string, string):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gggaaat&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">gggaaat</span>
<span class="go">   ttt</span>
</pre></div>
</div>
<p>If both watson and crick are given, but not ovhg an attempt
will be made to find the best annealing between the strands.
There are limitations to this. For long fragments it is quite
slow. The length of the annealing sequences have to be at least
half the length of the shortest of the strands.</p>
<p>Three arguments (string, string, ovhg=int):</p>
<p>The ovhg parameter is an integer describing the length of the
crick strand overhang in the 5’ end of the molecule.</p>
<p>The ovhg parameter controls the stagger at the five prime end:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dsDNA</span>       <span class="n">overhang</span>

  <span class="n">nnn</span><span class="o">...</span>    <span class="mi">2</span>
<span class="n">nnnnn</span><span class="o">...</span>

 <span class="n">nnnn</span><span class="o">...</span>    <span class="mi">1</span>
<span class="n">nnnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>    <span class="mi">0</span>
<span class="n">nnnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>   <span class="o">-</span><span class="mi">1</span>
 <span class="n">nnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>   <span class="o">-</span><span class="mi">2</span>
  <span class="n">nnn</span><span class="o">...</span>
</pre></div>
</div>
<p>Example of creating Dseq objects with different amounts of stagger:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span> <span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">agt</span>
<span class="go"> attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go"> agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">  agt</span>
<span class="go">attca</span>
</pre></div>
</div>
<p>If the ovhg parameter is specified a crick strand also
needs to be supplied, otherwise an exception is raised.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna_/dsdna.py&quot;</span>, line <span class="m">169</span>, in <span class="n">__init__</span>
<span class="w">    </span><span class="k">else</span><span class="p">:</span>
<span class="gr">ValueError</span>: <span class="n">ovhg defined without crick strand!</span>
</pre></div>
</div>
<p>The shape of the fragment is set by circular = True, False</p>
<p>Note that both ends of the DNA fragment has to be compatible to set
circular = True.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span> <span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;tttcccc&quot;</span><span class="p">,</span><span class="s2">&quot;aaacccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-11)</span>
<span class="go">    tttcccc</span>
<span class="go">ccccaaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">4</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ccccttt&quot;</span><span class="p">,</span><span class="s2">&quot;ccccaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-11)</span>
<span class="go">ccccttt</span>
<span class="go">    aaacccc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">-4</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>Coercing to string</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">&#39;ggggtttcccc&#39;</span>
</pre></div>
</div>
<p>A Dseq object can be longer that either the watson or crick strands.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>

<span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
      <span class="n">GATCCTTT</span>
<span class="n">AAAGCCCTA</span>
</pre></div>
</div>
<p>The slicing of a linear Dseq object works mostly as it does for a string.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;at&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;gac&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-1)</span>
<span class="go">a</span>
<span class="go">t</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">Dseq(-2)</span>
<span class="go">at</span>
<span class="go">ta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Dseq(-0)</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">Dseq(-3)</span>
<span class="go">gac</span>
<span class="go">ctg</span>
</pre></div>
</div>
<p>The slicing of a circular Dseq object has a slightly different meaning.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggAtCc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggAtCc</span>
<span class="go">ccTaGg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-5)</span>
<span class="go">CcggA</span>
<span class="go">GgccT</span>
</pre></div>
</div>
<p>The slice [X:X] produces an empty slice for a string, while this
will return the linearized sequence starting at X:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggatcc</span>
<span class="go">cctagg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-6)</span>
<span class="go">tccgga</span>
<span class="go">aggcct</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.T4">
<span class="sig-name descname"><span class="pre">T4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.T4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.T4" title="Link to this definition">#</a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.apply_cut">
<span class="sig-name descname"><span class="pre">apply_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.apply_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.apply_cut" title="Link to this definition">#</a></dt>
<dd><p>Extracts a subfragment of the sequence between two cuts.</p>
<p>For more detail see the documentation of get_cutsite_pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left_cut</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>, </em><em>None</em><em>]</em>)</p></li>
<li><p><strong>right_cut</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>, </em><em>None</em><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.all.Dseq" title="pydna.all.Dseq">Dseq</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaaGAATTCaa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">p3</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span>
<span class="go">(None, ((3, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p1</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">aaG</span>
<span class="go">ttCTTAA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span>
<span class="go">(((3, -4), EcoRI), ((11, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p2</span><span class="p">)</span>
<span class="go">Dseq(-12)</span>
<span class="go">AATTCaaG</span>
<span class="go">    GttCTTAA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p3</span>
<span class="go">(((11, -4), EcoRI), None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p3</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">AATTCaa</span>
<span class="go">    Gtt</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;TTCaaGAA&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((6, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair</span>
<span class="go">(((6, -4), EcoRI), ((6, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">pair</span><span class="p">)</span>
<span class="go">Dseq(-12)</span>
<span class="go">AATTCaaG</span>
<span class="go">    GttCTTAA</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.cas9">
<span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RNA</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#slice" title="(in Python v3.8)"><span class="pre">slice</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.cas9" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.cut">
<span class="sig-name descname"><span class="pre">cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">EnzymesType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><span class="pre">DseqType</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.cut" title="Link to this definition">#</a></dt>
<dd><p>Returns a list of linear Dseq fragments produced in the digestion.
If there are no cuts, an empty list is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzymes</strong> (<em>enzyme object</em><em> or </em><em>iterable</em><em> of </em><em>such objects</em>) – A Bio.Restriction.XXX restriction objects or iterable.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>frags</strong> – list of Dseq objects formed by the digestion</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ggatccnnngaattc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">BamHI</span><span class="p">,</span><span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">))</span>
<span class="go">&lt;class &#39;tuple&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">):</span> <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">frag</span><span class="p">))</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="go">Dseq(-14)</span>
<span class="go">gatccnnngaattc</span>
<span class="go">    gnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span> <span class="o">==</span>  <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">,</span> <span class="n">EcoRI</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.cutsite_is_valid">
<span class="sig-name descname"><span class="pre">cutsite_is_valid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cutsite_is_valid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.cutsite_is_valid" title="Link to this definition">#</a></dt>
<dd><p>Returns False if:
- Cut positions fall outside the sequence (could be moved to Biopython)
- Overhang is not double stranded
- Recognition site is not double stranded or is outside the sequence
- For enzymes that cut twice, it checks that at least one possibility is valid</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.cutters">
<span class="sig-name descname"><span class="pre">cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence at least once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.exo1_end">
<span class="sig-name descname"><span class="pre">exo1_end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.exo1_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.exo1_end" title="Link to this definition">#</a></dt>
<dd><p>5’-3’ resection at the end (right side) of the molecule.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.exo1_front">
<span class="sig-name descname"><span class="pre">exo1_front</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.exo1_front"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.exo1_front" title="Link to this definition">#</a></dt>
<dd><p>5’-3’ resection at the start (left side) of the molecule.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.fill_in">
<span class="sig-name descname"><span class="pre">fill_in</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.fill_in"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.fill_in" title="Link to this definition">#</a></dt>
<dd><p>Fill in of five prime protruding end with a DNA polymerase
that has only DNA polymerase activity (such as exo-klenow <a class="footnote-reference brackets" href="#id6" id="id9" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>)
and any combination of A, G, C or T. Default are all four
nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;tac&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id10" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">6</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment">http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sub</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_SeqAbstractBaseClass</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_sys.maxsize</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.find" title="Link to this definition">#</a></dt>
<dd><p>This method behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<p>Returns -1 if the subsequence is NOT found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sub</strong> (<em>string</em><em> or </em><em>Seq object</em>) – a string or another Seq object to look for.</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – slice start.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – slice end.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;atcgactgacgtgtt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">atcgactgacgtgtt</span>
<span class="go">tagctgactgcacaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;gac&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;taa&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.five_prime_end">
<span class="sig-name descname"><span class="pre">five_prime_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.five_prime_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.five_prime_end" title="Link to this definition">#</a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.three_prime_end" title="pydna.dseq.Dseq.three_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.three_prime_end</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.from_full_sequence_and_overhangs">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_full_sequence_and_overhangs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">full_sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick_ovhg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">watson_ovhg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.from_full_sequence_and_overhangs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.from_full_sequence_and_overhangs" title="Link to this definition">#</a></dt>
<dd><p>Create a linear Dseq object from a full sequence and the 3’ overhangs of each strand.</p>
<p>The order of the parameters is like this because the 3’ overhang of the crick strand is the one
on the left side of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>full_sequence</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The full sequence of the Dseq object.</p></li>
<li><p><strong>crick_ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The overhang of the crick strand in the 3’ end. Equivalent to Dseq.ovhg.</p></li>
<li><p><strong>watson_ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The overhang of the watson strand in the 5’ end.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Dseq object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.all.Dseq" title="pydna.all.Dseq">Dseq</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">  AAAA</span>
<span class="go">TTTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">AAAAAA</span>
<span class="go">  TT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">  AA</span>
<span class="go">TTTTTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">AAAA</span>
<span class="go">  TTTT</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.from_representation">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_representation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dsdna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.from_representation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.from_representation" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.from_string">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.from_string" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.get_cut_parameters">
<span class="sig-name descname"><span class="pre">get_cut_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_left</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.get_cut_parameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.get_cut_parameters" title="Link to this definition">#</a></dt>
<dd><p>For a given cut expressed as ((cut_watson, ovhg), enz), returns
a tuple (cut_watson, cut_crick, ovhg).</p>
<ul class="simple">
<li><p>cut_watson: see get_cutsites docs</p></li>
<li><p>cut_crick: equivalent of cut_watson in the crick strand</p></li>
<li><p>ovhg: see get_cutsites docs</p></li>
</ul>
<p>The cut can be None if it represents the left or right end of the sequence.
Then it will return the position of the watson and crick ends with respect
to the “full sequence”. The <cite>is_left</cite> parameter is only used in this case.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.get_cutsite_pairs">
<span class="sig-name descname"><span class="pre">get_cutsite_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.get_cutsite_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.get_cutsite_pairs" title="Link to this definition">#</a></dt>
<dd><p>Returns pairs of cutsites that render the edges of the resulting fragments.</p>
<p>A fragment produced by restriction is represented by a tuple of length 2 that
may contain cutsites or <cite>None</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><p>Two cutsites: represents the extraction of a fragment between those two
cutsites, in that orientation. To represent the opening of a circular
molecule with a single cutsite, we put the same cutsite twice.</p></li>
<li><p><cite>None</cite>, cutsite: represents the extraction of a fragment between the left
edge of linear sequence and the cutsite.</p></li>
<li><p>cutsite, <cite>None</cite>: represents the extraction of a fragment between the cutsite
and the right edge of a linear sequence.</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cutsites</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]|None],<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]|None]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaaGAATTCaa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="go">[(None, ((3, -4), EcoRI)), (((3, -4), EcoRI), ((11, -4), EcoRI)), (((11, -4), EcoRI), None)]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;TTCaaGAA&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((6, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="go">[(((6, -4), EcoRI), ((6, -4), EcoRI))]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.get_cutsites">
<span class="sig-name descname"><span class="pre">get_cutsites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">EnzymesType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.get_cutsites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.get_cutsites" title="Link to this definition">#</a></dt>
<dd><p>Returns a list of cutsites, represented represented as <cite>((cut_watson, ovhg), enz)</cite>:</p>
<ul class="simple">
<li><p><cite>cut_watson</cite> is a positive integer contained in <cite>[0,len(seq))</cite>, where <cite>seq</cite> is the sequence
that will be cut. It represents the position of the cut on the watson strand, using the full
sequence as a reference. By “full sequence” I mean the one you would get from <cite>str(Dseq)</cite>.</p></li>
<li><p><cite>ovhg</cite> is the overhang left after the cut. It has the same meaning as <cite>ovhg</cite> in
the <cite>Bio.Restriction</cite> enzyme objects, or pydna’s <cite>Dseq</cite> property.</p></li>
<li><dl class="simple">
<dt><cite>enz</cite> is the enzyme object. It’s not necessary to perform the cut, but can be</dt><dd><p>used to keep track of which enzyme was used.</p>
</dd>
</dl>
</li>
</ul>
<p>Cuts are only returned if the recognition site and overhang are on the double-strand
part of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzymes</strong> (<em>Union</em><em>[</em><em>_RestrictionBatch</em><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_AbstractCut</em><em>]</em><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;AAGAATTCAAGAATTC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">)</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
</pre></div>
</div>
<p><cite>cut_watson</cite> is defined with respect to the “full sequence”, not the
watson strand:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaa&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span>
<span class="go">Dseq(-10)</span>
<span class="go"> aGAATTCaa</span>
<span class="go">ttCTTAAGtt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[((3, -4), EcoRI)]</span>
</pre></div>
</div>
<p>Cuts are only returned if the recognition site and overhang are on the double-strand
part of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[((1, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.isblunt">
<span class="sig-name descname"><span class="pre">isblunt</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.isblunt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.isblunt" title="Link to this definition">#</a></dt>
<dd><p>isblunt.</p>
<p>Return True if Dseq is linear and blunt and
false if staggered or circular.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">gat</span>
<span class="go">cta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="s2">&quot;atcg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> gat</span>
<span class="go">gcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="s2">&quot;gatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">gat</span>
<span class="go">ctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o3)</span>
<span class="go">gat</span>
<span class="go">cta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.left_end_position">
<span class="sig-name descname"><span class="pre">left_end_position</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.left_end_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.left_end_position" title="Link to this definition">#</a></dt>
<dd><p>The index in the full sequence of the watson and crick start positions.</p>
<p>full sequence (str(self)) for all three cases is AAA</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AAA</span>              <span class="n">AA</span>               <span class="n">AAT</span>
 <span class="n">TT</span>             <span class="n">TTT</span>               <span class="n">TTT</span>
<span class="n">Returns</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="n">Returns</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>    <span class="n">Returns</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.looped">
<span class="sig-name descname"><span class="pre">looped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.looped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.looped" title="Link to this definition">#</a></dt>
<dd><p>Circularized Dseq object.</p>
<p>This can only be done if the two ends are compatible,
otherwise a TypeError is raised.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o7)</span>
<span class="go">catcgat</span>
<span class="go">gtagcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;</span>, line <span class="m">357</span>, in <span class="n">looped</span>
<span class="w">    </span><span class="k">if</span> <span class="n">type5</span> <span class="o">==</span> <span class="n">type3</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky5</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)):</span>
<span class="gr">TypeError</span>: <span class="n">DNA cannot be circularized.</span>
<span class="x">5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.lower">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.lower" title="Link to this definition">#</a></dt>
<dd><p>Return a lower case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseq object in lowercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.all.Dseq" title="pydna.all.Dseq">Dseq</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.upper" title="pydna.dseq.Dseq.upper"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.upper</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.mung">
<span class="sig-name descname"><span class="pre">mung</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.mung"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.mung" title="Link to this definition">#</a></dt>
<dd><p>Simulates treatment a nuclease with 5’-3’ and 3’-5’ single
strand specific exonuclease activity (such as mung bean nuclease <a class="footnote-reference brackets" href="#id8" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="n">ggatcc</span>    <span class="o">-&gt;</span>     <span class="n">gatcc</span>
     <span class="n">ctaggg</span>          <span class="n">ctagg</span>

     <span class="n">ggatcc</span>   <span class="o">-&gt;</span>      <span class="n">ggatc</span>
    <span class="n">tcctag</span>            <span class="n">cctag</span>

<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
<span class="n">caaa</span>
 <span class="n">tttc</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
 <span class="n">aaac</span>
<span class="n">gttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id12" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">7</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/Mung_bean_nuclease">http://en.wikipedia.org/wiki/Mung_bean_nuclease</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.mw">
<span class="sig-name descname"><span class="pre">mw</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.mw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.mw" title="Link to this definition">#</a></dt>
<dd><p>This method returns the molecular weight of the DNA molecule
in g/mol. The following formula is used:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MW</span> <span class="o">=</span> <span class="p">(</span><span class="n">A</span> <span class="n">x</span> <span class="mf">313.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">T</span> <span class="n">x</span> <span class="mf">304.2</span><span class="p">)</span> <span class="o">+</span>
     <span class="p">(</span><span class="n">C</span> <span class="n">x</span> <span class="mf">289.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">G</span> <span class="n">x</span> <span class="mf">329.2</span><span class="p">)</span> <span class="o">+</span>
     <span class="p">(</span><span class="n">N</span> <span class="n">x</span> <span class="mf">308.9</span><span class="p">)</span> <span class="o">+</span> <span class="mf">79.0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.n_cutters">
<span class="sig-name descname"><span class="pre">n_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.n_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.n_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting n times.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.no_cutters">
<span class="sig-name descname"><span class="pre">no_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.no_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.no_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch not cutting sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.once_cutters">
<span class="sig-name descname"><span class="pre">once_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="headerlink" href="#pydna.all.Dseq.once_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.quick">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">quick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">watson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">ovhg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.quick"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.quick" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="headerlink" href="#pydna.all.Dseq.rc" title="Link to this definition">#</a></dt>
<dd><p>Dseq object where watson and crick have switched places.</p>
<p>This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Dseq object where watson and crick have switched places.</p>
<p>This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.right_end_position">
<span class="sig-name descname"><span class="pre">right_end_position</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.right_end_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.right_end_position" title="Link to this definition">#</a></dt>
<dd><p>The index in the full sequence of the watson and crick end positions.</p>
<p>full sequence (str(self)) for all three cases is AAA</p>
<p><code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">AAA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AAA</span>
<span class="pre">TT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TTT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TTT</span>
<span class="pre">Returns</span> <span class="pre">(3,</span> <span class="pre">2)</span>&#160;&#160;&#160; <span class="pre">Returns</span> <span class="pre">(2,</span> <span class="pre">3)</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Returns</span> <span class="pre">(3,</span> <span class="pre">3)</span>
<span class="pre">`</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.seguid" title="Link to this definition">#</a></dt>
<dd><p>SEGUID checksum for the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.shifted">
<span class="sig-name descname"><span class="pre">shifted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.shifted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.shifted" title="Link to this definition">#</a></dt>
<dd><p>Shifted version of a circular Dseq object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.t4">
<span class="sig-name descname"><span class="pre">t4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="headerlink" href="#pydna.all.Dseq.t4" title="Link to this definition">#</a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.terminal_transferase">
<span class="sig-name descname"><span class="pre">terminal_transferase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.terminal_transferase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.terminal_transferase" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.three_prime_end">
<span class="sig-name descname"><span class="pre">three_prime_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.three_prime_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.three_prime_end" title="Link to this definition">#</a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.five_prime_end" title="pydna.dseq.Dseq.five_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.five_prime_end</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.tolinear">
<span class="sig-name descname"><span class="pre">tolinear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.tolinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.tolinear" title="Link to this definition">#</a></dt>
<dd><p>Returns a blunt, linear copy of a circular Dseq object. This can
only be done if the Dseq object is circular, otherwise a
TypeError is raised.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[:]</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.transcribe">
<span class="sig-name descname"><span class="pre">transcribe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><span class="pre">Seq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.transcribe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.transcribe" title="Link to this definition">#</a></dt>
<dd><p>Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.</p>
<p>Following the usual convention, the sequence is interpreted as the
coding strand of the DNA double helix, not the template strand. This
means we can get the RNA sequence just by switching T to U.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">transcribe</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
<p>Trying to transcribe an RNA sequence has no effect.
If you have a nucleotide sequence which might be DNA or RNA
(or even a mixture), calling the transcribe method will ensure
any T becomes U.</p>
<p>Trying to transcribe a protein sequence will replace any
T for Threonine with U for Selenocysteine, which has no
biologically plausible rational.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRU&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><span class="pre">Seq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.translate" title="Link to this definition">#</a></dt>
<dd><p>Translate..</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.all.Dseq.trunc">
<span class="sig-name descname"><span class="pre">trunc</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">30</span></em><a class="headerlink" href="#pydna.all.Dseq.trunc" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.twice_cutters">
<span class="sig-name descname"><span class="pre">twice_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.twice_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.twice_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence twice.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.unique_cutters">
<span class="sig-name descname"><span class="pre">unique_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.unique_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.unique_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.upper">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.upper" title="Link to this definition">#</a></dt>
<dd><p>Return an upper case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">AAA</span>
<span class="go">TTT</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseq object in uppercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.all.Dseq" title="pydna.all.Dseq">Dseq</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.lower" title="pydna.dseq.Dseq.lower"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.lower</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.all.Dseq.watson_ovhg">
<span class="sig-name descname"><span class="pre">watson_ovhg</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.watson_ovhg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.Dseq.watson_ovhg" title="Link to this definition">#</a></dt>
<dd><p>Returns the overhang of the watson strand at the three prime.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.read">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/readers.html#read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.read" title="Link to this definition">#</a></dt>
<dd><p>This function is similar the <a class="reference internal" href="#pydna.all.parse" title="pydna.all.parse"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code></a> function but expects one and only
one sequence or and exception is thrown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>string</em>) – see below</p></li>
<li><p><strong>ds</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Double stranded or single stranded DNA, if True return
Dseqrecord objects, else Bio.SeqRecord objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>contains the first Dseqrecord or SeqRecord object parsed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The data parameter is similar to the data parameter for <a class="reference internal" href="#pydna.all.parse" title="pydna.all.parse"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code></a>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.all.parse" title="pydna.all.parse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.read_primer">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">read_primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/readers.html#read_primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.read_primer" title="Link to this definition">#</a></dt>
<dd><p>Use this function to read a primer sequence from a string or a local file.
The usage is similar to the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse_primer()</span></code> function.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.parse">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/parsers.html#parse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.parse" title="Link to this definition">#</a></dt>
<dd><p>Return <em>all</em> DNA sequences found in data.</p>
<p>If no sequences are found, an empty list is returned. This is a greedy
function, use carefully.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>string</em><em> or </em><em>iterable</em>) – <p>The data parameter is a string containing:</p>
<ol class="arabic simple">
<li><p>an absolute path to a local file.
The file will be read in text
mode and parsed for EMBL, FASTA
and Genbank sequences. Can be
a string or a Path object.</p></li>
<li><p>a string containing one or more
sequences in EMBL, GENBANK,
or FASTA format. Mixed formats
are allowed.</p></li>
<li><p>data can be a list or other iterable where the elements are 1 or 2</p></li>
</ol>
</p></li>
<li><p><strong>ds</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – If True double stranded <a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a> objects are returned.
If False single stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> <a class="footnote-reference brackets" href="#id10" id="id13" role="doc-noteref"><span class="fn-bracket">[</span>6<span class="fn-bracket">]</span></a> objects are
returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>contains Dseqrecord or SeqRecord objects</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id14" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">8</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</aside>
</aside>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.all.read" title="pydna.all.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.parse_primers">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">parse_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/parsers.html#parse_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.parse_primers" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.primer_design">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">primer_design</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">55.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">estimate_function</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/design.html#primer_design"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.primer_design" title="Link to this definition">#</a></dt>
<dd><p>This function designs a forward primer and a reverse primer for PCR amplification
of a given template sequence.</p>
<p>The template argument is a Dseqrecord object or equivalent containing the template sequence.</p>
<p>The optional fp and rp arguments can contain an existing primer for the sequence (either the forward or reverse primer).
One or the other primers can be specified, not both (since then there is nothing to design!, use the pydna.amplify.pcr function instead).</p>
<p>The limit argument is the minimum length of the primer. The default value is 13.</p>
<p>If one of the primers is given, the other primer is designed to match in terms of Tm.
If both primers are designed, they will be designed to target_tm</p>
<p>tm_func is a function that takes an ascii string representing an oligonuceotide as argument and returns a float.
Some useful functions can be found in the <a class="reference internal" href="#module-pydna.tm" title="pydna.tm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.tm</span></code></a> module, but can be substituted for a custom made function.</p>
<p>estimate_function is a tm_func-like function that is used to get a first guess for the primer design, that is then used as starting
point for the final result. This is useful when the tm_func function is slow to calculate (e.g. it relies on an
external API, such as the NEB primer design API). The estimate_function should be faster than the tm_func function.
The default value is <cite>None</cite>.
To use the default <cite>tm_func</cite> as estimate function to get the NEB Tm faster, you can do:
<cite>primer_design(dseqr, target_tm=55, tm_func=tm_neb, estimate_function=tm_default)</cite>.</p>
<p>The function returns a pydna.amplicon.Amplicon class instance. This object has
the object.forward_primer and object.reverse_primer properties which contain the designed primers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><em>pydna.dseqrecord.Dseqrecord</em></a>) – a Dseqrecord object. The only required argument.</p></li>
<li><p><strong>fp</strong> (<a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><em>pydna.primer.Primer</em></a><em>, </em><em>optional</em>) – optional pydna.primer.Primer objects containing one primer each.</p></li>
<li><p><strong>rp</strong> (<a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><em>pydna.primer.Primer</em></a><em>, </em><em>optional</em>) – optional pydna.primer.Primer objects containing one primer each.</p></li>
<li><p><strong>target_tm</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a><em>, </em><em>optional</em>) – target tm for the primers, set to 55°C by default.</p></li>
<li><p><strong>tm_func</strong> (<em>function</em>) – Function used for tm calculation. This function takes an ascii string
representing an oligonuceotide as argument and returns a float.
Some useful functions can be found in the <a class="reference internal" href="#module-pydna.tm" title="pydna.tm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.tm</span></code></a> module, but can be
substituted for a custom made function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>result</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon">Amplicon</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">Dseqrecord(-64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.design</span><span class="w"> </span><span class="kn">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span>
<span class="go">Amplicon(64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">f64 17-mer:5&#39;-atgactgctaacccttc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 18-mer:5&#39;-catcgtaagtttcgaacg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ampl</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">5atgactgctaacccttc...cgttcgaaacttacgatg3</span>
<span class="go">                     ||||||||||||||||||</span>
<span class="go">                    3gcaagctttgaatgctac5</span>
<span class="go">5atgactgctaacccttc3</span>
<span class="go"> |||||||||||||||||</span>
<span class="go">3tactgacgattgggaag...gcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">f64 23-mer:5&#39;-GGATCCatgactgct..ttc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">r64 24-mer:5&#39;-GGATCCcatcgtaag..acg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.amplify</span><span class="w"> </span><span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span> <span class="n">pr</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">      5atgactgctaacccttc...cgttcgaaacttacgatg3</span>
<span class="go">                           ||||||||||||||||||</span>
<span class="go">                          3gcaagctttgaatgctacCCTAGG5</span>
<span class="go">5GGATCCatgactgctaacccttc3</span>
<span class="go">       |||||||||||||||||</span>
<span class="go">      3tactgacgattgggaag...gcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGGATCC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttg&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;myprimer&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">fp</span> <span class="o">=</span> <span class="n">pf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">myprimer 27-mer:5&#39;-atgactgctaaccct..ttg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 32-mer:5&#39;-catcgtaagtttcga..atc-3&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.assembly_fragments">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">assembly_fragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlink</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/design.html#assembly_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.assembly_fragments" title="Link to this definition">#</a></dt>
<dd><p>This function return a list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects where
primers have been modified with tails so that the fragments can be fused in
the order they appear in the list by for example Gibson assembly or homologous
recombination.</p>
<p>Given that we have two linear <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects a and b</p>
<p>we can modify the reverse primer of a and forward primer of b with tails to allow
fusion by fusion PCR, Gibson assembly or in-vivo homologous recombination.
The basic requirements for the primers for the three techniques are the same.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>
<span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>
                  <span class="o">|||||</span>
                 <span class="o">&lt;</span><span class="n">gacgt</span>
<span class="n">agcct</span><span class="o">&gt;</span>
<span class="o">|||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                        <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
                       <span class="o">/</span>                     \
                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                                         <span class="o">|||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
                       <span class="n">TTTAT</span><span class="o">&gt;</span>
                       <span class="o">|||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

<span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
\<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>Design tailed primers incorporating a part of the next or previous fragment to be assembled.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATA</span>

<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>

                <span class="n">ctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by any of the techniques mentioned earlier to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing
Amplicons and other similar objects.</p>
<p><strong>At least every second sequence object needs to be an Amplicon</strong></p>
<p>This rule exists because if a sequence object is that is not a PCR product
is to be fused with another fragment, that other fragment needs to be an Amplicon
so that the primer of the other object can be modified to include the whole stretch
of sequence homology needed for the fusion. See the example below where a is a
non-amplicon (a linear plasmid  vector for instance)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>           <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
<span class="o">/</span>                     \         <span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>   <span class="o">&lt;--&gt;</span>  <span class="n">TTTATATCGCATGACTCTTCTTT</span>
<span class="o">|||||||||||||||||||||||</span>         <span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>                          <span class="o">&lt;</span><span class="n">AGAAA</span>
                                <span class="n">TTTAT</span><span class="o">&gt;</span>
                                <span class="o">|||||||||||||||||||||||</span>
                          <span class="o">&lt;--&gt;</span>  <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

     <span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
     <span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
     <span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
     \<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>In this case only the forward primer of b is fitted with a tail with a part a:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
         <span class="n">tcttggtctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by for example Gibson assembly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaacca</span>

                             <span class="n">TCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing
Amplicons and other similar objects.</p>
<p>The overlap argument controls how many base pairs of overlap required between
adjacent sequence fragments. In the junction between Amplicons, tails with the
length of about half of this value is added to the two primers
closest to the junction.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;       &lt;
Amplicon1
         Amplicon2
         &gt;       &lt;

         ⇣

&gt;       &lt;-
Amplicon1
         Amplicon2
        -&gt;       &lt;
</pre></div>
</div>
<p>In the case of an Amplicon adjacent to a Dseqrecord object, the tail will
be twice as long (1*overlap) since the
recombining sequence is present entirely on this primer:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Dseqrecd1
         Amplicon1
         &gt;       &lt;

         ⇣

Dseqrecd1
         Amplicon1
       --&gt;       &lt;
</pre></div>
</div>
<p>Note that if the sequence of DNA fragments starts or stops with an Amplicon,
the very first and very last prinmer will not be modified i.e. assembles are
always assumed to be linear. There are simple tricks around that for circular
assemblies depicted in the last two examples below.</p>
<p>The maxlink arguments controls the cut off length for sequences that will be
synhtesized by adding them to primers for the adjacent fragment(s). The
argument list may contain short spacers (such as spacers between fusion proteins).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Example 1: Linear assembly of PCR products (pydna.amplicon.Amplicon class objects) ------

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon4
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;-       -&gt;       &lt;-                      pydna.assembly.Assembly
Amplicon1         Amplicon3
         Amplicon2         Amplicon4     ➤  Amplicon1Amplicon2Amplicon3Amplicon4
        -&gt;       &lt;-       -&gt;       &lt;

Example 2: Linear assembly of alternating Amplicons and other fragments

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;--     --&gt;       &lt;--                     pydna.assembly.Assembly
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2     ➤  Amplicon1Dseqrecd1Amplicon2Dseqrecd2

Example 3: Linear assembly of alternating Amplicons and other fragments

Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2
         &gt;       &lt;       --&gt;       &lt;

                     ⇣
             pydna.design.assembly_fragments
                     ⇣
                                                  pydna.assembly.Assembly
Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2     ➤  Dseqrecd1Amplicon1Dseqrecd2Amplicon2
       --&gt;       &lt;--     --&gt;       &lt;

Example 4: Circular assembly of alternating Amplicons and other fragments

                 -&gt;       &lt;==
Dseqrecd1         Amplicon2
         Amplicon1         Dseqrecd1
       --&gt;       &lt;-
                     ⇣
                     pydna.design.assembly_fragments
                     ⇣
                                                   pydna.assembly.Assembly
                 -&gt;       &lt;==
Dseqrecd1         Amplicon2                    -Dseqrecd1Amplicon1Amplicon2-
         Amplicon1                       ➤    |                             |
       --&gt;       &lt;-                            -----------------------------

------ Example 5: Circular assembly of Amplicons

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;=       -&gt;       &lt;-
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
        -&gt;       &lt;-       +&gt;       &lt;

                     ⇣
             make new Amplicon using the Amplicon1.template and
             the last fwd primer and the first rev primer.
                     ⇣
                                                   pydna.assembly.Assembly
+&gt;       &lt;=       -&gt;       &lt;-
 Amplicon1         Amplicon3                  -Amplicon1Amplicon2Amplicon3-
          Amplicon2                      ➤   |                             |
         -&gt;       &lt;-                          -----------------------------
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects) – list Amplicon and Dseqrecord object for which fusion primers should be constructed.</p></li>
<li><p><strong>overlap</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Length of required overlap between fragments.</p></li>
<li><p><strong>maxlink</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Maximum length of spacer sequences that may be present in f. These will be included in tails for designed primers.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, the assembly is circular. If False, the assembly is linear.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>seqs</strong> –</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Amplicon1</span><span class="p">,</span>
 <span class="n">Amplicon2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.design</span><span class="w"> </span><span class="kn">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ccaaacccaccaggtaccttatgtaagtacttcaagtcgccagaagacttcttggtcaagttgcc&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtactggtgctgaaccttgtatcaagttgggtgttgacgccattgccccaggtggtcgtttcgtt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.design</span><span class="w"> </span><span class="kn">import</span> <span class="n">assembly_fragments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># We would like a circular recombination, so the first sequence has to be repeated</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa1</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">,</span><span class="n">fa2</span> <span class="o">=</span> <span class="n">assembly_fragments</span><span class="p">([</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">a</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Since all fragments are Amplicons, we need to extract the rp of the 1st and fp of the last fragments.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.amplify</span><span class="w"> </span><span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">fa2</span><span class="o">.</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">fa1</span><span class="o">.</span><span class="n">reverse_primer</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">]</span>
<span class="go">[Amplicon(100), Amplicon(101), Amplicon(102)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">fb</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">fc</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;fa fb fc&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly</span><span class="w"> </span><span class="kn">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">([</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 100bp 101bp 102bp</span>
<span class="go">limit(bp)....: 25</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_linear</span><span class="p">()</span>
<span class="go">[Contig(-231), Contig(-166), Contig(-36)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;cdseguid=85t6tfcvWav0wnXEIb-lkUtrl4s&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;cdseguid=85t6tfcvWav0wnXEIb-lkUtrl4s&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go"> -|fa|36</span>
<span class="go">|     \/</span>
<span class="go">|     /\</span>
<span class="go">|     36|fb|36</span>
<span class="go">|           \/</span>
<span class="go">|           /\</span>
<span class="go">|           36|fc|36</span>
<span class="go">|                 \/</span>
<span class="go">|                 /\</span>
<span class="go">|                 36-</span>
<span class="go">|                    |</span>
<span class="go"> --------------------</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.eq">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">eq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#eq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.eq" title="Link to this definition">#</a></dt>
<dd><p>Compare two or more DNA sequences for equality.</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same double stranded DNA molecule.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>iterable</em>) – iterable containing sequences
args can be strings, Biopython Seq or SeqRecord, Dseqrecord
or dsDNA objects.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Consider all molecules circular or linear</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Consider all molecules circular or linear</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>eq</strong> – Returns True or False</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same DNA molecule.</p>
<p>Two linear sequences are considiered equal if either:</p>
<ol class="arabic simple">
<li><p>They have the same sequence (case insensitive)</p></li>
<li><p>One sequence is the reverse complement of the other</p></li>
</ol>
<p>Two circular sequences are considered equal if they are circular
permutations meaning that they have the same length and:</p>
<ol class="arabic simple">
<li><p>One sequence can be found in the concatenation of the other sequence with itself.</p></li>
<li><p>The reverse complement of one sequence can be found in the concatenation of the other sequence with itself.</p></li>
</ol>
<p>The topology for the comparison can be set using one of the keywords
linear or circular to True or False.</p>
<p>If circular or linear is not set, it will be deduced from the topology of
each sequence for sequences that have a linear or circular attribute
(like Dseq and Dseqrecord).</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">eq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aTaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCCa&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;tGGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.all.gbtext_clean">
<span class="sig-prename descclassname"><span class="pre">pydna.all.</span></span><span class="sig-name descname"><span class="pre">gbtext_clean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gbtext</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#gbtext_clean"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.all.gbtext_clean" title="Link to this definition">#</a></dt>
<dd><p>This function takes a string containing <strong>one</strong> genbank sequence
in Genbank format and returns a named tuple containing two fields,
the gbtext containing a string with the corrected genbank sequence and
jseq which contains the JSON intermediate.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; s = &#39;&#39;&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013
... DEFINITION  .
... ACCESSION
... VERSION
... SOURCE      .
...   ORGANISM  .
... COMMENT
... COMMENT     ApEinfo:methylated:1
... ORIGIN
...         1 aaa
... //&#39;&#39;&#39;
&gt;&gt;&gt; from pydna.readers import read
&gt;&gt;&gt; read(s)
... /site-packages/Bio/GenBank/Scanner.py:1388: BiopythonParserWarning: Malformed LOCUS line found - is this correct?
:&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013\n&#39;
  &quot;correct?\n:%r&quot; % line, BiopythonParserWarning)
Traceback (most recent call last):
  File &quot;... /pydna/readers.py&quot;, line 48, in read
    results = results.pop()
IndexError: pop from empty list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File &quot;&lt;stdin&gt;&quot;, line 1, in &lt;module&gt;
  File &quot;... /pydna/readers.py&quot;, line 50, in read
    raise ValueError(&quot;No sequences found in data:\n({})&quot;.format(data[:79]))
ValueError: No sequences found in data:
(LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013
DEFINITI)
&gt;&gt;&gt; from pydna.genbankfixer import gbtext_clean
&gt;&gt;&gt; s2, j2 = gbtext_clean(s)
&gt;&gt;&gt; print(s2)
LOCUS       New_DNA                    3 bp ds-DNA     circular SYN 19-JUN-2013
DEFINITION  .
ACCESSION
VERSION
SOURCE      .
ORGANISM  .
COMMENT
COMMENT     ApEinfo:methylated:1
FEATURES             Location/Qualifiers
ORIGIN
        1 aaa
//
&gt;&gt;&gt; s3 = read(s2)
&gt;&gt;&gt; s3
Dseqrecord(o3)
&gt;&gt;&gt; print(s3.format())
LOCUS       New_DNA                    3 bp    DNA     circular SYN 19-JUN-2013
DEFINITION  .
ACCESSION   New_DNA
VERSION     New_DNA
KEYWORDS    .
SOURCE
  ORGANISM  .
            .
COMMENT
            ApEinfo:methylated:1
FEATURES             Location/Qualifiers
ORIGIN
        1 aaa
//
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.amplicon">
<span id="pydna-amplicon-module"></span><h2>pydna.amplicon module<a class="headerlink" href="#module-pydna.amplicon" title="Link to this heading">#</a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Amplicon</span></code></a> class for PCR simulation.
This class is not meant to be use directly but is
used by the <code class="xref py py-mod docutils literal notranslate"><span class="pre">amplify</span></code> module</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.amplicon.</span></span><span class="sig-name descname"><span class="pre">Amplicon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_primer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<p>The Amplicon class holds information about a PCR reaction involving two
primers and one template. This class is used by the Anneal class and is not
meant to be instantiated directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_primer</strong> (<a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>(</em><em>Biopython</em><em>)</em>) – SeqRecord object holding the forward (sense) primer</p></li>
<li><p><strong>reverse_primer</strong> (<a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>(</em><em>Biopython</em><em>)</em>) – SeqRecord object holding the reverse (antisense) primer</p></li>
<li><p><strong>template</strong> (<a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><em>Dseqrecord</em></a>) – Dseqrecord object holding the template (circular or linear)</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.from_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.from_SeqRecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.amplicon.Amplicon.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.figure" title="Link to this definition">#</a></dt>
<dd><p>This method returns a simple figure of the two primers binding
to a part of the template.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">5</span><span class="n">tacactcaccgtctatcattatc</span><span class="o">...</span><span class="n">cgactgtatcatctgatagcac3</span>
                           <span class="o">||||||||||||||||||||||</span>
                          <span class="mi">3</span><span class="n">gctgacatagtagactatcgtg5</span>
<span class="mi">5</span><span class="n">tacactcaccgtctatcattatc3</span>
 <span class="o">|||||||||||||||||||||||</span>
<span class="mi">3</span><span class="n">atgtgagtggcagatagtaatag</span><span class="o">...</span><span class="n">gctgacatagtagactatcgtg5</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>figure</strong> – A string containing a text representation of the primers
annealing on the template (see example above).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.set_forward_primer_footprint">
<span class="sig-name descname"><span class="pre">set_forward_primer_footprint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.set_forward_primer_footprint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.set_forward_primer_footprint" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.set_reverse_primer_footprint">
<span class="sig-name descname"><span class="pre">set_reverse_primer_footprint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.set_reverse_primer_footprint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.set_reverse_primer_footprint" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.program">
<span class="sig-name descname"><span class="pre">program</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.program" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.dbd_program">
<span class="sig-name descname"><span class="pre">dbd_program</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.dbd_program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.dbd_program" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplicon.Amplicon.primers">
<span class="sig-name descname"><span class="pre">primers</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplicon.html#Amplicon.primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.primers" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-pydna.amplify">
<span id="pydna-amplify-module"></span><h2>pydna.amplify module<a class="headerlink" href="#module-pydna.amplify" title="Link to this heading">#</a></h2>
<p>This module provide the <a class="reference internal" href="#pydna.amplify.Anneal" title="pydna.amplify.Anneal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Anneal</span></code></a> class and the <a class="reference internal" href="#pydna.amplify.pcr" title="pydna.amplify.pcr"><code class="xref py py-func docutils literal notranslate"><span class="pre">pcr()</span></code></a> function
for PCR simulation. The pcr function is simpler to use, but expects only one
PCR product. The Anneal class should be used if more flexibility is required.</p>
<p>Primers with 5’ tails as well as inverse PCR on circular templates are handled
correctly.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.amplify.Anneal">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.amplify.</span></span><span class="sig-name descname"><span class="pre">Anneal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/amplify.html#Anneal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplify.Anneal" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The Anneal class has the following important attributes:</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.forward_primers">
<span class="sig-name descname"><span class="pre">forward_primers</span></span><a class="headerlink" href="#pydna.amplify.Anneal.forward_primers" title="Link to this definition">#</a></dt>
<dd><p>Description of <cite>forward_primers</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.reverse_primers">
<span class="sig-name descname"><span class="pre">reverse_primers</span></span><a class="headerlink" href="#pydna.amplify.Anneal.reverse_primers" title="Link to this definition">#</a></dt>
<dd><p>Description of <cite>reverse_primers</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.template">
<span class="sig-name descname"><span class="pre">template</span></span><a class="headerlink" href="#pydna.amplify.Anneal.template" title="Link to this definition">#</a></dt>
<dd><p>A copy of the template argument. Primers annealing sites has been
added as features that can be visualized in a seqence editor such as
ApE.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.limit">
<span class="sig-name descname"><span class="pre">limit</span></span><a class="headerlink" href="#pydna.amplify.Anneal.limit" title="Link to this definition">#</a></dt>
<dd><p>The limit of PCR primer annealing, default is 13 bp.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.products">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">products</span></span><a class="headerlink" href="#pydna.amplify.Anneal.products" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.amplify.Anneal.report">
<span class="sig-name descname"><span class="pre">report</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.amplify.Anneal.report" title="Link to this definition">#</a></dt>
<dd><p>returns a short report describing if or where primer
anneal on the template.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.amplify.pcr">
<span class="sig-prename descclassname"><span class="pre">pydna.amplify.</span></span><span class="sig-name descname"><span class="pre">pcr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><span class="pre">Amplicon</span></a></span></span><a class="reference internal" href="../_modules/pydna/amplify.html#pcr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.amplify.pcr" title="Link to this definition">#</a></dt>
<dd><p>pcr is a convenience function for the Anneal class to simplify its
usage, especially from the command line. If more than one or no PCR
product is formed, a ValueError is raised.</p>
<p>args is any iterable of Dseqrecords or an iterable of iterables of
Dseqrecords. args will be greedily flattened.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>iterable containing sequence objects</em>) – Several arguments are also accepted.</p></li>
<li><p><strong>limit</strong> (<em>int = 13</em><em>, </em><em>optional</em>) – limit length of the annealing part of the primers.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>sequences in args could be of type:</p>
<ul class="simple">
<li><p>string</p></li>
<li><p>Seq</p></li>
<li><p>SeqRecord (or subclass)</p></li>
<li><p>Dseqrecord (or sublcass)</p></li>
</ul>
<p>The last sequence will be assumed to be the template while
all preceeding sequences will be assumed to be primers.</p>
<p>This is a powerful function, use with care!</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>product</strong> – An <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> object representing the PCR
product. The direction of the PCR product will be the same as for
the template sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon">Amplicon</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.readers</span><span class="w"> </span><span class="kn">import</span> <span class="n">read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.amplify</span><span class="w"> </span><span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;cgactgtatcatctgatagcac&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">([</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.assembly">
<span id="pydna-assembly-module"></span><h2>pydna.assembly module<a class="headerlink" href="#module-pydna.assembly" title="Link to this heading">#</a></h2>
<p>Assembly of sequences by homologous recombination.</p>
<p>Should also be useful for related techniques such as Gibson assembly and fusion
PCR. Given a list of sequences (Dseqrecords), all sequences are analyzed for
shared homology longer than the set limit.</p>
<p>A graph is constructed where each overlapping region form a node and
sequences separating the overlapping regions form edges.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>            <span class="o">--</span> <span class="n">A</span> <span class="o">--</span>
<span class="n">catgatctacgtatcgtgt</span>     <span class="o">--</span> <span class="n">B</span> <span class="o">--</span>
            <span class="n">atcgtgtactgtcatattc</span>
                        <span class="n">catattcaaagttct</span>



<span class="o">--</span><span class="n">x</span><span class="o">--&gt;</span> <span class="n">A</span> <span class="o">--</span><span class="n">y</span><span class="o">--&gt;</span> <span class="n">B</span> <span class="o">--</span><span class="n">z</span><span class="o">--&gt;</span>   <span class="p">(</span><span class="n">Graph</span><span class="p">)</span>

<span class="n">Nodes</span><span class="p">:</span>

<span class="n">A</span> <span class="p">:</span> <span class="n">atcgtgt</span>
<span class="n">B</span> <span class="p">:</span> <span class="n">catattc</span>

<span class="n">Edges</span><span class="p">:</span>

<span class="n">x</span> <span class="p">:</span> <span class="n">catgatctacgt</span>
<span class="n">y</span> <span class="p">:</span> <span class="n">actgt</span>
<span class="n">z</span> <span class="p">:</span> <span class="n">aaagttct</span>
</pre></div>
</div>
<p>The NetworkX package is used to trace linear and circular paths through the
graph.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.assembly.Assembly">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.assembly.</span></span><span class="sig-name descname"><span class="pre">Assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">common_sub_strings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly.html#Assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly.Assembly" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Assembly of a list of linear DNA fragments into linear or circular
constructs. The Assembly is meant to replace the Assembly method as it
is easier to use. Accepts a list of Dseqrecords (source fragments) to
initiate an Assembly object. Several methods are available for analysis
of overlapping sequences, graph construction and assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fragments</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – a list of Dseqrecord objects.</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – The shortest shared homology to be considered</p></li>
<li><p><strong>algorithm</strong> (<em>function</em><em>, </em><em>optional</em>) – The algorithm used to determine the shared sequences.</p></li>
<li><p><strong>max_nodes</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The maximum number of nodes in the graph. This can be tweaked to
manage sequences with a high number of shared sub sequences.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly</span><span class="w"> </span><span class="kn">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;acgatgctatactgCCCCCtgtgctgtgctcta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtgctgtgctctaTTTTTtattctggctgtatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tattctggctgtatcGGGGGtacgatgctatactg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">((</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">),</span> <span class="n">limit</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 33bp 34bp 35bp</span>
<span class="go">limit(bp)....: 14</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()</span>
<span class="go">[Contig(o59), Contig(o59)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">watson</span>
<span class="go">&#39;acgatgctatactgCCCCCtgtgctgtgctctaTTTTTtattctggctgtatcGGGGGt&#39;</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly.Assembly.assemble_linear">
<span class="sig-name descname"><span class="pre">assemble_linear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.assembly.Assembly.assemble_linear" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly.Assembly.assemble_circular">
<span class="sig-name descname"><span class="pre">assemble_circular</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.assembly.Assembly.assemble_circular" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-pydna.assembly2">
<span id="pydna-assembly2-module"></span><h2>pydna.assembly2 module<a class="headerlink" href="#module-pydna.assembly2" title="Link to this heading">#</a></h2>
<p>Improved implementation of the assembly module. To see a list of issues with the previous implementation,
see [issues tagged with fixed-with-new-assembly-model](<a class="github reference external" href="https://github.com/pydna-group/pydna/issues?q=is%3Aissue%20state%3Aopen%20label%3Afixed-with-new-assembly-model">pydna-group/pydna#issues</a>)</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.gather_overlapping_locations">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">gather_overlapping_locations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragment_length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#gather_overlapping_locations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.gather_overlapping_locations" title="Link to this definition">#</a></dt>
<dd><p>Turn a list of locations into a list of tuples of those locations, where each tuple contains
locations that overlap. For example, if locs = [loc1, loc2, loc3], and loc1 and loc2 overlap,
the output will be [(loc1, loc2), (loc3,)].</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.ends_from_cutsite">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">ends_from_cutsite</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#ends_from_cutsite"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.ends_from_cutsite" title="Link to this definition">#</a></dt>
<dd><p>Get the sticky or blunt ends created by a restriction enzyme cut.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cutsite</strong> (<em>CutSiteType</em>) – A tuple ((cut_watson, ovhg), enzyme) describing where the cut occurs</p></li>
<li><p><strong>seq</strong> (<em>_Dseq</em>) – The DNA sequence being cut</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – If cutsite is None</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>tuple[tuple[str, str], tuple[str, str]]</em> – A tuple of two tuples, each containing the type of end (‘5’’, ‘3’’, or ‘blunt’)
and the sequence of the overhang. The first tuple is for the left end, second for the right end.</p></li>
<li><p><em>&gt;&gt;&gt; from Bio.Restriction import NotI</em></p></li>
<li><p><em>&gt;&gt;&gt; x = _Dseq(“ctcgGCGGCCGCcagcggccg”)</em></p></li>
<li><p><em>&gt;&gt;&gt; x.get_cutsites(NotI)</em></p></li>
<li><p><em>[((6, -4), NotI)]</em></p></li>
<li><p><em>&gt;&gt;&gt; ends_from_cutsite(x.get_cutsites(NotI)[0], x)</em></p></li>
<li><p><em>((“5’”, ‘ggcc’), (“5’”, ‘ggcc’))</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.restriction_ligation_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">restriction_ligation_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzymes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">RestrictionBatch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_blunt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#restriction_ligation_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.restriction_ligation_overlap" title="Link to this definition">#</a></dt>
<dd><p>Assembly algorithm to find overlaps that would result from restriction and ligation.</p>
<p>Like in sticky and gibson, the order matters (see example below of partial overlap)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqx</strong> (<em>_Dseqrecord</em>) – The first sequence</p></li>
<li><p><strong>seqy</strong> (<em>_Dseqrecord</em>) – The second sequence</p></li>
<li><p><strong>enzymes</strong> (<em>RestrictionBatch</em>) – The enzymes to use</p></li>
<li><p><strong>partial</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Whether to allow partial overlaps</p></li>
<li><p><strong>allow_blunt</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Whether to allow blunt ends</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>list[SequenceOverlap]</em> – A list of overlaps between the two sequences</p></li>
<li><p><em>&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord</em></p></li>
<li><p><em>&gt;&gt;&gt; from pydna.assembly2 import restriction_ligation_overlap</em></p></li>
<li><p><em>&gt;&gt;&gt; from Bio.Restriction import EcoRI, RgaI, DrdI, EcoRV</em></p></li>
<li><p><em>&gt;&gt;&gt; x = Dseqrecord(“ccGAATTCaa”)</em></p></li>
<li><p><em>&gt;&gt;&gt; y = Dseqrecord(“aaaaGAATTCgg”)</em></p></li>
<li><p><em>&gt;&gt;&gt; restriction_ligation_overlap(x, y, [EcoRI])</em></p></li>
<li><p><em>[(3, 5, 4)]</em></p></li>
<li><p><em>&gt;&gt;&gt; restriction_ligation_overlap(y, x, [EcoRI])</em></p></li>
<li><p><em>[(5, 3, 4)]</em></p></li>
<li><p><em>Partial overlap, note how it is not symmetric</em></p></li>
<li><p><em>&gt;&gt;&gt; x = Dseqrecord(“GACTAAAGGGTC”)</em></p></li>
<li><p><em>&gt;&gt;&gt; y = Dseqrecord(“AAGCGATCGCAAGCGATCGCAA”)</em></p></li>
<li><p><em>&gt;&gt;&gt; restriction_ligation_overlap(x, y, [RgaI, DrdI], partial=True)</em></p></li>
<li><p><em>[(6, 5, 1), (6, 15, 1)]</em></p></li>
<li><p><em>&gt;&gt;&gt; restriction_ligation_overlap(y, x, [RgaI, DrdI], partial=True)</em></p></li>
<li><p><em>[]</em></p></li>
<li><p><em>Blunt overlap, returns length of the overlap 0</em></p></li>
<li><p><em>&gt;&gt;&gt; x = Dseqrecord(“aaGATATCcc”)</em></p></li>
<li><p><em>&gt;&gt;&gt; y = Dseqrecord(“ttttGATATCaa”)</em></p></li>
<li><p><em>&gt;&gt;&gt; restriction_ligation_overlap(x, y, [EcoRV], allow_blunt=True)</em></p></li>
<li><p><em>[(5, 7, 0)]</em></p></li>
<li><p><em>&gt;&gt;&gt; restriction_ligation_overlap(y, x, [EcoRV], allow_blunt=True)</em></p></li>
<li><p><em>[(7, 5, 0)]</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.combine_algorithms">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">combine_algorithms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">algorithms</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#combine_algorithms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.combine_algorithms" title="Link to this definition">#</a></dt>
<dd><p>Combine assembly algorithms, if any of them returns a match, the match is returned.</p>
<p>This can be used for example in a ligation where you want to allow both sticky and blunt end ligation.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.blunt_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">blunt_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#blunt_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.blunt_overlap" title="Link to this definition">#</a></dt>
<dd><p>Assembly algorithm to find blunt overlaps. Used for blunt ligation.</p>
<p>It basically returns [(len(seqx), 0, 0)] if the right end of seqx is blunt and the
left end of seqy is blunt (compatible with blunt ligation). Otherwise, it returns an empty list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqx</strong> (<em>_Dseqrecord</em>) – The first sequence</p></li>
<li><p><strong>seqy</strong> (<em>_Dseqrecord</em>) – The second sequence</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – There for compatibility, but it is ignored</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>list[SequenceOverlap]</em> – A list of overlaps between the two sequences</p></li>
<li><p><em>&gt;&gt;&gt; from pydna.assembly2 import blunt_overlap</em></p></li>
<li><p><em>&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord</em></p></li>
<li><p><em>&gt;&gt;&gt; x = Dseqrecord(“AAAAAA”)</em></p></li>
<li><p><em>&gt;&gt;&gt; y = Dseqrecord(“TTTTTT”)</em></p></li>
<li><p><em>&gt;&gt;&gt; blunt_overlap(x, y)</em></p></li>
<li><p><em>[(6, 0, 0)]</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.common_sub_strings">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">common_sub_strings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#common_sub_strings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.common_sub_strings" title="Link to this definition">#</a></dt>
<dd><p>Assembly algorithm to find common substrings of length == limit. see the docs of
the function common_sub_strings_str for more details. It is case insensitive.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;TAAAAAAT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;CCaAaAaACC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">common_sub_strings</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">[(1, 2, 6), (1, 3, 5), (2, 2, 5)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.gibson_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">gibson_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#gibson_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.gibson_overlap" title="Link to this definition">#</a></dt>
<dd><p>Assembly algorithm to find terminal overlaps (e.g. for Gibson assembly).
The order matters, we want alignments like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seqx</span><span class="p">:</span>    <span class="n">oooo</span><span class="o">------</span><span class="n">xxxx</span>
<span class="n">seqy</span><span class="p">:</span>              <span class="n">xxxx</span><span class="o">------</span><span class="n">oooo</span>
<span class="n">Product</span><span class="p">:</span> <span class="n">oooo</span><span class="o">------</span><span class="n">xxxx</span><span class="o">------</span><span class="n">oooo</span>

<span class="n">Not</span> <span class="n">like</span><span class="p">:</span>

<span class="n">seqx</span><span class="p">:</span>               <span class="n">oooo</span><span class="o">------</span><span class="n">xxxx</span>
<span class="n">seqy</span><span class="p">:</span>     <span class="n">xxxx</span><span class="o">------</span><span class="n">oooo</span>
<span class="n">Product</span> <span class="p">(</span><span class="n">unwanted</span><span class="p">):</span> <span class="n">oooo</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqx</strong> (<em>_Dseqrecord</em>) – The first sequence</p></li>
<li><p><strong>seqy</strong> (<em>_Dseqrecord</em>) – The second sequence</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Minimum length of the overlap</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>list[SequenceOverlap]</em> – A list of overlaps between the two sequences</p></li>
<li><p><em>&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord</em></p></li>
<li><p><em>&gt;&gt;&gt; from pydna.assembly2 import gibson_overlap</em></p></li>
<li><p><em>&gt;&gt;&gt; x = Dseqrecord(“ttactaAAAAAA”)</em></p></li>
<li><p><em>&gt;&gt;&gt; y = Dseqrecord(“AAAAAAcgcacg”)</em></p></li>
<li><p><em>&gt;&gt;&gt; gibson_overlap(x, y, limit=5)</em></p></li>
<li><p><em>[(6, 0, 6), (7, 0, 5)]</em></p></li>
<li><p><em>&gt;&gt;&gt; gibson_overlap(y, x, limit=5)</em></p></li>
<li><p><em>[]</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.sticky_end_sub_strings">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">sticky_end_sub_strings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#sticky_end_sub_strings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.sticky_end_sub_strings" title="Link to this definition">#</a></dt>
<dd><p>Assembly algorithm for ligation of sticky ends.</p>
<p>For now, if limit 0 / False (default) only full overlaps are considered.
Otherwise, partial overlaps are also returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqx</strong> (<em>_Dseqrecord</em>) – The first sequence</p></li>
<li><p><strong>seqy</strong> (<em>_Dseqrecord</em>) – The second sequence</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Whether to allow partial overlaps</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of overlaps between the two sequences</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[SequenceOverlap]</p>
</dd>
</dl>
<p>Ligation of fully overlapping sticky ends, note how the order matters</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">sticky_end_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s2">&quot;AAAAAA&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s2">&quot;AAAAAA&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sticky_end_sub_strings</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">[(3, 0, 3)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sticky_end_sub_strings</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">[]</span>
</pre></div>
</div>
<p>Ligation of partially overlapping sticky ends, specified with limit=True</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s2">&quot;AAAAAA&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s2">&quot;AAAAAA&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sticky_end_sub_strings</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sticky_end_sub_strings</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">[(4, 0, 2)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.zip_match_leftwards">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">zip_match_leftwards</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><span class="pre">SeqRecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><span class="pre">SeqRecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#zip_match_leftwards"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.zip_match_leftwards" title="Link to this definition">#</a></dt>
<dd><p>Starting from the rightmost edge of the match, return a new match encompassing the max
number of bases. This can be used to return a longer match if a primer aligns for longer
than the limit or a shorter match if there are mismatches. This is convenient to maintain
as many features as possible. It is used in PCR assembly.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">_Dseqrecord</span><span class="p">(</span><span class="s1">&#39;AAAAACGTCCCGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">primer</span> <span class="o">=</span> <span class="n">_Dseqrecord</span><span class="p">(</span><span class="s1">&#39;ACGTCCCGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">match</span> <span class="o">=</span> <span class="p">(</span><span class="mi">13</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="c1"># an empty match at the end of each</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">zip_match_leftwards</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">primer</span><span class="p">,</span> <span class="n">match</span><span class="p">)</span>
<span class="go">(4, 0, 9)</span>
</pre></div>
</div>
<p>Works in circular molecules if the match spans the origin:
&gt;&gt;&gt; seq = _Dseqrecord(‘TCCCGTAAAAACG’, circular=True)
&gt;&gt;&gt; primer = _Dseqrecord(‘ACGTCCCGT’)
&gt;&gt;&gt; match = (6, 9, 0)
&gt;&gt;&gt; zip_match_leftwards(seq, primer, match)
(10, 0, 9)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.zip_match_rightwards">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">zip_match_rightwards</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#zip_match_rightwards"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.zip_match_rightwards" title="Link to this definition">#</a></dt>
<dd><p>Same as zip_match_leftwards, but towards the right.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.seqrecord2_uppercase_DNA_string">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">seqrecord2_uppercase_DNA_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><span class="pre">SeqRecord</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#seqrecord2_uppercase_DNA_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.seqrecord2_uppercase_DNA_string" title="Link to this definition">#</a></dt>
<dd><p>Transform a Dseqrecord to a sequence string where U is replaced by T, everything is upper case and
circular sequences are repeated twice. This is used for PCR, to support primers with U’s (e.g. for USER cloning).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.primer_template_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">primer_template_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#primer_template_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.primer_template_overlap" title="Link to this definition">#</a></dt>
<dd><p>Assembly algorithm to find overlaps between a primer and a template. It accepts mismatches.
When there are mismatches, it only returns the common part between the primer and the template.</p>
<p>If seqx is a primer and seqy is a template, it represents the binding of a forward primer.
If seqx is a template and seqy is a primer, it represents the binding of a reverse primer,
where the primer has been passed as its reverse complement (see examples).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqx</strong> (<em>_Dseqrecord</em><em> | </em><em>_Primer</em>) – The primer</p></li>
<li><p><strong>seqy</strong> (<em>_Dseqrecord</em><em> | </em><em>_Primer</em>) – The template</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Minimum length of the overlap</p></li>
<li><p><strong>mismatches</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Maximum number of mismatches (only substitutions, no deletion or insertion)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>list[SequenceOverlap]</em> – A list of overlaps between the primer and the template</p></li>
<li><p><em>&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord</em></p></li>
<li><p><em>&gt;&gt;&gt; from pydna.primer import Primer</em></p></li>
<li><p><em>&gt;&gt;&gt; from pydna.assembly2 import primer_template_overlap</em></p></li>
<li><p><em>&gt;&gt;&gt; template = Dseqrecord(“AATTAGCAGCGATCGAGT”, circular=True)</em></p></li>
<li><p><em>&gt;&gt;&gt; primer = Primer(“TTAGCAGC”)</em></p></li>
<li><p><em>&gt;&gt;&gt; primer_template_overlap(primer, template, limit=8, mismatches=0)</em></p></li>
<li><p><em>[(0, 2, 8)]</em></p></li>
<li><p>This actually represents the binding of the primer <code class="docutils literal notranslate"><span class="pre">GCTGCTAA</span></code> (reverse complement)</p></li>
<li><p><em>&gt;&gt;&gt; primer_template_overlap(template, primer, limit=8, mismatches=0)</em></p></li>
<li><p><em>[(2, 0, 8)]</em></p></li>
<li><p><em>&gt;&gt;&gt; primer_template_overlap(primer, template.reverse_complement(), limit=8, mismatches=0)</em></p></li>
<li><p><em>[]</em></p></li>
<li><p><em>&gt;&gt;&gt; primer_template_overlap(primer.reverse_complement(), template, limit=8, mismatches=0)</em></p></li>
<li><p><em>[]</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.fill_left">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">fill_left</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#fill_left"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.fill_left" title="Link to this definition">#</a></dt>
<dd><p>Fill the left overhang of a sequence with the complementary sequence.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.fill_right">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">fill_right</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#fill_right"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.fill_right" title="Link to this definition">#</a></dt>
<dd><p>Fill the right overhang of a sequence with the complementary sequence.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.fill_dseq">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">fill_dseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#fill_dseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.fill_dseq" title="Link to this definition">#</a></dt>
<dd><p>Fill the overhangs of a sequence with the complementary sequence.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.reverse_complement_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">reverse_complement_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#reverse_complement_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.reverse_complement_assembly" title="Link to this definition">#</a></dt>
<dd><p>Complement an assembly, i.e. reverse the order of the fragments and the orientation of the overlaps.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.filter_linear_subassemblies">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">filter_linear_subassemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">linear_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#filter_linear_subassemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.filter_linear_subassemblies" title="Link to this definition">#</a></dt>
<dd><p>Remove linear assemblies which are sub-assemblies of circular assemblies</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.remove_subassemblies">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">remove_subassemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#remove_subassemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.remove_subassemblies" title="Link to this definition">#</a></dt>
<dd><p>Filter out subassemblies, i.e. assemblies that are contained within another assembly.</p>
<dl class="simple">
<dt>For example:</dt><dd><p>[(1, 2, ‘1[8:14]:2[1:7]’), (2, 3, ‘2[10:17]:3[1:8]’)]
[(1, 2, ‘1[8:14]:2[1:7]’)]</p>
</dd>
</dl>
<p>The second one is a subassembly of the first one.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.assembly2str">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">assembly2str</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#assembly2str"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.assembly2str" title="Link to this definition">#</a></dt>
<dd><p>Convert an assembly to a string representation, for example:
((1, 2, [8:14], [1:7]),(2, 3, [10:17], [1:8]))
becomes:
(‘1[8:14]:2[1:7]’, ‘2[10:17]:3[1:8]’)</p>
<p>The reason for this is that by default, a feature ‘[8:14]’ when present in a tuple
is printed to the console as <code class="docutils literal notranslate"><span class="pre">SimpleLocation(ExactPosition(8),</span> <span class="pre">ExactPosition(14),</span> <span class="pre">strand=1)</span></code> (very long).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.assembly2str_tuple">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">assembly2str_tuple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#assembly2str_tuple"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.assembly2str_tuple" title="Link to this definition">#</a></dt>
<dd><p>Convert an assembly to a string representation, like
((1, 2, [8:14], [1:7]),(2, 3, [10:17], [1:8]))</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.assembly_has_mismatches">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">assembly_has_mismatches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#assembly_has_mismatches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.assembly_has_mismatches" title="Link to this definition">#</a></dt>
<dd><p>Check if an assembly has mismatches. This should never happen and if so it returns an error.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.assembly_is_circular">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">assembly_is_circular</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#assembly_is_circular"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.assembly_is_circular" title="Link to this definition">#</a></dt>
<dd><p>Based on the topology of the locations of an assembly, determine if it is circular.
This does not work for insertion assemblies, that’s why assemble takes the optional argument is_insertion.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.assemble">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">assemble</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_insertion</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#assemble"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.assemble" title="Link to this definition">#</a></dt>
<dd><p>Generate a Dseqrecord from an assembly and a list of fragments.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.annotate_primer_binding_sites">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">annotate_primer_binding_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_dseqr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#annotate_primer_binding_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.annotate_primer_binding_sites" title="Link to this definition">#</a></dt>
<dd><p>Annotate the primer binding sites in a Dseqrecord.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.edge_representation2subfragment_representation">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">edge_representation2subfragment_representation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#edge_representation2subfragment_representation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.edge_representation2subfragment_representation" title="Link to this definition">#</a></dt>
<dd><p>Turn this kind of edge representation fragment 1, fragment 2, right edge on 1, left edge on 2
a = [(1, 2, ‘loc1a’, ‘loc2a’), (2, 3, ‘loc2b’, ‘loc3b’), (3, 1, ‘loc3c’, ‘loc1c’)]
Into this: fragment 1, left edge on 1, right edge on 1
b = [(1, ‘loc1c’, ‘loc1a’), (2, ‘loc2a’, ‘loc2b’), (3, ‘loc3b’, ‘loc3c’)]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.subfragment_representation2edge_representation">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">subfragment_representation2edge_representation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#subfragment_representation2edge_representation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.subfragment_representation2edge_representation" title="Link to this definition">#</a></dt>
<dd><p>Turn this kind of subfragment representation fragment 1, left edge on 1, right edge on 1
a = [(1, ‘loc1c’, ‘loc1a’), (2, ‘loc2a’, ‘loc2b’), (3, ‘loc3b’, ‘loc3c’)]
Into this: fragment 1, fragment 2, right edge on 1, left edge on 2
b = [(1, 2, ‘loc1a’, ‘loc2a’), (2, 3, ‘loc2b’ ‘loc3b’), (3, 1, ‘loc3c’, ‘loc1c’)]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.get_assembly_subfragments">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">get_assembly_subfragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subfragment_representation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#get_assembly_subfragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.get_assembly_subfragments" title="Link to this definition">#</a></dt>
<dd><p>From the fragment representation returned by edge_representation2subfragment_representation, get the subfragments that are joined together.</p>
<p>Subfragments are the slices of the fragments that are joined together</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  <span class="o">--</span><span class="n">A</span><span class="o">--</span>
<span class="n">TACGTAAT</span>
  <span class="o">--</span><span class="n">B</span><span class="o">--</span>
 <span class="n">TCGTAACGA</span>

<span class="n">Gives</span><span class="p">:</span> <span class="n">TACGTAA</span> <span class="o">/</span> <span class="n">CGTAACGA</span>
</pre></div>
</div>
<p>To reproduce:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s1">&#39;TACGTAAT&#39;</span><span class="p">)</span>
<span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s1">&#39;TCGTAACGA&#39;</span><span class="p">)</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">],</span> <span class="n">limit</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">a0</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">get_linear_assemblies</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">assembly2str</span><span class="p">(</span><span class="n">a0</span><span class="p">))</span>
<span class="n">a0_subfragment_rep</span> <span class="o">=</span><span class="n">edge_representation2subfragment_representation</span><span class="p">(</span><span class="n">a0</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">get_assembly_subfragments</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">],</span> <span class="n">a0_subfragment_rep</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

<span class="c1"># prints TACGTAA and CGTAACGA</span>
</pre></div>
</div>
<p>Subfragments: <code class="docutils literal notranslate"><span class="pre">cccccgtatcgtgt</span></code>, <code class="docutils literal notranslate"><span class="pre">atcgtgtactgtcatattc</span></code></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.extract_subfragment">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">extract_subfragment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#extract_subfragment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.extract_subfragment" title="Link to this definition">#</a></dt>
<dd><p>Extract a subfragment from a sequence for an assembly, given the start and end locations of the subfragment.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.is_sublist">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">is_sublist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sublist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">my_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">my_list_is_cyclic</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#is_sublist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.is_sublist" title="Link to this definition">#</a></dt>
<dd><p>Returns True if argument sublist is a sublist of argument my_list (can be treated as cyclic), False otherwise.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_sublist</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="kc">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_sublist</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="kc">False</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
<p># See the case here for cyclic lists
&gt;&gt;&gt; is_sublist([3, 1], [1, 2, 3], False)
False
&gt;&gt;&gt; is_sublist([3, 1], [1, 2, 3], True)
True</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.circular_permutation_min_abs">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">circular_permutation_min_abs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#circular_permutation_min_abs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.circular_permutation_min_abs" title="Link to this definition">#</a></dt>
<dd><p>Returns the circular permutation of lst with the smallest absolute value first.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">circular_permutation_min_abs</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="go">[1, 2, 3]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">circular_permutation_min_abs</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">[1, 2, 3]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">Assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">common_sub_strings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_fragment_order</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_all_fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Assembly of a list of DNA fragments into linear or circular constructs.
Accepts a list of Dseqrecords (source fragments) to
initiate an Assembly object. Several methods are available for analysis
of overlapping sequences, graph construction and assembly.</p>
<p>The assembly contains a directed graph, where nodes represent fragments and
edges represent overlaps between fragments. :</p>
<ul class="simple">
<li><p>The node keys are integers, representing the index of the fragment in the
input list of fragments. The sign of the node key represents the orientation
of the fragment, positive for forward orientation, negative for reverse orientation.</p></li>
<li><dl class="simple">
<dt>The edges contain the locations of the overlaps in the fragments. For an edge (u, v, key):</dt><dd><ul>
<li><p>u and v are the nodes connected by the edge.</p></li>
<li><p>key is a string that represents the location of the overlap. In the format:
‘u[start:end](strand):v[start:end](strand)’.</p></li>
<li><p>Edges have a ‘locations’ attribute, which is a list of two FeatureLocation objects,
representing the location of the overlap in the u and v fragment, respectively.</p></li>
<li><p>You can think of an edge as a representation of the join of two fragments.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<p>If fragment 1 and 2 share a subsequence of 6bp, [8:14] in fragment 1 and [1:7] in fragment 2,
there will be 4 edges representing that overlap in the graph, for all possible
orientations of the fragments (see add_edges_from_match for details):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">(1,</span> <span class="pre">2,</span> <span class="pre">'1[8:14]:2[1:7]')</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">(2,</span> <span class="pre">1,</span> <span class="pre">'2[1:7]:1[8:14]')</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">(-1,</span> <span class="pre">-2,</span> <span class="pre">'-1[0:6]:-2[10:16]')</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">(-2,</span> <span class="pre">-1,</span> <span class="pre">'-2[10:16]:-1[0:6]')</span></code></p></li>
</ul>
<p>An assembly can be thought of as a tuple of graph edges, but instead of representing them with node indexes and keys, we represent them
as u, v, locu, locv, where u and v are the nodes connected by the edge, and locu and locv are the locations of the overlap in the first
and second fragment. Assemblies are then represented as:</p>
<ul class="simple">
<li><p>Linear: ((1, 2, [8:14], [1:7]), (2, 3, [10:17], [1:8]))</p></li>
<li><p>Circular: ((1, 2, [8:14], [1:7]), (2, 3, [10:17], [1:8]), (3, 1, [12:17], [1:6]))</p></li>
</ul>
<p>Note that the first and last fragment are the same in a circular assembly.</p>
<p>The following constrains are applied to remove duplicate assemblies:</p>
<ul class="simple">
<li><p>Circular assemblies: the first subfragment is not reversed, and has the smallest index in the input fragment list.
use_fragment_order is ignored.</p></li>
<li><dl class="simple">
<dt>Linear assemblies:</dt><dd><ul>
<li><p>Using uid (see add_edges_from_match) to identify unique edges.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – A list of Dseqrecord objects.</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – The shortest shared homology to be considered, this is passed as the third argument to the <code class="docutils literal notranslate"><span class="pre">algorithm</span></code> function.
For certain algorithms, this might be ignored.</p></li>
<li><p><strong>algorithm</strong> (<em>function</em><em>, </em><em>optional</em>) – The algorithm used to determine the shared sequences. It’s a function that takes two Dseqrecord objects as inputs,
and will get passed the third argument (limit), that may or may not be used. It must return a list of overlaps
(see common_sub_strings for an example).</p></li>
<li><p><strong>use_fragment_order</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – It’s set to True by default to reproduce legacy pydna behaviour: only assemblies that start with the first fragment and end with the last are considered.
You should set it to False.</p></li>
<li><p><strong>use_all_fragments</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Constrain the assembly to use all fragments.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>from assembly2 import Assembly, assembly2str
from pydna.dseqrecord import Dseqrecord</p>
<dl class="simple">
<dt>example_fragments = (</dt><dd><p>Dseqrecord(‘AacgatCAtgctcc’, name=’a’),
Dseqrecord(‘TtgctccTAAattctgc’, name=’b’),
Dseqrecord(‘CattctgcGAGGacgatG’, name=’c’),</p>
</dd>
</dl>
<p>)</p>
<p>asm = Assembly(example_fragments, limit=5, use_fragment_order=False)
print(‘Linear ===============’)
for assembly in asm.get_linear_assemblies():</p>
<blockquote>
<div><p>print(’ ‘, assembly2str(assembly))</p>
</div></blockquote>
<p>print(‘Circular =============’)
for assembly in asm.get_circular_assemblies():</p>
<blockquote>
<div><p>print(’ ‘, assembly2str(assembly))</p>
</div></blockquote>
<p># Prints
Linear ===============</p>
<blockquote>
<div><p>(‘1[8:14]:2[1:7]’, ‘2[10:17]:3[1:8]’)
(‘2[10:17]:3[1:8]’, ‘3[12:17]:1[1:6]’)
(‘3[12:17]:1[1:6]’, ‘1[8:14]:2[1:7]’)
(‘1[1:6]:3[12:17]’,)
(‘2[1:7]:1[8:14]’,)
(‘3[1:8]:2[10:17]’,)</p>
</div></blockquote>
<dl class="simple">
<dt>Circular =============</dt><dd><p>(‘1[8:14]:2[1:7]’, ‘2[10:17]:3[1:8]’, ‘3[12:17]:1[1:6]’)</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.assembly_is_valid">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">assembly_is_valid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_all_fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_insertion</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.assembly_is_valid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.assembly_is_valid" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the assembly is valid, False otherwise. See function comments for conditions tested.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.add_edges_from_match">
<span class="sig-name descname"><span class="pre">add_edges_from_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">first</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">secnd</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.add_edges_from_match"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.add_edges_from_match" title="Link to this definition">#</a></dt>
<dd><p>Add edges to the graph from a match returned by the <code class="docutils literal notranslate"><span class="pre">algorithm</span></code> function (see pydna.common_substrings). For
format of edges (see documentation of the Assembly class).</p>
<p>Matches are directional, because not all <code class="docutils literal notranslate"><span class="pre">algorithm</span></code> functions return the same match for (u,v) and (v,u). For example,
homologous recombination does but sticky end ligation does not. The function returns two edges:</p>
<ul class="simple">
<li><p>Fragments in the orientation they were passed, with locations of the match (u, v, loc_u, loc_v)</p></li>
<li><p>Reverse complement of the fragments with inverted order, with flipped locations (-v, -u, flip(loc_v), flip(loc_u))/</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.format_assembly_edge">
<span class="sig-name descname"><span class="pre">format_assembly_edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph_edge</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.format_assembly_edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.format_assembly_edge" title="Link to this definition">#</a></dt>
<dd><p>Go from the (u, v, key) to the (u, v, locu, locv) format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.get_linear_assemblies">
<span class="sig-name descname"><span class="pre">get_linear_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.get_linear_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.get_linear_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Get linear assemblies, applying the constrains described in __init__, ensuring that paths represent
real assemblies (see assembly_is_valid). Subassemblies are removed (see remove_subassemblies).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.node_path2assembly_list">
<span class="sig-name descname"><span class="pre">node_path2assembly_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cycle</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.node_path2assembly_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.node_path2assembly_list" title="Link to this definition">#</a></dt>
<dd><dl class="simple">
<dt>Convert a node path in the format [1, 2, 3] (as returned by _nx.cycles.simple_cycles) to a list of all</dt><dd><p>possible assemblies.</p>
</dd>
</dl>
<p>There may be multiple assemblies for a given node path, if there are several edges connecting two nodes,
for example two overlaps between 1 and 2, and single overlap between 2 and 3 should return 3 assemblies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.get_unique_linear_paths">
<span class="sig-name descname"><span class="pre">get_unique_linear_paths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G_with_begin_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">MultiDiGraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_paths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.get_unique_linear_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.get_unique_linear_paths" title="Link to this definition">#</a></dt>
<dd><p>Get unique linear paths from the graph, removing those that contain the same node twice.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.get_possible_assembly_number">
<span class="sig-name descname"><span class="pre">get_possible_assembly_number</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">paths</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.get_possible_assembly_number"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.get_possible_assembly_number" title="Link to this definition">#</a></dt>
<dd><p>Get the number of possible assemblies from a list of node paths. Basically, for each path
passed as a list of integers / nodes, we calculate the number of paths possible connecting
the nodes in that order, given the graph (all the edges connecting them).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.get_circular_assemblies">
<span class="sig-name descname"><span class="pre">get_circular_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.get_circular_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.get_circular_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Get circular assemblies, applying the constrains described in __init__, ensuring that paths represent
real assemblies (see assembly_is_valid).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.format_insertion_assembly">
<span class="sig-name descname"><span class="pre">format_insertion_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.format_insertion_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.format_insertion_assembly" title="Link to this definition">#</a></dt>
<dd><p>Sorts the fragment representing a cycle so that they represent an insertion assembly if possible,
else returns None.</p>
<p>Here we check if one of the joins between fragments represents the edges of an insertion assembly
The fragment must be linear, and the join must be as indicated below</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">--------</span>         <span class="o">-------</span>           <span class="n">Fragment</span> <span class="mi">1</span>
    <span class="o">||</span>            <span class="o">||</span>
    <span class="n">xxxxxxxx</span>      <span class="o">||</span>               <span class="n">Fragment</span> <span class="mi">2</span>
          <span class="o">||</span>      <span class="o">||</span>
          <span class="n">oooooooooo</span>               <span class="n">Fragment</span> <span class="mi">3</span>
</pre></div>
</div>
<p>The above example will be [(1, 2, [4:6], [0:2]), (2, 3, [6:8], [0:2]), (3, 1, [8:10], [9:11)])]</p>
<p>These could be returned in any order by simple_cycles, so we sort the edges so that the first
and last <code class="docutils literal notranslate"><span class="pre">u</span></code> and <code class="docutils literal notranslate"><span class="pre">v</span></code> match the fragment that gets the insertion (1 in the example above).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.format_insertion_assembly_edge_case">
<span class="sig-name descname"><span class="pre">format_insertion_assembly_edge_case</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.format_insertion_assembly_edge_case"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.format_insertion_assembly_edge_case" title="Link to this definition">#</a></dt>
<dd><p>Edge case from <a class="github reference external" href="https://github.com/manulera/OpenCloning_backend/issues/329">manulera/OpenCloning_backend#329</a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.get_insertion_assemblies">
<span class="sig-name descname"><span class="pre">get_insertion_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.get_insertion_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.get_insertion_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Assemblies that represent the insertion of a fragment or series of fragment inside a linear construct. For instance,
digesting CCCCGAATTCCCCGAATTC with EcoRI and inserting the fragment with two overhangs into the EcoRI site of AAAGAATTCAAA.
This is not so much meant for the use-case of linear fragments that represent actual linear fragments, but for linear
fragments that represent a genome region. This can then be used to simulate homologous recombination.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.assemble_linear">
<span class="sig-name descname"><span class="pre">assemble_linear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.assemble_linear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.assemble_linear" title="Link to this definition">#</a></dt>
<dd><p>Assemble linear constructs, from assemblies returned by self.get_linear_assemblies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.assemble_circular">
<span class="sig-name descname"><span class="pre">assemble_circular</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.assemble_circular"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.assemble_circular" title="Link to this definition">#</a></dt>
<dd><p>Assemble circular constructs, from assemblies returned by self.get_circular_assemblies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.assemble_insertion">
<span class="sig-name descname"><span class="pre">assemble_insertion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.assemble_insertion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.assemble_insertion" title="Link to this definition">#</a></dt>
<dd><p>Assemble insertion constructs, from assemblies returned by self.get_insertion_assemblies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.get_locations_on_fragments">
<span class="sig-name descname"><span class="pre">get_locations_on_fragments</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.get_locations_on_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.get_locations_on_fragments" title="Link to this definition">#</a></dt>
<dd><p>Get a dictionary where the keys are the nodes in the graph, and the values are dictionaries with keys
<code class="docutils literal notranslate"><span class="pre">left</span></code>, <code class="docutils literal notranslate"><span class="pre">right</span></code>, containing (for each fragment) the locations where the fragment is joined to another fragment on its left
and right side. The values in <code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> are often the same, except in restriction-ligation with partial overlap enabled,
where we can end up with a situation like this:</p>
<p>GGTCTCCCCAATT and aGGTCTCCAACCAA as fragments</p>
<p># Partial overlap in assembly 1[9:11]:2[8:10]
GGTCTCCxxAACCAA
CCAGAGGGGTTxxTT</p>
<p># Partial overlap in 2[10:12]:1[7:9]
aGGTCTCCxxCCAATT
tCCAGAGGTTGGxxAA</p>
<p>Would return:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="mi">1</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;left&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">7</span><span class="p">:</span><span class="mi">9</span><span class="p">],</span> <span class="s1">&#39;right&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">9</span><span class="p">:</span><span class="mi">11</span><span class="p">]},</span>
    <span class="mi">2</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;left&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">8</span><span class="p">:</span><span class="mi">10</span><span class="p">],</span> <span class="s1">&#39;right&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">12</span><span class="p">]},</span>
    <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;left&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;right&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">6</span><span class="p">]},</span>
    <span class="o">-</span><span class="mi">2</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;left&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;right&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">6</span><span class="p">]}</span>
<span class="p">}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.Assembly.assembly_uses_only_adjacent_edges">
<span class="sig-name descname"><span class="pre">assembly_uses_only_adjacent_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#Assembly.assembly_uses_only_adjacent_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.Assembly.assembly_uses_only_adjacent_edges" title="Link to this definition">#</a></dt>
<dd><p>Check whether only adjacent edges within each fragment are used in the assembly. This is useful to check if a cut and ligate assembly is valid,
and prevent including partially digested fragments. For example, imagine the following fragment being an input for a digestion
and ligation assembly, where the enzyme cuts at the sites indicated by the vertical lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>       <span class="n">x</span>       <span class="n">y</span>       <span class="n">z</span>
<span class="o">-------|-------|-------|---------</span>
</pre></div>
</div>
<p>We would only want assemblies that contain subfragments start-x, x-y, y-z, z-end, and not start-x, y-end, for instance.
The latter would indicate that the fragment was partially digested.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.assembly2.PCRAssembly">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">PCRAssembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#PCRAssembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.PCRAssembly" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.assembly2.Assembly" title="pydna.assembly2.Assembly"><code class="xref py py-class docutils literal notranslate"><span class="pre">Assembly</span></code></a></p>
<p>An assembly that represents a PCR, where <code class="docutils literal notranslate"><span class="pre">fragments</span></code> is a list of primer, template, primer (in that order).
It always uses the <code class="docutils literal notranslate"><span class="pre">primer_template_overlap</span></code> algorithm and accepts the <code class="docutils literal notranslate"><span class="pre">mismatches</span></code> argument to indicate
the number of mismatches allowed in the overlap. Only supports substitution mismatches, not indels.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.PCRAssembly.get_linear_assemblies">
<span class="sig-name descname"><span class="pre">get_linear_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#PCRAssembly.get_linear_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.PCRAssembly.get_linear_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Get linear assemblies, applying the constrains described in __init__, ensuring that paths represent
real assemblies (see assembly_is_valid). Subassemblies are removed (see remove_subassemblies).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.PCRAssembly.get_circular_assemblies">
<span class="sig-name descname"><span class="pre">get_circular_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#PCRAssembly.get_circular_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.PCRAssembly.get_circular_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Get circular assemblies, applying the constrains described in __init__, ensuring that paths represent
real assemblies (see assembly_is_valid).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.PCRAssembly.get_insertion_assemblies">
<span class="sig-name descname"><span class="pre">get_insertion_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#PCRAssembly.get_insertion_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.PCRAssembly.get_insertion_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Assemblies that represent the insertion of a fragment or series of fragment inside a linear construct. For instance,
digesting CCCCGAATTCCCCGAATTC with EcoRI and inserting the fragment with two overhangs into the EcoRI site of AAAGAATTCAAA.
This is not so much meant for the use-case of linear fragments that represent actual linear fragments, but for linear
fragments that represent a genome region. This can then be used to simulate homologous recombination.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.assembly2.SingleFragmentAssembly">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">SingleFragmentAssembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">frags:</span> <span class="pre">[&lt;class</span> <span class="pre">'pydna.dseqrecord.Dseqrecord'&gt;],</span> <span class="pre">limit=25,</span> <span class="pre">algorithm=common_sub_strings</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#SingleFragmentAssembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.SingleFragmentAssembly" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.assembly2.Assembly" title="pydna.assembly2.Assembly"><code class="xref py py-class docutils literal notranslate"><span class="pre">Assembly</span></code></a></p>
<p>An assembly that represents the circularisation or splicing of a single fragment.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.SingleFragmentAssembly.get_circular_assemblies">
<span class="sig-name descname"><span class="pre">get_circular_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#SingleFragmentAssembly.get_circular_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.SingleFragmentAssembly.get_circular_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Get circular assemblies, applying the constrains described in __init__, ensuring that paths represent
real assemblies (see assembly_is_valid).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.SingleFragmentAssembly.get_insertion_assemblies">
<span class="sig-name descname"><span class="pre">get_insertion_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">only_adjacent_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#SingleFragmentAssembly.get_insertion_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.SingleFragmentAssembly.get_insertion_assemblies" title="Link to this definition">#</a></dt>
<dd><p>This could be renamed splicing assembly, but the essence is similar</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.assembly2.SingleFragmentAssembly.get_linear_assemblies">
<span class="sig-name descname"><span class="pre">get_linear_assemblies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/assembly2.html#SingleFragmentAssembly.get_linear_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.SingleFragmentAssembly.get_linear_assemblies" title="Link to this definition">#</a></dt>
<dd><p>Get linear assemblies, applying the constrains described in __init__, ensuring that paths represent
real assemblies (see assembly_is_valid). Subassemblies are removed (see remove_subassemblies).</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.common_function_assembly_products">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">common_function_assembly_products</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_results_function</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#common_function_assembly_products"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.common_function_assembly_products" title="Link to this definition">#</a></dt>
<dd><p>Common function to avoid code duplication. Could be simplified further
once SingleFragmentAssembly and Assembly are merged.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em> or </em><em>None</em>) – Minimum overlap length required, or None if not applicable</p></li>
<li><p><strong>algorithm</strong> (<em>Callable</em>) – Function that determines valid overlaps between fragments</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – If True, only return circular assemblies</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.gibson_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">gibson_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#gibson_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.gibson_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for Gibson assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum overlap length required, by default 25</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.in_fusion_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">in_fusion_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#in_fusion_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.in_fusion_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for in-fusion assembly. This is the same as Gibson
assembly, but with a different name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum overlap length required, by default 25</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.fusion_pcr_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">fusion_pcr_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#fusion_pcr_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.fusion_pcr_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for fusion PCR assembly. This is the same as Gibson
assembly, but with a different name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum overlap length required, by default 25</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.in_vivo_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">in_vivo_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#in_vivo_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.in_vivo_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for in vivo assembly (IVA), which relies on homologous recombination between the fragments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum overlap length required, by default 25</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.restriction_ligation_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">restriction_ligation_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><span class="pre">_AbstractCut</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_blunt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#restriction_ligation_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.restriction_ligation_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for restriction ligation assembly:</p>
<ul class="simple">
<li><p>Finds cutsites in the fragments</p></li>
<li><p>Finds all products that could be assembled by ligating the fragments based on those cutsites</p></li>
<li><p>Will NOT return products that combine an existing end with an end generated by the same enzyme (see example below)</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>enzymes</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_AbstractCut</em><em>]</em>) – List of restriction enzymes to use</p></li>
<li><p><strong>allow_blunt</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, allow blunt end ligations, by default True</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>In the example below, we plan to assemble a plasmid from a backbone and an insert, using the EcoRI and SalI enzymes.
Note how 2 circular products are returned, one contains the insert (<code class="docutils literal notranslate"><span class="pre">acgt</span></code>)
and the desired part of the backbone (<code class="docutils literal notranslate"><span class="pre">cccccc</span></code>), the other contains the
reversed insert (<code class="docutils literal notranslate"><span class="pre">tgga</span></code>) and the cut-out part of the backbone (<code class="docutils literal notranslate"><span class="pre">aaa</span></code>).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">restriction_ligation_assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span><span class="p">,</span> <span class="n">SalI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">backbone</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;cccGAATTCaaaGTCGACccc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggGAATTCaggtGTCGACgg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">restriction_ligation_assembly</span><span class="p">([</span><span class="n">backbone</span><span class="p">,</span> <span class="n">insert</span><span class="p">],</span> <span class="p">[</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">SalI</span><span class="p">],</span> <span class="n">circular_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o22)</span>
<span class="go">TCGACccccccGAATTCaggtG</span>
<span class="go">AGCTGggggggCTTAAGtccaC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o19)</span>
<span class="go">AATTCaaaGTCGACacctG</span>
<span class="go">TTAAGtttCAGCTGtggaC</span>
</pre></div>
</div>
<p>Note that passing a pre-cut fragment will not work.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">restriction_products</span> <span class="o">=</span> <span class="n">insert</span><span class="o">.</span><span class="n">cut</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">SalI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cut_insert</span> <span class="o">=</span> <span class="n">restriction_products</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">restriction_ligation_assembly</span><span class="p">([</span><span class="n">backbone</span><span class="p">,</span> <span class="n">cut_insert</span><span class="p">],</span> <span class="p">[</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">SalI</span><span class="p">],</span> <span class="n">circular_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">[]</span>
</pre></div>
</div>
<p>It also works with a single fragment, for circularization:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;GAATTCaaaGAATTC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span><span class="n">restriction_ligation_assembly</span><span class="p">([</span><span class="n">seq</span><span class="p">],</span> <span class="p">[</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o9)</span>
<span class="go">AATTCaaaG</span>
<span class="go">TTAAGtttC</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.golden_gate_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">golden_gate_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><span class="pre">_AbstractCut</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_blunt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#golden_gate_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.golden_gate_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for Golden Gate assembly. This is the same as
restriction ligation assembly, but with a different name. Check the documentation
for <code class="docutils literal notranslate"><span class="pre">restriction_ligation_assembly</span></code> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>enzymes</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_AbstractCut</em><em>]</em>) – List of restriction enzymes to use</p></li>
<li><p><strong>allow_blunt</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, allow blunt end ligations, by default True</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>See the example for <code class="docutils literal notranslate"><span class="pre">restriction_ligation_assembly</span></code>.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.ligation_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">ligation_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_blunt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#ligation_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.ligation_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for ligation assembly, as inputs pass the fragments (digested if needed) that
will be ligated.</p>
<p>For most cases, you probably should use <code class="docutils literal notranslate"><span class="pre">restriction_ligation_assembly</span></code> instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>allow_blunt</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, allow blunt end ligations, by default False</p></li>
<li><p><strong>allow_partial_overlap</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, allow partial overlaps between sticky ends, by default False</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>In the example below, we plan to assemble a plasmid from a backbone and an insert,
using the EcoRI enzyme. The insert and insertion site in the backbone are flanked by
EcoRI sites, so there are two possible products depending on the orientation of the insert.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">ligation_assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">backbone</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;cccGAATTCaaaGAATTCccc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">backbone_cut</span> <span class="o">=</span> <span class="n">backbone</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggGAATTCaggtGAATTCgg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert_cut</span> <span class="o">=</span> <span class="n">insert</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">ligation_assembly</span><span class="p">([</span><span class="n">backbone_cut</span><span class="p">,</span> <span class="n">insert_cut</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o22)</span>
<span class="go">AATTCccccccGAATTCaggtG</span>
<span class="go">TTAAGggggggCTTAAGtccaC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o22)</span>
<span class="go">AATTCccccccGAATTCacctG</span>
<span class="go">TTAAGggggggCTTAAGtggaC</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.assembly_is_multi_site">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">assembly_is_multi_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">asm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#assembly_is_multi_site"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.assembly_is_multi_site" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the assembly is a multi-site assembly, False otherwise.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.gateway_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">gateway_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Literal" title="(in Python v3.8)"><span class="pre">Literal</span></a><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'BP'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'LR'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">greedy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multi_site_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#gateway_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.gateway_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for Gateway assembly / Gateway cloning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to assemble</p></li>
<li><p><strong>reaction_type</strong> (<em>Literal</em><em>[</em><em>'BP'</em><em>, </em><em>'LR'</em><em>]</em>) – Type of Gateway reaction</p></li>
<li><p><strong>greedy</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, use greedy gateway consensus sites, by default False</p></li>
<li><p><strong>circular_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return circular assemblies, by default False</p></li>
<li><p><strong>multi_site_only</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, only return products that where 2 sites recombined. Even if input sequences
contain multiple att sites (typically 2), a product could be generated where only one
site recombines. That’s typically not what you want, so you can set this to True to
only return products where both att sites recombined.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Below an example with dummy Gateway sequences, composed with minimal sequences and the consensus
att sites.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">gateway_assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attB1</span> <span class="o">=</span> <span class="s2">&quot;ACAACTTTGTACAAAAAAGCAGAAG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attP1</span> <span class="o">=</span> <span class="s2">&quot;AAAATAATGATTTTATTTGACTGATAGTGACCTGTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAACTTTGTACAAAAAAGCTGAACGAGAAGCGTAAAATGATATAAATATCAATATATTAAATTAGATTTTGCATAAAAAACAGACTACATAATACTGTAAAACACAACATATCCAGTCACTATGAATCAACTACTTAGATGGTATTAGTGACCTGTA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attR1</span> <span class="o">=</span> <span class="s2">&quot;ACAACTTTGTACAAAAAAGCTGAACGAGAAACGTAAAATGATATAAATATCAATATATTAAATTAGATTTTGCATAAAAAACAGACTACATAATACTGTAAAACACAACATATGCAGTCACTATG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attL1</span> <span class="o">=</span> <span class="s2">&quot;CAAATAATGATTTTATTTTGACTGATAGTGACCTGTTCGTTGCAACAAATTGATAAGCAATGCTTTCTTATAATGCCAACTTTGTACAAAAAAGCAGGCT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span> <span class="o">+</span> <span class="n">attB1</span> <span class="o">+</span> <span class="s2">&quot;ccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span> <span class="o">+</span> <span class="n">attP1</span> <span class="o">+</span> <span class="s2">&quot;ccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span> <span class="o">+</span> <span class="n">attR1</span> <span class="o">+</span> <span class="s2">&quot;ccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq4</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span> <span class="o">+</span> <span class="n">attL1</span> <span class="o">+</span> <span class="s2">&quot;ccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products_BP</span> <span class="o">=</span> <span class="n">gateway_assembly</span><span class="p">([</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">],</span> <span class="s2">&quot;BP&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products_LR</span> <span class="o">=</span> <span class="n">gateway_assembly</span><span class="p">([</span><span class="n">seq3</span><span class="p">,</span> <span class="n">seq4</span><span class="p">],</span> <span class="s2">&quot;LR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">products_BP</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">products_LR</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>Now let’s understand the <code class="docutils literal notranslate"><span class="pre">multi_site_only</span></code> parameter. Let’s consider a case where we are swapping fragments
between two plasmids using an LR reaction. Experimentally, we expect to obtain two plasmids, resulting from the
swapping between the two att sites. That’s what we get if we set <code class="docutils literal notranslate"><span class="pre">multi_site_only</span></code> to True.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">attL2</span> <span class="o">=</span> <span class="s1">&#39;aaataatgattttattttgactgatagtgacctgttcgttgcaacaaattgataagcaatgctttcttataatgccaactttgtacaagaaagctg&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attR2</span> <span class="o">=</span> <span class="s1">&#39;accactttgtacaagaaagctgaacgagaaacgtaaaatgatataaatatcaatatattaaattagattttgcataaaaaacagactacataatactgtaaaacacaacatatccagtcactatg&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;cccccc&quot;</span> <span class="o">+</span> <span class="n">attL1</span> <span class="o">+</span> <span class="s2">&quot;ccc&quot;</span> <span class="o">+</span> <span class="n">attL2</span> <span class="o">+</span> <span class="s2">&quot;cccccc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">backbone</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ttttt&quot;</span> <span class="o">+</span> <span class="n">attR1</span> <span class="o">+</span> <span class="s2">&quot;aaa&quot;</span> <span class="o">+</span> <span class="n">attR2</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">gateway_assembly</span><span class="p">([</span><span class="n">insert</span><span class="p">,</span> <span class="n">backbone</span><span class="p">],</span> <span class="s2">&quot;LR&quot;</span><span class="p">,</span> <span class="n">multi_site_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">products</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>However, if we set <code class="docutils literal notranslate"><span class="pre">multi_site_only</span></code> to False, we get 4 products, which also include the intermediate products
where the two plasmids are combined into a single one through recombination of a single att site. This is an
intermediate of the reaction, and typically we don’t want it:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">gateway_assembly</span><span class="p">([</span><span class="n">insert</span><span class="p">,</span> <span class="n">backbone</span><span class="p">],</span> <span class="s2">&quot;LR&quot;</span><span class="p">,</span> <span class="n">multi_site_only</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">products</span><span class="p">])</span>
<span class="go">[469, 237, 232, 469]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.common_function_integration_products">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">common_function_integration_products</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#common_function_integration_products"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.common_function_integration_products" title="Link to this definition">#</a></dt>
<dd><p>Common function to avoid code duplication for integration products.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frags</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – List of DNA fragments to integrate</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em> or </em><em>None</em>) – Minimum overlap length required, or None if not applicable</p></li>
<li><p><strong>algorithm</strong> (<em>Callable</em>) – Function that determines valid overlaps between fragments</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of integrated DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.common_handle_insertion_fragments">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">common_handle_insertion_fragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">inserts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#common_handle_insertion_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.common_handle_insertion_fragments" title="Link to this definition">#</a></dt>
<dd><p>Common function to handle / validate insertion fragments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p></li>
<li><p><strong>inserts</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>] or </em><em>_Dseqrecord</em>) – DNA fragment(s) to insert</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List containing genome and insert fragments</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.common_function_excision_products">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">common_function_excision_products</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Callable" title="(in Python v3.8)"><span class="pre">Callable</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#common_function_excision_products"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.common_function_excision_products" title="Link to this definition">#</a></dt>
<dd><p>Common function to avoid code duplication for excision products.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em> or </em><em>None</em>) – Minimum overlap length required, or None if not applicable</p></li>
<li><p><strong>algorithm</strong> (<em>Callable</em>) – Function that determines valid overlaps between fragments</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of excised DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.homologous_recombination_integration">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">homologous_recombination_integration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">inserts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">40</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#homologous_recombination_integration"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.homologous_recombination_integration" title="Link to this definition">#</a></dt>
<dd><p>Returns the products resulting from the integration of an insert (or inserts joined
through in vivo recombination) into the genome through homologous recombination.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p></li>
<li><p><strong>inserts</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – DNA fragment(s) to insert</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum homology length required, by default 40</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of integrated DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Below an example with a single insert.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">homologous_recombination_integration</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">homology</span> <span class="o">=</span> <span class="s2">&quot;AAGTCCGTTCGTTTTACCTG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;aaaaaa</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">ccccc</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">aaaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">gggg</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">homologous_recombination_integration</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="p">[</span><span class="n">insert</span><span class="p">],</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&#39;aaaaaaAAGTCCGTTCGTTTTACCTGggggAAGTCCGTTCGTTTTACCTGaaaaaa&#39;</span>
</pre></div>
</div>
<p>Below an example with two inserts joined through homology.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">homology2</span> <span class="o">=</span> <span class="s2">&quot;ATTACAGCATGGGAAGAAAGA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert_1</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">gggg</span><span class="si">{</span><span class="n">homology2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert_2</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">homology2</span><span class="si">}</span><span class="s2">cccc</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">homologous_recombination_integration</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="p">[</span><span class="n">insert_1</span><span class="p">,</span> <span class="n">insert_2</span><span class="p">],</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&#39;aaaaaaAAGTCCGTTCGTTTTACCTGggggATTACAGCATGGGAAGAAAGAccccAAGTCCGTTCGTTTTACCTGaaaaaa&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.homologous_recombination_excision">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">homologous_recombination_excision</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">40</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#homologous_recombination_excision"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.homologous_recombination_excision" title="Link to this definition">#</a></dt>
<dd><p>Returns the products resulting from the excision of a fragment from the genome through
homologous recombination.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum homology length required, by default 40</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List containing excised plasmid and remaining genome sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Example of a homologous recombination event, where a plasmid is excised from the
genome (circular sequence of 25 bp), and that part is removed from the genome,
leaving a shorter linear sequence (32 bp).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">homologous_recombination_excision</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">homology</span> <span class="o">=</span> <span class="s2">&quot;AAGTCCGTTCGTTTTACCTG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;aaaaaa</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">ccccc</span><span class="si">{</span><span class="n">homology</span><span class="si">}</span><span class="s2">aaaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span> <span class="o">=</span> <span class="n">homologous_recombination_excision</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">products</span>
<span class="go">[Dseqrecord(o25), Dseqrecord(-32)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.cre_lox_integration">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">cre_lox_integration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">inserts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#cre_lox_integration"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.cre_lox_integration" title="Link to this definition">#</a></dt>
<dd><p>Returns the products resulting from the integration of an insert (or inserts joined
through cre-lox recombination among them) into the genome through cre-lox integration.</p>
<p>Also works with lox66 and lox71 (see <code class="docutils literal notranslate"><span class="pre">pydna.cre_lox</span></code> for more details).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p></li>
<li><p><strong>inserts</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>] or </em><em>_Dseqrecord</em>) – DNA fragment(s) to insert</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of integrated DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Below an example of reversible integration and excision.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">cre_lox_integration</span><span class="p">,</span> <span class="n">cre_lox_excision</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.cre_lox</span><span class="w"> </span><span class="kn">import</span> <span class="n">LOXP_SEQUENCE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cccccc</span><span class="si">{</span><span class="n">LOXP_SEQUENCE</span><span class="si">}</span><span class="s2">aaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOXP_SEQUENCE</span><span class="si">}</span><span class="s2">bbbbb&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">]</span>
<span class="go">[Dseqrecord(-45), Dseqrecord(o39)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">cre_lox_integration</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="p">[</span><span class="n">b</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span>
<span class="go">[Dseqrecord(-84)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span> <span class="o">=</span> <span class="n">cre_lox_excision</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span>
<span class="go">[Dseqrecord(o39), Dseqrecord(-45)]</span>
</pre></div>
</div>
<p>Below an example with lox66 and lox71 (irreversible integration).
Here, the result of excision is still returned because there is a low
probability of it happening, but it’s considered a rare event.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lox66</span> <span class="o">=</span> <span class="s1">&#39;ATAACTTCGTATAGCATACATTATACGAACGGTA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lox71</span> <span class="o">=</span> <span class="s1">&#39;TACCGTTCGTATAGCATACATTATACGAAGTTAT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cccccc</span><span class="si">{</span><span class="n">lox66</span><span class="si">}</span><span class="s2">aaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lox71</span><span class="si">}</span><span class="s2">bbbbb&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">cre_lox_integration</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="p">[</span><span class="n">b</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span>
<span class="go">[Dseqrecord(-84)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span> <span class="o">=</span> <span class="n">cre_lox_excision</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span>
<span class="go">[Dseqrecord(o39), Dseqrecord(-45)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.cre_lox_excision">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">cre_lox_excision</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#cre_lox_excision"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.cre_lox_excision" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for CRE-lox excision.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List containing excised plasmid and remaining genome sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Below an example of reversible integration and excision.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">cre_lox_integration</span><span class="p">,</span> <span class="n">cre_lox_excision</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.cre_lox</span><span class="w"> </span><span class="kn">import</span> <span class="n">LOXP_SEQUENCE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cccccc</span><span class="si">{</span><span class="n">LOXP_SEQUENCE</span><span class="si">}</span><span class="s2">aaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOXP_SEQUENCE</span><span class="si">}</span><span class="s2">bbbbb&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">]</span>
<span class="go">[Dseqrecord(-45), Dseqrecord(o39)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">cre_lox_integration</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="p">[</span><span class="n">b</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span>
<span class="go">[Dseqrecord(-84)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span> <span class="o">=</span> <span class="n">cre_lox_excision</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span>
<span class="go">[Dseqrecord(o39), Dseqrecord(-45)]</span>
</pre></div>
</div>
<p>Below an example with lox66 and lox71 (irreversible integration).
Here, the result of excision is still returned because there is a low
probability of it happening, but it’s considered a rare event.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lox66</span> <span class="o">=</span> <span class="s1">&#39;ATAACTTCGTATAGCATACATTATACGAACGGTA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lox71</span> <span class="o">=</span> <span class="s1">&#39;TACCGTTCGTATAGCATACATTATACGAAGTTAT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cccccc</span><span class="si">{</span><span class="n">lox66</span><span class="si">}</span><span class="s2">aaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lox71</span><span class="si">}</span><span class="s2">bbbbb&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">cre_lox_integration</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="p">[</span><span class="n">b</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span>
<span class="go">[Dseqrecord(-84)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span> <span class="o">=</span> <span class="n">cre_lox_excision</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res2</span>
<span class="go">[Dseqrecord(o39), Dseqrecord(-45)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.crispr_integration">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">crispr_integration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">inserts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">40</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#crispr_integration"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.crispr_integration" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for CRISPR integration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>_Dseqrecord</em>) – Target genome sequence</p></li>
<li><p><strong>inserts</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Dseqrecord</em><em>]</em>) – DNA fragment(s) to insert</p></li>
<li><p><strong>guides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_Primer</em><em>]</em>) – List of guide RNAs as Primer objects. This may change in the future.</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum overlap length required, by default 40</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of integrated DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">crispr_integration</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaccggttcaatgcaaacagtaatgatggatgacattcaaagcac&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;genome&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaccggttAAAAAAAAAttcaaagcac&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;insert&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">guide</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;ttcaatgcaaacagtaatga&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;guide&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">product</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">crispr_integration</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="p">[</span><span class="n">insert</span><span class="p">],</span> <span class="p">[</span><span class="n">guide</span><span class="p">],</span> <span class="mi">8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">product</span>
<span class="go">Dseqrecord(-27)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.assembly2.pcr_assembly">
<span class="sig-prename descclassname"><span class="pre">pydna.assembly2.</span></span><span class="sig-name descname"><span class="pre">pcr_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">fwd_primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">rvs_primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_primer_features</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">14</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatches</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/assembly2.html#pcr_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.assembly2.pcr_assembly" title="Link to this definition">#</a></dt>
<dd><p>Returns the products for PCR assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<em>_Dseqrecord</em>) – Template sequence</p></li>
<li><p><strong>fwd_primer</strong> (<em>_Primer</em>) – Forward primer</p></li>
<li><p><strong>rvs_primer</strong> (<em>_Primer</em>) – Reverse primer</p></li>
<li><p><strong>add_primer_features</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, add primer features to the product, by default False</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Minimum overlap length required, by default 14</p></li>
<li><p><strong>mismatches</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Maximum number of mismatches, by default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of assembled DNA molecules</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[_Dseqrecord]</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-pydna.codon">
<span id="pydna-codon-module"></span><h2>pydna.codon module<a class="headerlink" href="#module-pydna.codon" title="Link to this heading">#</a></h2>
<p>docstring.</p>
</section>
<section id="module-pydna.common_sub_strings">
<span id="pydna-common-sub-strings-module"></span><h2>pydna.common_sub_strings module<a class="headerlink" href="#module-pydna.common_sub_strings" title="Link to this heading">#</a></h2>
<p>This module is based on the Py-rstr-max package that
was written by Romain Brixtel (rbrixtel_at_gmail_dot_com)
(<a class="reference external" href="https://brixtel.users.greyc.fr">https://brixtel.users.greyc.fr</a>) and is available from
<a class="reference external" href="https://code.google.com/p/py-rstr-max">https://code.google.com/p/py-rstr-max</a>
<a class="github reference external" href="https://github.com/gip0/py-rstr-max">gip0/py-rstr-max</a>
the original code was covered by an MIT licence.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.common_sub_strings.common_sub_strings">
<span class="sig-prename descclassname"><span class="pre">pydna.common_sub_strings.</span></span><span class="sig-name descname"><span class="pre">common_sub_strings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stringx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">stringy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/common_sub_strings.html#common_sub_strings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.common_sub_strings.common_sub_strings" title="Link to this definition">#</a></dt>
<dd><p>Finds all common substrings between stringx and stringy, and returns
them sorted by length.</p>
<p>This function is case sensitive.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stringx</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>stringy</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>[(startx1, starty1, length1),(startx2, starty2, length2), …]</p>
<p>startx1 = startposition in x, where substring 1 starts
starty1 = position in y where substring 1 starts
length1 = lenght of substring</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a> of <a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.common_sub_strings.terminal_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.common_sub_strings.</span></span><span class="sig-name descname"><span class="pre">terminal_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stringx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">stringy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/common_sub_strings.html#terminal_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.common_sub_strings.terminal_overlap" title="Link to this definition">#</a></dt>
<dd><p>Finds the the flanking common substrings between stringx and stringy
longer than limit. This means that the results only contains substrings
that starts or ends at the the ends of stringx and stringy.</p>
<p>This function is case sensitive.</p>
<p>returns a list of tuples describing the substrings
The list is sorted longest -&gt; shortest.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stringx</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>stringy</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p></li>
<li><p><strong>limit</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>[(startx1,starty1,length1),(startx2,starty2,length2), …]</p>
<p>startx1 = startposition in x, where substring 1 starts
starty1 = position in y where substring 1 starts
length1 = lenght of substring</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a> of <a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.common_sub_strings</span><span class="w"> </span><span class="kn">import</span> <span class="n">terminal_overlap</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">terminal_overlap</span><span class="p">(</span><span class="s2">&quot;agctatgtatcttgcatcgta&quot;</span><span class="p">,</span> <span class="s2">&quot;gcatcgtagtctatttgcttac&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="go">[(13, 0, 8)]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>             <span class="o">&lt;--</span> <span class="mi">8</span> <span class="o">-&gt;</span>
<span class="o">&lt;----</span> <span class="mi">13</span> <span class="o">---&gt;</span>
<span class="n">agctatgtatcttgcatcgta</span>                    <span class="n">stringx</span>
             <span class="n">gcatcgtagtctatttgcttac</span>      <span class="n">stringy</span>
             <span class="mi">0</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.conftest">
<span id="pydna-conftest-module"></span><h2>pydna.conftest module<a class="headerlink" href="#module-pydna.conftest" title="Link to this heading">#</a></h2>
<p>conftest.py</p>
</section>
<section id="module-pydna.contig">
<span id="pydna-contig-module"></span><h2>pydna.contig module<a class="headerlink" href="#module-pydna.contig" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.contig.Contig">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.contig.</span></span><span class="sig-name descname"><span class="pre">Contig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/contig.html#Contig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<p>This class holds information about a DNA assembly. This class is instantiated by
the <code class="xref py py-class docutils literal notranslate"><span class="pre">Assembly</span></code> class and is not meant to be used directly.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.from_string">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/contig.html#Contig.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.from_string" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.from_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/contig.html#Contig.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.from_SeqRecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/contig.html#Contig.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.contig.Contig.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.detailed_figure">
<span class="sig-name descname"><span class="pre">detailed_figure</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/contig.html#Contig.detailed_figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.detailed_figure" title="Link to this definition">#</a></dt>
<dd><p>Returns a text representation of the assembled fragments.</p>
<p>Linear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">acgatgctatactgCCCCCtgtgctgtgctcta</span>
                   <span class="n">TGTGCTGTGCTCTA</span>
                   <span class="n">tgtgctgtgctctaTTTTTtattctggctgtatc</span>
</pre></div>
</div>
<p>Circular:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">||||||||||||||</span>
<span class="n">acgatgctatactgCCCCCtgtgctgtgctcta</span>
                   <span class="n">TGTGCTGTGCTCTA</span>
                   <span class="n">tgtgctgtgctctaTTTTTtattctggctgtatc</span>
                                      <span class="n">TATTCTGGCTGTATC</span>
                                      <span class="n">tattctggctgtatcGGGGGtacgatgctatactg</span>
                                                           <span class="n">ACGATGCTATACTG</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.contig.Contig.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/contig.html#Contig.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.contig.Contig.figure" title="Link to this definition">#</a></dt>
<dd><p>Compact ascii representation of the assembled fragments.</p>
<p>Each fragment is represented by:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Size</span> <span class="n">of</span> <span class="n">common</span> <span class="mi">5</span><span class="s1">&#39; substring|Name and size of DNA fragment|</span>
<span class="n">Size</span> <span class="n">of</span> <span class="n">common</span> <span class="mi">5</span><span class="s1">&#39; substring</span>
</pre></div>
</div>
<p>Linear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">frag20</span><span class="o">|</span> <span class="mi">6</span>
       \\<span class="o">/</span>
       <span class="o">/</span>\\
        <span class="mi">6</span><span class="o">|</span><span class="n">frag23</span><span class="o">|</span> <span class="mi">6</span>
                 \\<span class="o">/</span>
                 <span class="o">/</span>\\
                  <span class="mi">6</span><span class="o">|</span><span class="n">frag14</span>
</pre></div>
</div>
<p>Circular:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="o">-|</span><span class="mi">2577</span><span class="o">|</span><span class="mi">61</span>
<span class="o">|</span>       \\<span class="o">/</span>
<span class="o">|</span>       <span class="o">/</span>\\
<span class="o">|</span>       <span class="mi">61</span><span class="o">|</span><span class="mi">5681</span><span class="o">|</span><span class="mi">98</span>
<span class="o">|</span>               \\<span class="o">/</span>
<span class="o">|</span>               <span class="o">/</span>\\
<span class="o">|</span>               <span class="mi">98</span><span class="o">|</span><span class="mi">2389</span><span class="o">|</span><span class="mi">557</span>
<span class="o">|</span>                       \\<span class="o">/</span>
<span class="o">|</span>                       <span class="o">/</span>\\
<span class="o">|</span>                       <span class="mi">557</span><span class="o">-</span>
<span class="o">|</span>                          <span class="o">|</span>
 <span class="o">--------------------------</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.cre_lox">
<span id="pydna-cre-lox-module"></span><h2>pydna.cre_lox module<a class="headerlink" href="#module-pydna.cre_lox" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.cre_lox.cre_loxP_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.cre_lox.</span></span><span class="sig-name descname"><span class="pre">cre_loxP_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">_l</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/cre_lox.html#cre_loxP_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.cre_lox.cre_loxP_overlap" title="Link to this definition">#</a></dt>
<dd><p>Find matching loxP sites between two sequences.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.cre_lox.get_regex_dict">
<span class="sig-prename descclassname"><span class="pre">pydna.cre_lox.</span></span><span class="sig-name descname"><span class="pre">get_regex_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">original_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/cre_lox.html#get_regex_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.cre_lox.get_regex_dict" title="Link to this definition">#</a></dt>
<dd><p>Get the regex dictionary for the original dictionary.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.cre_lox.find_loxP_sites">
<span class="sig-prename descclassname"><span class="pre">pydna.cre_lox.</span></span><span class="sig-name descname"><span class="pre">find_loxP_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/cre_lox.html#find_loxP_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.cre_lox.find_loxP_sites" title="Link to this definition">#</a></dt>
<dd><p>Find all loxP sites in a sequence and return a dictionary with the name and positions of the sites.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.cre_lox.annotate_loxP_sites">
<span class="sig-prename descclassname"><span class="pre">pydna.cre_lox.</span></span><span class="sig-name descname"><span class="pre">annotate_loxP_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/cre_lox.html#annotate_loxP_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.cre_lox.annotate_loxP_sites" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.crispr">
<span id="pydna-crispr-module"></span><h2>pydna.crispr module<a class="headerlink" href="#module-pydna.crispr" title="Link to this heading">#</a></h2>
<p>Provides the Dseq class for handling double stranded DNA sequences.</p>
<p>Dseq is a subclass of <a class="reference external" href="https://biopython.org/docs/latest/api/Bio.Seq.html#Bio.Seq.Seq" title="(in Biopython v1.86)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></a>. The Dseq class
is mostly useful as a part of the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a> class
which can hold more meta data.</p>
<p>The Dseq class support the notion of circular and linear DNA topology.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.crispr.cas9">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.crispr.</span></span><span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protospacer</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/crispr.html#cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.crispr.cas9" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">_cas</span></code></p>
<p>docstring.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="o">|----</span><span class="n">size</span><span class="o">----------|</span>

    <span class="o">---</span><span class="n">protospacer</span><span class="o">------</span>
                    <span class="o">-</span><span class="n">fst3</span>
    <span class="n">fst5</span>             <span class="o">|-|</span>
    <span class="o">|--------------|</span>
                        <span class="n">PAM</span>
<span class="mi">5</span><span class="o">-</span><span class="n">NNGGAAGAGTAATACACTA</span><span class="o">-</span><span class="n">AAANGGNN</span><span class="o">-</span><span class="mi">3</span>
<span class="o">|||||||||||||||||||</span> <span class="o">||||||||</span>
<span class="mi">3</span><span class="o">-</span><span class="n">NNCCTTCTCATTATGTGAT</span><span class="o">-</span><span class="n">TTTNCCNN</span><span class="o">-</span><span class="mi">5</span>
    <span class="o">|||||||||||||||||</span> <span class="o">|||</span>
<span class="mi">5</span><span class="o">-</span><span class="n">GGAAGAGTAATACACTA</span><span class="o">-</span><span class="n">AAAg</span><span class="o">-</span><span class="n">u</span><span class="o">-</span><span class="n">a</span><span class="o">-</span><span class="n">a</span><span class="o">-</span><span class="n">g</span><span class="o">-</span><span class="n">g</span>  <span class="n">Scaffold</span>
    <span class="o">---</span><span class="n">gRNA</span> <span class="n">spacer</span><span class="o">---</span>    <span class="n">u</span><span class="o">-</span><span class="n">a</span>
                        <span class="n">u</span><span class="o">-</span><span class="n">a</span>
                        <span class="n">u</span><span class="o">-</span><span class="n">a</span>
                        <span class="n">u</span><span class="o">-</span><span class="n">a</span>
                        <span class="n">a</span><span class="o">-</span><span class="n">u</span>
                        <span class="n">g</span><span class="o">-</span><span class="n">u</span><span class="o">-</span><span class="n">g</span>
                        <span class="n">a</span>    <span class="n">a</span>
                        <span class="n">g</span><span class="o">-</span><span class="n">c</span><span class="o">-</span><span class="n">a</span>
                        <span class="n">c</span><span class="o">-</span><span class="n">g</span>
                        <span class="n">u</span><span class="o">-</span><span class="n">a</span>
                        <span class="n">a</span><span class="o">-</span><span class="n">u</span>
                        <span class="n">g</span>   <span class="n">a</span>  <span class="n">tetraloop</span>
                        <span class="n">a</span><span class="o">-</span><span class="n">a</span>
</pre></div>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.crispr.cas9.scaffold">
<span class="sig-name descname"><span class="pre">scaffold</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGG'</span></em><a class="headerlink" href="#pydna.crispr.cas9.scaffold" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.crispr.cas9.pam">
<span class="sig-name descname"><span class="pre">pam</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'.GG'</span></em><a class="headerlink" href="#pydna.crispr.cas9.pam" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.crispr.cas9.size">
<span class="sig-name descname"><span class="pre">size</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">20</span></em><a class="headerlink" href="#pydna.crispr.cas9.size" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.crispr.cas9.fst5">
<span class="sig-name descname"><span class="pre">fst5</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">17</span></em><a class="headerlink" href="#pydna.crispr.cas9.fst5" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.crispr.cas9.fst3">
<span class="sig-name descname"><span class="pre">fst3</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">-3</span></em><a class="headerlink" href="#pydna.crispr.cas9.fst3" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.crispr.cas9.ovhg">
<span class="sig-name descname"><span class="pre">ovhg</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#pydna.crispr.cas9.ovhg" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.crispr.cas9.search">
<span class="sig-name descname"><span class="pre">search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linear</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/crispr.html#cas9.search"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.crispr.cas9.search" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.crispr.protospacer">
<span class="sig-prename descclassname"><span class="pre">pydna.crispr.</span></span><span class="sig-name descname"><span class="pre">protospacer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">guide_construct</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">cas9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/crispr.html#protospacer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.crispr.protospacer" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</section>
<section id="module-pydna.design">
<span id="pydna-design-module"></span><h2>pydna.design module<a class="headerlink" href="#module-pydna.design" title="Link to this heading">#</a></h2>
<p>This module contain functions for primer design for various purposes.</p>
<ul class="simple">
<li><p>:func:primer_design for designing primers for a sequence or a matching primer for an existing primer. Returns an <code class="xref py py-class docutils literal notranslate"><span class="pre">Amplicon</span></code> object (same as the <code class="xref py py-mod docutils literal notranslate"><span class="pre">amplify</span></code> module returns).</p></li>
<li><p>:func:assembly_fragments Adds tails to primers for a linear assembly through homologous recombination or Gibson assembly.</p></li>
<li><p>:func:circular_assembly_fragments Adds tails to primers for a circular assembly through homologous recombination or Gibson assembly.</p></li>
</ul>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.design.primer_design">
<span class="sig-prename descclassname"><span class="pre">pydna.design.</span></span><span class="sig-name descname"><span class="pre">primer_design</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">55.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">estimate_function</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/design.html#primer_design"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.design.primer_design" title="Link to this definition">#</a></dt>
<dd><p>This function designs a forward primer and a reverse primer for PCR amplification
of a given template sequence.</p>
<p>The template argument is a Dseqrecord object or equivalent containing the template sequence.</p>
<p>The optional fp and rp arguments can contain an existing primer for the sequence (either the forward or reverse primer).
One or the other primers can be specified, not both (since then there is nothing to design!, use the pydna.amplify.pcr function instead).</p>
<p>The limit argument is the minimum length of the primer. The default value is 13.</p>
<p>If one of the primers is given, the other primer is designed to match in terms of Tm.
If both primers are designed, they will be designed to target_tm</p>
<p>tm_func is a function that takes an ascii string representing an oligonuceotide as argument and returns a float.
Some useful functions can be found in the <a class="reference internal" href="#module-pydna.tm" title="pydna.tm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.tm</span></code></a> module, but can be substituted for a custom made function.</p>
<p>estimate_function is a tm_func-like function that is used to get a first guess for the primer design, that is then used as starting
point for the final result. This is useful when the tm_func function is slow to calculate (e.g. it relies on an
external API, such as the NEB primer design API). The estimate_function should be faster than the tm_func function.
The default value is <cite>None</cite>.
To use the default <cite>tm_func</cite> as estimate function to get the NEB Tm faster, you can do:
<cite>primer_design(dseqr, target_tm=55, tm_func=tm_neb, estimate_function=tm_default)</cite>.</p>
<p>The function returns a pydna.amplicon.Amplicon class instance. This object has
the object.forward_primer and object.reverse_primer properties which contain the designed primers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><em>pydna.dseqrecord.Dseqrecord</em></a>) – a Dseqrecord object. The only required argument.</p></li>
<li><p><strong>fp</strong> (<a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><em>pydna.primer.Primer</em></a><em>, </em><em>optional</em>) – optional pydna.primer.Primer objects containing one primer each.</p></li>
<li><p><strong>rp</strong> (<a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><em>pydna.primer.Primer</em></a><em>, </em><em>optional</em>) – optional pydna.primer.Primer objects containing one primer each.</p></li>
<li><p><strong>target_tm</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a><em>, </em><em>optional</em>) – target tm for the primers, set to 55°C by default.</p></li>
<li><p><strong>tm_func</strong> (<em>function</em>) – Function used for tm calculation. This function takes an ascii string
representing an oligonuceotide as argument and returns a float.
Some useful functions can be found in the <a class="reference internal" href="#module-pydna.tm" title="pydna.tm"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.tm</span></code></a> module, but can be
substituted for a custom made function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>result</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon">Amplicon</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">Dseqrecord(-64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.design</span><span class="w"> </span><span class="kn">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span>
<span class="go">Amplicon(64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">f64 17-mer:5&#39;-atgactgctaacccttc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 18-mer:5&#39;-catcgtaagtttcgaacg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ampl</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">5atgactgctaacccttc...cgttcgaaacttacgatg3</span>
<span class="go">                     ||||||||||||||||||</span>
<span class="go">                    3gcaagctttgaatgctac5</span>
<span class="go">5atgactgctaacccttc3</span>
<span class="go"> |||||||||||||||||</span>
<span class="go">3tactgacgattgggaag...gcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">f64 23-mer:5&#39;-GGATCCatgactgct..ttc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">r64 24-mer:5&#39;-GGATCCcatcgtaag..acg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.amplify</span><span class="w"> </span><span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span> <span class="n">pr</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">      5atgactgctaacccttc...cgttcgaaacttacgatg3</span>
<span class="go">                           ||||||||||||||||||</span>
<span class="go">                          3gcaagctttgaatgctacCCTAGG5</span>
<span class="go">5GGATCCatgactgctaacccttc3</span>
<span class="go">       |||||||||||||||||</span>
<span class="go">      3tactgacgattgggaag...gcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGGATCC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttg&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;myprimer&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">fp</span> <span class="o">=</span> <span class="n">pf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">myprimer 27-mer:5&#39;-atgactgctaaccct..ttg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 32-mer:5&#39;-catcgtaagtttcga..atc-3&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.design.assembly_fragments">
<span class="sig-prename descclassname"><span class="pre">pydna.design.</span></span><span class="sig-name descname"><span class="pre">assembly_fragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlink</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/design.html#assembly_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.design.assembly_fragments" title="Link to this definition">#</a></dt>
<dd><p>This function return a list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects where
primers have been modified with tails so that the fragments can be fused in
the order they appear in the list by for example Gibson assembly or homologous
recombination.</p>
<p>Given that we have two linear <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects a and b</p>
<p>we can modify the reverse primer of a and forward primer of b with tails to allow
fusion by fusion PCR, Gibson assembly or in-vivo homologous recombination.
The basic requirements for the primers for the three techniques are the same.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>
<span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>
                  <span class="o">|||||</span>
                 <span class="o">&lt;</span><span class="n">gacgt</span>
<span class="n">agcct</span><span class="o">&gt;</span>
<span class="o">|||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                        <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
                       <span class="o">/</span>                     \
                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                                         <span class="o">|||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
                       <span class="n">TTTAT</span><span class="o">&gt;</span>
                       <span class="o">|||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

<span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
\<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>Design tailed primers incorporating a part of the next or previous fragment to be assembled.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATA</span>

<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>

                <span class="n">ctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by any of the techniques mentioned earlier to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing
Amplicons and other similar objects.</p>
<p><strong>At least every second sequence object needs to be an Amplicon</strong></p>
<p>This rule exists because if a sequence object is that is not a PCR product
is to be fused with another fragment, that other fragment needs to be an Amplicon
so that the primer of the other object can be modified to include the whole stretch
of sequence homology needed for the fusion. See the example below where a is a
non-amplicon (a linear plasmid  vector for instance)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>           <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
<span class="o">/</span>                     \         <span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>   <span class="o">&lt;--&gt;</span>  <span class="n">TTTATATCGCATGACTCTTCTTT</span>
<span class="o">|||||||||||||||||||||||</span>         <span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>                          <span class="o">&lt;</span><span class="n">AGAAA</span>
                                <span class="n">TTTAT</span><span class="o">&gt;</span>
                                <span class="o">|||||||||||||||||||||||</span>
                          <span class="o">&lt;--&gt;</span>  <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

     <span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
     <span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
     <span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
     \<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>In this case only the forward primer of b is fitted with a tail with a part a:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
         <span class="n">tcttggtctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by for example Gibson assembly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaacca</span>

                             <span class="n">TCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing
Amplicons and other similar objects.</p>
<p>The overlap argument controls how many base pairs of overlap required between
adjacent sequence fragments. In the junction between Amplicons, tails with the
length of about half of this value is added to the two primers
closest to the junction.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;       &lt;
Amplicon1
         Amplicon2
         &gt;       &lt;

         ⇣

&gt;       &lt;-
Amplicon1
         Amplicon2
        -&gt;       &lt;
</pre></div>
</div>
<p>In the case of an Amplicon adjacent to a Dseqrecord object, the tail will
be twice as long (1*overlap) since the
recombining sequence is present entirely on this primer:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Dseqrecd1
         Amplicon1
         &gt;       &lt;

         ⇣

Dseqrecd1
         Amplicon1
       --&gt;       &lt;
</pre></div>
</div>
<p>Note that if the sequence of DNA fragments starts or stops with an Amplicon,
the very first and very last prinmer will not be modified i.e. assembles are
always assumed to be linear. There are simple tricks around that for circular
assemblies depicted in the last two examples below.</p>
<p>The maxlink arguments controls the cut off length for sequences that will be
synhtesized by adding them to primers for the adjacent fragment(s). The
argument list may contain short spacers (such as spacers between fusion proteins).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Example 1: Linear assembly of PCR products (pydna.amplicon.Amplicon class objects) ------

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon4
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;-       -&gt;       &lt;-                      pydna.assembly.Assembly
Amplicon1         Amplicon3
         Amplicon2         Amplicon4     ➤  Amplicon1Amplicon2Amplicon3Amplicon4
        -&gt;       &lt;-       -&gt;       &lt;

Example 2: Linear assembly of alternating Amplicons and other fragments

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;--     --&gt;       &lt;--                     pydna.assembly.Assembly
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2     ➤  Amplicon1Dseqrecd1Amplicon2Dseqrecd2

Example 3: Linear assembly of alternating Amplicons and other fragments

Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2
         &gt;       &lt;       --&gt;       &lt;

                     ⇣
             pydna.design.assembly_fragments
                     ⇣
                                                  pydna.assembly.Assembly
Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2     ➤  Dseqrecd1Amplicon1Dseqrecd2Amplicon2
       --&gt;       &lt;--     --&gt;       &lt;

Example 4: Circular assembly of alternating Amplicons and other fragments

                 -&gt;       &lt;==
Dseqrecd1         Amplicon2
         Amplicon1         Dseqrecd1
       --&gt;       &lt;-
                     ⇣
                     pydna.design.assembly_fragments
                     ⇣
                                                   pydna.assembly.Assembly
                 -&gt;       &lt;==
Dseqrecd1         Amplicon2                    -Dseqrecd1Amplicon1Amplicon2-
         Amplicon1                       ➤    |                             |
       --&gt;       &lt;-                            -----------------------------

------ Example 5: Circular assembly of Amplicons

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣

&gt;       &lt;=       -&gt;       &lt;-
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
        -&gt;       &lt;-       +&gt;       &lt;

                     ⇣
             make new Amplicon using the Amplicon1.template and
             the last fwd primer and the first rev primer.
                     ⇣
                                                   pydna.assembly.Assembly
+&gt;       &lt;=       -&gt;       &lt;-
 Amplicon1         Amplicon3                  -Amplicon1Amplicon2Amplicon3-
          Amplicon2                      ➤   |                             |
         -&gt;       &lt;-                          -----------------------------
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects) – list Amplicon and Dseqrecord object for which fusion primers should be constructed.</p></li>
<li><p><strong>overlap</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Length of required overlap between fragments.</p></li>
<li><p><strong>maxlink</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – Maximum length of spacer sequences that may be present in f. These will be included in tails for designed primers.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, the assembly is circular. If False, the assembly is linear.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>seqs</strong> –</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Amplicon1</span><span class="p">,</span>
 <span class="n">Amplicon2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.design</span><span class="w"> </span><span class="kn">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ccaaacccaccaggtaccttatgtaagtacttcaagtcgccagaagacttcttggtcaagttgcc&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtactggtgctgaaccttgtatcaagttgggtgttgacgccattgccccaggtggtcgtttcgtt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.design</span><span class="w"> </span><span class="kn">import</span> <span class="n">assembly_fragments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># We would like a circular recombination, so the first sequence has to be repeated</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa1</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">,</span><span class="n">fa2</span> <span class="o">=</span> <span class="n">assembly_fragments</span><span class="p">([</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">a</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Since all fragments are Amplicons, we need to extract the rp of the 1st and fp of the last fragments.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.amplify</span><span class="w"> </span><span class="kn">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">fa2</span><span class="o">.</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">fa1</span><span class="o">.</span><span class="n">reverse_primer</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">]</span>
<span class="go">[Amplicon(100), Amplicon(101), Amplicon(102)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">fb</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">fc</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;fa fb fc&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly</span><span class="w"> </span><span class="kn">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">([</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 100bp 101bp 102bp</span>
<span class="go">limit(bp)....: 25</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_linear</span><span class="p">()</span>
<span class="go">[Contig(-231), Contig(-166), Contig(-36)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;cdseguid=85t6tfcvWav0wnXEIb-lkUtrl4s&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;cdseguid=85t6tfcvWav0wnXEIb-lkUtrl4s&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go"> -|fa|36</span>
<span class="go">|     \/</span>
<span class="go">|     /\</span>
<span class="go">|     36|fb|36</span>
<span class="go">|           \/</span>
<span class="go">|           /\</span>
<span class="go">|           36|fc|36</span>
<span class="go">|                 \/</span>
<span class="go">|                 /\</span>
<span class="go">|                 36-</span>
<span class="go">|                    |</span>
<span class="go"> --------------------</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.design.circular_assembly_fragments">
<span class="sig-prename descclassname"><span class="pre">pydna.design.</span></span><span class="sig-name descname"><span class="pre">circular_assembly_fragments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlink</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/design.html#circular_assembly_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.design.circular_assembly_fragments" title="Link to this definition">#</a></dt>
<dd><p>Equivalent to <cite>assembly_fragments</cite> with <cite>circular=True</cite>.</p>
<p>Deprecated, kept for backward compatibility. Use <cite>assembly_fragments</cite> with <cite>circular=True</cite> instead.</p>
</dd></dl>

</section>
<section id="module-pydna.download">
<span id="pydna-download-module"></span><h2>pydna.download module<a class="headerlink" href="#module-pydna.download" title="Link to this heading">#</a></h2>
<p>Provides a function for downloading online text files.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.download.download_text">
<span class="sig-prename descclassname"><span class="pre">pydna.download.</span></span><span class="sig-name descname"><span class="pre">download_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/download.html#download_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.download.download_text" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</section>
<section id="module-pydna.dseq">
<span id="pydna-dseq-module"></span><h2>pydna.dseq module<a class="headerlink" href="#module-pydna.dseq" title="Link to this heading">#</a></h2>
<p>Provides the Dseq class for handling double stranded DNA sequences.</p>
<p>Dseq is a subclass of <a class="reference external" href="https://biopython.org/docs/latest/api/Bio.Seq.html#Bio.Seq.Seq" title="(in Biopython v1.86)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></a>. The Dseq class
is mostly useful as a part of the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a> class
which can hold more meta data.</p>
<p>The Dseq class support the notion of circular and linear DNA topology.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.dseq.Dseq">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.dseq.</span></span><span class="sig-name descname"><span class="pre">Dseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">watson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ovhg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><code class="xref py py-class docutils literal notranslate"><span class="pre">Seq</span></code></a></p>
<p>Dseq holds information for a double stranded DNA fragment.</p>
<p>Dseq also holds information describing the topology of
the DNA fragment (linear or circular).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>watson</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a string representing the watson (sense) DNA strand.</p></li>
<li><p><strong>crick</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><em>optional</em>) – a string representing the crick (antisense) DNA strand.</p></li>
<li><p><strong>ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – A positive or negative number to describe the stagger between the
watson and crick strands.
see below for a detailed explanation.</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True indicates that sequence is linear, False that it is circular.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True indicates that sequence is circular, False that it is linear.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Dseq is a subclass of the Biopython Seq object. It stores two
strings representing the watson (sense) and crick(antisense) strands.
two properties called linear and circular, and a numeric value ovhg
(overhang) describing the stagger for the watson and crick strand
in the 5’ end of the fragment.</p>
<p>The most common usage is probably to create a Dseq object as a
part of a Dseqrecord object (see <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a>).</p>
<p>There are three ways of creating a Dseq object directly listed below, but you can also
use the function Dseq.from_full_sequence_and_overhangs() to create a Dseq:</p>
<p>Only one argument (string):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>The given string will be interpreted as the watson strand of a
blunt, linear double stranded sequence object. The crick strand
is created automatically from the watson strand.</p>
<p>Two arguments (string, string):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gggaaat&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">gggaaat</span>
<span class="go">   ttt</span>
</pre></div>
</div>
<p>If both watson and crick are given, but not ovhg an attempt
will be made to find the best annealing between the strands.
There are limitations to this. For long fragments it is quite
slow. The length of the annealing sequences have to be at least
half the length of the shortest of the strands.</p>
<p>Three arguments (string, string, ovhg=int):</p>
<p>The ovhg parameter is an integer describing the length of the
crick strand overhang in the 5’ end of the molecule.</p>
<p>The ovhg parameter controls the stagger at the five prime end:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">dsDNA</span>       <span class="n">overhang</span>

  <span class="n">nnn</span><span class="o">...</span>    <span class="mi">2</span>
<span class="n">nnnnn</span><span class="o">...</span>

 <span class="n">nnnn</span><span class="o">...</span>    <span class="mi">1</span>
<span class="n">nnnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>    <span class="mi">0</span>
<span class="n">nnnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>   <span class="o">-</span><span class="mi">1</span>
 <span class="n">nnnn</span><span class="o">...</span>

<span class="n">nnnnn</span><span class="o">...</span>   <span class="o">-</span><span class="mi">2</span>
  <span class="n">nnn</span><span class="o">...</span>
</pre></div>
</div>
<p>Example of creating Dseq objects with different amounts of stagger:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span> <span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">agt</span>
<span class="go"> attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go"> agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">  agt</span>
<span class="go">attca</span>
</pre></div>
</div>
<p>If the ovhg parameter is specified a crick strand also
needs to be supplied, otherwise an exception is raised.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna_/dsdna.py&quot;</span>, line <span class="m">169</span>, in <span class="n">__init__</span>
<span class="w">    </span><span class="k">else</span><span class="p">:</span>
<span class="gr">ValueError</span>: <span class="n">ovhg defined without crick strand!</span>
</pre></div>
</div>
<p>The shape of the fragment is set by circular = True, False</p>
<p>Note that both ends of the DNA fragment has to be compatible to set
circular = True.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span> <span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;tttcccc&quot;</span><span class="p">,</span><span class="s2">&quot;aaacccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-11)</span>
<span class="go">    tttcccc</span>
<span class="go">ccccaaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">4</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ccccttt&quot;</span><span class="p">,</span><span class="s2">&quot;ccccaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-11)</span>
<span class="go">ccccttt</span>
<span class="go">    aaacccc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">-4</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>Coercing to string</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">&#39;ggggtttcccc&#39;</span>
</pre></div>
</div>
<p>A Dseq object can be longer that either the watson or crick strands.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>

<span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
      <span class="n">GATCCTTT</span>
<span class="n">AAAGCCCTA</span>
</pre></div>
</div>
<p>The slicing of a linear Dseq object works mostly as it does for a string.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;at&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;gac&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-1)</span>
<span class="go">a</span>
<span class="go">t</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">Dseq(-2)</span>
<span class="go">at</span>
<span class="go">ta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Dseq(-0)</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">Dseq(-3)</span>
<span class="go">gac</span>
<span class="go">ctg</span>
</pre></div>
</div>
<p>The slicing of a circular Dseq object has a slightly different meaning.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggAtCc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggAtCc</span>
<span class="go">ccTaGg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-5)</span>
<span class="go">CcggA</span>
<span class="go">GgccT</span>
</pre></div>
</div>
<p>The slice [X:X] produces an empty slice for a string, while this
will return the linearized sequence starting at X:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggatcc</span>
<span class="go">cctagg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-6)</span>
<span class="go">tccgga</span>
<span class="go">aggcct</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.trunc">
<span class="sig-name descname"><span class="pre">trunc</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">30</span></em><a class="headerlink" href="#pydna.dseq.Dseq.trunc" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.quick">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">quick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">watson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">ovhg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.quick"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.quick" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.from_string">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_string" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.from_representation">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_representation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dsdna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.from_representation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_representation" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.from_full_sequence_and_overhangs">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_full_sequence_and_overhangs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">full_sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">crick_ovhg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">watson_ovhg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.from_full_sequence_and_overhangs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_full_sequence_and_overhangs" title="Link to this definition">#</a></dt>
<dd><p>Create a linear Dseq object from a full sequence and the 3’ overhangs of each strand.</p>
<p>The order of the parameters is like this because the 3’ overhang of the crick strand is the one
on the left side of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>full_sequence</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The full sequence of the Dseq object.</p></li>
<li><p><strong>crick_ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The overhang of the crick strand in the 3’ end. Equivalent to Dseq.ovhg.</p></li>
<li><p><strong>watson_ovhg</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The overhang of the watson strand in the 5’ end.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Dseq object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">  AAAA</span>
<span class="go">TTTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">AAAAAA</span>
<span class="go">  TT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">  AA</span>
<span class="go">TTTTTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;AAAAAA&#39;</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">AAAA</span>
<span class="go">  TTTT</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.mw">
<span class="sig-name descname"><span class="pre">mw</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.mw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.mw" title="Link to this definition">#</a></dt>
<dd><p>This method returns the molecular weight of the DNA molecule
in g/mol. The following formula is used:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MW</span> <span class="o">=</span> <span class="p">(</span><span class="n">A</span> <span class="n">x</span> <span class="mf">313.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">T</span> <span class="n">x</span> <span class="mf">304.2</span><span class="p">)</span> <span class="o">+</span>
     <span class="p">(</span><span class="n">C</span> <span class="n">x</span> <span class="mf">289.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">G</span> <span class="n">x</span> <span class="mf">329.2</span><span class="p">)</span> <span class="o">+</span>
     <span class="p">(</span><span class="n">N</span> <span class="n">x</span> <span class="mf">308.9</span><span class="p">)</span> <span class="o">+</span> <span class="mf">79.0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.upper">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.upper" title="Link to this definition">#</a></dt>
<dd><p>Return an upper case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">AAA</span>
<span class="go">TTT</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseq object in uppercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.lower" title="pydna.dseq.Dseq.lower"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.lower</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.lower">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.lower" title="Link to this definition">#</a></dt>
<dd><p>Return a lower case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseq object in lowercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.upper" title="pydna.dseq.Dseq.upper"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.upper</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sub</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_SeqAbstractBaseClass</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_sys.maxsize</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.find" title="Link to this definition">#</a></dt>
<dd><p>This method behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<p>Returns -1 if the subsequence is NOT found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sub</strong> (<em>string</em><em> or </em><em>Seq object</em>) – a string or another Seq object to look for.</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – slice start.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – slice end.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;atcgactgacgtgtt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">atcgactgacgtgtt</span>
<span class="go">tagctgactgcacaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;gac&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;taa&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Dseq object where watson and crick have switched places.</p>
<p>This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="headerlink" href="#pydna.dseq.Dseq.rc" title="Link to this definition">#</a></dt>
<dd><p>Dseq object where watson and crick have switched places.</p>
<p>This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.shifted">
<span class="sig-name descname"><span class="pre">shifted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.shifted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.shifted" title="Link to this definition">#</a></dt>
<dd><p>Shifted version of a circular Dseq object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.looped">
<span class="sig-name descname"><span class="pre">looped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.looped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.looped" title="Link to this definition">#</a></dt>
<dd><p>Circularized Dseq object.</p>
<p>This can only be done if the two ends are compatible,
otherwise a TypeError is raised.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o7)</span>
<span class="go">catcgat</span>
<span class="go">gtagcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;</span>, line <span class="m">357</span>, in <span class="n">looped</span>
<span class="w">    </span><span class="k">if</span> <span class="n">type5</span> <span class="o">==</span> <span class="n">type3</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky5</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)):</span>
<span class="gr">TypeError</span>: <span class="n">DNA cannot be circularized.</span>
<span class="x">5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.tolinear">
<span class="sig-name descname"><span class="pre">tolinear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.tolinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.tolinear" title="Link to this definition">#</a></dt>
<dd><p>Returns a blunt, linear copy of a circular Dseq object. This can
only be done if the Dseq object is circular, otherwise a
TypeError is raised.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[:]</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.five_prime_end">
<span class="sig-name descname"><span class="pre">five_prime_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.five_prime_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.five_prime_end" title="Link to this definition">#</a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.three_prime_end" title="pydna.dseq.Dseq.three_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.three_prime_end</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.three_prime_end">
<span class="sig-name descname"><span class="pre">three_prime_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.three_prime_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.three_prime_end" title="Link to this definition">#</a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.five_prime_end" title="pydna.dseq.Dseq.five_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.five_prime_end</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.watson_ovhg">
<span class="sig-name descname"><span class="pre">watson_ovhg</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.watson_ovhg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.watson_ovhg" title="Link to this definition">#</a></dt>
<dd><p>Returns the overhang of the watson strand at the three prime.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.fill_in">
<span class="sig-name descname"><span class="pre">fill_in</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.fill_in"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.fill_in" title="Link to this definition">#</a></dt>
<dd><p>Fill in of five prime protruding end with a DNA polymerase
that has only DNA polymerase activity (such as exo-klenow <a class="footnote-reference brackets" href="#id12" id="id15" role="doc-noteref"><span class="fn-bracket">[</span>7<span class="fn-bracket">]</span></a>)
and any combination of A, G, C or T. Default are all four
nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;tac&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id16" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id19">9</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment">http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.transcribe">
<span class="sig-name descname"><span class="pre">transcribe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><span class="pre">Seq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.transcribe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.transcribe" title="Link to this definition">#</a></dt>
<dd><p>Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.</p>
<p>Following the usual convention, the sequence is interpreted as the
coding strand of the DNA double helix, not the template strand. This
means we can get the RNA sequence just by switching T to U.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">transcribe</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
<p>Trying to transcribe an RNA sequence has no effect.
If you have a nucleotide sequence which might be DNA or RNA
(or even a mixture), calling the transcribe method will ensure
any T becomes U.</p>
<p>Trying to transcribe a protein sequence will replace any
T for Threonine with U for Selenocysteine, which has no
biologically plausible rational.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRU&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><span class="pre">Seq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.translate" title="Link to this definition">#</a></dt>
<dd><p>Translate..</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.mung">
<span class="sig-name descname"><span class="pre">mung</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.mung"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.mung" title="Link to this definition">#</a></dt>
<dd><p>Simulates treatment a nuclease with 5’-3’ and 3’-5’ single
strand specific exonuclease activity (such as mung bean nuclease <a class="footnote-reference brackets" href="#id14" id="id17" role="doc-noteref"><span class="fn-bracket">[</span>8<span class="fn-bracket">]</span></a>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="n">ggatcc</span>    <span class="o">-&gt;</span>     <span class="n">gatcc</span>
     <span class="n">ctaggg</span>          <span class="n">ctagg</span>

     <span class="n">ggatcc</span>   <span class="o">-&gt;</span>      <span class="n">ggatc</span>
    <span class="n">tcctag</span>            <span class="n">cctag</span>

<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
<span class="n">caaa</span>
 <span class="n">tttc</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
 <span class="n">aaac</span>
<span class="n">gttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id18" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id21">10</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/Mung_bean_nuclease">http://en.wikipedia.org/wiki/Mung_bean_nuclease</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.T4">
<span class="sig-name descname"><span class="pre">T4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.T4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.T4" title="Link to this definition">#</a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.t4">
<span class="sig-name descname"><span class="pre">t4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="headerlink" href="#pydna.dseq.Dseq.t4" title="Link to this definition">#</a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>nucleotides</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.exo1_front">
<span class="sig-name descname"><span class="pre">exo1_front</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.exo1_front"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.exo1_front" title="Link to this definition">#</a></dt>
<dd><p>5’-3’ resection at the start (left side) of the molecule.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.exo1_end">
<span class="sig-name descname"><span class="pre">exo1_end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DseqType</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.exo1_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.exo1_end" title="Link to this definition">#</a></dt>
<dd><p>5’-3’ resection at the end (right side) of the molecule.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.no_cutters">
<span class="sig-name descname"><span class="pre">no_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.no_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.no_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch not cutting sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.unique_cutters">
<span class="sig-name descname"><span class="pre">unique_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.unique_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.unique_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.once_cutters">
<span class="sig-name descname"><span class="pre">once_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="headerlink" href="#pydna.dseq.Dseq.once_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.twice_cutters">
<span class="sig-name descname"><span class="pre">twice_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.twice_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.twice_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence twice.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.n_cutters">
<span class="sig-name descname"><span class="pre">n_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.n_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.n_cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting n times.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cutters">
<span class="sig-name descname"><span class="pre">cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">RestrictionBatch</span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cutters" title="Link to this definition">#</a></dt>
<dd><p>Enzymes in a RestrictionBatch cutting sequence at least once.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.seguid" title="Link to this definition">#</a></dt>
<dd><p>SEGUID checksum for the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.isblunt">
<span class="sig-name descname"><span class="pre">isblunt</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.isblunt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.isblunt" title="Link to this definition">#</a></dt>
<dd><p>isblunt.</p>
<p>Return True if Dseq is linear and blunt and
false if staggered or circular.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">gat</span>
<span class="go">cta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="s2">&quot;atcg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> gat</span>
<span class="go">gcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="s2">&quot;gatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">gat</span>
<span class="go">ctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o3)</span>
<span class="go">gat</span>
<span class="go">cta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isblunt</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cas9">
<span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RNA</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#slice" title="(in Python v3.8)"><span class="pre">slice</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cas9" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.terminal_transferase">
<span class="sig-name descname"><span class="pre">terminal_transferase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.terminal_transferase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.terminal_transferase" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cut">
<span class="sig-name descname"><span class="pre">cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">EnzymesType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><span class="pre">DseqType</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cut" title="Link to this definition">#</a></dt>
<dd><p>Returns a list of linear Dseq fragments produced in the digestion.
If there are no cuts, an empty list is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzymes</strong> (<em>enzyme object</em><em> or </em><em>iterable</em><em> of </em><em>such objects</em>) – A Bio.Restriction.XXX restriction objects or iterable.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>frags</strong> – list of Dseq objects formed by the digestion</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ggatccnnngaattc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">BamHI</span><span class="p">,</span><span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">))</span>
<span class="go">&lt;class &#39;tuple&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">):</span> <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">frag</span><span class="p">))</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="go">Dseq(-14)</span>
<span class="go">gatccnnngaattc</span>
<span class="go">    gnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span> <span class="o">==</span>  <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">,</span> <span class="n">EcoRI</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.cutsite_is_valid">
<span class="sig-name descname"><span class="pre">cutsite_is_valid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.cutsite_is_valid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.cutsite_is_valid" title="Link to this definition">#</a></dt>
<dd><p>Returns False if:
- Cut positions fall outside the sequence (could be moved to Biopython)
- Overhang is not double stranded
- Recognition site is not double stranded or is outside the sequence
- For enzymes that cut twice, it checks that at least one possibility is valid</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.get_cutsites">
<span class="sig-name descname"><span class="pre">get_cutsites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">EnzymesType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.get_cutsites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.get_cutsites" title="Link to this definition">#</a></dt>
<dd><p>Returns a list of cutsites, represented represented as <cite>((cut_watson, ovhg), enz)</cite>:</p>
<ul class="simple">
<li><p><cite>cut_watson</cite> is a positive integer contained in <cite>[0,len(seq))</cite>, where <cite>seq</cite> is the sequence
that will be cut. It represents the position of the cut on the watson strand, using the full
sequence as a reference. By “full sequence” I mean the one you would get from <cite>str(Dseq)</cite>.</p></li>
<li><p><cite>ovhg</cite> is the overhang left after the cut. It has the same meaning as <cite>ovhg</cite> in
the <cite>Bio.Restriction</cite> enzyme objects, or pydna’s <cite>Dseq</cite> property.</p></li>
<li><dl class="simple">
<dt><cite>enz</cite> is the enzyme object. It’s not necessary to perform the cut, but can be</dt><dd><p>used to keep track of which enzyme was used.</p>
</dd>
</dl>
</li>
</ul>
<p>Cuts are only returned if the recognition site and overhang are on the double-strand
part of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>enzymes</strong> (<em>Union</em><em>[</em><em>_RestrictionBatch</em><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><em>_AbstractCut</em><em>]</em><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;AAGAATTCAAGAATTC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">)</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
</pre></div>
</div>
<p><cite>cut_watson</cite> is defined with respect to the “full sequence”, not the
watson strand:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaa&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span>
<span class="go">Dseq(-10)</span>
<span class="go"> aGAATTCaa</span>
<span class="go">ttCTTAAGtt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[((3, -4), EcoRI)]</span>
</pre></div>
</div>
<p>Cuts are only returned if the recognition site and overhang are on the double-strand
part of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[((1, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="o">.</span><span class="n">from_full_sequence_and_overhangs</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="go">[]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.left_end_position">
<span class="sig-name descname"><span class="pre">left_end_position</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.left_end_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.left_end_position" title="Link to this definition">#</a></dt>
<dd><p>The index in the full sequence of the watson and crick start positions.</p>
<p>full sequence (str(self)) for all three cases is AAA</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AAA</span>              <span class="n">AA</span>               <span class="n">AAT</span>
 <span class="n">TT</span>             <span class="n">TTT</span>               <span class="n">TTT</span>
<span class="n">Returns</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="n">Returns</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>    <span class="n">Returns</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.right_end_position">
<span class="sig-name descname"><span class="pre">right_end_position</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.right_end_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.right_end_position" title="Link to this definition">#</a></dt>
<dd><p>The index in the full sequence of the watson and crick end positions.</p>
<p>full sequence (str(self)) for all three cases is AAA</p>
<p><code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">AAA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AA</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">AAA</span>
<span class="pre">TT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TTT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TTT</span>
<span class="pre">Returns</span> <span class="pre">(3,</span> <span class="pre">2)</span>&#160;&#160;&#160; <span class="pre">Returns</span> <span class="pre">(2,</span> <span class="pre">3)</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Returns</span> <span class="pre">(3,</span> <span class="pre">3)</span>
<span class="pre">`</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.get_cut_parameters">
<span class="sig-name descname"><span class="pre">get_cut_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_left</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.get_cut_parameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.get_cut_parameters" title="Link to this definition">#</a></dt>
<dd><p>For a given cut expressed as ((cut_watson, ovhg), enz), returns
a tuple (cut_watson, cut_crick, ovhg).</p>
<ul class="simple">
<li><p>cut_watson: see get_cutsites docs</p></li>
<li><p>cut_crick: equivalent of cut_watson in the crick strand</p></li>
<li><p>ovhg: see get_cutsites docs</p></li>
</ul>
<p>The cut can be None if it represents the left or right end of the sequence.
Then it will return the position of the watson and crick ends with respect
to the “full sequence”. The <cite>is_left</cite> parameter is only used in this case.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.apply_cut">
<span class="sig-name descname"><span class="pre">apply_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><span class="pre">Dseq</span></a></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.apply_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.apply_cut" title="Link to this definition">#</a></dt>
<dd><p>Extracts a subfragment of the sequence between two cuts.</p>
<p>For more detail see the documentation of get_cutsite_pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left_cut</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>, </em><em>None</em><em>]</em>)</p></li>
<li><p><strong>right_cut</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>, </em><em>None</em><em>]</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq">Dseq</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaaGAATTCaa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">p3</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span>
<span class="go">(None, ((3, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p1</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">aaG</span>
<span class="go">ttCTTAA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span>
<span class="go">(((3, -4), EcoRI), ((11, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p2</span><span class="p">)</span>
<span class="go">Dseq(-12)</span>
<span class="go">AATTCaaG</span>
<span class="go">    GttCTTAA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p3</span>
<span class="go">(((11, -4), EcoRI), None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">p3</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">AATTCaa</span>
<span class="go">    Gtt</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;TTCaaGAA&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((6, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair</span>
<span class="go">(((6, -4), EcoRI), ((6, -4), EcoRI))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">pair</span><span class="p">)</span>
<span class="go">Dseq(-12)</span>
<span class="go">AATTCaaG</span>
<span class="go">    GttCTTAA</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseq.Dseq.get_cutsite_pairs">
<span class="sig-name descname"><span class="pre">get_cutsite_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutsites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AbstractCut</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_cas</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/dseq.html#Dseq.get_cutsite_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseq.Dseq.get_cutsite_pairs" title="Link to this definition">#</a></dt>
<dd><p>Returns pairs of cutsites that render the edges of the resulting fragments.</p>
<p>A fragment produced by restriction is represented by a tuple of length 2 that
may contain cutsites or <cite>None</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><p>Two cutsites: represents the extraction of a fragment between those two
cutsites, in that orientation. To represent the opening of a circular
molecule with a single cutsite, we put the same cutsite twice.</p></li>
<li><p><cite>None</cite>, cutsite: represents the extraction of a fragment between the left
edge of linear sequence and the cutsite.</p></li>
<li><p>cutsite, <cite>None</cite>: represents the extraction of a fragment between the cutsite
and the right edge of a linear sequence.</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cutsites</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><em>tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>, </em><em>_AbstractCut</em><em>]</em><em>]</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]|None],<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)">tuple</a>[<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)">int</a>], _AbstractCut]|None]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;aaGAATTCaaGAATTCaa&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="go">[(None, ((3, -4), EcoRI)), (((3, -4), EcoRI), ((11, -4), EcoRI)), (((11, -4), EcoRI), None)]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s1">&#39;TTCaaGAA&#39;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span> <span class="o">=</span> <span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">([</span><span class="n">EcoRI</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cutsites</span>
<span class="go">[((6, -4), EcoRI)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseq</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
<span class="go">[(((6, -4), EcoRI), ((6, -4), EcoRI))]</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.dseqrecord">
<span id="pydna-dseqrecord-module"></span><h2>pydna.dseqrecord module<a class="headerlink" href="#module-pydna.dseqrecord" title="Link to this heading">#</a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a> class, for handling double stranded
DNA sequences. The Dseqrecord holds sequence information in the form of a <a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseq.Dseq</span></code></a>
object. The Dseq and Dseqrecord classes are subclasses of Biopythons
Seq and SeqRecord classes, respectively.</p>
<p>The Dseq and Dseqrecord classes support the notion of circular and linear DNA topology.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.dseqrecord.</span></span><span class="sig-name descname"><span class="pre">Dseqrecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<p>Dseqrecord is a double stranded version of the Biopython SeqRecord <a class="footnote-reference brackets" href="#id16" id="id19" role="doc-noteref"><span class="fn-bracket">[</span>9<span class="fn-bracket">]</span></a> class.
The Dseqrecord object holds a Dseq object describing the sequence.
Additionally, Dseqrecord hold meta information about the sequence in the
from of a list of SeqFeatures, in the same way as the SeqRecord does.</p>
<p>The Dseqrecord can be initialized with a string, Seq, Dseq, SeqRecord
or another Dseqrecord. The sequence information will be stored in a
Dseq object in all cases.</p>
<p>Dseqrecord objects can be read or parsed from sequences in FASTA, EMBL or Genbank formats.
See the <a class="reference internal" href="#module-pydna.readers" title="pydna.readers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.readers</span></code></a> and <a class="reference internal" href="#module-pydna.parsers" title="pydna.parsers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.parsers</span></code></a> modules for further information.</p>
<p>There is a short representation associated with the Dseqrecord.
<code class="docutils literal notranslate"><span class="pre">Dseqrecord(-3)</span></code> represents a linear sequence of length 2
while <code class="docutils literal notranslate"><span class="pre">Dseqrecord(o7)</span></code>
represents a circular sequence of length 7.</p>
<p>Dseqrecord and Dseq share the same concept of length. This length can be larger
than each strand alone if they are staggered as in the example below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<em>string</em><em>, </em><a class="reference internal" href="#pydna.seq.Seq" title="pydna.seq.Seq"><em>Seq</em></a><em>, </em><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><em>SeqRecord</em></a><em>, </em><a class="reference internal" href="#pydna.all.Dseq" title="pydna.all.Dseq"><em>Dseq</em></a><em> or </em><em>other Dseqrecord object</em>) – This data will be used to form the seq property</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True or False reflecting the shape of the DNA molecule</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – True or False reflecting the shape of the DNA molecule</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id20" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">11</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</aside>
</aside>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.source">
<span class="sig-name descname"><span class="pre">source</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#pydna.opencloning_models.Source" title="pydna.opencloning_models.Source"><span class="pre">Source</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.source" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.from_string">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.from_string" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.from_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><span class="pre">SeqRecord</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5e-14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.from_SeqRecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.circular">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">circular</span></span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.circular" title="Link to this definition">#</a></dt>
<dd><p>The circular property can not be set directly.
Use <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.looped" title="pydna.dseqrecord.Dseqrecord.looped"><code class="xref py py-meth docutils literal notranslate"><span class="pre">looped()</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.m">
<span class="sig-name descname"><span class="pre">m</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.m"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.m" title="Link to this definition">#</a></dt>
<dd><p>This method returns the mass of the DNA molecule in grams. This is
calculated as the product between the molecular weight of the Dseq object
and the</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.extract_feature">
<span class="sig-name descname"><span class="pre">extract_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.extract_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.extract_feature" title="Link to this definition">#</a></dt>
<dd><p>Extracts a feature and creates a new Dseqrecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates the feature to extract</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-2)</span>
<span class="go">gt</span>
<span class="go">ca</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'misc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.add_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.add_feature" title="Link to this definition">#</a></dt>
<dd><p>Add a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates start of the feature</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates end of the feature</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4), strand=1), type=&#39;misc&#39;, qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.seguid" title="Link to this definition">#</a></dt>
<dd><p>Url safe SEGUID for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be part of a URL.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;ldseguid=TEwydy0ugvGXh3VJnVwgtxoyDQA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.looped">
<span class="sig-name descname"><span class="pre">looped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.looped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.looped" title="Link to this definition">#</a></dt>
<dd><p>Circular version of the Dseqrecord object.</p>
<p>The underlying linear Dseq object has to have compatible ends.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.looped" title="pydna.dseq.Dseq.looped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.looped</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.tolinear">
<span class="sig-name descname"><span class="pre">tolinear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.tolinear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.tolinear" title="Link to this definition">#</a></dt>
<dd><p>Returns a linear, blunt copy of a circular Dseqrecord object. The
underlying Dseq object has to be circular.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="p">[:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.terminal_transferase">
<span class="sig-name descname"><span class="pre">terminal_transferase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.terminal_transferase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.terminal_transferase" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.format">
<span class="sig-name descname"><span class="pre">format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gb'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.format" title="Link to this definition">#</a></dt>
<dd><p>Returns the sequence as a string using a format supported by Biopython
SeqIO <a class="footnote-reference brackets" href="#id18" id="id21" role="doc-noteref"><span class="fn-bracket">[</span>10<span class="fn-bracket">]</span></a>. Default is “gb” which is short for Genbank.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;gb&quot;</span><span class="p">))</span>
<span class="go">LOCUS       name                       3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  description.</span>
<span class="go">ACCESSION   id</span>
<span class="go">VERSION     id</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id22" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id28">12</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gb'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.write" title="Link to this definition">#</a></dt>
<dd><p>Writes the Dseqrecord to a file using the format f, which must
be a format supported by Biopython SeqIO for writing <a class="footnote-reference brackets" href="#id20" id="id23" role="doc-noteref"><span class="fn-bracket">[</span>11<span class="fn-bracket">]</span></a>. Default
is “gb” which is short for Genbank. Note that Biopython SeqIO reads
more formats than it writes.</p>
<p>Filename is the path to the file where the sequece is to be
written. The filename is optional, if it is not given, the
description property (string) is used together with the format.</p>
<p>If obj is the Dseqrecord object, the default file name will be:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;.&lt;f&gt;</span></code></p>
<p>Where &lt;f&gt; is “gb” by default. If the filename already exists and
AND the sequence it contains is different, a new file name will be
used so that the old file is not lost:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;_NEW.&lt;f&gt;</span></code></p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id24" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id30">13</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.find_aminoacids">
<span class="sig-name descname"><span class="pre">find_aminoacids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.find_aminoacids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find_aminoacids" title="Link to this definition">#</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.find_aa">
<span class="sig-name descname"><span class="pre">find_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find_aa" title="Link to this definition">#</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.map_trace_files">
<span class="sig-name descname"><span class="pre">map_trace_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pth</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.map_trace_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.map_trace_files" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.linearize">
<span class="sig-name descname"><span class="pre">linearize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.linearize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.linearize" title="Link to this definition">#</a></dt>
<dd><p>Similar to <cite>:func:cut</cite>.</p>
<p>Throws an exception if there is not excactly one cut
i.e. none or more than one digestion products.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.no_cutters">
<span class="sig-name descname"><span class="pre">no_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.no_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.no_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.unique_cutters">
<span class="sig-name descname"><span class="pre">unique_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.unique_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.unique_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.once_cutters">
<span class="sig-name descname"><span class="pre">once_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.once_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.once_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.twice_cutters">
<span class="sig-name descname"><span class="pre">twice_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.twice_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.twice_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.n_cutters">
<span class="sig-name descname"><span class="pre">n_cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.n_cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.n_cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.cutters">
<span class="sig-name descname"><span class="pre">cutters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">RestrictionBatch</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.cutters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cutters" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.number_of_cuts">
<span class="sig-name descname"><span class="pre">number_of_cuts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.number_of_cuts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.number_of_cuts" title="Link to this definition">#</a></dt>
<dd><p>The number of cuts by digestion with the Restriction enzymes
contained in the iterable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.cas9">
<span class="sig-name descname"><span class="pre">cas9</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RNA</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.cas9"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cas9" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.synced">
<span class="sig-name descname"><span class="pre">synced</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.synced"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.synced" title="Link to this definition">#</a></dt>
<dd><p>This method returns a new circular sequence (Dseqrecord object), which has been rotated
in such a way that there is maximum overlap between the sequence and
ref, which may be a string, Biopython Seq, SeqRecord object or
another Dseqrecord object.</p>
<p>The reason for using this could be to rotate a new recombinant plasmid so
that it starts at the same position after cloning. See the example below:</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;gaat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">gaat</span>
<span class="go">ctta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">a</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-4)</span>
<span class="go">atga</span>
<span class="go">tact</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;CCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span> <span class="o">=</span> <span class="p">(</span><span class="n">d</span><span class="o">+</span><span class="n">insert</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">atgaCCC</span>
<span class="go">tactGGG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">synced</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">gaCCCat</span>
<span class="go">ctGGGta</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.upper">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.upper" title="Link to this definition">#</a></dt>
<dd><p>Returns an uppercase copy.
&gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord
&gt;&gt;&gt; my_seq = Dseqrecord(“aAa”)
&gt;&gt;&gt; my_seq.seq
Dseq(-3)
aAa
tTt
&gt;&gt;&gt; upper = my_seq.upper()
&gt;&gt;&gt; upper.seq
Dseq(-3)
AAA
TTT
&gt;&gt;&gt;</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseqrecord object in uppercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.lower" title="pydna.dseqrecord.Dseqrecord.lower"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.lower</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.lower">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.lower" title="Link to this definition">#</a></dt>
<dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aAa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aAa</span>
<span class="go">tTt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">upper</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">upper</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">AAA</span>
<span class="go">TTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lower</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lower</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dseqrecord object in lowercase</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.upper" title="pydna.dseqrecord.Dseqrecord.upper"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.upper</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.orfs">
<span class="sig-name descname"><span class="pre">orfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.orfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.orfs" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.orfs_to_features">
<span class="sig-name descname"><span class="pre">orfs_to_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.orfs_to_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.orfs_to_features" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard">
<span class="sig-name descname"><span class="pre">copy_gb_to_clipboard</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.copy_gb_to_clipboard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard">
<span class="sig-name descname"><span class="pre">copy_fasta_to_clipboard</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.copy_fasta_to_clipboard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.figure">
<span class="sig-name descname"><span class="pre">figure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\x1b[48;5;11m'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\x1b[0m'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.figure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.figure" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.shifted">
<span class="sig-name descname"><span class="pre">shifted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.shifted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.shifted" title="Link to this definition">#</a></dt>
<dd><p>Circular Dseqrecord with a new origin &lt;shift&gt;.</p>
<p>This only works on circular Dseqrecords. If we consider the following
circular sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">GAAAT</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">watson</span> <span class="pre">strand</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">CTTTA</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">crick</span> <span class="pre">strand</span></code></div>
</div>
<p>The T and the G on the watson strand are linked together as well
as the A and the C of the of the crick strand.</p>
<p>if <code class="docutils literal notranslate"><span class="pre">shift</span></code> is 1, this indicates a new origin at position 1:</p>
<div class="line-block">
<div class="line-block">
<div class="line">new origin at the | symbol:</div>
<div class="line"><br /></div>
</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">G|AAAT</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">C|TTTA</span></code></div>
</div>
<p>new sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">AAATG</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">TTTAC</span></code></div>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aaat</span>
<span class="go">ttta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">shifted</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aata</span>
<span class="go">ttat</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.cut">
<span class="sig-name descname"><span class="pre">cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">enzymes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cut" title="Link to this definition">#</a></dt>
<dd><p>Digest a Dseqrecord object with one or more restriction enzymes.</p>
<p>returns a list of linear Dseqrecords. If there are no cuts, an empty
list is returned.</p>
<p>See also <code class="xref py py-func docutils literal notranslate"><span class="pre">Dseq.cut()</span></code>
:param enzymes: A Bio.Restriction.XXX restriction object or iterable of such.
:type enzymes: enzyme object or iterable of such objects</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>Dseqrecord_frags</strong> – list of Dseqrecord objects formed by the digestion</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">BamHI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="go">(Dseqrecord(-5), Dseqrecord(-5))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="p">,</span> <span class="n">frag2</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag2</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">gatcc</span>
<span class="go">    g</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.annotations">
<span class="sig-name descname"><span class="pre">annotations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">_AnnotationsDict</span></em><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.annotations" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.dbxrefs">
<span class="sig-name descname"><span class="pre">dbxrefs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.dbxrefs" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.apply_cut">
<span class="sig-name descname"><span class="pre">apply_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.apply_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.apply_cut" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.dseqrecord.Dseqrecord.history">
<span class="sig-name descname"><span class="pre">history</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/dseqrecord.html#Dseqrecord.history"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.history" title="Link to this definition">#</a></dt>
<dd><p>Returns a string representation of the cloning history of the sequence.
Returns an empty string if the sequence has no source.</p>
<p>Check the documentation notebooks for extensive examples.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>str</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>A string representation of the cloning history of the sequence.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.assembly2</span><span class="w"> </span><span class="kn">import</span> <span class="n">gibson_assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fragments</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>   <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;TTTTacgatAAtgctccCCCC&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment1&quot;</span><span class="p">),</span>
<span class="gp">... </span>   <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;CCCCtcatGGGG&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment2&quot;</span><span class="p">),</span>
<span class="gp">... </span>   <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;GGGGatataTTTT&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fragment3&quot;</span><span class="p">),</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">product</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">gibson_assembly</span><span class="p">(</span><span class="n">fragments</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">product</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;product_name&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">product</span><span class="o">.</span><span class="n">history</span><span class="p">())</span>
<span class="go">╙── product_name (Dseqrecord(o34))</span>
<span class="go">    └─╼ GibsonAssemblySource</span>
<span class="go">        ├─╼ fragment1 (Dseqrecord(-21))</span>
<span class="go">        ├─╼ fragment2 (Dseqrecord(-12))</span>
<span class="go">        └─╼ fragment3 (Dseqrecord(-13))</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.fakeseq">
<span id="pydna-fakeseq-module"></span><h2>pydna.fakeseq module<a class="headerlink" href="#module-pydna.fakeseq" title="Link to this heading">#</a></h2>
<p>docstring.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.fakeseq.FakeSeq">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.fakeseq.</span></span><span class="sig-name descname"><span class="pre">FakeSeq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50e-15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/fakeseq.html#FakeSeq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.fakeseq.FakeSeq" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>docstring.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.fakeseq.FakeSeq.m">
<span class="sig-name descname"><span class="pre">m</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/fakeseq.html#FakeSeq.m"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.fakeseq.FakeSeq.m" title="Link to this definition">#</a></dt>
<dd><p>Mass of the DNA molecule in grams.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.fakeseq.FakeSeq.M">
<span class="sig-name descname"><span class="pre">M</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/fakeseq.html#FakeSeq.M"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.fakeseq.FakeSeq.M" title="Link to this definition">#</a></dt>
<dd><p>M grams/mol.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.fusionpcr">
<span id="pydna-fusionpcr-module"></span><h2>pydna.fusionpcr module<a class="headerlink" href="#module-pydna.fusionpcr" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.fusionpcr.fuse_by_pcr">
<span class="sig-prename descclassname"><span class="pre">pydna.fusionpcr.</span></span><span class="sig-name descname"><span class="pre">fuse_by_pcr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/fusionpcr.html#fuse_by_pcr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.fusionpcr.fuse_by_pcr" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.fusionpcr.list_parts">
<span class="sig-prename descclassname"><span class="pre">pydna.fusionpcr.</span></span><span class="sig-name descname"><span class="pre">list_parts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fusion_pcr_fragment</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/fusionpcr.html#list_parts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.fusionpcr.list_parts" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.gateway">
<span id="pydna-gateway-module"></span><h2>pydna.gateway module<a class="headerlink" href="#module-pydna.gateway" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.gateway.gateway_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.gateway.</span></span><span class="sig-name descname"><span class="pre">gateway_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">greedy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/gateway.html#gateway_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gateway.gateway_overlap" title="Link to this definition">#</a></dt>
<dd><p>Find gateway overlaps. If greedy is True, it uses a more greedy consensus site to find attP sites,
which might give false positives</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.gateway.find_gateway_sites">
<span class="sig-prename descclassname"><span class="pre">pydna.gateway.</span></span><span class="sig-name descname"><span class="pre">find_gateway_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">greedy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.SimpleLocation" title="(in Biopython v1.86)"><span class="pre">SimpleLocation</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/gateway.html#find_gateway_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gateway.find_gateway_sites" title="Link to this definition">#</a></dt>
<dd><p>Find all gateway sites in a sequence and return a dictionary with the name and positions of the sites.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.gateway.annotate_gateway_sites">
<span class="sig-prename descclassname"><span class="pre">pydna.gateway.</span></span><span class="sig-name descname"><span class="pre">annotate_gateway_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">greedy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/gateway.html#annotate_gateway_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gateway.annotate_gateway_sites" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.gel">
<span id="pydna-gel-module"></span><h2>pydna.gel module<a class="headerlink" href="#module-pydna.gel" title="Link to this heading">#</a></h2>
<p>docstring.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.gel.interpolator">
<span class="sig-prename descclassname"><span class="pre">pydna.gel.</span></span><span class="sig-name descname"><span class="pre">interpolator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mwstd</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/gel.html#interpolator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gel.interpolator" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.gel.gel">
<span class="sig-prename descclassname"><span class="pre">pydna.gel.</span></span><span class="sig-name descname"><span class="pre">gel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gel_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">margin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interpolator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">interpolator(mwstd=_mwstd)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/gel.html#gel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.gel.gel" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.genbank">
<span id="pydna-genbank-module"></span><h2>pydna.genbank module<a class="headerlink" href="#module-pydna.genbank" title="Link to this heading">#</a></h2>
<p>This module provides a class for downloading sequences from genbank
called Genbank and an function that does the same thing called genbank.</p>
<p>The function can be used if the environmental variable <strong>pydna_email</strong> has
been set to a valid email address. The easiest way to do this permanantly is to edit the
<cite>pydna.ini</cite> file. See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.genbank.Genbank">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.genbank.</span></span><span class="sig-name descname"><span class="pre">Genbank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">users_email</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">tool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pydna'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbank.html#Genbank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbank.Genbank" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Class to facilitate download from genbank. It is easier and
quicker to use the <a class="reference internal" href="#pydna.genbank.genbank" title="pydna.genbank.genbank"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.genbank()</span></code></a> function directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>users_email</strong> (<em>string</em>) – Has to be a valid email address. You should always tell
Genbanks who you are, so that they can contact you.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.genbank</span><span class="w"> </span><span class="kn">import</span> <span class="n">Genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gb</span><span class="o">=</span><span class="n">Genbank</span><span class="p">(</span><span class="s2">&quot;bjornjobb@gmail.com&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span> <span class="o">=</span> <span class="n">gb</span><span class="o">.</span><span class="n">nucleotide</span><span class="p">(</span><span class="s2">&quot;LP002422.1&quot;</span><span class="p">)</span>   <span class="c1"># &lt;- entry from genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbank.Genbank.nucleotide">
<span class="sig-name descname"><span class="pre">nucleotide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Literal" title="(in Python v3.8)"><span class="pre">Literal</span></a><span class="p"><span class="pre">[</span></span><span class="m"><span class="pre">1</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="m"><span class="pre">2</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord" title="pydna.genbankrecord.GenbankRecord"><span class="pre">GenbankRecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/genbank.html#Genbank.nucleotide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbank.Genbank.nucleotide" title="Link to this definition">#</a></dt>
<dd><p>This method downloads a genbank nuclotide record from genbank. This method is
cached by default. This can be controlled by editing the <strong>pydna_cached_funcs</strong> environment
variable. The best way to do this permanently is to edit the edit the
<cite>pydna.ini</cite> file. See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<p>Item is a string containing one genbank accession number
for a nucleotide file. Genbank nucleotide accession numbers have this format:</p>
<div class="line-block">
<div class="line">A12345   = 1 letter  + 5 numerals</div>
<div class="line">AB123456 = 2 letters + 6 numerals</div>
</div>
<p>The accession number is sometimes followed by a point and version number</p>
<div class="line-block">
<div class="line">BK006936.2</div>
</div>
<p>Item can also contain optional interval information in the following formats:</p>
<div class="line-block">
<div class="line">BK006936.2 REGION: complement(613900..615202)</div>
<div class="line">NM_005546 REGION: 1..100</div>
<div class="line">NM_005546 REGION: complement(1..100)</div>
<div class="line">21614549:1-100</div>
<div class="line">21614549:c100-1</div>
<div class="line">21614549 1-100</div>
<div class="line">21614549 c100-1</div>
</div>
<p>It is useful to set an interval for large genbank records to limit the download time.
The items above containing interval information and can be obtained directly by
looking up an entry in Genbank and setting the <cite>Change region shown</cite> on the
upper right side of the page. The <cite>ACCESSION</cite> line of the displayed Genbank
file will have the formatting shown.</p>
<p>Alternatively, seq_start and seq_stop can be set explicitly to the sequence intervals to be
downloaded.</p>
<p>If strand is 2. “c”, “C”, “crick”, “Crick”, “antisense”,”Antisense”,
“2”, 2, “-” or “-1”, the antisense (Crick) strand is returned, otherwise
the sense (Watson) strand is returned.</p>
<p>Result is returned as a <a class="reference internal" href="#pydna.genbankrecord.GenbankRecord" title="pydna.genbankrecord.GenbankRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.genbankrecord.GenbankRecord</span></code></a> object.</p>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id25" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id32">14</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm">http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm</a></p>
</aside>
<aside class="footnote brackets" id="id26" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id34">15</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch</a></p>
</aside>
</aside>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbank.genbank">
<span class="sig-prename descclassname"><span class="pre">pydna.genbank.</span></span><span class="sig-name descname"><span class="pre">genbank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">accession</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CS570233.1'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.genbankrecord.GenbankRecord" title="pydna.genbankrecord.GenbankRecord"><span class="pre">GenbankRecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/genbank.html#genbank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbank.genbank" title="Link to this definition">#</a></dt>
<dd><p>Download a genbank nuclotide record.</p>
<p>This function takes the same paramenters as the
:func:pydna.genbank.Genbank.nucleotide method. The email address stored
in the <cite>pydna_email</cite> environment variable is used. The easiest way set
this permanantly is to edit the <cite>pydna.ini</cite> file.
See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<p>if no accession is given, a <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/CS570233.1">very short Genbank
entry</a>
is used as an example (see below). This can be useful for testing the
connection to Genbank.</p>
<p>Please note that this result is also cached by default by settings in
the pydna.ini file.
See the documentation of <code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">LOCUS</span>       <span class="n">CS570233</span>                  <span class="mi">14</span> <span class="n">bp</span>    <span class="n">DNA</span>     <span class="n">linear</span>   <span class="n">PAT</span> <span class="mi">18</span><span class="o">-</span><span class="n">MAY</span><span class="o">-</span><span class="mi">2007</span>
<span class="n">DEFINITION</span>  <span class="n">Sequence</span> <span class="mi">6</span> <span class="kn">from</span><span class="w"> </span><span class="nn">Patent</span> <span class="n">WO2007025016</span><span class="o">.</span>
<span class="n">ACCESSION</span>   <span class="n">CS570233</span>
<span class="n">VERSION</span>     <span class="n">CS570233</span><span class="mf">.1</span>
<span class="n">KEYWORDS</span>    <span class="o">.</span>
<span class="n">SOURCE</span>      <span class="n">synthetic</span> <span class="n">construct</span>
  <span class="n">ORGANISM</span>  <span class="n">synthetic</span> <span class="n">construct</span>
            <span class="n">other</span> <span class="n">sequences</span><span class="p">;</span> <span class="n">artificial</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">REFERENCE</span>   <span class="mi">1</span>
  <span class="n">AUTHORS</span>   <span class="n">Shaw</span><span class="p">,</span><span class="n">R</span><span class="o">.</span><span class="n">W</span><span class="o">.</span> <span class="ow">and</span> <span class="n">Cottenoir</span><span class="p">,</span><span class="n">M</span><span class="o">.</span>
  <span class="n">TITLE</span>     <span class="n">Inhibition</span> <span class="n">of</span> <span class="n">metallo</span><span class="o">-</span><span class="n">beta</span><span class="o">-</span><span class="n">lactamase</span> <span class="n">by</span> <span class="n">double</span><span class="o">-</span><span class="n">stranded</span> <span class="n">dna</span>
  <span class="n">JOURNAL</span>   <span class="n">Patent</span><span class="p">:</span> <span class="n">WO</span> <span class="mi">2007025016</span><span class="o">-</span><span class="n">A1</span> <span class="mi">6</span> <span class="mi">01</span><span class="o">-</span><span class="n">MAR</span><span class="o">-</span><span class="mi">2007</span><span class="p">;</span>
            <span class="n">Texas</span> <span class="n">Tech</span> <span class="n">University</span> <span class="n">System</span> <span class="p">(</span><span class="n">US</span><span class="p">)</span>
<span class="n">FEATURES</span>             <span class="n">Location</span><span class="o">/</span><span class="n">Qualifiers</span>
     <span class="n">source</span>          <span class="mf">1..14</span>
                     <span class="o">/</span><span class="n">organism</span><span class="o">=</span><span class="s2">&quot;synthetic construct&quot;</span>
                     <span class="o">/</span><span class="n">mol_type</span><span class="o">=</span><span class="s2">&quot;unassigned DNA&quot;</span>
                     <span class="o">/</span><span class="n">db_xref</span><span class="o">=</span><span class="s2">&quot;taxon:32630&quot;</span>
                     <span class="o">/</span><span class="n">note</span><span class="o">=</span><span class="s2">&quot;This is a 14bp aptamer inhibitor.&quot;</span>
<span class="n">ORIGIN</span>
        <span class="mi">1</span> <span class="n">atgttcctac</span> <span class="n">atga</span>
<span class="o">//</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.genbankfile">
<span id="pydna-genbankfile-module"></span><h2>pydna.genbankfile module<a class="headerlink" href="#module-pydna.genbankfile" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.genbankfile.</span></span><span class="sig-name descname"><span class="pre">GenbankFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfile.html#GenbankFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile.from_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfile.html#GenbankFile.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile.from_SeqRecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfile.html#GenbankFile.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankfile.GenbankFile.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbankfile.GenbankFile.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.genbankfixer">
<span id="pydna-genbankfixer-module"></span><h2>pydna.genbankfixer module<a class="headerlink" href="#module-pydna.genbankfixer" title="Link to this heading">#</a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.genbankfixer.gbtext_clean" title="pydna.genbankfixer.gbtext_clean"><code class="xref py py-func docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a> function which can clean up broken Genbank files enough to
pass the BioPython Genbank parser</p>
<p>Almost all of this code was lifted from BioJSON (<a class="github reference external" href="https://github.com/levskaya/BioJSON">levskaya/BioJSON</a>) by Anselm Levskaya.
The original code was not accompanied by any software licence. This parser is based on pyparsing.</p>
<p>There are some modifications to deal with fringe cases.</p>
<p>The parser first produces JSON as an intermediate format which is then formatted back into a
string in Genbank format.</p>
<p>The parser is not complete, so some fields do not survive the roundtrip (see below).
This should not be a difficult fix. The returned result has two properties,
.jseq which is the intermediate JSON produced by the parser and .gbtext
which is the formatted genbank string.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.parseGBLoc">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">parseGBLoc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#parseGBLoc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.parseGBLoc" title="Link to this definition">#</a></dt>
<dd><p>retwingles parsed genbank location strings, assumes no joins of RC and FWD sequences</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.strip_multiline">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">strip_multiline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#strip_multiline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.strip_multiline" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.toInt">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">toInt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#toInt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.toInt" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.strip_indent">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">strip_indent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#strip_indent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.strip_indent" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.concat_dict">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">concat_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dlist</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#concat_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.concat_dict" title="Link to this definition">#</a></dt>
<dd><p>more or less dict(list of string pairs) but merges
vals with the same keys so no duplicates occur</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.toJSON">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">toJSON</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gbkstring</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#toJSON"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.toJSON" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.wrapstring">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">wrapstring</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">str_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rowstart</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rowend</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padfirst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#wrapstring"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.wrapstring" title="Link to this definition">#</a></dt>
<dd><p>wraps the provided string in lines of length rowend-rowstart
and padded on the left by rowstart.
-&gt; if padfirst is false the first line is not padded</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.locstr">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">locstr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#locstr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.locstr" title="Link to this definition">#</a></dt>
<dd><p>genbank formatted location string, assumes no join’d combo of rev and fwd seqs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.originstr">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">originstr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#originstr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.originstr" title="Link to this definition">#</a></dt>
<dd><p>formats dna sequence as broken, numbered lines ala genbank</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.toGB">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">toGB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jseq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#toGB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.toGB" title="Link to this definition">#</a></dt>
<dd><p>parses json jseq data and prints out ApE compatible genbank</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.genbankfixer.gbtext_clean">
<span class="sig-prename descclassname"><span class="pre">pydna.genbankfixer.</span></span><span class="sig-name descname"><span class="pre">gbtext_clean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gbtext</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankfixer.html#gbtext_clean"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankfixer.gbtext_clean" title="Link to this definition">#</a></dt>
<dd><p>This function takes a string containing <strong>one</strong> genbank sequence
in Genbank format and returns a named tuple containing two fields,
the gbtext containing a string with the corrected genbank sequence and
jseq which contains the JSON intermediate.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; s = &#39;&#39;&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013
... DEFINITION  .
... ACCESSION
... VERSION
... SOURCE      .
...   ORGANISM  .
... COMMENT
... COMMENT     ApEinfo:methylated:1
... ORIGIN
...         1 aaa
... //&#39;&#39;&#39;
&gt;&gt;&gt; from pydna.readers import read
&gt;&gt;&gt; read(s)
... /site-packages/Bio/GenBank/Scanner.py:1388: BiopythonParserWarning: Malformed LOCUS line found - is this correct?
:&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013\n&#39;
  &quot;correct?\n:%r&quot; % line, BiopythonParserWarning)
Traceback (most recent call last):
  File &quot;... /pydna/readers.py&quot;, line 48, in read
    results = results.pop()
IndexError: pop from empty list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File &quot;&lt;stdin&gt;&quot;, line 1, in &lt;module&gt;
  File &quot;... /pydna/readers.py&quot;, line 50, in read
    raise ValueError(&quot;No sequences found in data:\n({})&quot;.format(data[:79]))
ValueError: No sequences found in data:
(LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013
DEFINITI)
&gt;&gt;&gt; from pydna.genbankfixer import gbtext_clean
&gt;&gt;&gt; s2, j2 = gbtext_clean(s)
&gt;&gt;&gt; print(s2)
LOCUS       New_DNA                    3 bp ds-DNA     circular SYN 19-JUN-2013
DEFINITION  .
ACCESSION
VERSION
SOURCE      .
ORGANISM  .
COMMENT
COMMENT     ApEinfo:methylated:1
FEATURES             Location/Qualifiers
ORIGIN
        1 aaa
//
&gt;&gt;&gt; s3 = read(s2)
&gt;&gt;&gt; s3
Dseqrecord(o3)
&gt;&gt;&gt; print(s3.format())
LOCUS       New_DNA                    3 bp    DNA     circular SYN 19-JUN-2013
DEFINITION  .
ACCESSION   New_DNA
VERSION     New_DNA
KEYWORDS    .
SOURCE
  ORGANISM  .
            .
COMMENT
            ApEinfo:methylated:1
FEATURES             Location/Qualifiers
ORIGIN
        1 aaa
//
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-pydna.genbankrecord">
<span id="pydna-genbankrecord-module"></span><h2>pydna.genbankrecord module<a class="headerlink" href="#module-pydna.genbankrecord" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.genbankrecord.</span></span><span class="sig-name descname"><span class="pre">GenbankRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankrecord.html#GenbankRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.from_string">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankrecord.html#GenbankRecord.from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.from_string" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.from_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankrecord.html#GenbankRecord.from_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.from_SeqRecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankrecord.html#GenbankRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="gp">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.pydna_code">
<span class="sig-name descname"><span class="pre">pydna_code</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankrecord.html#GenbankRecord.pydna_code"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.pydna_code" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.genbankrecord.GenbankRecord.biopython_code">
<span class="sig-name descname"><span class="pre">biopython_code</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/genbankrecord.html#GenbankRecord.biopython_code"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.biopython_code" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.goldengate">
<span id="pydna-goldengate-module"></span><h2>pydna.goldengate module<a class="headerlink" href="#module-pydna.goldengate" title="Link to this heading">#</a></h2>
<p>Assembly of sequences by GoldenGate ligation assembly.</p>
</section>
<section id="module-pydna.ladders">
<span id="pydna-ladders-module"></span><h2>pydna.ladders module<a class="headerlink" href="#module-pydna.ladders" title="Link to this heading">#</a></h2>
<p>Agarose gel DNA ladders.</p>
<p>A DNA ladder is a list of FakeSeq objects that has to be initiated with
Size (bp), amount of substance (mol) and Relative mobility (Rf).</p>
<p>Rf is a float value between 0.000 and 1.000. These are used together with
the cubic spline interpolator in the gel module to calculate migartion
distance from fragment length. The Rf values are calculated manually from
a gel image. Exampel can be found in scripts/molecular_weight_standards.ods.</p>
</section>
<section id="module-pydna.ligate">
<span id="pydna-ligate-module"></span><h2>pydna.ligate module<a class="headerlink" href="#module-pydna.ligate" title="Link to this heading">#</a></h2>
<p>docstring.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.ligate.ligate">
<span class="sig-prename descclassname"><span class="pre">pydna.ligate.</span></span><span class="sig-name descname"><span class="pre">ligate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/ligate.html#ligate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.ligate.ligate" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</section>
<section id="module-pydna.opencloning_models">
<span id="pydna-opencloning-models-module"></span><h2>pydna.opencloning_models module<a class="headerlink" href="#module-pydna.opencloning_models" title="Link to this heading">#</a></h2>
<p>This module provides classes that roughly map to the <a class="reference external" href="https://opencloning.org">OpenCloning</a>
data model, which is defined using <cite>LinkML &lt;https://linkml.io&gt;</cite>, and available as a python
package <a class="reference external" href="https://pypi.org/project/opencloning-linkml/">opencloning-linkml</a>. These classes
are documented there, and the ones in this module essentially replace the fields pointing to
sequences and primers (which use ids in the data model) to <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> and <code class="docutils literal notranslate"><span class="pre">Primer</span></code>
objects, respectively. Similarly, it uses Location from <code class="docutils literal notranslate"><span class="pre">Biopython</span></code> instead of a string,
which is what the data model uses.</p>
<p>When using pydna to plan cloning, it stores the provenance of <code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects in
their <code class="docutils literal notranslate"><span class="pre">source</span></code> attribute. Not all methods generate sources so far, so refer to the
documentation notebooks for examples on how to use this feature. The <code class="docutils literal notranslate"><span class="pre">history</span></code> method of
<code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects can be used to get a string representation of the provenance of the
sequence. You can also use the <code class="docutils literal notranslate"><span class="pre">CloningStrategy</span></code> class to create a JSON representation of
the cloning strategy. That <code class="docutils literal notranslate"><span class="pre">CloningStrategy</span></code> can be loaded in the OpenCloning web interface
to see a representation of the cloning strategy.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.opencloning_models.id_mode">
<span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">id_mode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_python_internal_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#id_mode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.id_mode" title="Link to this definition">#</a></dt>
<dd><p>Context manager that is used to determine how ids are assigned to objects when
mapping them to the OpenCloning data model. If <code class="docutils literal notranslate"><span class="pre">use_python_internal_id</span></code> is True,
the built-in python <code class="docutils literal notranslate"><span class="pre">id()</span></code> function is used to assign ids to objects. That function
produces a unique integer for each object in python, so it’s guaranteed to be unique.
If <code class="docutils literal notranslate"><span class="pre">use_python_internal_id</span></code> is False, the object’s <code class="docutils literal notranslate"><span class="pre">.id</span></code> attribute (must be a string integer)
is used to assign ids to objects. This is useful when the objects already have meaningful ids,
and you want to keep references to them in <code class="docutils literal notranslate"><span class="pre">SourceInput</span></code> objects (which sequences and
primers are used in a particular source).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>use_python_internal_id</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – If True, use Python’s built-in id() function.
If False, use the object’s .id attribute (must be a string integer).</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.opencloning_models</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_id</span><span class="p">,</span> <span class="n">id_mode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseqr</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ATGC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseqr</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;my_sequence&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dseqr</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;123&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_id</span><span class="p">(</span><span class="n">dseqr</span><span class="p">)</span> <span class="o">==</span> <span class="nb">id</span><span class="p">(</span><span class="n">dseqr</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">id_mode</span><span class="p">(</span><span class="n">use_python_internal_id</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">get_id</span><span class="p">(</span><span class="n">dseqr</span><span class="p">)</span>
<span class="go">123</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.opencloning_models.get_id">
<span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">get_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj:</span> <span class="pre">Primer'</span> <span class="pre">|</span> <span class="pre">'Dseqrecord</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#get_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.get_id" title="Link to this definition">#</a></dt>
<dd><p>Get ID using the current strategy from thread-local storage (see id_mode)
:param obj: The object to get the id of
:type obj: Primer | Dseqrecord</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>int</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>The id of the object</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceLocationStr">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">SequenceLocationStr</span></span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceLocationStr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceLocationStr" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></p>
<p>A string representation of a sequence location, genbank-like.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceLocationStr.from_biopython_location">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_biopython_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceLocationStr.from_biopython_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceLocationStr.from_biopython_location" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceLocationStr.to_biopython_location">
<span class="sig-name descname"><span class="pre">to_biopython_location</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a></span></span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceLocationStr.to_biopython_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceLocationStr.to_biopython_location" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceLocationStr.field_validator">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">field_validator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceLocationStr.field_validator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceLocationStr.field_validator" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceLocationStr.from_start_and_end">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_start_and_end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceLocationStr.from_start_and_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceLocationStr.from_start_and_end" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.ConfiguredBaseModel">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">ConfiguredBaseModel</span></span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#ConfiguredBaseModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.ConfiguredBaseModel" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">BaseModel</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.ConfiguredBaseModel.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.ConfiguredBaseModel.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.TextFileSequence">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">TextFileSequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Literal" title="(in Python v3.8)"><span class="pre">Literal</span></a><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'TextFileSequence'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'TextFileSequence'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_file_format</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SequenceFileFormat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang_crick_3prime</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overhang_watson_3prime</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_content</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#TextFileSequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.TextFileSequence" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">TextFileSequence</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.TextFileSequence.from_dseqrecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_dseqrecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dseqr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#TextFileSequence.from_dseqrecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.TextFileSequence.from_dseqrecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.TextFileSequence.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.TextFileSequence.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.PrimerModel">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">PrimerModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Literal" title="(in Python v3.8)"><span class="pre">Literal</span></a><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'Primer'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Primer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#PrimerModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.PrimerModel" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Primer</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.PrimerModel.from_primer">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#PrimerModel.from_primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.PrimerModel.from_primer" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.PrimerModel.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.PrimerModel.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.SourceInput">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">SourceInput</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><span class="pre">object</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SourceInput"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SourceInput" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.ConfiguredBaseModel" title="pydna.opencloning_models.ConfiguredBaseModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">ConfiguredBaseModel</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SourceInput.sequence">
<span class="sig-name descname"><span class="pre">sequence</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><span class="pre">object</span></a></em><a class="headerlink" href="#pydna.opencloning_models.SourceInput.sequence" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SourceInput.to_pydantic_model">
<span class="sig-name descname"><span class="pre">to_pydantic_model</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">SourceInput</span></span></span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SourceInput.to_pydantic_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SourceInput.to_pydantic_model" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SourceInput.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.SourceInput.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">AssemblyFragment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#object" title="(in Python v3.8)"><span class="pre">object</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_complemented</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#AssemblyFragment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.SourceInput" title="pydna.opencloning_models.SourceInput"><code class="xref py py-class docutils literal notranslate"><span class="pre">SourceInput</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment.left_location">
<span class="sig-name descname"><span class="pre">left_location</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment.left_location" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment.right_location">
<span class="sig-name descname"><span class="pre">right_location</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Location</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment.right_location" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment.reverse_complemented">
<span class="sig-name descname"><span class="pre">reverse_complemented</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></em><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment.reverse_complemented" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment.from_biopython_location">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_biopython_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#AssemblyFragment.from_biopython_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment.from_biopython_location" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment.to_pydantic_model">
<span class="sig-name descname"><span class="pre">to_pydantic_model</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AssemblyFragment</span></span></span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#AssemblyFragment.to_pydantic_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment.to_pydantic_model" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblyFragment.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.AssemblyFragment.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">Source</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#Source"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.Source" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.ConfiguredBaseModel" title="pydna.opencloning_models.ConfiguredBaseModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">ConfiguredBaseModel</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.input">
<span class="sig-name descname"><span class="pre">input</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.opencloning_models.SourceInput" title="pydna.opencloning_models.SourceInput"><span class="pre">SourceInput</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#pydna.opencloning_models.AssemblyFragment" title="pydna.opencloning_models.AssemblyFragment"><span class="pre">AssemblyFragment</span></a><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pydna.opencloning_models.Source.input" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.Source.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">Source</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.input_models">
<span class="sig-name descname"><span class="pre">input_models</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#Source.input_models"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.Source.input_models" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.to_pydantic_model">
<span class="sig-name descname"><span class="pre">to_pydantic_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#Source.to_pydantic_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.Source.to_pydantic_model" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.add_to_history_graph">
<span class="sig-name descname"><span class="pre">add_to_history_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">history_graph</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">nx.DiGraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#Source.add_to_history_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.Source.add_to_history_graph" title="Link to this definition">#</a></dt>
<dd><p>Add the source to the history graph.</p>
<p>It does not use the get_id function, because it just uses it to have unique identifiers
for graph nodes, not to store them anywhere.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.history_string">
<span class="sig-name descname"><span class="pre">history_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#Source.history_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.Source.history_string" title="Link to this definition">#</a></dt>
<dd><p>Returns a string representation of the cloning history of the sequence.
See dseqrecord.history() for examples.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.Source.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.Source.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblySource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">AssemblySource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#AssemblySource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.AssemblySource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.Source" title="pydna.opencloning_models.Source"><code class="xref py py-class docutils literal notranslate"><span class="pre">Source</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblySource.circular">
<span class="sig-name descname"><span class="pre">circular</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></em><a class="headerlink" href="#pydna.opencloning_models.AssemblySource.circular" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblySource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.AssemblySource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblySource.to_pydantic_model">
<span class="sig-name descname"><span class="pre">to_pydantic_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#AssemblySource.to_pydantic_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.AssemblySource.to_pydantic_model" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblySource.from_subfragment_representation">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_subfragment_representation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SubFragmentRepresentationAssembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'Dseqrecord'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#AssemblySource.from_subfragment_representation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.AssemblySource.from_subfragment_representation" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.AssemblySource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.AssemblySource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.RestrictionAndLigationSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">RestrictionAndLigationSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">*,</span> <span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;,</span> <span class="pre">circular:</span> <span class="pre">bool,</span> <span class="pre">restriction_enzymes:</span> <span class="pre">list[~Bio.Restriction.Restriction.AbstractCut]</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#RestrictionAndLigationSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.RestrictionAndLigationSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.RestrictionAndLigationSource.restriction_enzymes">
<span class="sig-name descname"><span class="pre">restriction_enzymes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><span class="pre">AbstractCut</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pydna.opencloning_models.RestrictionAndLigationSource.restriction_enzymes" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.RestrictionAndLigationSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.RestrictionAndLigationSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">RestrictionAndLigationSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.RestrictionAndLigationSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.RestrictionAndLigationSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.GibsonAssemblySource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">GibsonAssemblySource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#GibsonAssemblySource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.GibsonAssemblySource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.GibsonAssemblySource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.GibsonAssemblySource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">GibsonAssemblySource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.GibsonAssemblySource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.GibsonAssemblySource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.InFusionSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">InFusionSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#InFusionSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.InFusionSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.InFusionSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.InFusionSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">InFusionSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.InFusionSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.InFusionSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.OverlapExtensionPCRLigationSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">OverlapExtensionPCRLigationSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#OverlapExtensionPCRLigationSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.OverlapExtensionPCRLigationSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.OverlapExtensionPCRLigationSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.InVivoAssemblySource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">InVivoAssemblySource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#InVivoAssemblySource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.InVivoAssemblySource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.InVivoAssemblySource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.InVivoAssemblySource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">InVivoAssemblySource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.InVivoAssemblySource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.InVivoAssemblySource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.LigationSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">LigationSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#LigationSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.LigationSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.LigationSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.LigationSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">LigationSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.LigationSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.LigationSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.GatewaySource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">GatewaySource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction_type:</span> <span class="pre">~opencloning_linkml.datamodel._models.GatewayReactionType</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">greedy:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#GatewaySource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.GatewaySource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.GatewaySource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.GatewaySource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">GatewaySource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.GatewaySource.reaction_type">
<span class="sig-name descname"><span class="pre">reaction_type</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">GatewayReactionType</span></em><a class="headerlink" href="#pydna.opencloning_models.GatewaySource.reaction_type" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.GatewaySource.greedy">
<span class="sig-name descname"><span class="pre">greedy</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></em><a class="headerlink" href="#pydna.opencloning_models.GatewaySource.greedy" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.GatewaySource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.GatewaySource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.HomologousRecombinationSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">HomologousRecombinationSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#HomologousRecombinationSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.HomologousRecombinationSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.HomologousRecombinationSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.HomologousRecombinationSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">HomologousRecombinationSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.HomologousRecombinationSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.HomologousRecombinationSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.CRISPRSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">CRISPRSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CRISPRSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CRISPRSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.HomologousRecombinationSource" title="pydna.opencloning_models.HomologousRecombinationSource"><code class="xref py py-class docutils literal notranslate"><span class="pre">HomologousRecombinationSource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.CRISPRSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.CRISPRSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">CRISPRSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.CRISPRSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.CRISPRSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.CreLoxRecombinationSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">CreLoxRecombinationSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CreLoxRecombinationSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CreLoxRecombinationSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.CreLoxRecombinationSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.CreLoxRecombinationSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">CreLoxRecombinationSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.CreLoxRecombinationSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.CreLoxRecombinationSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.PCRSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">PCRSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular:</span> <span class="pre">bool</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_primer_features:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#PCRSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.PCRSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.AssemblySource" title="pydna.opencloning_models.AssemblySource"><code class="xref py py-class docutils literal notranslate"><span class="pre">AssemblySource</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.PCRSource.TARGET_MODEL">
<span class="sig-name descname"><span class="pre">TARGET_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.PCRSource.TARGET_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">PCRSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.PCRSource.add_primer_features">
<span class="sig-name descname"><span class="pre">add_primer_features</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></em><a class="headerlink" href="#pydna.opencloning_models.PCRSource.add_primer_features" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.PCRSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.PCRSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">SequenceCutSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">*,</span> <span class="pre">input:</span> <span class="pre">list[~pydna.opencloning_models.SourceInput</span> <span class="pre">|</span> <span class="pre">~pydna.opencloning_models.AssemblyFragment]</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;,</span> <span class="pre">left_edge:</span> <span class="pre">~typing.Tuple[~typing.Tuple[int,</span> <span class="pre">int],</span> <span class="pre">~Bio.Restriction.Restriction.AbstractCut</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">|</span> <span class="pre">~pydna.crispr._cas]</span> <span class="pre">|</span> <span class="pre">None,</span> <span class="pre">right_edge:</span> <span class="pre">~typing.Tuple[~typing.Tuple[int,</span> <span class="pre">int],</span> <span class="pre">~Bio.Restriction.Restriction.AbstractCut</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">|</span> <span class="pre">~pydna.crispr._cas]</span> <span class="pre">|</span> <span class="pre">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceCutSource"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.opencloning_models.Source" title="pydna.opencloning_models.Source"><code class="xref py py-class docutils literal notranslate"><span class="pre">Source</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.left_edge">
<span class="sig-name descname"><span class="pre">left_edge</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">CutSiteType</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.left_edge" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.right_edge">
<span class="sig-name descname"><span class="pre">right_edge</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">CutSiteType</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.right_edge" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.BASE_MODEL">
<span class="sig-name descname"><span class="pre">BASE_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.BASE_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">SequenceCutSource</span></code></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.ENZYME_MODEL">
<span class="sig-name descname"><span class="pre">ENZYME_MODEL</span></span><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.ENZYME_MODEL" title="Link to this definition">#</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">RestrictionEnzymeDigestionSource</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.from_parent">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_edge</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CutSiteType</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_edge</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CutSiteType</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceCutSource.from_parent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.from_parent" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.to_pydantic_model">
<span class="sig-name descname"><span class="pre">to_pydantic_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#SequenceCutSource.to_pydantic_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.to_pydantic_model" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.SequenceCutSource.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.SequenceCutSource.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.opencloning_models.</span></span><span class="sig-name descname"><span class="pre">CloningStrategy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">*,</span> <span class="pre">sequences:</span> <span class="pre">list[~opencloning_linkml.datamodel._models.Sequence</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.TemplateSequence</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.TextFileSequence</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.Primer],</span> <span class="pre">sources:</span> <span class="pre">list[~opencloning_linkml.datamodel._models.Source</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.DatabaseSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.CollectionSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.ManuallyTypedSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.UploadedFileSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.RepositoryIdSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.GenomeCoordinatesSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.SequenceCutSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.AssemblySource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.OligoHybridizationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.PolymeraseExtensionSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.AnnotationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.ReverseComplementSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.PCRSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.LigationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.HomologousRecombinationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.GibsonAssemblySource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.InFusionSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.OverlapExtensionPCRLigationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.InVivoAssemblySource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.RestrictionAndLigationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.GatewaySource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.CreLoxRecombinationSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.CRISPRSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.RestrictionEnzymeDigestionSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.AddgeneIdSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.WekWikGeneIdSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.SEVASource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.BenchlingUrlSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.SnapGenePlasmidSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.EuroscarfSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.IGEMSource</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.OpenDNACollectionsSource],</span> <span class="pre">primers:</span> <span class="pre">~typing.List[~pydna.opencloning_models.PrimerModel]</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">&lt;factory&gt;,</span> <span class="pre">description:</span> <span class="pre">str</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">files:</span> <span class="pre">list[~opencloning_linkml.datamodel._models.AssociatedFile</span> <span class="pre">|</span> <span class="pre">~opencloning_linkml.datamodel._models.SequencingFile]</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">schema_version:</span> <span class="pre">str</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">'0.4.5',</span> <span class="pre">backend_version:</span> <span class="pre">str</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">frontend_version:</span> <span class="pre">str</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">CloningStrategy</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.primers">
<span class="sig-name descname"><span class="pre">primers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.opencloning_models.PrimerModel" title="pydna.opencloning_models.PrimerModel"><span class="pre">PrimerModel</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></em><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.primers" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.add_primer">
<span class="sig-name descname"><span class="pre">add_primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><span class="pre">Primer</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy.add_primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.add_primer" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.add_dseqrecord">
<span class="sig-name descname"><span class="pre">add_dseqrecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dseqr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy.add_dseqrecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.add_dseqrecord" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.reassign_ids">
<span class="sig-name descname"><span class="pre">reassign_ids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy.reassign_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.reassign_ids" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.from_dseqrecords">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_dseqrecords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dseqrs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'Dseqrecord'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy.from_dseqrecords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.from_dseqrecords" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.model_dump_json">
<span class="sig-name descname"><span class="pre">model_dump_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy.model_dump_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.model_dump_json" title="Link to this definition">#</a></dt>
<dd><dl class="simple">
<dt>!!! abstract “Usage Documentation”</dt><dd><p>[<cite>model_dump_json</cite>](../concepts/serialization.md#json-mode)</p>
</dd>
</dl>
<p>Generates a JSON representation of the model using Pydantic’s <cite>to_json</cite> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>indent</strong> – Indentation to use in the JSON output. If None is passed, the output will be compact.</p></li>
<li><p><strong>ensure_ascii</strong> – If <cite>True</cite>, the output is guaranteed to have all incoming non-ASCII characters escaped.
If <cite>False</cite> (the default), these characters will be output as-is.</p></li>
<li><p><strong>include</strong> – Field(s) to include in the JSON output.</p></li>
<li><p><strong>exclude</strong> – Field(s) to exclude from the JSON output.</p></li>
<li><p><strong>context</strong> – Additional context to pass to the serializer.</p></li>
<li><p><strong>by_alias</strong> – Whether to serialize using field aliases.</p></li>
<li><p><strong>exclude_unset</strong> – Whether to exclude fields that have not been explicitly set.</p></li>
<li><p><strong>exclude_defaults</strong> – Whether to exclude fields that are set to their default value.</p></li>
<li><p><strong>exclude_none</strong> – Whether to exclude fields that have a value of <cite>None</cite>.</p></li>
<li><p><strong>exclude_computed_fields</strong> – Whether to exclude computed fields.
While this can be useful for round-tripping, it is usually recommended to use the dedicated
<cite>round_trip</cite> parameter instead.</p></li>
<li><p><strong>round_trip</strong> – If True, dumped values should be valid as input for non-idempotent types such as Json[T].</p></li>
<li><p><strong>warnings</strong> – How to handle serialization errors. False/”none” ignores them, True/”warn” logs errors,
“error” raises a [<cite>PydanticSerializationError</cite>][pydantic_core.PydanticSerializationError].</p></li>
<li><p><strong>fallback</strong> – A function to call when an unknown value is encountered. If not provided,
a [<cite>PydanticSerializationError</cite>][pydantic_core.PydanticSerializationError] error is raised.</p></li>
<li><p><strong>serialize_as_any</strong> – Whether to serialize fields with duck-typing serialization behavior.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A JSON string representation of the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.model_dump">
<span class="sig-name descname"><span class="pre">model_dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/opencloning_models.html#CloningStrategy.model_dump"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.model_dump" title="Link to this definition">#</a></dt>
<dd><dl class="simple">
<dt>!!! abstract “Usage Documentation”</dt><dd><p>[<cite>model_dump</cite>](../concepts/serialization.md#python-mode)</p>
</dd>
</dl>
<p>Generate a dictionary representation of the model, optionally specifying which fields to include or exclude.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> – The mode in which <cite>to_python</cite> should run.
If mode is ‘json’, the output will only contain JSON serializable types.
If mode is ‘python’, the output may contain non-JSON-serializable Python objects.</p></li>
<li><p><strong>include</strong> – A set of fields to include in the output.</p></li>
<li><p><strong>exclude</strong> – A set of fields to exclude from the output.</p></li>
<li><p><strong>context</strong> – Additional context to pass to the serializer.</p></li>
<li><p><strong>by_alias</strong> – Whether to use the field’s alias in the dictionary key if defined.</p></li>
<li><p><strong>exclude_unset</strong> – Whether to exclude fields that have not been explicitly set.</p></li>
<li><p><strong>exclude_defaults</strong> – Whether to exclude fields that are set to their default value.</p></li>
<li><p><strong>exclude_none</strong> – Whether to exclude fields that have a value of <cite>None</cite>.</p></li>
<li><p><strong>exclude_computed_fields</strong> – Whether to exclude computed fields.
While this can be useful for round-tripping, it is usually recommended to use the dedicated
<cite>round_trip</cite> parameter instead.</p></li>
<li><p><strong>round_trip</strong> – If True, dumped values should be valid as input for non-idempotent types such as Json[T].</p></li>
<li><p><strong>warnings</strong> – How to handle serialization errors. False/”none” ignores them, True/”warn” logs errors,
“error” raises a [<cite>PydanticSerializationError</cite>][pydantic_core.PydanticSerializationError].</p></li>
<li><p><strong>fallback</strong> – A function to call when an unknown value is encountered. If not provided,
a [<cite>PydanticSerializationError</cite>][pydantic_core.PydanticSerializationError] error is raised.</p></li>
<li><p><strong>serialize_as_any</strong> – Whether to serialize fields with duck-typing serialization behavior.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary representation of the model.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pydna.opencloning_models.CloningStrategy.model_config">
<span class="sig-name descname"><span class="pre">model_config</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ClassVar</span><span class="p"><span class="pre">[</span></span><span class="pre">ConfigDict</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'arbitrary_types_allowed':</span> <span class="pre">True,</span> <span class="pre">'extra':</span> <span class="pre">'forbid',</span> <span class="pre">'strict':</span> <span class="pre">False,</span> <span class="pre">'use_enum_values':</span> <span class="pre">True,</span> <span class="pre">'validate_assignment':</span> <span class="pre">True,</span> <span class="pre">'validate_default':</span> <span class="pre">True}</span></em><a class="headerlink" href="#pydna.opencloning_models.CloningStrategy.model_config" title="Link to this definition">#</a></dt>
<dd><p>Configuration for the model, should be a dictionary conforming to [<cite>ConfigDict</cite>][pydantic.config.ConfigDict].</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.parsers">
<span id="pydna-parsers-module"></span><h2>pydna.parsers module<a class="headerlink" href="#module-pydna.parsers" title="Link to this heading">#</a></h2>
<p>Provides two functions, parse and parse_primers</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.extract_from_text">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">extract_from_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/parsers.html#extract_from_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.extract_from_text" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.embl_gb_fasta">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">embl_gb_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/parsers.html#embl_gb_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.embl_gb_fasta" title="Link to this definition">#</a></dt>
<dd><p>Parse embl, genbank or fasta format from text.</p>
<p>Returns list of Bio.SeqRecord.SeqRecord</p>
<p>annotations[“molecule_type”]
annotations[“topology”]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.parse">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/parsers.html#parse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.parse" title="Link to this definition">#</a></dt>
<dd><p>Return <em>all</em> DNA sequences found in data.</p>
<p>If no sequences are found, an empty list is returned. This is a greedy
function, use carefully.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>string</em><em> or </em><em>iterable</em>) – <p>The data parameter is a string containing:</p>
<ol class="arabic simple">
<li><p>an absolute path to a local file.
The file will be read in text
mode and parsed for EMBL, FASTA
and Genbank sequences. Can be
a string or a Path object.</p></li>
<li><p>a string containing one or more
sequences in EMBL, GENBANK,
or FASTA format. Mixed formats
are allowed.</p></li>
<li><p>data can be a list or other iterable where the elements are 1 or 2</p></li>
</ol>
</p></li>
<li><p><strong>ds</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – If True double stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects are returned.
If False single stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> <a class="footnote-reference brackets" href="#id22" id="id28" role="doc-noteref"><span class="fn-bracket">[</span>12<span class="fn-bracket">]</span></a> objects are
returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>contains Dseqrecord or SeqRecord objects</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)">list</a></p>
</dd>
</dl>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id29" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id36">16</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</aside>
</aside>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.parse_primers">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">parse_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/parsers.html#parse_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.parse_primers" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.parsers.parse_snapgene">
<span class="sig-prename descclassname"><span class="pre">pydna.parsers.</span></span><span class="sig-name descname"><span class="pre">parse_snapgene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/parsers.html#parse_snapgene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.parsers.parse_snapgene" title="Link to this definition">#</a></dt>
<dd><p>Parse a SnapGene file and return a Dseqrecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path to the SnapGene file to parse.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The parsed SnapGene file as a Dseqrecord object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-pydna.primer">
<span id="pydna-primer-module"></span><h2>pydna.primer module<a class="headerlink" href="#module-pydna.primer" title="Link to this heading">#</a></h2>
<p>This module provide the Primer class that is a subclass of the biopython SeqRecord.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.primer.Primer">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.primer.</span></span><span class="sig-name descname"><span class="pre">Primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amplicon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">footprint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/primer.html#Primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.primer.Primer" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<p>Primer and its position on a template, footprint and tail.</p>
<dl class="py property">
<dt class="sig sig-object py" id="pydna.primer.Primer.footprint">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">footprint</span></span><a class="headerlink" href="#pydna.primer.Primer.footprint" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.primer.Primer.tail">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tail</span></span><a class="headerlink" href="#pydna.primer.Primer.tail" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.primer.Primer.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/primer.html#Primer.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.primer.Primer.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.readers">
<span id="pydna-readers-module"></span><h2>pydna.readers module<a class="headerlink" href="#module-pydna.readers" title="Link to this heading">#</a></h2>
<p>Provides two functions, read and read_primer.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.readers.read">
<span class="sig-prename descclassname"><span class="pre">pydna.readers.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/readers.html#read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.readers.read" title="Link to this definition">#</a></dt>
<dd><p>This function is similar the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code> function but expects one and only
one sequence or and exception is thrown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>string</em>) – see below</p></li>
<li><p><strong>ds</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Double stranded or single stranded DNA, if True return
Dseqrecord objects, else Bio.SeqRecord objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>contains the first Dseqrecord or SeqRecord object parsed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pydna.all.Dseqrecord" title="pydna.all.Dseqrecord">Dseqrecord</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The data parameter is similar to the data parameter for <code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.readers.read_primer">
<span class="sig-prename descclassname"><span class="pre">pydna.readers.</span></span><span class="sig-name descname"><span class="pre">read_primer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/readers.html#read_primer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.readers.read_primer" title="Link to this definition">#</a></dt>
<dd><p>Use this function to read a primer sequence from a string or a local file.
The usage is similar to the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse_primer()</span></code> function.</p>
</dd></dl>

</section>
<section id="module-pydna.seq">
<span id="pydna-seq-module"></span><h2>pydna.seq module<a class="headerlink" href="#module-pydna.seq" title="Link to this heading">#</a></h2>
<p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.seq.Seq">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seq.</span></span><span class="sig-name descname"><span class="pre">Seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytearray" title="(in Python v3.8)"><span class="pre">bytearray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_SeqAbstractBaseClass</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.Seq.html#Bio.Seq.SequenceDataAbstractBaseClass" title="(in Biopython v1.86)"><span class="pre">SequenceDataAbstractBaseClass</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#Seq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://biopython.org/docs/latest/api/Bio.Seq.html#Bio.Seq.Seq" title="(in Biopython v1.86)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Seq</span></code></a></p>
<p>docstring.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.seq.ProteinSeq" title="pydna.seq.ProteinSeq"><span class="pre">ProteinSeq</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.translate" title="Link to this definition">#</a></dt>
<dd><p>Translate..</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.gc" title="Link to this definition">#</a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.cai" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#slice" title="(in Python v3.8)"><span class="pre">slice</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.rarecodons" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.startcodon" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.stopcodon" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PrettyTable</span></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.express" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.orfs2">
<span class="sig-name descname"><span class="pre">orfs2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.orfs2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.orfs2" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.orfs">
<span class="sig-name descname"><span class="pre">orfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">minsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.Tuple" title="(in Python v3.8)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.orfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.orfs" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.seguid" title="Link to this definition">#</a></dt>
<dd><p>Url safe SEGUID <a class="footnote-reference brackets" href="#id24" id="id30" role="doc-noteref"><span class="fn-bracket">[</span>13<span class="fn-bracket">]</span></a> for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be part of a URL.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttjU&#39;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id31" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id38">17</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#Seq.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.Seq.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement as a DNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>Any U in the sequence is treated as a T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;AATCG&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">reverse_complement_rna</span></code> returns an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">reverse_complement</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with
<code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.Seq.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seq.Seq.rc" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement as a DNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>Any U in the sequence is treated as a T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;AATCG&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">reverse_complement_rna</span></code> returns an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">reverse_complement</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with
<code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seq.</span></span><span class="sig-name descname"><span class="pre">ProteinSeq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytes" title="(in Python v3.8)"><span class="pre">bytes</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#bytearray" title="(in Python v3.8)"><span class="pre">bytearray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_SeqAbstractBaseClass</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.Seq.html#Bio.Seq.SequenceDataAbstractBaseClass" title="(in Biopython v1.86)"><span class="pre">SequenceDataAbstractBaseClass</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://biopython.org/docs/latest/api/Bio.Seq.html#Bio.Seq.Seq" title="(in Biopython v1.86)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Seq</span></code></a></p>
<p>docstring.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.translate" title="Link to this definition">#</a></dt>
<dd><p>Turn a nucleotide sequence into a protein sequence by creating a new sequence object.</p>
<p>This method will translate DNA or RNA sequences. It should not
be used on protein sequences as any result will be biologically
meaningless.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>- table - Which codon table to use?  This can be either a</em>) – (string), an NCBI identifier (integer), or a CodonTable
object (useful for non-standard genetic codes).  This
defaults to the “Standard” table.</p></li>
<li><p><strong>string</strong> (<em>- stop_symbol - Single character</em>) – terminators.  This defaults to the asterisk, “*”.</p></li>
<li><p><strong>for</strong> (<em>what to use</em>) – terminators.  This defaults to the asterisk, “*”.</p></li>
<li><p><strong>Boolean</strong> (<em>- cds -</em>) – translation continuing on past any stop codons (translated as the
specified stop_symbol).  If True, translation is terminated at
the first in frame stop codon (and the stop_symbol is not
appended to the returned protein sequence).</p></li>
<li><p><strong>full</strong> (<em>defaults to False meaning do a</em>) – translation continuing on past any stop codons (translated as the
specified stop_symbol).  If True, translation is terminated at
the first in frame stop codon (and the stop_symbol is not
appended to the returned protein sequence).</p></li>
<li><p><strong>Boolean</strong> – this checks the sequence starts with a valid alternative start
codon (which will be translated as methionine, M), that the
sequence length is a multiple of three, and that there is a
single in frame stop codon at the end (this will be excluded
from the protein sequence, regardless of the to_stop option).
If these tests fail, an exception is raised.</p></li>
<li><p><strong>True</strong> (<em>indicates this is a complete CDS.  If</em>) – this checks the sequence starts with a valid alternative start
codon (which will be translated as methionine, M), that the
sequence length is a multiple of three, and that there is a
single in frame stop codon at the end (this will be excluded
from the protein sequence, regardless of the to_stop option).
If these tests fail, an exception is raised.</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>:param<span class="classifier">this checks the sequence starts with a valid alternative start</span></dt><dd><p>codon (which will be translated as methionine, M), that the
sequence length is a multiple of three, and that there is a
single in frame stop codon at the end (this will be excluded
from the protein sequence, regardless of the to_stop option).
If these tests fail, an exception is raised.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gaps.</strong> (<em>- gap - Single character string to denote symbol used for</em>) – Defaults to the minus sign.</p>
</dd>
</dl>
<p>A <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object is returned if <code class="docutils literal notranslate"><span class="pre">translate</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code>
object; a <code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code> object is returned if <code class="docutils literal notranslate"><span class="pre">translate</span></code> is called
pn a <code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code> object.</p>
<p>e.g. Using the standard table:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;VAIVMGR*KGAR*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">stop_symbol</span><span class="o">=</span><span class="s2">&quot;@&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGR@KGAR@&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGR&#39;)</span>
</pre></div>
</div>
<p>Now using NCBI table 2, where TGA is not a stop codon:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGRWKGAR*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGRWKGAR&#39;)</span>
</pre></div>
</div>
<p>In fact, GTG is an alternative start codon under NCBI table 2, meaning
this sequence could be a complete CDS:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;MAIVMGRWKGAR&#39;)</span>
</pre></div>
</div>
<p>It isn’t a valid CDS under NCBI table 1, due to both the start codon
and also the in frame stop codons:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">Bio.Data.CodonTable.TranslationError</span>: <span class="n">First codon &#39;GTG&#39; is not a start codon</span>
</pre></div>
</div>
<p>If the sequence has no in-frame stop codon, then the to_stop argument
has no effect:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TTGGCCATTGTAATGGGCCGC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LAIVMGR&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;LAIVMGR&#39;)</span>
</pre></div>
</div>
<p>NOTE - Ambiguous codons like “TAN” or “NNN” could be an amino acid
or a stop codon.  These are translated as “X”.  Any invalid codon
(e.g. “TA?” or “T-A”) will throw a TranslationError.</p>
<p>NOTE - This does NOT behave like the python string’s translate
method.  For that use str(my_seq).translate(…) instead</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.complement">
<span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.complement" title="Link to this definition">#</a></dt>
<dd><p>Return the complement as a DNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GCT&#39;)</span>
</pre></div>
</div>
<p>Any U in the sequence is treated as a T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GCTAA&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">complement_rna</span></code> returns an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;GCUAA&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;GCT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">complement</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;GCT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;GCT&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">complement_rna</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.complement_rna">
<span class="sig-name descname"><span class="pre">complement_rna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.complement_rna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.complement_rna" title="Link to this definition">#</a></dt>
<dd><p>Return the complement as an RNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;GCU&#39;)</span>
</pre></div>
</div>
<p>Any T in the sequence is treated as a U:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;GCUAA&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">complement</span></code> returns a DNA sequence by default:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GCT&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;GCU&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;GCU&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;GCU&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">complement_rna</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement as a DNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>Any U in the sequence is treated as a T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;AATCG&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">reverse_complement_rna</span></code> returns an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">reverse_complement</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with
<code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seq.ProteinSeq.rc" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement as a DNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>Any U in the sequence is treated as a T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;AATCG&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">reverse_complement_rna</span></code> returns an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">reverse_complement</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with
<code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.reverse_complement_rna">
<span class="sig-name descname"><span class="pre">reverse_complement_rna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.reverse_complement_rna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.reverse_complement_rna" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement as an RNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>Any T in the sequence is treated as a U:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;AAUCG&#39;)</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">reverse_complement</span></code> returns a DNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;TCG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;UCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CGA&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;UCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">reverse_complement_rna</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with
<code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.transcribe">
<span class="sig-name descname"><span class="pre">transcribe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.transcribe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.transcribe" title="Link to this definition">#</a></dt>
<dd><p>Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.</p>
<p>Following the usual convention, the sequence is interpreted as the
coding strand of the DNA double helix, not the template strand. This
means we can get the RNA sequence just by switching T to U.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">transcribe</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">transcribe</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
<p>Trying to transcribe an RNA sequence has no effect.
If you have a nucleotide sequence which might be DNA or RNA
(or even a mixture), calling the transcribe method will ensure
any T becomes U.</p>
<p>Trying to transcribe a protein sequence will replace any
T for Threonine with U for Selenocysteine, which has no
biologically plausible rational.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRU&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.back_transcribe">
<span class="sig-name descname"><span class="pre">back_transcribe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.back_transcribe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.back_transcribe" title="Link to this definition">#</a></dt>
<dd><p>Return the DNA sequence from an RNA sequence by creating a new Seq object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">messenger_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">messenger_rna</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">messenger_rna</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<p>The sequence is modified in-place and returned if inplace is True:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span>
<span class="go">MutableSeq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<p>As <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are immutable, a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code> is raised if
<code class="docutils literal notranslate"><span class="pre">transcribe</span></code> is called on a <code class="docutils literal notranslate"><span class="pre">Seq</span></code> object with <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p>
<p>Trying to back-transcribe DNA has no effect, If you have a nucleotide
sequence which might be DNA or RNA (or even a mixture), calling the
back-transcribe method will ensure any U becomes T.</p>
<p>Trying to back-transcribe a protein sequence will replace any U for
Selenocysteine with T for Threonine, which is biologically meaningless.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRU&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRT&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.seguid" title="Link to this definition">#</a></dt>
<dd><p>Url safe SEGUID <a class="footnote-reference brackets" href="#id25" id="id32" role="doc-noteref"><span class="fn-bracket">[</span>14<span class="fn-bracket">]</span></a> for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be part of a URL.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">ProteinSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ProteinSeq</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttjU&#39;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id33" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>18<span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.molecular_weight">
<span class="sig-name descname"><span class="pre">molecular_weight</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.molecular_weight"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.molecular_weight" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.pI">
<span class="sig-name descname"><span class="pre">pI</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.pI"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.pI" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seq.ProteinSeq.instability_index">
<span class="sig-name descname"><span class="pre">instability_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><span class="pre">float</span></a></span></span><a class="reference internal" href="../_modules/pydna/seq.html#ProteinSeq.instability_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seq.ProteinSeq.instability_index" title="Link to this definition">#</a></dt>
<dd><p>Instability index according to Guruprasad et al.</p>
<p>Value above 40 means the protein is has a short half life.</p>
<p>Guruprasad K., Reddy B.V.B., Pandit M.W. Protein Engineering 4:155-161(1990).</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.seqrecord">
<span id="pydna-seqrecord-module"></span><h2>pydna.seqrecord module<a class="headerlink" href="#module-pydna.seqrecord" title="Link to this heading">#</a></h2>
<p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py class">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seqrecord.</span></span><span class="sig-name descname"><span class="pre">SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'description'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dbxrefs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqRecord.html#Bio.SeqRecord.SeqRecord" title="(in Biopython v1.86)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<p>A subclass of the Biopython SeqRecord class.</p>
<p>Has a number of extra methods and uses
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a
nicer output in the IPython shell.</p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.from_Bio_SeqRecord">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_Bio_SeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqRecord.html#Bio.SeqRecord.SeqRecord" title="(in Biopython v1.86)"><span class="pre">SeqRecord</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.from_Bio_SeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord" title="Link to this definition">#</a></dt>
<dd><p>Creates a pydnaSeqRecord from a Biopython SeqRecord.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.locus">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">locus</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.locus" title="Link to this definition">#</a></dt>
<dd><p>Alias for name property.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.accession">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">accession</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.accession" title="Link to this definition">#</a></dt>
<dd><p>Alias for id property.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.definition">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">definition</span></span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.definition" title="Link to this definition">#</a></dt>
<dd><p>Alias for description property.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seqrecord.SeqRecord.rc" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.isorf">
<span class="sig-name descname"><span class="pre">isorf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.isorf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.isorf" title="Link to this definition">#</a></dt>
<dd><p>Detect if sequence is an open reading frame (orf) in the 5’-3’.</p>
<p>direction.</p>
<p>Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id26" id="id34" role="doc-noteref"><span class="fn-bracket">[</span>15<span class="fn-bracket">]</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Sets the translation table, default is 1 (standard code)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence is an orf, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id35" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>19<span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.translate" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape">
<span class="sig-name descname"><span class="pre">add_colors_to_features_for_ape</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.add_colors_to_features_for_ape"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape" title="Link to this definition">#</a></dt>
<dd><p>Assign colors to features.</p>
<p>compatible with
the <a class="reference external" href="http://jorgensen.biology.utah.edu/wayned/ape/">ApE editor</a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'misc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.add_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_feature" title="Link to this definition">#</a></dt>
<dd><p>Add a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates start of the feature</p></li>
<li><p><strong>y</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Indicates end of the feature</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2),</span>
<span class="go">                           ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;,</span>
<span class="go">            qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.list_features">
<span class="sig-name descname"><span class="pre">list_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.list_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.list_features" title="Link to this definition">#</a></dt>
<dd><p>Print ASCII table with all features.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">list_features</span><span class="p">())</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">| Ft# | Label or Note | Dir | Sta | End | Len | type | orf? |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">|   0 | L:ft2         | --&gt; | 2   | 4   |   2 | misc |  no  |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.extract_feature">
<span class="sig-name descname"><span class="pre">extract_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.extract_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.extract_feature" title="Link to this definition">#</a></dt>
<dd><p>Extract feature and return a new SeqRecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>)</p></li>
<li><p><strong>extract</strong> (<em>Indicates the feature to</em>)</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">SeqRecord(seq=Seq(&#39;gt&#39;), id=&#39;ft2&#39;, name=&#39;part_name&#39;,</span>
<span class="go">          description=&#39;description&#39;, dbxrefs=[])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.sorted_features">
<span class="sig-name descname"><span class="pre">sorted_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.sorted_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.sorted_features" title="Link to this definition">#</a></dt>
<dd><p>Return a list of the features sorted by start position.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(3), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...),</span>
<span class="go"> SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">sorted_features</span><span class="p">())</span>
<span class="go">[SeqFeature(SimpleLocation(ExactPosition(2), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...),</span>
<span class="go"> SeqFeature(SimpleLocation(ExactPosition(3), ExactPosition(4),</span>
<span class="go">                           strand=1),</span>
<span class="go">            type=&#39;misc&#39;, qualifiers=...)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.seguid">
<span class="sig-name descname"><span class="pre">seguid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.seguid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.seguid" title="Link to this definition">#</a></dt>
<dd><p>Return the url safe SEGUID <a class="footnote-reference brackets" href="#id29" id="id36" role="doc-noteref"><span class="fn-bracket">[</span>16<span class="fn-bracket">]</span></a> for the sequence.</p>
<p>This checksum is the same as seguid but with base64.urlsafe
encoding instead of the normal base 64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be a part of and URL or a filename.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;gattaca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span> <span class="c1"># original seguid is +bKGnebMkia5kNg/gF7IORXMnIU</span>
<span class="go">&#39;lsseguid=tp2jzeCM2e3W4yxtrrx09CMKa_8&#39;</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id37" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>20<span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.comment">
<span class="sig-name descname"><span class="pre">comment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">newcomment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.comment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.comment" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.datefunction">
<span class="sig-name descname"><span class="pre">datefunction</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.datefunction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.datefunction" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.stamp">
<span class="sig-name descname"><span class="pre">stamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">now</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">datefunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tool</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pydna'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'</span> <span class="pre">'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.stamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stamp" title="Link to this definition">#</a></dt>
<dd><p>Add seguid checksum to COMMENTS sections</p>
<p>The checksum is stored in object.annotations[“comment”].
This shows in the COMMENTS section of a formatted genbank file.</p>
<p>For blunt linear sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">SEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p>For circular sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">cSEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p>Fore linear sequences which are not blunt:</p>
<p><code class="docutils literal notranslate"><span class="pre">lSEGUID</span> <span class="pre">&lt;seguid&gt;</span></code></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="go">&#39;lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttjU&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;comment&quot;</span><span class="p">][:</span><span class="mi">41</span><span class="p">]</span>
<span class="go">&#39;pydna lsseguid=gBw0Jp907Tg_yX3jNgS4qQWttj&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.lcs">
<span class="sig-name descname"><span class="pre">lcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.lcs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.lcs" title="Link to this definition">#</a></dt>
<dd><p>Return the longest common substring between the sequence.</p>
<p>and another sequence (other). The other sequence can be a string,
Seq, SeqRecord, Dseq or DseqRecord.
The method returns a SeqFeature with type “read” as this method
is mostly used to map sequence reads to the sequence. This can be
changed by passing a type as keyword with some other string value.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">SeqFeature(SimpleLocation(ExactPosition(0),</span>
<span class="go">                          ExactPosition(6), strand=1),</span>
<span class="go">                          type=&#39;read&#39;,</span>
<span class="go">                          qualifiers=...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GATC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">SeqFeature(SimpleLocation(ExactPosition(1),</span>
<span class="go">                          ExactPosition(5), strand=1),</span>
<span class="go">                          type=&#39;read&#39;,</span>
<span class="go">                          qualifiers=...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;CCCCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lcs</span><span class="p">(</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">SeqFeature(None)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.gc" title="Link to this definition">#</a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.cai" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.rarecodons" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.startcodon" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stopcodon" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.express" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.copy" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.SeqRecord.dump">
<span class="sig-name descname"><span class="pre">dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protocol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#SeqRecord.dump"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.dump" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pydna.seqrecord.</span></span><span class="sig-name descname"><span class="pre">ProteinSeqRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'id'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'description'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dbxrefs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.rc" title="Link to this definition">#</a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.isorf">
<span class="sig-name descname"><span class="pre">isorf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.isorf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.isorf" title="Link to this definition">#</a></dt>
<dd><p>Detect if sequence is an open reading frame (orf) in the 5’-3’.</p>
<p>direction.</p>
<p>Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id31" id="id38" role="doc-noteref"><span class="fn-bracket">[</span>17<span class="fn-bracket">]</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>table</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – Sets the translation table, default is 1 (standard code)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence is an orf, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id39" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>21<span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</aside>
</aside>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.gc">
<span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.gc" title="Link to this definition">#</a></dt>
<dd><p>Return GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.cai">
<span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.cai" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.rarecodons">
<span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.startcodon">
<span class="sig-name descname"><span class="pre">startcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.startcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.startcodon" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.stopcodon">
<span class="sig-name descname"><span class="pre">stopcodon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.stopcodon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pydna.seqrecord.ProteinSeqRecord.express">
<span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/seqrecord.html#ProteinSeqRecord.express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.seqrecord.ProteinSeqRecord.express" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pydna.sequence_picker">
<span id="pydna-sequence-picker-module"></span><h2>pydna.sequence_picker module<a class="headerlink" href="#module-pydna.sequence_picker" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.sequence_picker.genbank_accession">
<span class="sig-prename descclassname"><span class="pre">pydna.sequence_picker.</span></span><span class="sig-name descname"><span class="pre">genbank_accession</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></span><a class="reference internal" href="../_modules/pydna/sequence_picker.html#genbank_accession"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.sequence_picker.genbank_accession" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

</section>
<section id="module-pydna.sequence_regex">
<span id="pydna-sequence-regex-module"></span><h2>pydna.sequence_regex module<a class="headerlink" href="#module-pydna.sequence_regex" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.sequence_regex.compute_regex_site">
<span class="sig-prename descclassname"><span class="pre">pydna.sequence_regex.</span></span><span class="sig-name descname"><span class="pre">compute_regex_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/sequence_regex.html#compute_regex_site"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.sequence_regex.compute_regex_site" title="Link to this definition">#</a></dt>
<dd><p>Creates a regex pattern from a string that may contain degenerate bases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>site</strong> – The string to convert to a regex pattern.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The regex pattern.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.sequence_regex.dseqrecord_finditer">
<span class="sig-prename descclassname"><span class="pre">pydna.sequence_regex.</span></span><span class="sig-name descname"><span class="pre">dseqrecord_finditer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><span class="pre">Dseqrecord</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#list" title="(in Python v3.8)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><span class="pre">Match</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pydna/sequence_regex.html#dseqrecord_finditer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.sequence_regex.dseqrecord_finditer" title="Link to this definition">#</a></dt>
<dd><p>Finds all matches of a regex pattern in a Dseqrecord.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pattern</strong> – The regex pattern to search for.</p></li>
<li><p><strong>seq</strong> – The Dseqrecord to search in.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of matches.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-pydna.threading_timer_decorator_exit">
<span id="pydna-threading-timer-decorator-exit-module"></span><h2>pydna.threading_timer_decorator_exit module<a class="headerlink" href="#module-pydna.threading_timer_decorator_exit" title="Link to this heading">#</a></h2>
<p>MIT License</p>
<p>Copyright (c) 2015 Aaron Hall</p>
<p>Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the “Software”), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:</p>
<p>The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.</p>
<p>THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.cdquit">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">cdquit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fn_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#cdquit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.cdquit" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.exit_after">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">exit_after</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#exit_after"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.exit_after" title="Link to this definition">#</a></dt>
<dd><p>use as decorator to exit process if
function takes longer than s seconds</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.a">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">a</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#a"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.a" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.b">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">b</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#b"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.b" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.c">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">c</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#c"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.c" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.d">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.d" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.countdown">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">countdown</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#countdown"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.countdown" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.threading_timer_decorator_exit.main">
<span class="sig-prename descclassname"><span class="pre">pydna.threading_timer_decorator_exit.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/threading_timer_decorator_exit.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.threading_timer_decorator_exit.main" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.tm">
<span id="pydna-tm-module"></span><h2>pydna.tm module<a class="headerlink" href="#module-pydna.tm" title="Link to this heading">#</a></h2>
<p>This module provide functions for melting temperature calculations.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_default">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.DNA_NN4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">de_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span> <span class="pre">/</span> <span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span> <span class="pre">/</span> <span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selfcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Tris</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">75.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Mg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dNTPs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saltcorr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.Tm_NN</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#tm_default"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_default" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_dbd">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_dbd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.DNA_NN3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">de_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">250</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">250</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selfcomp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Tris</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Mg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dNTPs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saltcorr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">_mt.Tm_NN</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#tm_dbd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_dbd" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_product">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_product</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">K</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.050</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#tm_product"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_product" title="Link to this definition">#</a></dt>
<dd><p>Tm calculation for the amplicon.</p>
<p>according to:</p>
<p>Rychlik, Spencer, and Rhoads, 1990, Optimization of the anneal
ing temperature for DNA amplification in vitro
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/2243783">http://www.ncbi.nlm.nih.gov/pubmed/2243783</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.ta_default">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">ta_default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_product</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_product</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#ta_default"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.ta_default" title="Link to this definition">#</a></dt>
<dd><p>Ta calculation.</p>
<p>according to:</p>
<p>Rychlik, Spencer, and Rhoads, 1990, Optimization of the anneal
ing temperature for DNA amplification in vitro
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/2243783">http://www.ncbi.nlm.nih.gov/pubmed/2243783</a></p>
<p>The formula described uses the length and GC content of the product and
salt concentration (monovalent cations)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.ta_dbd">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">ta_dbd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_dbd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_product</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#ta_dbd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.ta_dbd" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_default</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.program" title="Link to this definition">#</a></dt>
<dd><p>Returns a string containing a text representation of a suggested
PCR program using Taq or similar polymerase.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|95°C|95°C               |    |tmf:59.5
|____|_____          72°C|72°C|tmr:59.7
|3min|30s  \ 59.1°C _____|____|60s/kb
|    |      \______/ 0:32|5min|GC 51%
|    |       30s         |    |1051bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.taq_program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">taq_program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_default</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_default</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.tm.taq_program" title="Link to this definition">#</a></dt>
<dd><p>Returns a string containing a text representation of a suggested
PCR program using Taq or similar polymerase.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|95°C|95°C               |    |tmf:59.5
|____|_____          72°C|72°C|tmr:59.7
|3min|30s  \ 59.1°C _____|____|60s/kb
|    |      \______/ 0:32|5min|GC 51%
|    |       30s         |    |1051bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.dbd_program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">dbd_program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_dbd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_dbd</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#dbd_program"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.dbd_program" title="Link to this definition">#</a></dt>
<dd><p>Text representation of a suggested PCR program.</p>
<p>Using a polymerase with a DNA binding domain such as Pfu-Sso7d.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|98°C|98°C               |    |tmf:53.8
|____|_____          72°C|72°C|tmr:54.8
|30s |10s  \ 57.0°C _____|____|15s/kb
|    |      \______/ 0:15|5min|GC 51%
|    |       10s         |    |1051bp

|98°C|98°C      |    |tmf:82.5
|____|____      |    |tmr:84.4
|30s |10s \ 72°C|72°C|15s/kb
|    |     \____|____|GC 52%
|    |      3:45|5min|15058bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.pfu_sso7d_program">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">pfu_sso7d_program</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">tm_dbd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">ta_dbd</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.tm.pfu_sso7d_program" title="Link to this definition">#</a></dt>
<dd><p>Text representation of a suggested PCR program.</p>
<p>Using a polymerase with a DNA binding domain such as Pfu-Sso7d.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|98°C|98°C               |    |tmf:53.8
|____|_____          72°C|72°C|tmr:54.8
|30s |10s  \ 57.0°C _____|____|15s/kb
|    |      \______/ 0:15|5min|GC 51%
|    |       10s         |    |1051bp

|98°C|98°C      |    |tmf:82.5
|____|____      |    |tmr:84.4
|30s |10s \ 72°C|72°C|15s/kb
|    |     \____|____|GC 52%
|    |      3:45|5min|15058bp
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.Q5">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">Q5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#Q5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.Q5" title="Link to this definition">#</a></dt>
<dd><p>For Q5 Ta they take the lower of the two Tms and add 1C
(up to 72C). For Phusion they take the lower of the two
and add 3C (up to 72C).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tmbresluc">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tmbresluc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primerc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saltc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#tmbresluc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tmbresluc" title="Link to this definition">#</a></dt>
<dd><p>Returns the tm for a primer using a formula adapted to polymerases
with a DNA binding domain, such as the Phusion polymerase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>primer</strong> (<em>string</em>) – primer sequence 5’-3’</p></li>
<li><p><strong>primerc</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – primer concentration in nM), set to 500.0 nm by default.</p></li>
<li><p><strong>saltc</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a><em>, </em><em>optional</em>) – Monovalent cation concentration in mM, set to 50.0 mM by default.</p></li>
<li><p><strong>thermodynamics</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – prints details of the thermodynamic data to stdout. For
debugging only.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tm</strong> – the tm of the primer</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#float" title="(in Python v3.8)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.tm.tm_neb">
<span class="sig-prename descclassname"><span class="pre">pydna.tm.</span></span><span class="sig-name descname"><span class="pre">tm_neb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prodcode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'q5-0'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/tm.html#tm_neb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.tm.tm_neb" title="Link to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-pydna.types">
<span id="pydna-types-module"></span><h2>pydna.types module<a class="headerlink" href="#module-pydna.types" title="Link to this heading">#</a></h2>
<p>Types used in the pydna package.</p>
</section>
<section id="module-pydna.user_cloning">
<span id="pydna-user-cloning-module"></span><h2>pydna.user_cloning module<a class="headerlink" href="#module-pydna.user_cloning" title="Link to this heading">#</a></h2>
<p>Nicking &amp; USER cloning</p>
<p>CGAuGTCGACTTAGATCTCACAGGCTTTTTTCAAGaCGGCCTTGAATTCAGTCATTTGGATCCGGCCGATC
GCTACAGCTGAATCTAGAGTGTCCGAAAAAAGTTCTGCCGGAACTTAAGTCAGTAAACCTAGGCCGGCuAG</p>
<ol class="arabic simple">
<li><p>Digest both strands</p></li>
<li><p>Collect all linear ssDNA</p></li>
<li><p>Anneal all combinations</p></li>
<li><p>Keep ones present in original molecule</p></li>
</ol>
<p>5. Rank by stability
6.</p>
</section>
<section id="module-pydna.utils">
<span id="pydna-utils-module"></span><h2>pydna.utils module<a class="headerlink" href="#module-pydna.utils" title="Link to this heading">#</a></h2>
<p>Miscellaneous functions.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.three_frame_orfs">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">three_frame_orfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">startcodons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('ATG',)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stopcodons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('TAG',</span> <span class="pre">'TAA',</span> <span class="pre">'TGA')</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#three_frame_orfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.three_frame_orfs" title="Link to this definition">#</a></dt>
<dd><p>Overlapping orfs in three frames.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.shift_location">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">shift_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">original_location</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#shift_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.shift_location" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.shift_feature">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">shift_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#shift_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.shift_feature" title="Link to this definition">#</a></dt>
<dd><p>Return a new feature with shifted location.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.smallest_rotation">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">smallest_rotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#smallest_rotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.smallest_rotation" title="Link to this definition">#</a></dt>
<dd><p>Smallest rotation of a string.</p>
<p>Algorithm described in Pierre Duval, Jean. 1983. Factorizing Words
over an Ordered Alphabet. Journal of Algorithms &amp; Computational Technology
4 (4) (December 1): 363–381. and Algorithms on strings and sequences based
on Lyndon words, David Eppstein 2011.
<a class="reference external" href="https://gist.github.com/dvberkel/1950267">https://gist.github.com/dvberkel/1950267</a></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">smallest_rotation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">smallest_rotation</span><span class="p">(</span><span class="s2">&quot;taaa&quot;</span><span class="p">)</span>
<span class="go">&#39;aaat&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.cai">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sce'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#dict" title="(in Python v3.8)"><span class="pre">dict</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">_weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#cai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.cai" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.rarecodons">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">rarecodons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#rarecodons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.rarecodons" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.express">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">express</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sce'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#express"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.express" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
<p><strong>NOT IMPLEMENTED YET</strong></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.open_folder">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">open_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pth</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#open_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.open_folder" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.rc">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">StrOrBytes</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">StrOrBytes</span></span></span><a class="reference internal" href="../_modules/pydna/utils.html#rc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.rc" title="Link to this definition">#</a></dt>
<dd><p>Reverse complement.</p>
<p>accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.complement">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.complement" title="Link to this definition">#</a></dt>
<dd><p>Complement.</p>
<p>accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.identifier_from_string">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">identifier_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/pydna/utils.html#identifier_from_string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.identifier_from_string" title="Link to this definition">#</a></dt>
<dd><p>Return a valid python identifier.</p>
<p>based on the argument s or an empty string</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.flatten">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">flatten</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/typing.html#typing.List" title="(in Python v3.8)"><span class="pre">List</span></a></span></span><a class="reference internal" href="../_modules/pydna/utils.html#flatten"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.flatten" title="Link to this definition">#</a></dt>
<dd><p>Flattens an iterable of iterables.</p>
<p>Down to str, bytes, bytearray or any of the pydna or Biopython seq objects</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.seq31">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">seq31</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#seq31"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.seq31" title="Link to this definition">#</a></dt>
<dd><p>Turn a three letter code protein sequence into one with one letter code.</p>
<p>The single input argument ‘seq’ should be a protein sequence using single
letter codes, as a python string.</p>
<p>This function returns the amino acid sequence as a string using the one
letter amino acid codes. Output follows the IUPAC standard (including
ambiguous characters B for “Asx”, J for “Xle” and X for “Xaa”, and also U
for “Sel” and O for “Pyl”) plus “Ter” for a terminator given as an
asterisk.</p>
<p>Any unknown
character (including possible gap characters), is changed into ‘Xaa’.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqUtils</span><span class="w"> </span><span class="kn">import</span> <span class="n">seq3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span><span class="p">(</span><span class="s2">&quot;MAIVMGRWKGAR*&quot;</span><span class="p">)</span>
<span class="go">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">seq31</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq31</span><span class="p">(</span><span class="s1">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;</span><span class="p">)</span>
<span class="go">&#39;M  A  I  V  M  G  R  W  K  G  A  R  *&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomRNA">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#randomRNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomRNA" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomDNA">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#randomDNA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomDNA" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomORF">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomORF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#randomORF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomORF" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.randomprot">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">randomprot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxlength</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#randomprot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.randomprot" title="Link to this definition">#</a></dt>
<dd><p>docstring.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.eq">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">eq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#eq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.eq" title="Link to this definition">#</a></dt>
<dd><p>Compare two or more DNA sequences for equality.</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same double stranded DNA molecule.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>iterable</em>) – iterable containing sequences
args can be strings, Biopython Seq or SeqRecord, Dseqrecord
or dsDNA objects.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Consider all molecules circular or linear</p></li>
<li><p><strong>linear</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a><em>, </em><em>optional</em>) – Consider all molecules circular or linear</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>eq</strong> – Returns True or False</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)">bool</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same DNA molecule.</p>
<p>Two linear sequences are considiered equal if either:</p>
<ol class="arabic simple">
<li><p>They have the same sequence (case insensitive)</p></li>
<li><p>One sequence is the reverse complement of the other</p></li>
</ol>
<p>Two circular sequences are considered equal if they are circular
permutations meaning that they have the same length and:</p>
<ol class="arabic simple">
<li><p>One sequence can be found in the concatenation of the other sequence with itself.</p></li>
<li><p>The reverse complement of one sequence can be found in the concatenation of the other sequence with itself.</p></li>
</ol>
<p>The topology for the comparison can be set using one of the keywords
linear or circular to True or False.</p>
<p>If circular or linear is not set, it will be deduced from the topology of
each sequence for sequences that have a linear or circular attribute
(like Dseq and Dseqrecord).</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">eq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aTaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCCa&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;tGGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.cuts_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">cuts_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_cut</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#cuts_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.cuts_overlap" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.location_boundaries">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">location_boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.SimpleLocation" title="(in Biopython v1.86)"><span class="pre">SimpleLocation</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.CompoundLocation" title="(in Biopython v1.86)"><span class="pre">CompoundLocation</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#location_boundaries"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.location_boundaries" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.locations_overlap">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">locations_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">loc1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.SimpleLocation" title="(in Biopython v1.86)"><span class="pre">SimpleLocation</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.CompoundLocation" title="(in Biopython v1.86)"><span class="pre">CompoundLocation</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">loc2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.SimpleLocation" title="(in Biopython v1.86)"><span class="pre">SimpleLocation</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.CompoundLocation" title="(in Biopython v1.86)"><span class="pre">CompoundLocation</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#locations_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.locations_overlap" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.sum_is_sticky">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">sum_is_sticky</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">three_prime_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">five_prime_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#tuple" title="(in Python v3.8)"><span class="pre">tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partial</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/pydna/utils.html#sum_is_sticky"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.sum_is_sticky" title="Link to this definition">#</a></dt>
<dd><p>Return the overlap length if the 3’ end of seq1 and 5’ end of seq2 ends are sticky and compatible for ligation.
Return 0 if they are not compatible.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.limit_iterator">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">limit_iterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">iterator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pydna/utils.html#limit_iterator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.limit_iterator" title="Link to this definition">#</a></dt>
<dd><p>Call the function with an iterator to raise an error if the number of items is greater than the limit.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pydna.utils.create_location">
<span class="sig-prename descclassname"><span class="pre">pydna.utils.</span></span><span class="sig-name descname"><span class="pre">create_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.8/library/constants.html#None" title="(in Python v3.8)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://biopython.org/docs/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.Location" title="(in Biopython v1.86)"><span class="pre">Location</span></a></span></span><a class="reference internal" href="../_modules/pydna/utils.html#create_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pydna.utils.create_location" title="Link to this definition">#</a></dt>
<dd><p>Create a location object from a start and end position.
If the end position is less than the start position, the location is circular. It handles negative positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The start position of the location.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The end position of the location.</p></li>
<li><p><strong>lim</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The length of the sequence.</p></li>
<li><p><strong>strand</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – The strand of the location. None, 1 or -1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>location</strong> – The location object. Can be a SimpleLocation or a CompoundLocation if the feature spans the origin of
a circular sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Location</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">create_location</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">create_location</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">&#39;[0:5](-)&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">create_location</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<span class="go">&#39;[0:5](+)&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">create_location</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="go">&#39;[0:5]&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">create_location</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="go">&#39;join{[8:10], [0:2]}&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">create_location</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">&#39;join{[0:2](-), [8:10](-)}&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">create_location</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="go">&#39;join{[8:10], [0:2]}&#39;</span>
</pre></div>
</div>
<p>Note this special case, 0 is the same as len(seq)
&gt;&gt;&gt; str(create_location(5, 0, 10))
‘[5:10]’</p>
<p>Note the special case where if start and end are the same,
the location spans the entire sequence (it’s not empty).
&gt;&gt;&gt; str(create_location(5, 5, 10))
‘join{[5:10], [0:5]}’</p>
</dd></dl>

</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../notebooks/new_assemblies.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">How assemblies work</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna">pydna</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna-package-layout">pydna package layout</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-use-the-documentation">How to use the documentation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna-source-code">pydna source code</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-get-more-help">How to get more help</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#examples-of-pydna-in-use">Examples of pydna in use</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.get_env"><code class="docutils literal notranslate"><span class="pre">get_env()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.logo"><code class="docutils literal notranslate"><span class="pre">logo()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#submodules">Submodules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.all">pydna.all module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal"><code class="docutils literal notranslate"><span class="pre">Anneal</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.forward_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.forward_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.reverse_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.reverse_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.template"><code class="docutils literal notranslate"><span class="pre">Anneal.template</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.limit"><code class="docutils literal notranslate"><span class="pre">Anneal.limit</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.products"><code class="docutils literal notranslate"><span class="pre">Anneal.products</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Anneal.report"><code class="docutils literal notranslate"><span class="pre">Anneal.report()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.pcr"><code class="docutils literal notranslate"><span class="pre">pcr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.genbank"><code class="docutils literal notranslate"><span class="pre">genbank()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Genbank"><code class="docutils literal notranslate"><span class="pre">Genbank</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Genbank.nucleotide"><code class="docutils literal notranslate"><span class="pre">Genbank.nucleotide()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.download_text"><code class="docutils literal notranslate"><span class="pre">download_text()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.add_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.add_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.cas9"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.circular"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.circular</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.copy_fasta_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_fasta_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.copy_gb_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_gb_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.extract_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.figure"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.find"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.find_aa"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aa()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.find_aminoacids"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aminoacids()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.format"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.format()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.from_string"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.history"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.history()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.linearize"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.linearize()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.looped"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.lower"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.m"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.m()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.map_trace_files"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.map_trace_files()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.number_of_cuts"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.number_of_cuts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.orfs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.orfs_to_features"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs_to_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.rc"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.seguid"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.shifted"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.source"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.source</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.synced"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.synced()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.upper"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseqrecord.write"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.write()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq"><code class="docutils literal notranslate"><span class="pre">Dseq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.T4"><code class="docutils literal notranslate"><span class="pre">Dseq.T4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseq.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cas9"><code class="docutils literal notranslate"><span class="pre">Dseq.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cut"><code class="docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cutsite_is_valid"><code class="docutils literal notranslate"><span class="pre">Dseq.cutsite_is_valid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.exo1_end"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.exo1_front"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_front()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.fill_in"><code class="docutils literal notranslate"><span class="pre">Dseq.fill_in()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.find"><code class="docutils literal notranslate"><span class="pre">Dseq.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.five_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.five_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.from_full_sequence_and_overhangs"><code class="docutils literal notranslate"><span class="pre">Dseq.from_full_sequence_and_overhangs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.from_representation"><code class="docutils literal notranslate"><span class="pre">Dseq.from_representation()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.from_string"><code class="docutils literal notranslate"><span class="pre">Dseq.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.get_cut_parameters"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cut_parameters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.get_cutsite_pairs"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsite_pairs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.get_cutsites"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsites()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.isblunt"><code class="docutils literal notranslate"><span class="pre">Dseq.isblunt()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.left_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.left_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.looped"><code class="docutils literal notranslate"><span class="pre">Dseq.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.lower"><code class="docutils literal notranslate"><span class="pre">Dseq.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.mung"><code class="docutils literal notranslate"><span class="pre">Dseq.mung()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.mw"><code class="docutils literal notranslate"><span class="pre">Dseq.mw()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.quick"><code class="docutils literal notranslate"><span class="pre">Dseq.quick()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.rc"><code class="docutils literal notranslate"><span class="pre">Dseq.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.right_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.right_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.seguid"><code class="docutils literal notranslate"><span class="pre">Dseq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.shifted"><code class="docutils literal notranslate"><span class="pre">Dseq.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.t4"><code class="docutils literal notranslate"><span class="pre">Dseq.t4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseq.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.three_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.three_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseq.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.transcribe"><code class="docutils literal notranslate"><span class="pre">Dseq.transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.translate"><code class="docutils literal notranslate"><span class="pre">Dseq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.trunc"><code class="docutils literal notranslate"><span class="pre">Dseq.trunc</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.upper"><code class="docutils literal notranslate"><span class="pre">Dseq.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.Dseq.watson_ovhg"><code class="docutils literal notranslate"><span class="pre">Dseq.watson_ovhg()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.read"><code class="docutils literal notranslate"><span class="pre">read()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.read_primer"><code class="docutils literal notranslate"><span class="pre">read_primer()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.parse"><code class="docutils literal notranslate"><span class="pre">parse()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.parse_primers"><code class="docutils literal notranslate"><span class="pre">parse_primers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.primer_design"><code class="docutils literal notranslate"><span class="pre">primer_design()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.assembly_fragments"><code class="docutils literal notranslate"><span class="pre">assembly_fragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.eq"><code class="docutils literal notranslate"><span class="pre">eq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.all.gbtext_clean"><code class="docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.amplicon">pydna.amplicon module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon"><code class="docutils literal notranslate"><span class="pre">Amplicon</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Amplicon.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Amplicon.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.rc"><code class="docutils literal notranslate"><span class="pre">Amplicon.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.figure"><code class="docutils literal notranslate"><span class="pre">Amplicon.figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.set_forward_primer_footprint"><code class="docutils literal notranslate"><span class="pre">Amplicon.set_forward_primer_footprint()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.set_reverse_primer_footprint"><code class="docutils literal notranslate"><span class="pre">Amplicon.set_reverse_primer_footprint()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.program"><code class="docutils literal notranslate"><span class="pre">Amplicon.program()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.dbd_program"><code class="docutils literal notranslate"><span class="pre">Amplicon.dbd_program()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplicon.Amplicon.primers"><code class="docutils literal notranslate"><span class="pre">Amplicon.primers()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.amplify">pydna.amplify module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal"><code class="docutils literal notranslate"><span class="pre">Anneal</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.forward_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.forward_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.reverse_primers"><code class="docutils literal notranslate"><span class="pre">Anneal.reverse_primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.template"><code class="docutils literal notranslate"><span class="pre">Anneal.template</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.limit"><code class="docutils literal notranslate"><span class="pre">Anneal.limit</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.products"><code class="docutils literal notranslate"><span class="pre">Anneal.products</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.Anneal.report"><code class="docutils literal notranslate"><span class="pre">Anneal.report()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.amplify.pcr"><code class="docutils literal notranslate"><span class="pre">pcr()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.assembly">pydna.assembly module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.assembly2">pydna.assembly2 module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gather_overlapping_locations"><code class="docutils literal notranslate"><span class="pre">gather_overlapping_locations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.ends_from_cutsite"><code class="docutils literal notranslate"><span class="pre">ends_from_cutsite()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.restriction_ligation_overlap"><code class="docutils literal notranslate"><span class="pre">restriction_ligation_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.combine_algorithms"><code class="docutils literal notranslate"><span class="pre">combine_algorithms()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.blunt_overlap"><code class="docutils literal notranslate"><span class="pre">blunt_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_sub_strings"><code class="docutils literal notranslate"><span class="pre">common_sub_strings()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gibson_overlap"><code class="docutils literal notranslate"><span class="pre">gibson_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.sticky_end_sub_strings"><code class="docutils literal notranslate"><span class="pre">sticky_end_sub_strings()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.zip_match_leftwards"><code class="docutils literal notranslate"><span class="pre">zip_match_leftwards()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.zip_match_rightwards"><code class="docutils literal notranslate"><span class="pre">zip_match_rightwards()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.seqrecord2_uppercase_DNA_string"><code class="docutils literal notranslate"><span class="pre">seqrecord2_uppercase_DNA_string()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.primer_template_overlap"><code class="docutils literal notranslate"><span class="pre">primer_template_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fill_left"><code class="docutils literal notranslate"><span class="pre">fill_left()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fill_right"><code class="docutils literal notranslate"><span class="pre">fill_right()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fill_dseq"><code class="docutils literal notranslate"><span class="pre">fill_dseq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.reverse_complement_assembly"><code class="docutils literal notranslate"><span class="pre">reverse_complement_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.filter_linear_subassemblies"><code class="docutils literal notranslate"><span class="pre">filter_linear_subassemblies()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.remove_subassemblies"><code class="docutils literal notranslate"><span class="pre">remove_subassemblies()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly2str"><code class="docutils literal notranslate"><span class="pre">assembly2str()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly2str_tuple"><code class="docutils literal notranslate"><span class="pre">assembly2str_tuple()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly_has_mismatches"><code class="docutils literal notranslate"><span class="pre">assembly_has_mismatches()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly_is_circular"><code class="docutils literal notranslate"><span class="pre">assembly_is_circular()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assemble"><code class="docutils literal notranslate"><span class="pre">assemble()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.annotate_primer_binding_sites"><code class="docutils literal notranslate"><span class="pre">annotate_primer_binding_sites()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.edge_representation2subfragment_representation"><code class="docutils literal notranslate"><span class="pre">edge_representation2subfragment_representation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.subfragment_representation2edge_representation"><code class="docutils literal notranslate"><span class="pre">subfragment_representation2edge_representation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.get_assembly_subfragments"><code class="docutils literal notranslate"><span class="pre">get_assembly_subfragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.extract_subfragment"><code class="docutils literal notranslate"><span class="pre">extract_subfragment()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.is_sublist"><code class="docutils literal notranslate"><span class="pre">is_sublist()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.circular_permutation_min_abs"><code class="docutils literal notranslate"><span class="pre">circular_permutation_min_abs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly"><code class="docutils literal notranslate"><span class="pre">Assembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assembly_is_valid"><code class="docutils literal notranslate"><span class="pre">Assembly.assembly_is_valid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.add_edges_from_match"><code class="docutils literal notranslate"><span class="pre">Assembly.add_edges_from_match()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.format_assembly_edge"><code class="docutils literal notranslate"><span class="pre">Assembly.format_assembly_edge()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_linear_assemblies"><code class="docutils literal notranslate"><span class="pre">Assembly.get_linear_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.node_path2assembly_list"><code class="docutils literal notranslate"><span class="pre">Assembly.node_path2assembly_list()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_unique_linear_paths"><code class="docutils literal notranslate"><span class="pre">Assembly.get_unique_linear_paths()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_possible_assembly_number"><code class="docutils literal notranslate"><span class="pre">Assembly.get_possible_assembly_number()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_circular_assemblies"><code class="docutils literal notranslate"><span class="pre">Assembly.get_circular_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.format_insertion_assembly"><code class="docutils literal notranslate"><span class="pre">Assembly.format_insertion_assembly()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.format_insertion_assembly_edge_case"><code class="docutils literal notranslate"><span class="pre">Assembly.format_insertion_assembly_edge_case()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_insertion_assemblies"><code class="docutils literal notranslate"><span class="pre">Assembly.get_insertion_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assemble_linear"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_linear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assemble_circular"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_circular()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assemble_insertion"><code class="docutils literal notranslate"><span class="pre">Assembly.assemble_insertion()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.get_locations_on_fragments"><code class="docutils literal notranslate"><span class="pre">Assembly.get_locations_on_fragments()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.Assembly.assembly_uses_only_adjacent_edges"><code class="docutils literal notranslate"><span class="pre">Assembly.assembly_uses_only_adjacent_edges()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly"><code class="docutils literal notranslate"><span class="pre">PCRAssembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly.get_linear_assemblies"><code class="docutils literal notranslate"><span class="pre">PCRAssembly.get_linear_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly.get_circular_assemblies"><code class="docutils literal notranslate"><span class="pre">PCRAssembly.get_circular_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.PCRAssembly.get_insertion_assemblies"><code class="docutils literal notranslate"><span class="pre">PCRAssembly.get_insertion_assemblies()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly.get_circular_assemblies"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly.get_circular_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly.get_insertion_assemblies"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly.get_insertion_assemblies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.SingleFragmentAssembly.get_linear_assemblies"><code class="docutils literal notranslate"><span class="pre">SingleFragmentAssembly.get_linear_assemblies()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_function_assembly_products"><code class="docutils literal notranslate"><span class="pre">common_function_assembly_products()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gibson_assembly"><code class="docutils literal notranslate"><span class="pre">gibson_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.in_fusion_assembly"><code class="docutils literal notranslate"><span class="pre">in_fusion_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.fusion_pcr_assembly"><code class="docutils literal notranslate"><span class="pre">fusion_pcr_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.in_vivo_assembly"><code class="docutils literal notranslate"><span class="pre">in_vivo_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.restriction_ligation_assembly"><code class="docutils literal notranslate"><span class="pre">restriction_ligation_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.golden_gate_assembly"><code class="docutils literal notranslate"><span class="pre">golden_gate_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.ligation_assembly"><code class="docutils literal notranslate"><span class="pre">ligation_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.assembly_is_multi_site"><code class="docutils literal notranslate"><span class="pre">assembly_is_multi_site()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.gateway_assembly"><code class="docutils literal notranslate"><span class="pre">gateway_assembly()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_function_integration_products"><code class="docutils literal notranslate"><span class="pre">common_function_integration_products()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_handle_insertion_fragments"><code class="docutils literal notranslate"><span class="pre">common_handle_insertion_fragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.common_function_excision_products"><code class="docutils literal notranslate"><span class="pre">common_function_excision_products()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.homologous_recombination_integration"><code class="docutils literal notranslate"><span class="pre">homologous_recombination_integration()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.homologous_recombination_excision"><code class="docutils literal notranslate"><span class="pre">homologous_recombination_excision()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.cre_lox_integration"><code class="docutils literal notranslate"><span class="pre">cre_lox_integration()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.cre_lox_excision"><code class="docutils literal notranslate"><span class="pre">cre_lox_excision()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.crispr_integration"><code class="docutils literal notranslate"><span class="pre">crispr_integration()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.assembly2.pcr_assembly"><code class="docutils literal notranslate"><span class="pre">pcr_assembly()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.codon">pydna.codon module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.common_sub_strings">pydna.common_sub_strings module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.common_sub_strings.common_sub_strings"><code class="docutils literal notranslate"><span class="pre">common_sub_strings()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.common_sub_strings.terminal_overlap"><code class="docutils literal notranslate"><span class="pre">terminal_overlap()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.conftest">pydna.conftest module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.contig">pydna.contig module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig"><code class="docutils literal notranslate"><span class="pre">Contig</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.from_string"><code class="docutils literal notranslate"><span class="pre">Contig.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Contig.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Contig.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.rc"><code class="docutils literal notranslate"><span class="pre">Contig.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.detailed_figure"><code class="docutils literal notranslate"><span class="pre">Contig.detailed_figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.contig.Contig.figure"><code class="docutils literal notranslate"><span class="pre">Contig.figure()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.cre_lox">pydna.cre_lox module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.cre_loxP_overlap"><code class="docutils literal notranslate"><span class="pre">cre_loxP_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.get_regex_dict"><code class="docutils literal notranslate"><span class="pre">get_regex_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.find_loxP_sites"><code class="docutils literal notranslate"><span class="pre">find_loxP_sites()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.cre_lox.annotate_loxP_sites"><code class="docutils literal notranslate"><span class="pre">annotate_loxP_sites()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.crispr">pydna.crispr module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9"><code class="docutils literal notranslate"><span class="pre">cas9</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.scaffold"><code class="docutils literal notranslate"><span class="pre">cas9.scaffold</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.pam"><code class="docutils literal notranslate"><span class="pre">cas9.pam</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.size"><code class="docutils literal notranslate"><span class="pre">cas9.size</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.fst5"><code class="docutils literal notranslate"><span class="pre">cas9.fst5</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.fst3"><code class="docutils literal notranslate"><span class="pre">cas9.fst3</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.ovhg"><code class="docutils literal notranslate"><span class="pre">cas9.ovhg</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.cas9.search"><code class="docutils literal notranslate"><span class="pre">cas9.search()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.crispr.protospacer"><code class="docutils literal notranslate"><span class="pre">protospacer()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.design">pydna.design module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.design.primer_design"><code class="docutils literal notranslate"><span class="pre">primer_design()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.design.assembly_fragments"><code class="docutils literal notranslate"><span class="pre">assembly_fragments()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.design.circular_assembly_fragments"><code class="docutils literal notranslate"><span class="pre">circular_assembly_fragments()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.download">pydna.download module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.download.download_text"><code class="docutils literal notranslate"><span class="pre">download_text()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.dseq">pydna.dseq module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq"><code class="docutils literal notranslate"><span class="pre">Dseq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.trunc"><code class="docutils literal notranslate"><span class="pre">Dseq.trunc</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.quick"><code class="docutils literal notranslate"><span class="pre">Dseq.quick()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.from_string"><code class="docutils literal notranslate"><span class="pre">Dseq.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.from_representation"><code class="docutils literal notranslate"><span class="pre">Dseq.from_representation()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.from_full_sequence_and_overhangs"><code class="docutils literal notranslate"><span class="pre">Dseq.from_full_sequence_and_overhangs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.mw"><code class="docutils literal notranslate"><span class="pre">Dseq.mw()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.upper"><code class="docutils literal notranslate"><span class="pre">Dseq.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.lower"><code class="docutils literal notranslate"><span class="pre">Dseq.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.find"><code class="docutils literal notranslate"><span class="pre">Dseq.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.rc"><code class="docutils literal notranslate"><span class="pre">Dseq.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.shifted"><code class="docutils literal notranslate"><span class="pre">Dseq.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.looped"><code class="docutils literal notranslate"><span class="pre">Dseq.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseq.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.five_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.five_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.three_prime_end"><code class="docutils literal notranslate"><span class="pre">Dseq.three_prime_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.watson_ovhg"><code class="docutils literal notranslate"><span class="pre">Dseq.watson_ovhg()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.fill_in"><code class="docutils literal notranslate"><span class="pre">Dseq.fill_in()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.transcribe"><code class="docutils literal notranslate"><span class="pre">Dseq.transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.translate"><code class="docutils literal notranslate"><span class="pre">Dseq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.mung"><code class="docutils literal notranslate"><span class="pre">Dseq.mung()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.T4"><code class="docutils literal notranslate"><span class="pre">Dseq.T4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.t4"><code class="docutils literal notranslate"><span class="pre">Dseq.t4()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.exo1_front"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_front()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.exo1_end"><code class="docutils literal notranslate"><span class="pre">Dseq.exo1_end()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cutters"><code class="docutils literal notranslate"><span class="pre">Dseq.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.seguid"><code class="docutils literal notranslate"><span class="pre">Dseq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.isblunt"><code class="docutils literal notranslate"><span class="pre">Dseq.isblunt()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cas9"><code class="docutils literal notranslate"><span class="pre">Dseq.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseq.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cut"><code class="docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.cutsite_is_valid"><code class="docutils literal notranslate"><span class="pre">Dseq.cutsite_is_valid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.get_cutsites"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsites()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.left_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.left_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.right_end_position"><code class="docutils literal notranslate"><span class="pre">Dseq.right_end_position()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.get_cut_parameters"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cut_parameters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseq.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseq.Dseq.get_cutsite_pairs"><code class="docutils literal notranslate"><span class="pre">Dseq.get_cutsite_pairs()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.dseqrecord">pydna.dseqrecord module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.source"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.source</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.from_string"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.circular"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.circular</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.m"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.m()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.extract_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.add_feature"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.add_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.seguid"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.looped"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.looped()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.tolinear"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.tolinear()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.terminal_transferase"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.terminal_transferase()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.format"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.format()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.write"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.write()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.find"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.find_aminoacids"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aminoacids()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.find_aa"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.find_aa()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.map_trace_files"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.map_trace_files()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.linearize"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.linearize()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.no_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.no_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.unique_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.unique_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.once_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.once_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.twice_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.twice_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.n_cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.n_cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.cutters"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cutters()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.number_of_cuts"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.number_of_cuts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.cas9"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cas9()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.rc"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.synced"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.synced()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.upper"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.upper()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.lower"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.lower()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.orfs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.orfs_to_features"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.orfs_to_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.copy_gb_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_gb_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.copy_fasta_to_clipboard"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.copy_fasta_to_clipboard()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.figure"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.figure()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.shifted"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.shifted()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.annotations"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.annotations</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.dbxrefs"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.dbxrefs</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.apply_cut"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.apply_cut()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.dseqrecord.Dseqrecord.history"><code class="docutils literal notranslate"><span class="pre">Dseqrecord.history()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.fakeseq">pydna.fakeseq module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fakeseq.FakeSeq"><code class="docutils literal notranslate"><span class="pre">FakeSeq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fakeseq.FakeSeq.m"><code class="docutils literal notranslate"><span class="pre">FakeSeq.m()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fakeseq.FakeSeq.M"><code class="docutils literal notranslate"><span class="pre">FakeSeq.M()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.fusionpcr">pydna.fusionpcr module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fusionpcr.fuse_by_pcr"><code class="docutils literal notranslate"><span class="pre">fuse_by_pcr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.fusionpcr.list_parts"><code class="docutils literal notranslate"><span class="pre">list_parts()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.gateway">pydna.gateway module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gateway.gateway_overlap"><code class="docutils literal notranslate"><span class="pre">gateway_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gateway.find_gateway_sites"><code class="docutils literal notranslate"><span class="pre">find_gateway_sites()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gateway.annotate_gateway_sites"><code class="docutils literal notranslate"><span class="pre">annotate_gateway_sites()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.gel">pydna.gel module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gel.interpolator"><code class="docutils literal notranslate"><span class="pre">interpolator()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.gel.gel"><code class="docutils literal notranslate"><span class="pre">gel()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbank">pydna.genbank module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbank.Genbank"><code class="docutils literal notranslate"><span class="pre">Genbank</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbank.Genbank.nucleotide"><code class="docutils literal notranslate"><span class="pre">Genbank.nucleotide()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbank.genbank"><code class="docutils literal notranslate"><span class="pre">genbank()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbankfile">pydna.genbankfile module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile"><code class="docutils literal notranslate"><span class="pre">GenbankFile</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">GenbankFile.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile.reverse_complement"><code class="docutils literal notranslate"><span class="pre">GenbankFile.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfile.GenbankFile.rc"><code class="docutils literal notranslate"><span class="pre">GenbankFile.rc()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbankfixer">pydna.genbankfixer module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.parseGBLoc"><code class="docutils literal notranslate"><span class="pre">parseGBLoc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.strip_multiline"><code class="docutils literal notranslate"><span class="pre">strip_multiline()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.toInt"><code class="docutils literal notranslate"><span class="pre">toInt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.strip_indent"><code class="docutils literal notranslate"><span class="pre">strip_indent()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.concat_dict"><code class="docutils literal notranslate"><span class="pre">concat_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.toJSON"><code class="docutils literal notranslate"><span class="pre">toJSON()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.wrapstring"><code class="docutils literal notranslate"><span class="pre">wrapstring()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.locstr"><code class="docutils literal notranslate"><span class="pre">locstr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.originstr"><code class="docutils literal notranslate"><span class="pre">originstr()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.toGB"><code class="docutils literal notranslate"><span class="pre">toGB()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankfixer.gbtext_clean"><code class="docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.genbankrecord">pydna.genbankrecord module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord"><code class="docutils literal notranslate"><span class="pre">GenbankRecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.from_string"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.from_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.from_SeqRecord"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.from_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.rc"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.pydna_code"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.pydna_code()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.genbankrecord.GenbankRecord.biopython_code"><code class="docutils literal notranslate"><span class="pre">GenbankRecord.biopython_code()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.goldengate">pydna.goldengate module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.ladders">pydna.ladders module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.ligate">pydna.ligate module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.ligate.ligate"><code class="docutils literal notranslate"><span class="pre">ligate()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.opencloning_models">pydna.opencloning_models module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.id_mode"><code class="docutils literal notranslate"><span class="pre">id_mode()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.get_id"><code class="docutils literal notranslate"><span class="pre">get_id()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.from_biopython_location"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.from_biopython_location()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.to_biopython_location"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.to_biopython_location()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.field_validator"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.field_validator()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceLocationStr.from_start_and_end"><code class="docutils literal notranslate"><span class="pre">SequenceLocationStr.from_start_and_end()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.ConfiguredBaseModel"><code class="docutils literal notranslate"><span class="pre">ConfiguredBaseModel</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.ConfiguredBaseModel.model_config"><code class="docutils literal notranslate"><span class="pre">ConfiguredBaseModel.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.TextFileSequence"><code class="docutils literal notranslate"><span class="pre">TextFileSequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.TextFileSequence.from_dseqrecord"><code class="docutils literal notranslate"><span class="pre">TextFileSequence.from_dseqrecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.TextFileSequence.model_config"><code class="docutils literal notranslate"><span class="pre">TextFileSequence.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PrimerModel"><code class="docutils literal notranslate"><span class="pre">PrimerModel</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PrimerModel.from_primer"><code class="docutils literal notranslate"><span class="pre">PrimerModel.from_primer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PrimerModel.model_config"><code class="docutils literal notranslate"><span class="pre">PrimerModel.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput"><code class="docutils literal notranslate"><span class="pre">SourceInput</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput.sequence"><code class="docutils literal notranslate"><span class="pre">SourceInput.sequence</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">SourceInput.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SourceInput.model_config"><code class="docutils literal notranslate"><span class="pre">SourceInput.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.left_location"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.left_location</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.right_location"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.right_location</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.reverse_complemented"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.reverse_complemented</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.from_biopython_location"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.from_biopython_location()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblyFragment.model_config"><code class="docutils literal notranslate"><span class="pre">AssemblyFragment.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source"><code class="docutils literal notranslate"><span class="pre">Source</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.input"><code class="docutils literal notranslate"><span class="pre">Source.input</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">Source.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.input_models"><code class="docutils literal notranslate"><span class="pre">Source.input_models()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">Source.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.add_to_history_graph"><code class="docutils literal notranslate"><span class="pre">Source.add_to_history_graph()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.history_string"><code class="docutils literal notranslate"><span class="pre">Source.history_string()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.Source.model_config"><code class="docutils literal notranslate"><span class="pre">Source.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource"><code class="docutils literal notranslate"><span class="pre">AssemblySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.circular"><code class="docutils literal notranslate"><span class="pre">AssemblySource.circular</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">AssemblySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">AssemblySource.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.from_subfragment_representation"><code class="docutils literal notranslate"><span class="pre">AssemblySource.from_subfragment_representation()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.AssemblySource.model_config"><code class="docutils literal notranslate"><span class="pre">AssemblySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource.restriction_enzymes"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource.restriction_enzymes</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.RestrictionAndLigationSource.model_config"><code class="docutils literal notranslate"><span class="pre">RestrictionAndLigationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GibsonAssemblySource"><code class="docutils literal notranslate"><span class="pre">GibsonAssemblySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GibsonAssemblySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">GibsonAssemblySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GibsonAssemblySource.model_config"><code class="docutils literal notranslate"><span class="pre">GibsonAssemblySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InFusionSource"><code class="docutils literal notranslate"><span class="pre">InFusionSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InFusionSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">InFusionSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InFusionSource.model_config"><code class="docutils literal notranslate"><span class="pre">InFusionSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource"><code class="docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.OverlapExtensionPCRLigationSource.model_config"><code class="docutils literal notranslate"><span class="pre">OverlapExtensionPCRLigationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InVivoAssemblySource"><code class="docutils literal notranslate"><span class="pre">InVivoAssemblySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InVivoAssemblySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">InVivoAssemblySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.InVivoAssemblySource.model_config"><code class="docutils literal notranslate"><span class="pre">InVivoAssemblySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.LigationSource"><code class="docutils literal notranslate"><span class="pre">LigationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.LigationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">LigationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.LigationSource.model_config"><code class="docutils literal notranslate"><span class="pre">LigationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource"><code class="docutils literal notranslate"><span class="pre">GatewaySource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">GatewaySource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.reaction_type"><code class="docutils literal notranslate"><span class="pre">GatewaySource.reaction_type</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.greedy"><code class="docutils literal notranslate"><span class="pre">GatewaySource.greedy</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.GatewaySource.model_config"><code class="docutils literal notranslate"><span class="pre">GatewaySource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.HomologousRecombinationSource"><code class="docutils literal notranslate"><span class="pre">HomologousRecombinationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.HomologousRecombinationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">HomologousRecombinationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.HomologousRecombinationSource.model_config"><code class="docutils literal notranslate"><span class="pre">HomologousRecombinationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CRISPRSource"><code class="docutils literal notranslate"><span class="pre">CRISPRSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CRISPRSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">CRISPRSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CRISPRSource.model_config"><code class="docutils literal notranslate"><span class="pre">CRISPRSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CreLoxRecombinationSource"><code class="docutils literal notranslate"><span class="pre">CreLoxRecombinationSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CreLoxRecombinationSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">CreLoxRecombinationSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CreLoxRecombinationSource.model_config"><code class="docutils literal notranslate"><span class="pre">CreLoxRecombinationSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource"><code class="docutils literal notranslate"><span class="pre">PCRSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource.TARGET_MODEL"><code class="docutils literal notranslate"><span class="pre">PCRSource.TARGET_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource.add_primer_features"><code class="docutils literal notranslate"><span class="pre">PCRSource.add_primer_features</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.PCRSource.model_config"><code class="docutils literal notranslate"><span class="pre">PCRSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.left_edge"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.left_edge</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.right_edge"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.right_edge</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.BASE_MODEL"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.BASE_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.ENZYME_MODEL"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.ENZYME_MODEL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.from_parent"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.from_parent()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.to_pydantic_model"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.to_pydantic_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.SequenceCutSource.model_config"><code class="docutils literal notranslate"><span class="pre">SequenceCutSource.model_config</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy"><code class="docutils literal notranslate"><span class="pre">CloningStrategy</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.primers"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.primers</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.add_primer"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.add_primer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.add_dseqrecord"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.add_dseqrecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.reassign_ids"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.reassign_ids()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.from_dseqrecords"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.from_dseqrecords()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.model_dump_json"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.model_dump_json()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.model_dump"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.model_dump()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.opencloning_models.CloningStrategy.model_config"><code class="docutils literal notranslate"><span class="pre">CloningStrategy.model_config</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.parsers">pydna.parsers module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.extract_from_text"><code class="docutils literal notranslate"><span class="pre">extract_from_text()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.embl_gb_fasta"><code class="docutils literal notranslate"><span class="pre">embl_gb_fasta()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.parse"><code class="docutils literal notranslate"><span class="pre">parse()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.parse_primers"><code class="docutils literal notranslate"><span class="pre">parse_primers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.parsers.parse_snapgene"><code class="docutils literal notranslate"><span class="pre">parse_snapgene()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.primer">pydna.primer module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer"><code class="docutils literal notranslate"><span class="pre">Primer</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer.footprint"><code class="docutils literal notranslate"><span class="pre">Primer.footprint</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer.tail"><code class="docutils literal notranslate"><span class="pre">Primer.tail</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.primer.Primer.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Primer.reverse_complement()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.readers">pydna.readers module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.readers.read"><code class="docutils literal notranslate"><span class="pre">read()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.readers.read_primer"><code class="docutils literal notranslate"><span class="pre">read_primer()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.seq">pydna.seq module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq"><code class="docutils literal notranslate"><span class="pre">Seq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.translate"><code class="docutils literal notranslate"><span class="pre">Seq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.gc"><code class="docutils literal notranslate"><span class="pre">Seq.gc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.cai"><code class="docutils literal notranslate"><span class="pre">Seq.cai()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.rarecodons"><code class="docutils literal notranslate"><span class="pre">Seq.rarecodons()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.startcodon"><code class="docutils literal notranslate"><span class="pre">Seq.startcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.stopcodon"><code class="docutils literal notranslate"><span class="pre">Seq.stopcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.express"><code class="docutils literal notranslate"><span class="pre">Seq.express()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.orfs2"><code class="docutils literal notranslate"><span class="pre">Seq.orfs2()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.orfs"><code class="docutils literal notranslate"><span class="pre">Seq.orfs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.seguid"><code class="docutils literal notranslate"><span class="pre">Seq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Seq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.Seq.rc"><code class="docutils literal notranslate"><span class="pre">Seq.rc()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq"><code class="docutils literal notranslate"><span class="pre">ProteinSeq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.translate"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.complement_rna"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.complement_rna()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.rc"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.reverse_complement_rna"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.reverse_complement_rna()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.transcribe"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.back_transcribe"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.back_transcribe()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.seguid"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.molecular_weight"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.molecular_weight()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.pI"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.pI()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seq.ProteinSeq.instability_index"><code class="docutils literal notranslate"><span class="pre">ProteinSeq.instability_index()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.seqrecord">pydna.seqrecord module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.from_Bio_SeqRecord"><code class="docutils literal notranslate"><span class="pre">SeqRecord.from_Bio_SeqRecord()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.locus"><code class="docutils literal notranslate"><span class="pre">SeqRecord.locus</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.accession"><code class="docutils literal notranslate"><span class="pre">SeqRecord.accession</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.definition"><code class="docutils literal notranslate"><span class="pre">SeqRecord.definition</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">SeqRecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">SeqRecord.isorf()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.translate"><code class="docutils literal notranslate"><span class="pre">SeqRecord.translate()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_colors_to_features_for_ape()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.add_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.add_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.list_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.list_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.extract_feature"><code class="docutils literal notranslate"><span class="pre">SeqRecord.extract_feature()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.sorted_features"><code class="docutils literal notranslate"><span class="pre">SeqRecord.sorted_features()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.seguid"><code class="docutils literal notranslate"><span class="pre">SeqRecord.seguid()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.comment"><code class="docutils literal notranslate"><span class="pre">SeqRecord.comment()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.datefunction"><code class="docutils literal notranslate"><span class="pre">SeqRecord.datefunction()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.stamp"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stamp()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.lcs"><code class="docutils literal notranslate"><span class="pre">SeqRecord.lcs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">SeqRecord.gc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">SeqRecord.cai()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">SeqRecord.rarecodons()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.startcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">SeqRecord.stopcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.express"><code class="docutils literal notranslate"><span class="pre">SeqRecord.express()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.copy"><code class="docutils literal notranslate"><span class="pre">SeqRecord.copy()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.SeqRecord.dump"><code class="docutils literal notranslate"><span class="pre">SeqRecord.dump()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.reverse_complement"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.reverse_complement()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.rc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.isorf"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.isorf()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.gc"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.gc()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.cai"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.cai()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.rarecodons"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.rarecodons()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.startcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.startcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.stopcodon"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.stopcodon()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.seqrecord.ProteinSeqRecord.express"><code class="docutils literal notranslate"><span class="pre">ProteinSeqRecord.express()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.sequence_picker">pydna.sequence_picker module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.sequence_picker.genbank_accession"><code class="docutils literal notranslate"><span class="pre">genbank_accession()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.sequence_regex">pydna.sequence_regex module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.sequence_regex.compute_regex_site"><code class="docutils literal notranslate"><span class="pre">compute_regex_site()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.sequence_regex.dseqrecord_finditer"><code class="docutils literal notranslate"><span class="pre">dseqrecord_finditer()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.threading_timer_decorator_exit">pydna.threading_timer_decorator_exit module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.cdquit"><code class="docutils literal notranslate"><span class="pre">cdquit()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.exit_after"><code class="docutils literal notranslate"><span class="pre">exit_after()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.a"><code class="docutils literal notranslate"><span class="pre">a()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.b"><code class="docutils literal notranslate"><span class="pre">b()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.c"><code class="docutils literal notranslate"><span class="pre">c()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.d"><code class="docutils literal notranslate"><span class="pre">d()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.countdown"><code class="docutils literal notranslate"><span class="pre">countdown()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.threading_timer_decorator_exit.main"><code class="docutils literal notranslate"><span class="pre">main()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.tm">pydna.tm module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_default"><code class="docutils literal notranslate"><span class="pre">tm_default()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_dbd"><code class="docutils literal notranslate"><span class="pre">tm_dbd()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_product"><code class="docutils literal notranslate"><span class="pre">tm_product()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.ta_default"><code class="docutils literal notranslate"><span class="pre">ta_default()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.ta_dbd"><code class="docutils literal notranslate"><span class="pre">ta_dbd()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.program"><code class="docutils literal notranslate"><span class="pre">program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.taq_program"><code class="docutils literal notranslate"><span class="pre">taq_program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.dbd_program"><code class="docutils literal notranslate"><span class="pre">dbd_program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.pfu_sso7d_program"><code class="docutils literal notranslate"><span class="pre">pfu_sso7d_program()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.Q5"><code class="docutils literal notranslate"><span class="pre">Q5()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tmbresluc"><code class="docutils literal notranslate"><span class="pre">tmbresluc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.tm.tm_neb"><code class="docutils literal notranslate"><span class="pre">tm_neb()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.types">pydna.types module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.user_cloning">pydna.user_cloning module</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-pydna.utils">pydna.utils module</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.three_frame_orfs"><code class="docutils literal notranslate"><span class="pre">three_frame_orfs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.shift_location"><code class="docutils literal notranslate"><span class="pre">shift_location()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.shift_feature"><code class="docutils literal notranslate"><span class="pre">shift_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.smallest_rotation"><code class="docutils literal notranslate"><span class="pre">smallest_rotation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.cai"><code class="docutils literal notranslate"><span class="pre">cai()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.rarecodons"><code class="docutils literal notranslate"><span class="pre">rarecodons()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.express"><code class="docutils literal notranslate"><span class="pre">express()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.open_folder"><code class="docutils literal notranslate"><span class="pre">open_folder()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.rc"><code class="docutils literal notranslate"><span class="pre">rc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.complement"><code class="docutils literal notranslate"><span class="pre">complement()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.identifier_from_string"><code class="docutils literal notranslate"><span class="pre">identifier_from_string()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.flatten"><code class="docutils literal notranslate"><span class="pre">flatten()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.seq31"><code class="docutils literal notranslate"><span class="pre">seq31()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomRNA"><code class="docutils literal notranslate"><span class="pre">randomRNA()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomDNA"><code class="docutils literal notranslate"><span class="pre">randomDNA()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomORF"><code class="docutils literal notranslate"><span class="pre">randomORF()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.randomprot"><code class="docutils literal notranslate"><span class="pre">randomprot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.eq"><code class="docutils literal notranslate"><span class="pre">eq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.cuts_overlap"><code class="docutils literal notranslate"><span class="pre">cuts_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.location_boundaries"><code class="docutils literal notranslate"><span class="pre">location_boundaries()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.locations_overlap"><code class="docutils literal notranslate"><span class="pre">locations_overlap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.sum_is_sticky"><code class="docutils literal notranslate"><span class="pre">sum_is_sticky()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.limit_iterator"><code class="docutils literal notranslate"><span class="pre">limit_iterator()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pydna.utils.create_location"><code class="docutils literal notranslate"><span class="pre">create_location()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Björn F. Johansson
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Björn F. Johansson.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>