
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pydna.dseq &#8212; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=b530ca2f" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eea9577"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pydna/dseq';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="icon" href="../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pydna 6.0.0a24.post17+b7b559bd66 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/pydna-group/pydna" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../index.html">
                    Pydna
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../getting_started.html">Getting started</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq.html">Representing sequences in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq_Features.html">Working with Features using the Dseqrecord class</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Importing_Seqs.html">Importing and viewing sequence files in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Restrict_Ligate_Cloning.html">Restriction and Ligation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/PCR.html">PCR</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../notebooks/primer_design.html">Primer design</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Gibson.html">Gibson Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/CRISPR.html">CRISPR-Cas9</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/history.html">Cloning history / Cloning strategy</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../example_gallery.html">Example gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Restriction.html">Plasmid Restriction/Ligation Cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Gibson.html">Gibson Assembly from the original publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_CRISPR.html">Designing a CRISPR-Cas9 toolbox for genome engineering.</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../how_pydna_works.html">How pydna works</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/new_assemblies.html">How assemblies work</a></li>


</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../reference/pydna.html">pydna package</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/pydna-group/pydna" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pydna.dseq</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="sd">&quot;&quot;&quot;Provides the Dseq class for handling double stranded DNA sequences.</span>

<span class="sd">Dseq is a subclass of :class:`Bio.Seq.Seq`. The Dseq class</span>
<span class="sd">is mostly useful as a part of the :class:`pydna.dseqrecord.Dseqrecord` class</span>
<span class="sd">which can hold more meta data.</span>

<span class="sd">The Dseq class support the notion of circular and linear DNA topology.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">copy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_copy</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">math</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_math</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">inspect</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_inspect</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span> <span class="k">as</span> <span class="n">_List</span><span class="p">,</span> <span class="n">Tuple</span> <span class="k">as</span> <span class="n">_Tuple</span><span class="p">,</span> <span class="n">Union</span> <span class="k">as</span> <span class="n">_Union</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">RestrictionBatch</span> <span class="k">as</span> <span class="n">_RestrictionBatch</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Restriction</span><span class="w"> </span><span class="kn">import</span> <span class="n">CommOnly</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">seguid</span><span class="w"> </span><span class="kn">import</span> <span class="n">ldseguid</span> <span class="k">as</span> <span class="n">_ldseguid</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">seguid</span><span class="w"> </span><span class="kn">import</span> <span class="n">cdseguid</span> <span class="k">as</span> <span class="n">_cdseguid</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span> <span class="k">as</span> <span class="n">_Seq</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">_SeqAbstractBaseClass</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Data.IUPACData</span><span class="w"> </span><span class="kn">import</span> <span class="n">unambiguous_dna_weights</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Data.IUPACData</span><span class="w"> </span><span class="kn">import</span> <span class="n">unambiguous_rna_weights</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Data.IUPACData</span><span class="w"> </span><span class="kn">import</span> <span class="n">atom_weights</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna._pretty</span><span class="w"> </span><span class="kn">import</span> <span class="n">pretty_str</span> <span class="k">as</span> <span class="n">_pretty_str</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">rc</span> <span class="k">as</span> <span class="n">_rc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">flatten</span> <span class="k">as</span> <span class="n">_flatten</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">cuts_overlap</span> <span class="k">as</span> <span class="n">_cuts_overlap</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">basepair_dict</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">dscode_to_watson_table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">dscode_to_crick_table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">regex_ds_melt_factory</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">regex_ss_melt_factory</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">dscode_to_full_sequence_table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">dscode_to_watson_tail_table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">dscode_to_crick_tail_table</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">complement_table_for_dscode</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">letters_not_in_dscode</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_parts</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">representation_tuple</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.alphabet</span><span class="w"> </span><span class="kn">import</span> <span class="n">dsbreaks</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.common_sub_strings</span><span class="w"> </span><span class="kn">import</span> <span class="n">common_sub_strings</span> <span class="k">as</span> <span class="n">_common_sub_strings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.types</span><span class="w"> </span><span class="kn">import</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">EnzymesType</span><span class="p">,</span> <span class="n">CutSiteType</span>


<span class="c1"># Sequences larger than this gets a truncated representation.</span>
<span class="n">length_limit_for_repr</span> <span class="o">=</span> <span class="mi">30</span>
<span class="n">placeholder</span> <span class="o">=</span> <span class="n">letters_not_in_dscode</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>


<div class="viewcode-block" id="CircularBytes">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.CircularBytes">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">CircularBytes</span><span class="p">(</span><span class="nb">bytes</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A circular bytes sequence: indexing and slicing wrap around index 0.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="nb">bytes</span> <span class="o">|</span> <span class="nb">bytearray</span> <span class="o">|</span> <span class="nb">memoryview</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">(</span><span class="n">value</span><span class="p">))</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">slice</span><span class="p">):</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="sa">b</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="s2">&quot;CircularBytes index out of range (empty bytes)&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">(</span><span class="n">key</span> <span class="o">%</span> <span class="n">n</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">slice</span><span class="p">):</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">step</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span><span class="p">,</span> <span class="n">key</span><span class="o">.</span><span class="n">step</span>
            <span class="n">step</span> <span class="o">=</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">step</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">step</span>
            <span class="k">if</span> <span class="n">step</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;slice step cannot be zero&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">step</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">start</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="n">n</span> <span class="k">if</span> <span class="n">stop</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">stop</span>
                <span class="k">while</span> <span class="n">stop</span> <span class="o">&lt;=</span> <span class="n">start</span><span class="p">:</span>
                    <span class="n">stop</span> <span class="o">+=</span> <span class="n">n</span>
                <span class="n">rng</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">step</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">start</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">start</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="k">if</span> <span class="n">stop</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">stop</span>
                <span class="k">while</span> <span class="n">stop</span> <span class="o">&gt;=</span> <span class="n">start</span><span class="p">:</span>
                    <span class="n">stop</span> <span class="o">-=</span> <span class="n">n</span>
                <span class="n">rng</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">step</span><span class="p">)</span>

            <span class="n">limit</span> <span class="o">=</span> <span class="n">n</span> <span class="k">if</span> <span class="n">step</span> <span class="o">%</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">n</span> <span class="o">*</span> <span class="mi">2</span>
            <span class="n">out</span> <span class="o">=</span> <span class="nb">bytearray</span><span class="p">()</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rng</span><span class="p">:</span>
                <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="n">n</span><span class="p">))</span>
                <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">count</span> <span class="o">&gt;</span> <span class="n">limit</span><span class="p">:</span>
                    <span class="k">break</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">out</span><span class="p">))</span>

        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>

<div class="viewcode-block" id="CircularBytes.cutaround">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.CircularBytes.cutaround">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cutaround</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bytes</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a circular slice of given length starting at index `start`.</span>
<span class="sd">        Can exceed len(self), wrapping around as needed.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        s = CircularBytes(b&quot;ABCDE&quot;)</span>
<span class="sd">        assert s.cutaround(3, 7) == b&quot;DEABCDE&quot;</span>
<span class="sd">        assert s.cutaround(-1, 4) == b&quot;EABC&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">length</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="sa">b</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="n">start</span> <span class="o">%=</span> <span class="n">n</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">bytearray</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">):</span>
            <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="p">[(</span><span class="n">start</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">%</span> <span class="n">n</span><span class="p">])</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">out</span><span class="p">))</span></div>


<div class="viewcode-block" id="CircularBytes.find">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.CircularBytes.find">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">find</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sub</span><span class="p">:</span> <span class="nb">bytes</span> <span class="o">|</span> <span class="nb">bytearray</span> <span class="o">|</span> <span class="nb">memoryview</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find a subsequence in the circular sequence, possibly</span>
<span class="sd">        wrapping across the origin.</span>
<span class="sd">        Returns -1 if not found.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>

        <span class="n">end</span> <span class="o">=</span> <span class="n">n</span> <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">min</span><span class="p">(</span><span class="n">end</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
        <span class="n">doubled</span> <span class="o">=</span> <span class="bp">self</span> <span class="o">+</span> <span class="bp">self</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">sub</span> <span class="o">=</span> <span class="n">sub</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="n">pos</span> <span class="o">=</span> <span class="n">doubled</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">sub</span><span class="p">),</span> <span class="n">start</span><span class="p">,</span> <span class="n">n</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">pos</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="ow">or</span> <span class="n">pos</span> <span class="o">&gt;=</span> <span class="n">n</span><span class="p">:</span>
            <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">return</span> <span class="n">pos</span></div>
</div>



<div class="viewcode-block" id="Dseq">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Dseq</span><span class="p">(</span><span class="n">_Seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Dseq describes a double stranded DNA fragment, linear or circular.</span>

<span class="sd">    Dseq can be initiated in two ways, using two strings, each representing the</span>
<span class="sd">    Watson (upper, sense) strand, the Crick (lower, antisense) strand and an</span>
<span class="sd">    optional value describing the stagger betwen the strands on the left side (ovhg).</span>

<span class="sd">    Alternatively, a single string represenation using dsIUPAC codes can be used.</span>
<span class="sd">    If a single string is used, the letters of that string are interpreted as base</span>
<span class="sd">    pairs rather than single bases. For example &quot;A&quot; would indicate the basepair</span>
<span class="sd">    &quot;A/T&quot;. An expanded IUPAC code is used where the letters PEXI have been assigned</span>
<span class="sd">    to GATC on the Watson strand with no paring base on the Crick strand G/&quot;&quot;, A/&quot;&quot;,</span>
<span class="sd">    T/&quot;&quot; and C/&quot;&quot;. The letters QFZJ have been assigned the opposite base pairs with</span>
<span class="sd">    an empty Watson strand &quot;&quot;/G, &quot;&quot;/A, &quot;&quot;/T, and &quot;&quot;/C.</span>

<span class="sd">    ::</span>

<span class="sd">        PEXIGATCQFZJ  would indicate the linear double-stranded fragment:</span>

<span class="sd">        GATCGATC</span>
<span class="sd">            CTAGCTAG</span>



<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    watson : str</span>
<span class="sd">        a string representing the Watson (sense) DNA strand or a basepair</span>
<span class="sd">        represenation.</span>

<span class="sd">    crick : str, optional</span>
<span class="sd">        a string representing the Crick (antisense) DNA strand.</span>

<span class="sd">    ovhg : int, optional</span>
<span class="sd">        A positive or negative number to describe the stagger between the</span>
<span class="sd">        Watson and Crick strands.</span>
<span class="sd">        see below for a detailed explanation.</span>

<span class="sd">    circular : bool, optional</span>
<span class="sd">        True indicates that sequence is circular, False that it is linear.</span>


<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    Dseq is a subclass of the Biopython Bio.Seq.Seq class. The constructor</span>
<span class="sd">    can accept two strings representing the Watson (sense) and Crick(antisense)</span>
<span class="sd">    DNA strands. These are interpreted as single stranded DNA. There is a check</span>
<span class="sd">    for complementarity between the strands.</span>

<span class="sd">    If the DNA molecule is staggered on the left side, an integer ovhg</span>
<span class="sd">    (overhang) must be given, describing the stagger between the Watson and Crick strand</span>
<span class="sd">    in the 5&#39; end of the fragment.</span>

<span class="sd">    Additionally, the optional boolean parameter circular can be given to indicate if the</span>
<span class="sd">    DNA molecule is circular.</span>

<span class="sd">    The most common usage of the Dseq class is probably not to use it directly, but to</span>
<span class="sd">    create it as part of a Dseqrecord object (see :class:`pydna.dseqrecord.Dseqrecord`).</span>
<span class="sd">    This works in the same way as for the relationship between the :class:`Bio.Seq.Seq` and</span>
<span class="sd">    :class:`Bio.SeqRecord.SeqRecord` classes in Biopython.</span>

<span class="sd">    There are multiple ways of creating a Dseq object directly listed below, but you can also</span>
<span class="sd">    use the function Dseq.from_full_sequence_and_overhangs() to create a Dseq:</span>

<span class="sd">    Two arguments (string, string), no overhang provided:</span>

<span class="sd">    &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">    &gt;&gt;&gt; Dseq(&quot;gggaaat&quot;,&quot;ttt&quot;)</span>
<span class="sd">    Dseq(-7)</span>
<span class="sd">    gggaaat</span>
<span class="sd">       ttt</span>

<span class="sd">    If Watson and Crick are given, but not ovhg, an attempt will be made to find the best annealing</span>
<span class="sd">    between the strands. There are important limitations to this. If there are several ways to</span>
<span class="sd">    anneal the strands, this will fail. For long fragments it is quite slow.</span>

<span class="sd">    Three arguments (string, string, ovhg=int):</span>

<span class="sd">    The ovhg parameter is an integer describing the length of the Crick strand overhang on the</span>
<span class="sd">    left side (the 5&#39; end of Watson strand).</span>

<span class="sd">    The ovhg parameter controls the stagger at the five prime end::</span>

<span class="sd">        dsDNA       overhang</span>

<span class="sd">          nnn...    2</span>
<span class="sd">        nnnnn...</span>

<span class="sd">         nnnn...    1</span>
<span class="sd">        nnnnn...</span>

<span class="sd">        nnnnn...    0</span>
<span class="sd">        nnnnn...</span>

<span class="sd">        nnnnn...   -1</span>
<span class="sd">         nnnn...</span>

<span class="sd">        nnnnn...   -2</span>
<span class="sd">          nnn...</span>

<span class="sd">    Example of creating Dseq objects with different amounts of stagger:</span>

<span class="sd">    &gt;&gt;&gt; Dseq(watson=&quot;att&quot;, crick=&quot;acata&quot;, ovhg=-2)</span>
<span class="sd">    Dseq(-7)</span>
<span class="sd">    att</span>
<span class="sd">      ataca</span>
<span class="sd">    &gt;&gt;&gt; Dseq(watson=&quot;ata&quot;,crick=&quot;acata&quot;,ovhg=-1)</span>
<span class="sd">    Dseq(-6)</span>
<span class="sd">    ata</span>
<span class="sd">     ataca</span>
<span class="sd">    &gt;&gt;&gt; Dseq(watson=&quot;taa&quot;,crick=&quot;actta&quot;,ovhg=0)</span>
<span class="sd">    Dseq(-5)</span>
<span class="sd">    taa</span>
<span class="sd">    attca</span>
<span class="sd">    &gt;&gt;&gt; Dseq(watson=&quot;aag&quot;,crick=&quot;actta&quot;,ovhg=1)</span>
<span class="sd">    Dseq(-5)</span>
<span class="sd">     aag</span>
<span class="sd">    attca</span>
<span class="sd">    &gt;&gt;&gt; Dseq(watson=&quot;agt&quot;,crick=&quot;actta&quot;,ovhg=2)</span>
<span class="sd">    Dseq(-5)</span>
<span class="sd">      agt</span>
<span class="sd">    attca</span>

<span class="sd">    If the ovhg parameter is specified a Crick strand also needs to be supplied, or</span>
<span class="sd">    an exception is raised.</span>

<span class="sd">    &gt;&gt;&gt; Dseq(watson=&quot;agt&quot;, ovhg=2)</span>
<span class="sd">    Traceback (most recent call last):</span>
<span class="sd">        ...</span>
<span class="sd">    ValueError: ovhg (overhang) defined without a crick strand.</span>


<span class="sd">    The shape or topology of the fragment is set by the circular parameter, True or False (default).</span>

<span class="sd">    &gt;&gt;&gt; Dseq(&quot;aaa&quot;, &quot;ttt&quot;, ovhg = 0)  # A linear sequence by default</span>
<span class="sd">    Dseq(-3)</span>
<span class="sd">    aaa</span>
<span class="sd">    ttt</span>
<span class="sd">    &gt;&gt;&gt; Dseq(&quot;aaa&quot;, &quot;ttt&quot;, ovhg = 0, circular = False)  # A linear sequence if circular is False</span>
<span class="sd">    Dseq(-3)</span>
<span class="sd">    aaa</span>
<span class="sd">    ttt</span>
<span class="sd">    &gt;&gt;&gt; Dseq(&quot;aaa&quot;, &quot;ttt&quot;, ovhg = 0, circular = True)  # A circular sequence</span>
<span class="sd">    Dseq(o3)</span>
<span class="sd">    aaa</span>
<span class="sd">    ttt</span>
<span class="sd">    &gt;&gt;&gt; Dseq(&quot;aaa&quot;, &quot;ttt&quot;, ovhg=1, circular = False)</span>
<span class="sd">    Dseq(-4)</span>
<span class="sd">     aaa</span>
<span class="sd">    ttt</span>
<span class="sd">    &gt;&gt;&gt; Dseq(&quot;aaa&quot;,&quot;ttt&quot;,ovhg=-1)</span>
<span class="sd">    Dseq(-4)</span>
<span class="sd">    aaa</span>
<span class="sd">     ttt</span>
<span class="sd">    &gt;&gt;&gt; Dseq(&quot;aaa&quot;, &quot;ttt&quot;, circular = True , ovhg=0)</span>
<span class="sd">    Dseq(o3)</span>
<span class="sd">    aaa</span>
<span class="sd">    ttt</span>

<span class="sd">    &gt;&gt;&gt; a=Dseq(&quot;tttcccc&quot;,&quot;aaacccc&quot;)</span>
<span class="sd">    &gt;&gt;&gt; a</span>
<span class="sd">    Dseq(-11)</span>
<span class="sd">        tttcccc</span>
<span class="sd">    ccccaaa</span>
<span class="sd">    &gt;&gt;&gt; a.ovhg</span>
<span class="sd">    4</span>

<span class="sd">    &gt;&gt;&gt; b=Dseq(&quot;ccccttt&quot;,&quot;ccccaaa&quot;)</span>
<span class="sd">    &gt;&gt;&gt; b</span>
<span class="sd">    Dseq(-11)</span>
<span class="sd">    ccccttt</span>
<span class="sd">        aaacccc</span>
<span class="sd">    &gt;&gt;&gt; b.ovhg</span>
<span class="sd">    -4</span>
<span class="sd">    &gt;&gt;&gt;</span>


<span class="sd">    dsIUPAC [#]_ is an nn extension to the IUPAC alphabet used to describe ss regions:</span>

<span class="sd">    ::</span>

<span class="sd">            aaaGATC       GATCccc          ad-hoc representations</span>
<span class="sd">        CTAGttt               gggCTAG</span>

<span class="sd">        QFZJaaaPEXI       PEXIcccQFZJ      dsIUPAC</span>



<span class="sd">    Coercing to string</span>

<span class="sd">    &gt;&gt;&gt; str(a)</span>
<span class="sd">    &#39;ggggtttcccc&#39;</span>

<span class="sd">    A Dseq object can be longer that either the watson or crick strands.</span>

<span class="sd">    ::</span>

<span class="sd">        &lt;-- length --&gt;</span>
<span class="sd">        GATCCTTT</span>
<span class="sd">             AAAGCCTAG</span>

<span class="sd">        &lt;-- length --&gt;</span>
<span class="sd">              GATCCTTT</span>
<span class="sd">        AAAGCCCTA</span>

<span class="sd">    The slicing of a linear Dseq object works mostly as it does for a string.</span>

<span class="sd">    &gt;&gt;&gt; s=&quot;ggatcc&quot;</span>
<span class="sd">    &gt;&gt;&gt; s[2:3]</span>
<span class="sd">    &#39;a&#39;</span>
<span class="sd">    &gt;&gt;&gt; s[2:4]</span>
<span class="sd">    &#39;at&#39;</span>
<span class="sd">    &gt;&gt;&gt; s[2:4:-1]</span>
<span class="sd">    &#39;&#39;</span>
<span class="sd">    &gt;&gt;&gt; s[::2]</span>
<span class="sd">    &#39;gac&#39;</span>
<span class="sd">    &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">    &gt;&gt;&gt; d=Dseq(s, circular=False)</span>
<span class="sd">    &gt;&gt;&gt; d[2:3]</span>
<span class="sd">    Dseq(-1)</span>
<span class="sd">    a</span>
<span class="sd">    t</span>
<span class="sd">    &gt;&gt;&gt; d[2:4]</span>
<span class="sd">    Dseq(-2)</span>
<span class="sd">    at</span>
<span class="sd">    ta</span>
<span class="sd">    &gt;&gt;&gt; d[2:4:-1]</span>
<span class="sd">    Dseq(-0)</span>
<span class="sd">    &lt;BLANKLINE&gt;</span>
<span class="sd">    &lt;BLANKLINE&gt;</span>
<span class="sd">    &gt;&gt;&gt; d[::2]</span>
<span class="sd">    Dseq(-3)</span>
<span class="sd">    gac</span>
<span class="sd">    ctg</span>


<span class="sd">    The slicing of a circular Dseq object has a slightly different meaning.</span>


<span class="sd">    &gt;&gt;&gt; s=&quot;ggAtCc&quot;</span>
<span class="sd">    &gt;&gt;&gt; d=Dseq(s, circular=True)</span>
<span class="sd">    &gt;&gt;&gt; d</span>
<span class="sd">    Dseq(o6)</span>
<span class="sd">    ggAtCc</span>
<span class="sd">    ccTaGg</span>
<span class="sd">    &gt;&gt;&gt; d[4:3]</span>
<span class="sd">    Dseq(-5)</span>
<span class="sd">    CcggA</span>
<span class="sd">    GgccT</span>


<span class="sd">    The slice [X:X] produces an empty slice for a string, while this</span>
<span class="sd">    will return the linearized sequence starting at X:</span>

<span class="sd">    &gt;&gt;&gt; s=&quot;ggatcc&quot;</span>
<span class="sd">    &gt;&gt;&gt; d=Dseq(s, circular=True)</span>
<span class="sd">    &gt;&gt;&gt; d</span>
<span class="sd">    Dseq(o6)</span>
<span class="sd">    ggatcc</span>
<span class="sd">    cctagg</span>
<span class="sd">    &gt;&gt;&gt; d[3:3]</span>
<span class="sd">    Dseq(-6)</span>
<span class="sd">    tccgga</span>
<span class="sd">    aggcct</span>
<span class="sd">    &gt;&gt;&gt;</span>


<span class="sd">    See Also</span>
<span class="sd">    --------</span>
<span class="sd">    pydna.dseqrecord.Dseqrecord</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">watson</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">],</span>
        <span class="n">crick</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ovhg</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">pos</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">watson</span><span class="p">,</span> <span class="p">(</span><span class="nb">bytes</span><span class="p">,</span> <span class="nb">bytearray</span><span class="p">)):</span>
            <span class="c1"># watson is decoded to a string if needed.</span>
            <span class="n">watson</span> <span class="o">=</span> <span class="n">watson</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">crick</span><span class="p">,</span> <span class="p">(</span><span class="nb">bytes</span><span class="p">,</span> <span class="nb">bytearray</span><span class="p">)):</span>
            <span class="c1"># crick is decoded to a string if needed.</span>
            <span class="n">crick</span> <span class="o">=</span> <span class="n">crick</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">crick</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ovhg</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;ovhg (overhang) defined without a crick strand.&quot;</span><span class="p">)</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Giving only the watson string implies inferring the Crick complementary strand</span>
<span class="sd">            from the Watson sequence. The watson string can contain dscode letters wich will</span>
<span class="sd">            be interpreted as outlined in the pydna.alphabet module.</span>

<span class="sd">            The _data property must be a byte string for compatibility with</span>
<span class="sd">            Biopython Bio.Seq.Seq</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">watson</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Crick strand given, ovhg is optional. An important consequence is that the</span>
<span class="sd">            watson and crick strands are interpreted as single stranded DNA that is</span>
<span class="sd">            supposed to anneal.</span>

<span class="sd">            If ovhg was not given, we try to guess the value below. This will fail</span>
<span class="sd">            if there are two or more ways to anneal with equal length of the double</span>
<span class="sd">            stranded part.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="k">if</span> <span class="n">ovhg</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># ovhg not given, try to guess from sequences</span>
                <span class="n">limit</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">_math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">watson</span><span class="p">))</span> <span class="o">/</span> <span class="n">_math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">4</span><span class="p">))</span>
                <span class="n">olaps</span> <span class="o">=</span> <span class="n">_common_sub_strings</span><span class="p">(</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">watson</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">_rc</span><span class="p">(</span><span class="n">crick</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()),</span>
                    <span class="n">limit</span><span class="p">,</span>
                <span class="p">)</span>

<span class="w">                </span><span class="sd">&quot;&quot;&quot;No overlaps found, strands do not anneal&quot;&quot;&quot;</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">olaps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Could not anneal the two strands.&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot; looked for annealing with at least </span><span class="si">{</span><span class="n">limit</span><span class="si">}</span><span class="s2"> basepairs&quot;</span>
                        <span class="s2">&quot; Please provide and overhang value (ovhg parameter)&quot;</span>
                    <span class="p">)</span>

<span class="w">                </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                We extract the positions and length of the first (longest) overlap,</span>
<span class="sd">                since common_sub_strings sorts the overlaps by length, longest first.</span>
<span class="sd">                &quot;&quot;&quot;</span>

                <span class="p">(</span><span class="n">pos_watson</span><span class="p">,</span> <span class="n">pos_crick</span><span class="p">,</span> <span class="n">longest_olap_length</span><span class="p">),</span> <span class="o">*</span><span class="n">rest</span> <span class="o">=</span> <span class="n">olaps</span>

<span class="w">                </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                We see if there is another overlap of the same length</span>
<span class="sd">                This means that annealing is ambigous. User should provide</span>
<span class="sd">                and ovhg value.</span>
<span class="sd">                &quot;&quot;&quot;</span>
                <span class="k">if</span> <span class="nb">any</span><span class="p">(</span>
                    <span class="n">olap_length</span> <span class="o">&gt;=</span> <span class="n">longest_olap_length</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">olap_length</span> <span class="ow">in</span> <span class="n">rest</span>
                <span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;More than one way of annealing the&quot;</span>
                        <span class="s2">&quot; strands. Please provide ovhg value&quot;</span>
                    <span class="p">)</span>

                <span class="n">ovhg</span> <span class="o">=</span> <span class="n">pos_crick</span> <span class="o">-</span> <span class="n">pos_watson</span>

<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Pad both strands on left side ovhg spaces</span>
<span class="sd">            a negative number gives no padding,</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">sense</span> <span class="o">=</span> <span class="n">ovhg</span> <span class="o">*</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">watson</span>
            <span class="n">antisense</span> <span class="o">=</span> <span class="o">-</span><span class="n">ovhg</span> <span class="o">*</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">crick</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

            <span class="n">max_len</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sense</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">antisense</span><span class="p">))</span>

<span class="w">            </span><span class="sd">&quot;&quot;&quot;pad both strands on right side to same size.&quot;&quot;&quot;</span>
            <span class="n">sense</span> <span class="o">=</span> <span class="n">sense</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">max_len</span><span class="p">)</span>
            <span class="n">antisense</span> <span class="o">=</span> <span class="n">antisense</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">max_len</span><span class="p">)</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;both strands padded so that bsepairs align&quot;&quot;&quot;</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">sense</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">antisense</span><span class="p">)</span>

            <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">w</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sense</span><span class="p">,</span> <span class="n">antisense</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">basepair_dict</span><span class="p">[</span><span class="n">w</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span>
                <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Base mismatch in representation </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Base mismatch in representation: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">circular</span> <span class="o">=</span> <span class="n">circular</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">pos</span>

        <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">+=</span> <span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">dsb</span> <span class="o">=</span> <span class="n">dsbreaks</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">dsb</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dsb</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Molecule is internally split in </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">dsb</span><span class="p">)</span><span class="si">}</span><span class="s2"> location(s):</span><span class="se">\n\n</span><span class="si">{</span><span class="n">msg</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="p">)</span>

<div class="viewcode-block" id="Dseq.quick">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.quick">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">quick</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="nb">bytes</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pos</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Fastest way to instantiate an object of the Dseq class.</span>

<span class="sd">        No checks of parameters are made.</span>
<span class="sd">        Does not call Bio.Seq.Seq.__init__() which has lots of time consuming checks.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">obj</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>
        <span class="n">obj</span><span class="o">.</span><span class="n">circular</span> <span class="o">=</span> <span class="n">circular</span>
        <span class="n">obj</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="n">obj</span><span class="o">.</span><span class="n">_data</span> <span class="o">=</span> <span class="n">data</span>

        <span class="k">return</span> <span class="n">obj</span></div>


<div class="viewcode-block" id="Dseq.from_representation">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.from_representation">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_representation</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">dsdna</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="n">obj</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>
        <span class="n">obj</span><span class="o">.</span><span class="n">circular</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">obj</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">clean</span> <span class="o">=</span> <span class="n">_inspect</span><span class="o">.</span><span class="n">cleandoc</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">dsdna</span><span class="p">)</span>
        <span class="n">watson</span><span class="p">,</span> <span class="n">crick</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">ln</span>
            <span class="k">for</span> <span class="n">ln</span> <span class="ow">in</span> <span class="n">clean</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">ln</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">ln</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;Dseq(&quot;</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="n">ovhgw</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">watson</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">watson</span><span class="o">.</span><span class="n">lstrip</span><span class="p">())</span>
        <span class="n">ovhgc</span> <span class="o">=</span> <span class="o">-</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">crick</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">crick</span><span class="o">.</span><span class="n">lstrip</span><span class="p">()))</span>

        <span class="n">ovhg</span> <span class="o">=</span> <span class="n">ovhgw</span> <span class="ow">or</span> <span class="n">ovhgc</span>

        <span class="n">watson</span> <span class="o">=</span> <span class="n">watson</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">crick</span> <span class="o">=</span> <span class="n">crick</span><span class="o">.</span><span class="n">strip</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="p">,</span> <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.from_full_sequence_and_overhangs">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.from_full_sequence_and_overhangs">[docs]</a>
    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_full_sequence_and_overhangs</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">full_sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">crick_ovhg</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">watson_ovhg</span><span class="p">:</span> <span class="nb">int</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a linear Dseq object from a full sequence and the 3&#39; overhangs of each strand.</span>

<span class="sd">        The order of the parameters is like this because the 3&#39; overhang of the crick strand is the one</span>
<span class="sd">        on the left side of the sequence.</span>


<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        full_sequence: str</span>
<span class="sd">            The full sequence of the Dseq object.</span>

<span class="sd">        crick_ovhg: int</span>
<span class="sd">            The overhang of the crick strand in the 3&#39; end. Equivalent to Dseq.ovhg.</span>

<span class="sd">        watson_ovhg: int</span>
<span class="sd">            The overhang of the watson strand in the 5&#39; end.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Dseq</span>
<span class="sd">            A Dseq object.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; Dseq.from_full_sequence_and_overhangs(&#39;AAAAAA&#39;, crick_ovhg=2, watson_ovhg=2)</span>
<span class="sd">        Dseq(-6)</span>
<span class="sd">          AAAA</span>
<span class="sd">        TTTT</span>
<span class="sd">        &gt;&gt;&gt; Dseq.from_full_sequence_and_overhangs(&#39;AAAAAA&#39;, crick_ovhg=-2, watson_ovhg=2)</span>
<span class="sd">        Dseq(-6)</span>
<span class="sd">        AAAAAA</span>
<span class="sd">          TT</span>
<span class="sd">        &gt;&gt;&gt; Dseq.from_full_sequence_and_overhangs(&#39;AAAAAA&#39;, crick_ovhg=2, watson_ovhg=-2)</span>
<span class="sd">        Dseq(-6)</span>
<span class="sd">          AA</span>
<span class="sd">        TTTTTT</span>
<span class="sd">        &gt;&gt;&gt; Dseq.from_full_sequence_and_overhangs(&#39;AAAAAA&#39;, crick_ovhg=-2, watson_ovhg=-2)</span>
<span class="sd">        Dseq(-6)</span>
<span class="sd">        AAAA</span>
<span class="sd">          TTTT</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">full_sequence_rev</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Dseq</span><span class="p">(</span><span class="n">full_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span>
        <span class="n">watson</span> <span class="o">=</span> <span class="n">full_sequence</span>
        <span class="n">crick</span> <span class="o">=</span> <span class="n">full_sequence_rev</span>

        <span class="c1"># If necessary, we trim the left side</span>
        <span class="k">if</span> <span class="n">crick_ovhg</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">crick</span> <span class="o">=</span> <span class="n">crick</span><span class="p">[:</span><span class="n">crick_ovhg</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">crick_ovhg</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">watson</span> <span class="o">=</span> <span class="n">watson</span><span class="p">[</span><span class="n">crick_ovhg</span><span class="p">:]</span>

        <span class="c1"># If necessary, we trim the right side</span>
        <span class="k">if</span> <span class="n">watson_ovhg</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">watson</span> <span class="o">=</span> <span class="n">watson</span><span class="p">[:</span><span class="n">watson_ovhg</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">watson_ovhg</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">crick</span> <span class="o">=</span> <span class="n">crick</span><span class="p">[</span><span class="n">watson_ovhg</span><span class="p">:]</span>

        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="p">,</span> <span class="n">crick</span><span class="o">=</span><span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="n">crick_ovhg</span><span class="p">)</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">watson</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The watson (upper) strand of the double stranded fragment 5&#39;-3&#39;.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        TYPE</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">crick</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The crick (lower) strand of the double stranded fragment 5&#39;-3&#39;.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        TYPE</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">left_ovhg</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The 5&#39; overhang of the lower strand compared the the upper.</span>

<span class="sd">        See module docstring for more information.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        TYPE</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_watson</span> <span class="ow">or</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_crick</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="o">.</span><span class="n">sticky_left5</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="o">.</span><span class="n">sticky_left3</span><span class="p">)</span>

    <span class="n">ovhg</span> <span class="o">=</span> <span class="n">left_ovhg</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">right_ovhg</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Overhang at the right side (end).&quot;&quot;&quot;</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_watson</span> <span class="ow">or</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_crick</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="o">.</span><span class="n">sticky_right5</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="o">.</span><span class="n">sticky_right3</span><span class="p">)</span>

    <span class="n">watson_ovhg</span> <span class="o">=</span> <span class="n">right_ovhg</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        A string representation of the sequence. The returned string</span>
<span class="sd">        is the watson strand of a blunt version of the sequence.</span>

<span class="sd">        &gt;&gt;&gt; ds = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ... GAATTC</span>
<span class="sd">        ...   TAA</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>

<span class="sd">        &gt;&gt;&gt; str(ds)</span>
<span class="sd">        &#39;GAATTC&#39;</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ...   ATT</span>
<span class="sd">        ... CTTAAG</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>

<span class="sd">        &gt;&gt;&gt; str(ds)</span>
<span class="sd">        &#39;GAATTC&#39;</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            A string representation of the sequence.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">bytes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>

    <span class="n">to_blunt_string</span> <span class="o">=</span> <span class="fm">__str__</span>  <span class="c1"># alias of __str__ # TODO: consider removing</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__bytes__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bytes</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_full_sequence_table</span><span class="p">)</span>

<div class="viewcode-block" id="Dseq.mw">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.mw">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">mw</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;The molecular weight of the DNA/RNA molecule in g/mol.</span>

<span class="sd">        The molecular weight data in Biopython Bio.Data.IUPACData</span>
<span class="sd">        is used. The DNA is assumed to have a 5&#39;-phosphate as many</span>
<span class="sd">        DNA fragments from restriction digestion do:</span>

<span class="sd">        ::</span>

<span class="sd">             P - G-A-T-T-A-C-A - OH</span>
<span class="sd">                 | | | | | | |</span>
<span class="sd">            OH - C-T-A-A-T-G-T - P</span>

<span class="sd">        The molecular weights listed in the unambiguous_dna_weights</span>
<span class="sd">        dictionary refers to free monophosphate nucleotides.</span>
<span class="sd">        One water molecule is removed for every phopshodiester bond</span>
<span class="sd">        formed between nucleotides. For linear molecules, the weight</span>
<span class="sd">        of one water molecule is added to account for the terminal</span>
<span class="sd">        hydroxyl group and a hydrogen on the 5&#39; terminal phosphate</span>
<span class="sd">        group.</span>

<span class="sd">        ::</span>

<span class="sd">             P - G---A---T - OH  P - C---A - OH</span>
<span class="sd">                 |   |   |           |   |</span>
<span class="sd">            OH - C---T---A---A---T---G---T - P</span>

<span class="sd">        If the DNA is discontinuous, the internal 5&#39;- end is assumed</span>
<span class="sd">        to have a phosphate and the 3&#39;- a hydroxyl group:</span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds_lin_obj = Dseq(&quot;GATTACA&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds_lin_obj</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        GATTACA</span>
<span class="sd">        CTAATGT</span>
<span class="sd">        &gt;&gt;&gt; round(ds_lin_obj.mw(), 1)</span>
<span class="sd">        4359.8</span>
<span class="sd">        &gt;&gt;&gt; ds_circ_obj = Dseq(&quot;GATTACA&quot;, circular = True)</span>
<span class="sd">        &gt;&gt;&gt; round(ds_circ_obj.mw(), 1)</span>
<span class="sd">        4323.8</span>
<span class="sd">        &gt;&gt;&gt; ssobj = Dseq(&quot;PEXXEIE&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ssobj</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        GATTACA</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; round(ssobj.mw(), 1)</span>
<span class="sd">        2184.4</span>
<span class="sd">        &gt;&gt;&gt; ds_lin_obj2 = Dseq(&quot;GATZFCA&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds_lin_obj2</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        GAT  CA</span>
<span class="sd">        CTAATGT</span>
<span class="sd">        &gt;&gt;&gt; round(ds_lin_obj2.mw(), 1)</span>
<span class="sd">        3724.4</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">h2o</span> <span class="o">=</span> <span class="n">atom_weights</span><span class="p">[</span><span class="s2">&quot;H&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">atom_weights</span><span class="p">[</span><span class="s2">&quot;O&quot;</span><span class="p">]</span>

        <span class="n">mwd</span> <span class="o">=</span> <span class="n">unambiguous_rna_weights</span> <span class="o">|</span> <span class="n">unambiguous_dna_weights</span> <span class="o">|</span> <span class="p">{</span><span class="s2">&quot; &quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>

        <span class="n">watsn_weight</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">mwd</span><span class="p">[</span><span class="n">nt</span><span class="p">]</span> <span class="o">-</span> <span class="n">h2o</span> <span class="k">for</span> <span class="n">nt</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="n">crick_weight</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">mwd</span><span class="p">[</span><span class="n">nt</span><span class="p">]</span> <span class="o">-</span> <span class="n">h2o</span> <span class="k">for</span> <span class="n">nt</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>

        <span class="n">watsn_weight</span> <span class="o">+=</span> <span class="n">h2o</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot; +&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="p">))</span>
        <span class="n">crick_weight</span> <span class="o">+=</span> <span class="n">h2o</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot; +&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">watsn_weight</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">watsn_weight</span> <span class="o">+=</span> <span class="n">h2o</span>

        <span class="k">if</span> <span class="n">crick_weight</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">crick_weight</span> <span class="o">+=</span> <span class="n">h2o</span>

        <span class="k">return</span> <span class="n">watsn_weight</span> <span class="o">+</span> <span class="n">crick_weight</span></div>


<div class="viewcode-block" id="Dseq.find">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.find">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">find</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">sub</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_SeqAbstractBaseClass</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">],</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">_sys</span><span class="o">.</span><span class="n">maxsize</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;This method behaves like the python string method of the same name.</span>

<span class="sd">        Returns an integer, the index of the first occurrence of substring</span>
<span class="sd">        argument sub in the (sub)sequence given by [start:end].</span>

<span class="sd">        Returns -1 if the subsequence is NOT found.</span>

<span class="sd">        The search is case sensitive.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        sub : string or Seq object</span>
<span class="sd">            a string or another Seq object to look for.</span>

<span class="sd">        start : int, optional</span>
<span class="sd">            slice start.</span>

<span class="sd">        end : int, optional</span>
<span class="sd">            slice end.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; seq = Dseq(&quot;agtaagt&quot;)</span>
<span class="sd">        &gt;&gt;&gt; seq</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        agtaagt</span>
<span class="sd">        tcattca</span>
<span class="sd">        &gt;&gt;&gt; seq.find(&quot;taa&quot;)</span>
<span class="sd">        2</span>
<span class="sd">        &gt;&gt;&gt; seq = Dseq(watson=&quot;agta&quot;,crick=&quot;actta&quot;,ovhg=-2)</span>
<span class="sd">        &gt;&gt;&gt; seq</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        agta</span>
<span class="sd">          attca</span>
<span class="sd">        &gt;&gt;&gt; seq.find(&quot;taa&quot;)</span>
<span class="sd">        -1</span>
<span class="sd">        &gt;&gt;&gt; seq = Dseq(watson=&quot;agta&quot;,crick=&quot;actta&quot;,ovhg=-2)</span>
<span class="sd">        &gt;&gt;&gt; seq</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        agta</span>
<span class="sd">          attca</span>
<span class="sd">        &gt;&gt;&gt; seq.find(&quot;ta&quot;)</span>
<span class="sd">        2</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">CircularBytes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">)</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">result</span></div>


    <span class="k">def</span><span class="w"> </span><span class="fm">__contains__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sub</span><span class="p">:</span> <span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">sub</span><span class="p">)</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sl</span><span class="p">:</span> <span class="p">[</span><span class="nb">slice</span><span class="p">,</span> <span class="nb">int</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sl</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">sl</span> <span class="o">=</span> <span class="nb">slice</span><span class="p">(</span><span class="n">sl</span><span class="p">,</span> <span class="n">sl</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">sl</span> <span class="o">=</span> <span class="nb">slice</span><span class="p">(</span><span class="n">sl</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">sl</span><span class="o">.</span><span class="n">stop</span><span class="p">,</span> <span class="n">sl</span><span class="o">.</span><span class="n">step</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">cb</span> <span class="o">=</span> <span class="n">CircularBytes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">quick</span><span class="p">(</span><span class="n">cb</span><span class="p">[</span><span class="n">sl</span><span class="p">])</span>
        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">(</span><span class="n">sl</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare to another Dseq object OR an object that implements</span>
<span class="sd">        watson, crick and ovhg properties.</span>

<span class="sd">        This comparison is case insensitive.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">same</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">other</span><span class="o">.</span><span class="n">watson</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">crick</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span>
                <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">circular</span>
            <span class="p">)</span>
            <span class="c1"># Also test for alphabet ?</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">same</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="n">same</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">lim</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">length_limit_for_repr</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_pretty_str</span><span class="p">:</span>

        <span class="n">header</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">(</span><span class="si">{</span><span class="p">({</span><span class="kc">False</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="w"> </span><span class="kc">True</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;o&#39;</span><span class="p">}[</span><span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">])</span><span class="si">}{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span>

        <span class="n">w</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="n">representation_tuple</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">),</span> <span class="n">length_limit_for_repr</span><span class="o">=</span><span class="n">length_limit_for_repr</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">_pretty_str</span><span class="p">(</span><span class="n">header</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">w</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">c</span><span class="p">)</span>

<div class="viewcode-block" id="Dseq.reverse_complement">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.reverse_complement">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">reverse_complement</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;Dseq&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Dseq object where watson and crick have switched places.</span>

<span class="sd">        This represents the same double stranded sequence.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;catcgatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-8)</span>
<span class="sd">        catcgatc</span>
<span class="sd">        gtagctag</span>
<span class="sd">        &gt;&gt;&gt; b=a.reverse_complement()</span>
<span class="sd">        &gt;&gt;&gt; b</span>
<span class="sd">        Dseq(-8)</span>
<span class="sd">        gatcgatg</span>
<span class="sd">        ctagctac</span>
<span class="sd">        &gt;&gt;&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">quick</span><span class="p">(</span><span class="n">_rc</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">),</span> <span class="n">circular</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">)</span></div>


    <span class="n">rc</span> <span class="o">=</span> <span class="n">reverse_complement</span>  <span class="c1"># alias for reverse_complement</span>

<div class="viewcode-block" id="Dseq.shifted">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.shifted">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">shifted</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Shifted copy of a circular Dseq object.</span>

<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;TAAG&quot;, circular = True)</span>
<span class="sd">        &gt;&gt;&gt; ds.shifted(1) # First bp moved to right side:</span>
<span class="sd">        Dseq(o4)</span>
<span class="sd">        AAGT</span>
<span class="sd">        TTCA</span>
<span class="sd">        &gt;&gt;&gt; ds.shifted(-1) # Last bp moved to left side:</span>
<span class="sd">        Dseq(o4)</span>
<span class="sd">        GTAA</span>
<span class="sd">        CATT</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;DNA is not circular.&quot;</span><span class="p">)</span>
        <span class="n">shift</span> <span class="o">=</span> <span class="n">shift</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">shift</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">_copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">shift</span><span class="p">:]</span> <span class="o">+</span> <span class="bp">self</span><span class="p">[:</span><span class="n">shift</span><span class="p">])</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span></div>


<div class="viewcode-block" id="Dseq.looped">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.looped">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">looped</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Circularized Dseq object.</span>

<span class="sd">        This can only be done if the two ends are compatible,</span>
<span class="sd">        otherwise a TypeError is raised.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;catcgatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-8)</span>
<span class="sd">        catcgatc</span>
<span class="sd">        gtagctag</span>
<span class="sd">        &gt;&gt;&gt; a.looped()</span>
<span class="sd">        Dseq(o8)</span>
<span class="sd">        catcgatc</span>
<span class="sd">        gtagctag</span>
<span class="sd">        &gt;&gt;&gt; b = Dseq(&quot;iatcgatj&quot;)</span>
<span class="sd">        &gt;&gt;&gt; b</span>
<span class="sd">        Dseq(-8)</span>
<span class="sd">        catcgat</span>
<span class="sd">         tagctag</span>
<span class="sd">        &gt;&gt;&gt; b.looped()</span>
<span class="sd">        Dseq(o7)</span>
<span class="sd">        catcgat</span>
<span class="sd">        gtagcta</span>
<span class="sd">        &gt;&gt;&gt; c = Dseq(&quot;jatcgati&quot;)</span>
<span class="sd">        &gt;&gt;&gt; c</span>
<span class="sd">        Dseq(-8)</span>
<span class="sd">         atcgatc</span>
<span class="sd">        gtagcta</span>
<span class="sd">        &gt;&gt;&gt; c.looped()</span>
<span class="sd">        Dseq(o7)</span>
<span class="sd">        catcgat</span>
<span class="sd">        gtagcta</span>
<span class="sd">        &gt;&gt;&gt; d = Dseq(&quot;ietcgazj&quot;)</span>
<span class="sd">        &gt;&gt;&gt; d</span>
<span class="sd">        Dseq(-8)</span>
<span class="sd">        catcga</span>
<span class="sd">          agctag</span>
<span class="sd">        &gt;&gt;&gt; d.looped()</span>
<span class="sd">        Traceback (most recent call last):</span>
<span class="sd">          File &quot;&lt;stdin&gt;&quot;, line 1, in &lt;module&gt;</span>
<span class="sd">          File &quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;, line 357, in looped</span>
<span class="sd">            if type5 == type3 and str(sticky5) == str(rc(sticky3)):</span>
<span class="sd">        TypeError: DNA cannot be circularized.</span>
<span class="sd">        5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="sd">        &gt;&gt;&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">_copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

        <span class="n">type5</span><span class="p">,</span> <span class="n">sticky5</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
        <span class="n">type3</span><span class="p">,</span> <span class="n">sticky3</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>

        <span class="n">err</span> <span class="o">=</span> <span class="ne">TypeError</span><span class="p">(</span>
            <span class="s2">&quot;DNA cannot be circularized.</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="s2">&quot;5&#39; and 3&#39; sticky ends not compatible!&quot;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">type5</span> <span class="o">!=</span> <span class="n">type3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">err</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Test if sticky ends are compatible</span>
            <span class="bp">self</span> <span class="o">+</span> <span class="bp">self</span>
        <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">err</span>

        <span class="n">new</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cast_to_ds_left</span><span class="p">()[:</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)]</span>

        <span class="n">new</span><span class="o">.</span><span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="n">new</span></div>


<div class="viewcode-block" id="Dseq.five_prime_end">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.five_prime_end">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">five_prime_end</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a 2-tuple of trings describing the structure of the 5&#39; end of</span>
<span class="sd">        the DNA fragment.</span>

<span class="sd">        The tuple contains (type , sticky) where type is eiter &quot;5&#39;&quot; or &quot;3&#39;&quot;.</span>
<span class="sd">        sticky is always in lower case and contains the sequence of the</span>
<span class="sd">        protruding end in 5&#39;-3&#39; direction.</span>

<span class="sd">        See examples below:</span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;aa&quot;, &quot;tttg&quot;, ovhg=2)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">          aa</span>
<span class="sd">        gttt</span>
<span class="sd">        &gt;&gt;&gt; a.five_prime_end()</span>
<span class="sd">        (&quot;3&#39;&quot;, &#39;tg&#39;)</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;caaa&quot;, &quot;tt&quot;, ovhg=-2)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        caaa</span>
<span class="sd">          tt</span>
<span class="sd">        &gt;&gt;&gt; a.five_prime_end()</span>
<span class="sd">        (&quot;5&#39;&quot;, &#39;ca&#39;)</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;aa&quot;, &quot;tt&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        aa</span>
<span class="sd">        tt</span>
<span class="sd">        &gt;&gt;&gt; a.five_prime_end()</span>
<span class="sd">        (&#39;blunt&#39;, &#39;&#39;)</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>
<span class="sd">        pydna.dseq.Dseq.three_prime_end</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># See docstring for function pydna.utils.get_parts for details</span>
        <span class="c1"># on what is contained in parts.</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>

        <span class="n">sticky5</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">sticky_left5</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span>

        <span class="n">sticky3</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">sticky_left3</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">single_watson</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_watson</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span>

        <span class="n">single_crick</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_crick</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># The walrus operator returns the value being assigned, so</span>
        <span class="c1"># we can test if it is empty or not.</span>
        <span class="k">if</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">single_watson</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;single&quot;</span>
        <span class="k">elif</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">single_crick</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;single&quot;</span>
        <span class="k">elif</span> <span class="n">sticky5</span> <span class="o">==</span> <span class="n">sticky3</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">type_</span><span class="p">,</span> <span class="n">sticky</span> <span class="o">=</span> <span class="s2">&quot;blunt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
        <span class="k">elif</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">sticky5</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;5&#39;&quot;</span>
        <span class="k">elif</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">sticky3</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;3&#39;&quot;</span>

        <span class="k">return</span> <span class="n">type_</span><span class="p">,</span> <span class="n">sticky</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span></div>


<div class="viewcode-block" id="Dseq.three_prime_end">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.three_prime_end">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">three_prime_end</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a tuple describing the structure of the 5&#39; end of</span>
<span class="sd">        the DNA fragment</span>

<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;aa&quot;, &quot;gttt&quot;, ovhg=0)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        aa</span>
<span class="sd">        tttg</span>
<span class="sd">        &gt;&gt;&gt; a.three_prime_end()</span>
<span class="sd">        (&quot;5&#39;&quot;, &#39;gt&#39;)</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;aaac&quot;, &quot;tt&quot;, ovhg=0)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        aaac</span>
<span class="sd">        tt</span>
<span class="sd">        &gt;&gt;&gt; a.three_prime_end()</span>
<span class="sd">        (&quot;3&#39;&quot;, &#39;ac&#39;)</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;aaa&quot;, &quot;ttt&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        aaa</span>
<span class="sd">        ttt</span>
<span class="sd">        &gt;&gt;&gt; a.three_prime_end()</span>
<span class="sd">        (&#39;blunt&#39;, &#39;&#39;)</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>
<span class="sd">        pydna.dseq.Dseq.five_prime_end</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># See docstring for function pydna.utils.get_parts for details</span>
        <span class="c1"># on what is contained in parts.</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>

        <span class="n">sticky5</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">sticky_right5</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">sticky3</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">sticky_right3</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span>

        <span class="n">single_watson</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_watson</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span>

        <span class="n">single_crick</span> <span class="o">=</span> <span class="n">parts</span><span class="o">.</span><span class="n">single_crick</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># The walrus operator returns the value being assigned, so</span>
        <span class="c1"># we can test if it is empty or not.</span>
        <span class="k">if</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">single_watson</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;single&quot;</span>
        <span class="k">elif</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">single_crick</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;single&quot;</span>
        <span class="k">elif</span> <span class="n">sticky5</span> <span class="o">==</span> <span class="n">sticky3</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">type_</span><span class="p">,</span> <span class="n">sticky</span> <span class="o">=</span> <span class="s2">&quot;blunt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
        <span class="k">elif</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">sticky5</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;5&#39;&quot;</span>
        <span class="k">elif</span> <span class="n">sticky</span> <span class="o">:=</span> <span class="n">sticky3</span><span class="p">:</span>
            <span class="n">type_</span> <span class="o">=</span> <span class="s2">&quot;3&#39;&quot;</span>

        <span class="k">return</span> <span class="n">type_</span><span class="p">,</span> <span class="n">sticky</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span></div>


    <span class="k">def</span><span class="w"> </span><span class="fm">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="p">[</span><span class="n">DseqType</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adding two Dseq objects together.</span>

<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;a&quot;, &quot;t&quot;, ovhg=0)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-1)</span>
<span class="sd">        a</span>
<span class="sd">        t</span>
<span class="sd">        &gt;&gt;&gt; ds + ds</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        aa</span>
<span class="sd">        tt</span>
<span class="sd">        &gt;&gt;&gt; &quot;g&quot; + ds # adding a string of left side returns a Dseq</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        ga</span>
<span class="sd">        ct</span>
<span class="sd">        &gt;&gt;&gt; ds + &quot;c&quot; # adding a string of right side returns a Dseq</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        ac</span>
<span class="sd">        tg</span>


<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : [DseqType, str, bytes]</span>
<span class="sd">            Object to be added.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            Preventing adding to a circular sequence.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        DseqType</span>
<span class="sd">            A new Dseq object.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;circular DNA cannot be ligated!&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">other</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;circular DNA cannot be ligated!&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="c1"># If other evaluates to False, return a copy of self.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">other</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">_copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="c1"># If self evaluates to False, return a copy of other.</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">_copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>

        <span class="c1"># get right side end properties for self.</span>
        <span class="n">self_type</span><span class="p">,</span> <span class="n">self_tail</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">other_type</span><span class="p">,</span> <span class="n">other_tail</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="c1"># if other does not have the expected properties</span>
            <span class="c1"># most likely it is a string that can be cast as</span>
            <span class="c1"># a Dseq.</span>
            <span class="n">other_type</span><span class="p">,</span> <span class="n">other_tail</span> <span class="o">=</span> <span class="s2">&quot;blunt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
            <span class="n">other</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">other</span><span class="p">)</span>

        <span class="n">err</span> <span class="o">=</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;sticky ends not compatible!&quot;</span><span class="p">)</span>

        <span class="c1"># The sticky ends has to be of the same type</span>
        <span class="c1"># or</span>
        <span class="c1"># one or both of is &quot;single&quot; indicating a stranded molecule.</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">self_type</span> <span class="o">!=</span> <span class="n">other_type</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;single&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="n">self_type</span><span class="p">,</span> <span class="n">other_type</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="n">err</span>

        <span class="c1"># tail length has to be equal for two phosphdiester bonds to form</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">self_tail</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">other_tail</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">err</span>

        <span class="c1"># Each basepair is checked against the pydna.alphabet basepair_dict</span>
        <span class="c1"># which contains the permitted base pairings.</span>
        <span class="k">for</span> <span class="n">w</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">self_tail</span><span class="p">,</span> <span class="n">other_tail</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">basepair_dict</span><span class="p">[(</span><span class="n">w</span><span class="p">,</span> <span class="n">c</span><span class="p">)]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">err</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">watson</span> <span class="o">+</span> <span class="n">other</span><span class="o">.</span><span class="n">watson</span><span class="p">,</span> <span class="n">other</span><span class="o">.</span><span class="n">crick</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__mul__</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">number</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;TypeError: can&#39;t multiply Dseq&quot;</span> <span class="sa">f</span><span class="s2">&quot; by non-int of type </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">number</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number</span><span class="p">)))</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_fill_in_left</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">nucleotides</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">stuffer</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="nb">type</span><span class="p">,</span> <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s2">&quot;5&#39;&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_rc</span><span class="p">(</span><span class="n">se</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">nucleotides</span><span class="p">:</span>
                    <span class="n">stuffer</span> <span class="o">+=</span> <span class="n">n</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span> <span class="o">+</span> <span class="n">stuffer</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">stuffer</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_fill_in_right</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">nucleotides</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">stuffer</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="nb">type</span><span class="p">,</span> <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s2">&quot;5&#39;&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">_rc</span><span class="p">(</span><span class="n">se</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">nucleotides</span><span class="p">:</span>
                    <span class="n">stuffer</span> <span class="o">+=</span> <span class="n">n</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson</span> <span class="o">+</span> <span class="n">stuffer</span>

<div class="viewcode-block" id="Dseq.fill_in">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.fill_in">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">fill_in</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nucleotides</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Fill in of five prime protruding end with a DNA polymerase</span>
<span class="sd">        that has only DNA polymerase activity (such as Exo-Klenow [#]_).</span>
<span class="sd">        Exo-Klenow is a modified version of the Klenow fragment of E.</span>
<span class="sd">        coli DNA polymerase I, which has been engineered to lack both</span>
<span class="sd">        3-5 proofreading and 5-3 exonuclease activities.</span>

<span class="sd">        and any combination of A, G, C or T. Default are all four</span>
<span class="sd">        nucleotides together.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        nucleotides : str</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; b=Dseq(&quot;caaa&quot;, &quot;cttt&quot;)</span>
<span class="sd">        &gt;&gt;&gt; b</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">        caaa</span>
<span class="sd">         tttc</span>
<span class="sd">        &gt;&gt;&gt; b.fill_in()</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">        caaag</span>
<span class="sd">        gtttc</span>
<span class="sd">        &gt;&gt;&gt; b.fill_in(&quot;g&quot;)</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">        caaag</span>
<span class="sd">        gtttc</span>
<span class="sd">        &gt;&gt;&gt; b.fill_in(&quot;tac&quot;)</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">        caaa</span>
<span class="sd">         tttc</span>
<span class="sd">        &gt;&gt;&gt; c=Dseq(&quot;aaac&quot;, &quot;tttg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; c</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">         aaac</span>
<span class="sd">        gttt</span>
<span class="sd">        &gt;&gt;&gt; c.fill_in()</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">         aaac</span>
<span class="sd">        gttt</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;aaa&quot;, &quot;ttt&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        aaa</span>
<span class="sd">        ttt</span>
<span class="sd">        &gt;&gt;&gt; a.fill_in()</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        aaa</span>
<span class="sd">        ttt</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [#] http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">nucleotides</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">nucleotides</span> <span class="o">=</span> <span class="s2">&quot;GATCRYWSMKHBVDN&quot;</span>

        <span class="n">nucleotides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">nucleotides</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">+</span> <span class="n">nucleotides</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fill_in_left</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
        <span class="n">watson</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fill_in_right</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="p">,</span> <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span></div>


    <span class="n">klenow</span> <span class="o">=</span> <span class="n">fill_in</span>  <span class="c1"># alias</span>

<div class="viewcode-block" id="Dseq.nibble_to_blunt">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.nibble_to_blunt">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">nibble_to_blunt</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Simulates treatment a nuclease with both 5&#39;-3&#39; and 3&#39;-5&#39; single</span>
<span class="sd">        strand specific exonuclease activity (such as mung bean nuclease [#]_)</span>

<span class="sd">        Mung bean nuclease is a nuclease enzyme derived from mung bean sprouts</span>
<span class="sd">        that preferentially degrades single-stranded DNA and RNA into</span>
<span class="sd">        5&#39;-phosphate- and 3&#39;-hydroxyl-containing nucleotides.</span>

<span class="sd">        Treatment results in blunt DNA, regardless of wheter the protruding end</span>
<span class="sd">        is 5&#39; or 3&#39;.</span>

<span class="sd">        ::</span>

<span class="sd">             ggatcc    -&gt;     gatcc</span>
<span class="sd">              ctaggg          ctagg</span>

<span class="sd">              ggatcc   -&gt;     ggatc</span>
<span class="sd">             tcctag           cctag</span>

<span class="sd">         &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">         &gt;&gt;&gt; b=Dseq(&quot;caaa&quot;, &quot;cttt&quot;)</span>
<span class="sd">         &gt;&gt;&gt; b</span>
<span class="sd">         Dseq(-5)</span>
<span class="sd">         caaa</span>
<span class="sd">          tttc</span>
<span class="sd">         &gt;&gt;&gt; b.mung()</span>
<span class="sd">         Dseq(-3)</span>
<span class="sd">         aaa</span>
<span class="sd">         ttt</span>
<span class="sd">         &gt;&gt;&gt; c=Dseq(&quot;aaac&quot;, &quot;tttg&quot;)</span>
<span class="sd">         &gt;&gt;&gt; c</span>
<span class="sd">         Dseq(-5)</span>
<span class="sd">          aaac</span>
<span class="sd">         gttt</span>
<span class="sd">         &gt;&gt;&gt; c.mung()</span>
<span class="sd">         Dseq(-3)</span>
<span class="sd">         aaa</span>
<span class="sd">         ttt</span>



<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [#] http://en.wikipedia.org/wiki/Mung_bean_nuclease</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">parts</span><span class="o">.</span><span class="n">middle</span><span class="p">)</span></div>


    <span class="n">mung</span> <span class="o">=</span> <span class="n">nibble_to_blunt</span>

<div class="viewcode-block" id="Dseq.T4">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.T4">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">T4</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nucleotides</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fill in 5&#39; protruding ends and nibble 3&#39; protruding ends.</span>

<span class="sd">        This is done using a DNA polymerase providing 3&#39;-5&#39; nuclease activity</span>
<span class="sd">        such as T4 DNA polymerase. This can be done in presence of any</span>
<span class="sd">        combination of the four nucleotides A, G, C or T.</span>

<span class="sd">        T4 DNA polymerase is widely used to polish DNA ends because of its</span>
<span class="sd">        strong 3-5 exonuclease activity in the absence of dNTPs, it chews</span>
<span class="sd">        back 3 overhangs to create blunt ends; in the presence of limiting</span>
<span class="sd">        dNTPs, it can fill in 5 overhangs; and by carefully controlling</span>
<span class="sd">        reaction time, temperature, and nucleotide supply, you can generate</span>
<span class="sd">        defined recessed or blunt termini.</span>

<span class="sd">        Tuning the nucleotide set can facilitate engineering of partial</span>
<span class="sd">        sticky ends. Default are all four nucleotides together.</span>

<span class="sd">        ::</span>

<span class="sd">                  aaagatc-3        aaa      3&#39; ends are always removed.</span>
<span class="sd">                  |||       ---&gt;   |||      A and T needed or the molecule will</span>
<span class="sd">            3-ctagttt              ttt      degrade completely.</span>



<span class="sd">            5-gatcaaa              gatcaaaGATC      5&#39; ends are filled in the</span>
<span class="sd">                  |||       ---&gt;   |||||||||||      presence of GATC</span>
<span class="sd">                  tttctag-5        CTAGtttctag</span>



<span class="sd">            5-gatcaaa              gatcaaaGAT       5&#39; ends are partially filled in the</span>
<span class="sd">                  |||       ---&gt;    |||||||||       presence of GAT to produce a 1 nt</span>
<span class="sd">                  tttctag-5         TAGtttctag      5&#39; overhang</span>



<span class="sd">            5-gatcaaa              gatcaaaGA       5&#39; ends are partially filled in the</span>
<span class="sd">                  |||       ---&gt;     |||||||       presence of GA to produce a 2 nt</span>
<span class="sd">                  tttctag-5          AGtttctag     5&#39; overhang</span>



<span class="sd">            5-gatcaaa              gatcaaaG        5&#39; ends are partially filled in the</span>
<span class="sd">                  |||       ---&gt;      |||||        presence of G to produce a 3 nt</span>
<span class="sd">                  tttctag-5           Gtttctag     5&#39; overhang</span>



<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        nucleotides : str</span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ... gatcaaa</span>
<span class="sd">        ...     tttctag</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-11)</span>
<span class="sd">        gatcaaa</span>
<span class="sd">            tttctag</span>
<span class="sd">        &gt;&gt;&gt; a.T4()</span>
<span class="sd">        Dseq(-11)</span>
<span class="sd">        gatcaaagatc</span>
<span class="sd">        ctagtttctag</span>
<span class="sd">        &gt;&gt;&gt; a.T4(&quot;GAT&quot;)</span>
<span class="sd">        Dseq(-11)</span>
<span class="sd">        gatcaaagat</span>
<span class="sd">         tagtttctag</span>
<span class="sd">        &gt;&gt;&gt; a.T4(&quot;GA&quot;)</span>
<span class="sd">        Dseq(-11)</span>
<span class="sd">        gatcaaaga</span>
<span class="sd">          agtttctag</span>
<span class="sd">        &gt;&gt;&gt; a.T4(&quot;G&quot;)</span>
<span class="sd">        Dseq(-11)</span>
<span class="sd">        gatcaaag</span>
<span class="sd">           gtttctag</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">nucleotides</span><span class="p">:</span>
            <span class="n">nucleotides</span> <span class="o">=</span> <span class="s2">&quot;GATCRYWSMKHBVDN&quot;</span>
        <span class="n">nucleotides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">nucleotides</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">+</span> <span class="n">nucleotides</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="nb">type</span><span class="p">,</span> <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s2">&quot;5&#39;&quot;</span><span class="p">:</span>
            <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fill_in_left</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s2">&quot;3&#39;&quot;</span><span class="p">:</span>
                <span class="n">ovhg</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">crick</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="p">[:</span> <span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">se</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ovhg</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">crick</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">crick</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">crick</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="ow">in</span> <span class="n">nucleotides</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">x</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="n">ovhg</span> <span class="o">=</span> <span class="n">x</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">crick</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">ovhg</span>
        <span class="n">crick</span> <span class="o">=</span> <span class="n">crick</span><span class="p">[:</span> <span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">crick</span><span class="p">:</span>
            <span class="n">ovhg</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">watson</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson</span>
        <span class="nb">type</span><span class="p">,</span> <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s2">&quot;5&#39;&quot;</span><span class="p">:</span>
            <span class="n">watson</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fill_in_right</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s2">&quot;3&#39;&quot;</span><span class="p">:</span>
                <span class="n">watson</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="p">[:</span> <span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">se</span><span class="p">)]</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">watson</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">watson</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="ow">in</span> <span class="n">nucleotides</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">x</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="n">watson</span> <span class="o">=</span> <span class="n">watson</span><span class="p">[:</span> <span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="p">,</span> <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span></div>


    <span class="n">t4</span> <span class="o">=</span> <span class="n">T4</span>  <span class="c1"># alias for the T4 method.</span>

<div class="viewcode-block" id="Dseq.nibble_five_prime_left">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.nibble_five_prime_left">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">nibble_five_prime_left</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        5&#39; =&gt; 3&#39;  resection at the left side (start) of the molecule.</span>

<span class="sd">        The argument n indicate the number of nucleotides that are to be</span>
<span class="sd">        removed. The outcome of this depend on the structure of the molecule.</span>
<span class="sd">        See the two examples below:</span>

<span class="sd">        The figure below indicates a recess of length two from a blunt DNA</span>
<span class="sd">        fragment. The resulting DNA fragment has a 3&#39; protruding single strand.</span>

<span class="sd">        ::</span>

<span class="sd">            gatc           tc</span>
<span class="sd">            ||||   --&gt;     ||</span>
<span class="sd">            ctag         ctag</span>


<span class="sd">        The figure below indicates a recess of length two from a DNA fragment</span>
<span class="sd">        with a 5&#39; sticky end resulting in a blunt sequence.</span>

<span class="sd">        ::</span>

<span class="sd">          ttgatc         gatc</span>
<span class="sd">            ||||   --&gt;   ||||</span>
<span class="sd">            ctag         ctag</span>


<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;gatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_left(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">          tc</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_left(3)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">           c</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_left(4)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ... GGgatc</span>
<span class="sd">        ...   ctag</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-6)</span>
<span class="sd">        GGgatc</span>
<span class="sd">          ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_left(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        n : int, optional</span>
<span class="sd">            The default is 1. This is the number of nucleotides removed.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        DseqType</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">recessed</span> <span class="o">=</span> <span class="n">_copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">+=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span> <span class="ow">or</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">recessed</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">n</span><span class="p">]</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">complement_table_for_dscode</span><span class="p">)</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_tail_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">lstrip</span><span class="p">()</span>
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">n</span><span class="p">:]</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">recessed</span></div>


<div class="viewcode-block" id="Dseq.nibble_five_prime_right">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.nibble_five_prime_right">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">nibble_five_prime_right</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        5&#39; =&gt; 3&#39;  resection at the right side (end) of the molecule.</span>

<span class="sd">        The argument n indicate the number of nucleotides that are to be</span>
<span class="sd">        removed. The outcome of this depend on the structure of the molecule.</span>
<span class="sd">        See the two examples below:</span>

<span class="sd">        The figure below indicates a recess of length two from a blunt DNA</span>
<span class="sd">        fragment. The resulting DNA fragment has a 3&#39; protruding single strand.</span>

<span class="sd">        ::</span>

<span class="sd">            gatc         gatc</span>
<span class="sd">            ||||   --&gt;   ||</span>
<span class="sd">            ctag         ct</span>

<span class="sd">        The figure below indicates a recess of length two from a DNA fragment</span>
<span class="sd">        with a 5&#39; sticky end resulting in a blunt sequence.</span>

<span class="sd">        ::</span>

<span class="sd">            gatc         gatc</span>
<span class="sd">            ||||   --&gt;   ||||</span>
<span class="sd">            ctagtt       ctag</span>


<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;gatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_right(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ct</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_right(3)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        c</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_right(4)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ... gatc</span>
<span class="sd">        ... ctagGG</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_five_prime_right(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">recessed</span> <span class="o">=</span> <span class="n">_copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="n">n</span>
        <span class="n">ovhg</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">right_ovhg</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">right_ovhg</span>
        <span class="n">n</span> <span class="o">-=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span>
        <span class="n">recessed</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">n</span><span class="p">]</span>
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">n</span><span class="p">:]</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_tail_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">lstrip</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">recessed</span></div>


    <span class="n">exo1_front</span> <span class="o">=</span> <span class="n">nibble_five_prime_left</span>  <span class="c1"># TODO: consider using the new names</span>
    <span class="n">exo1_end</span> <span class="o">=</span> <span class="n">nibble_five_prime_right</span>  <span class="c1"># TODO: consider using the new names</span>

<div class="viewcode-block" id="Dseq.nibble_three_prime_left">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.nibble_three_prime_left">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">nibble_three_prime_left</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        3&#39; =&gt; 5&#39; resection at the left side (beginning) of the molecule.</span>

<span class="sd">        The argument n indicate the number of nucleotides that are to be</span>
<span class="sd">        removed. The outcome of this depend on the structure of the molecule.</span>
<span class="sd">        See the two examples below:</span>

<span class="sd">        The figure below indicates a recess of length two from a blunt DNA</span>
<span class="sd">        fragment. The resulting DNA fragment has a 5&#39; protruding single strand.</span>

<span class="sd">        ::</span>

<span class="sd">            gatc         gatc</span>
<span class="sd">            ||||   --&gt;     ||</span>
<span class="sd">            ctag           ag</span>

<span class="sd">        The figure below indicates a recess of length two from a DNA fragment</span>
<span class="sd">        with a 3&#39; sticky end resulting in a blunt sequence.</span>

<span class="sd">        ::</span>

<span class="sd">            gatc         gatc</span>
<span class="sd">            ||||   --&gt;   ||||</span>
<span class="sd">          ttctag         ctag</span>


<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;gatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_left(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">          ag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_left(3)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">           g</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_left(4)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ...   gatc</span>
<span class="sd">        ... CCctag</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-6)</span>
<span class="sd">          gatc</span>
<span class="sd">        CCctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_left(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ovhg</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span>
        <span class="n">n</span> <span class="o">-=</span> <span class="nb">min</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span>
        <span class="n">recessed</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">n</span><span class="p">]</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_tail_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">lstrip</span><span class="p">()</span>
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">n</span><span class="p">:]</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">recessed</span></div>


<div class="viewcode-block" id="Dseq.nibble_three_prime_right">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.nibble_three_prime_right">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">nibble_three_prime_right</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        3&#39; =&gt; 5&#39; resection at the right side (end) of the molecule.</span>

<span class="sd">        The argument n indicate the number of nucleotides that are to be</span>
<span class="sd">        removed. The outcome of this depend on the structure of the molecule.</span>
<span class="sd">        See the two examples below:</span>

<span class="sd">        The figure below indicates a recess of length two from a blunt DNA</span>
<span class="sd">        fragment. The resulting DNA fragment has a 5&#39; protruding single strand.</span>

<span class="sd">        ::</span>

<span class="sd">            gatc         ga</span>
<span class="sd">            ||||   --&gt;   ||</span>
<span class="sd">            ctag         ctag</span>

<span class="sd">        The figure below indicates a recess of length two from a DNA fragment</span>
<span class="sd">        with a 3&#39; sticky end resulting in a blunt sequence.</span>

<span class="sd">        ::</span>

<span class="sd">            gatctt       gatc</span>
<span class="sd">            ||||   --&gt;   ||||</span>
<span class="sd">            ctag         ctag</span>


<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;gatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_right(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        ga</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_right(3)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        g</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_right(4)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq.from_representation(</span>
<span class="sd">        ... &#39;&#39;&#39;</span>
<span class="sd">        ... gatcCC</span>
<span class="sd">        ... ctag</span>
<span class="sd">        ... &#39;&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; ds.nibble_three_prime_right(2)</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gatc</span>
<span class="sd">        ctag</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="n">n</span>
        <span class="n">ovhg</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">right_ovhg</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">right_ovhg</span>
        <span class="n">n</span> <span class="o">+=</span> <span class="nb">min</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span>
        <span class="n">recessed</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">n</span><span class="p">]</span>
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">n</span><span class="p">:]</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">complement_table_for_dscode</span><span class="p">)</span>
            <span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_tail_table</span><span class="p">)</span>
            <span class="o">.</span><span class="n">lstrip</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">recessed</span></div>


<div class="viewcode-block" id="Dseq.no_cutters">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.no_cutters">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">no_cutters</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_RestrictionBatch</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_RestrictionBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Enzymes in a RestrictionBatch not cutting sequence.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">batch</span> <span class="o">=</span> <span class="n">CommOnly</span>
        <span class="n">ana</span> <span class="o">=</span> <span class="n">batch</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">ncut</span> <span class="o">=</span> <span class="p">{</span><span class="n">enz</span><span class="p">:</span> <span class="n">sitelist</span> <span class="k">for</span> <span class="p">(</span><span class="n">enz</span><span class="p">,</span> <span class="n">sitelist</span><span class="p">)</span> <span class="ow">in</span> <span class="n">ana</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">sitelist</span><span class="p">}</span>
        <span class="k">return</span> <span class="n">_RestrictionBatch</span><span class="p">(</span><span class="n">ncut</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.unique_cutters">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.unique_cutters">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">unique_cutters</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_RestrictionBatch</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_RestrictionBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Enzymes in a RestrictionBatch cutting sequence once.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">batch</span> <span class="o">=</span> <span class="n">CommOnly</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_cutters</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">batch</span><span class="o">=</span><span class="n">batch</span><span class="p">)</span></div>


    <span class="n">once_cutters</span> <span class="o">=</span> <span class="n">unique_cutters</span>  <span class="c1"># alias for unique_cutters</span>

<div class="viewcode-block" id="Dseq.twice_cutters">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.twice_cutters">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">twice_cutters</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_RestrictionBatch</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_RestrictionBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Enzymes in a RestrictionBatch cutting sequence twice.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">batch</span> <span class="o">=</span> <span class="n">CommOnly</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_cutters</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">batch</span><span class="o">=</span><span class="n">batch</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.n_cutters">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.n_cutters">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">n_cutters</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">batch</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_RestrictionBatch</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_RestrictionBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Enzymes in a RestrictionBatch cutting n times.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">batch</span> <span class="o">=</span> <span class="n">CommOnly</span>
        <span class="n">ana</span> <span class="o">=</span> <span class="n">batch</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">ncut</span> <span class="o">=</span> <span class="p">{</span><span class="n">enz</span><span class="p">:</span> <span class="n">sitelist</span> <span class="k">for</span> <span class="p">(</span><span class="n">enz</span><span class="p">,</span> <span class="n">sitelist</span><span class="p">)</span> <span class="ow">in</span> <span class="n">ana</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sitelist</span><span class="p">)</span> <span class="o">==</span> <span class="n">n</span><span class="p">}</span>
        <span class="k">return</span> <span class="n">_RestrictionBatch</span><span class="p">(</span><span class="n">ncut</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.cutters">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.cutters">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cutters</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_RestrictionBatch</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_RestrictionBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Enzymes in a RestrictionBatch cutting sequence at least once.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">batch</span> <span class="o">=</span> <span class="n">CommOnly</span>
        <span class="n">ana</span> <span class="o">=</span> <span class="n">batch</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">ncut</span> <span class="o">=</span> <span class="p">{</span><span class="n">enz</span><span class="p">:</span> <span class="n">sitelist</span> <span class="k">for</span> <span class="p">(</span><span class="n">enz</span><span class="p">,</span> <span class="n">sitelist</span><span class="p">)</span> <span class="ow">in</span> <span class="n">ana</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">sitelist</span><span class="p">}</span>
        <span class="k">return</span> <span class="n">_RestrictionBatch</span><span class="p">(</span><span class="n">ncut</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.seguid">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.seguid">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">seguid</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;SEGUID checksum for the sequence.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">cs</span> <span class="o">=</span> <span class="n">_cdseguid</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="n">alphabet</span><span class="o">=</span><span class="s2">&quot;{DNA-extended}&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
            <span class="n">w</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="s1">&#39;-&#39;</span><span class="si">}{</span><span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="si">}{</span><span class="s1">&#39;-&#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="n">c</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="s1">&#39;-&#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">watson</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="p">))</span><span class="si">}{</span><span class="bp">self</span><span class="o">.</span><span class="n">crick</span><span class="si">}{</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="s1">&#39;-&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="n">cs</span> <span class="o">=</span> <span class="n">_ldseguid</span><span class="p">(</span><span class="n">w</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="s2">&quot;{DNA-extended}&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cs</span></div>


<div class="viewcode-block" id="Dseq.isblunt">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.isblunt">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">isblunt</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;isblunt.</span>

<span class="sd">        Return True if Dseq is linear and blunt and</span>
<span class="sd">        false if staggered or circular.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;gat&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        gat</span>
<span class="sd">        cta</span>
<span class="sd">        &gt;&gt;&gt; a.isblunt()</span>
<span class="sd">        True</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;gat&quot;, &quot;atcg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">         gat</span>
<span class="sd">        gcta</span>
<span class="sd">        &gt;&gt;&gt; a.isblunt()</span>
<span class="sd">        False</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;gat&quot;, &quot;gatc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-4)</span>
<span class="sd">        gat</span>
<span class="sd">        ctag</span>
<span class="sd">        &gt;&gt;&gt; a.isblunt()</span>
<span class="sd">        False</span>
<span class="sd">        &gt;&gt;&gt; a=Dseq(&quot;gat&quot;, circular=True)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(o3)</span>
<span class="sd">        gat</span>
<span class="sd">        cta</span>
<span class="sd">        &gt;&gt;&gt; a.isblunt()</span>
<span class="sd">        False</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>

        <span class="k">return</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">parts</span><span class="o">.</span><span class="n">sticky_right5</span><span class="p">,</span>
                <span class="n">parts</span><span class="o">.</span><span class="n">sticky_right3</span><span class="p">,</span>
                <span class="n">parts</span><span class="o">.</span><span class="n">sticky_left3</span><span class="p">,</span>
                <span class="n">parts</span><span class="o">.</span><span class="n">sticky_left5</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.terminal_transferase">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.terminal_transferase">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">terminal_transferase</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nucleotides</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Terminal deoxynucleotidyl transferase (TdT) is a template-independent</span>
<span class="sd">        DNA polymerase that adds nucleotides to the 3-OH ends of DNA, typically</span>
<span class="sd">        single-stranded or recessed 3 ends. In cloning, its classically used</span>
<span class="sd">        to create homopolymer tails (e.g. poly-dG on a vector and poly-dC on an insert)</span>
<span class="sd">        so that fragments can anneal via complementary overhangs (tailing cloning).</span>

<span class="sd">        This activity ia also present in some DNA polymerases, such as Taq polymerase.</span>
<span class="sd">        This property is used in the populat T/A cloning protocol ([#]_).</span>

<span class="sd">        ::</span>

<span class="sd">            gct          gcta</span>
<span class="sd">            |||   --&gt;    |||</span>
<span class="sd">            cga         acga</span>



<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;aa&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;gct&quot;)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        gct</span>
<span class="sd">        cga</span>
<span class="sd">        &gt;&gt;&gt; a.terminal_transferase()</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">         gcta</span>
<span class="sd">        acga</span>
<span class="sd">        &gt;&gt;&gt; a.terminal_transferase(&quot;G&quot;)</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">         gctG</span>
<span class="sd">        Gcga</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        nucleotides : str, optional</span>
<span class="sd">            The default is &quot;a&quot;.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        DseqType</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>
<span class="sd">        .. [#] https://en.wikipedia.org/wiki/TA_cloning</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ovhg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">ovhg</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">watson</span> <span class="o">+</span> <span class="n">nucleotides</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">crick</span> <span class="o">+</span> <span class="n">nucleotides</span><span class="p">,</span> <span class="n">ovhg</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.user">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.user">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">user</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DseqType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        USER Enzyme treatment.</span>

<span class="sd">        USER Enzyme is a mixture of Uracil DNA glycosylase (UDG) and the</span>
<span class="sd">        DNA glycosylase-lyase Endonuclease VIII.</span>

<span class="sd">        UDG catalyses the excision of an uracil base, forming an abasic</span>
<span class="sd">        or apyrimidinic site (AP site). Endonuclease VIII removes the AP</span>
<span class="sd">        site creating a DNA gap.</span>

<span class="sd">        ::</span>

<span class="sd">            tagaagtaggUat          tagaagtagg at</span>
<span class="sd">            |||||||||||||  ---&gt;    |||||||||| ||</span>
<span class="sd">            atcUtcatccata          atc tcatccata</span>



<span class="sd">        &gt;&gt;&gt; a = Dseq(&quot;tagaagtaggUat&quot;, &quot;atcUtcatccata&quot;[::-1], 0)</span>
<span class="sd">        &gt;&gt;&gt; a</span>
<span class="sd">        Dseq(-13)</span>
<span class="sd">        tagaagtaggUat</span>
<span class="sd">        atcutcatccAta</span>
<span class="sd">        &gt;&gt;&gt; a.user()</span>
<span class="sd">        Dseq(-13)</span>
<span class="sd">        tagaagtagg at</span>
<span class="sd">        atc tcatccAta</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        DseqType</span>
<span class="sd">            DNA fragment with uracile bases removed.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="nb">bytes</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="sa">b</span><span class="s2">&quot;UuOo&quot;</span><span class="p">,</span> <span class="sa">b</span><span class="s2">&quot;ZzEe&quot;</span><span class="p">)))</span></div>


<div class="viewcode-block" id="Dseq.cut">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.cut">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cut</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="o">*</span><span class="n">enzymes</span><span class="p">:</span> <span class="n">EnzymesType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Tuple</span><span class="p">[</span><span class="n">DseqType</span><span class="p">,</span> <span class="o">...</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a list of linear Dseq fragments produced in the digestion.</span>
<span class="sd">        If there are no cuts, an empty list is returned.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        enzymes : enzyme object or iterable of such objects</span>
<span class="sd">            A Bio.Restriction.XXX restriction objects or iterable.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        frags : list</span>
<span class="sd">            list of Dseq objects formed by the digestion</span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; seq=Dseq(&quot;ggatccnnngaattc&quot;)</span>
<span class="sd">        &gt;&gt;&gt; seq</span>
<span class="sd">        Dseq(-15)</span>
<span class="sd">        ggatccnnngaattc</span>
<span class="sd">        cctaggnnncttaag</span>
<span class="sd">        &gt;&gt;&gt; from Bio.Restriction import BamHI,EcoRI</span>
<span class="sd">        &gt;&gt;&gt; type(seq.cut(BamHI))</span>
<span class="sd">        &lt;class &#39;tuple&#39;&gt;</span>
<span class="sd">        &gt;&gt;&gt; for frag in seq.cut(BamHI): print(repr(frag))</span>
<span class="sd">        Dseq(-5)</span>
<span class="sd">        g</span>
<span class="sd">        cctag</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        gatccnnngaattc</span>
<span class="sd">            gnnncttaag</span>
<span class="sd">        &gt;&gt;&gt; seq.cut(EcoRI, BamHI) ==  seq.cut(BamHI, EcoRI)</span>
<span class="sd">        True</span>
<span class="sd">        &gt;&gt;&gt; a,b,c = seq.cut(EcoRI, BamHI)</span>
<span class="sd">        &gt;&gt;&gt; a+b+c</span>
<span class="sd">        Dseq(-15)</span>
<span class="sd">        ggatccnnngaattc</span>
<span class="sd">        cctaggnnncttaag</span>
<span class="sd">        &gt;&gt;&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">cutsites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cutsites</span><span class="p">(</span><span class="o">*</span><span class="n">enzymes</span><span class="p">)</span>
        <span class="n">cutsite_pairs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">cs</span><span class="p">)</span> <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">cutsite_pairs</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.cutsite_is_valid">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.cutsite_is_valid">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cutsite_is_valid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cutsite</span><span class="p">:</span> <span class="n">CutSiteType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check is a cutsite is valid.</span>

<span class="sd">        A cutsite is a nested 2-tuple with this form:</span>

<span class="sd">        ((cut_watson, ovhg), enz), for example ((396, -4), EcoRI)</span>

<span class="sd">        The cut_watson (positive integer) is the cut position of the sequence as for example</span>
<span class="sd">        returned by the Bio.Restriction module.</span>

<span class="sd">        The ovhg (overhang, positive or negative integer or 0) has the same meaning as</span>
<span class="sd">        for restriction enzymes in the Bio.Restriction module and for</span>
<span class="sd">        pydna.dseq.Dseq objects (see docstring for this module and example below)</span>

<span class="sd">        Enzyme can be None.</span>

<span class="sd">        ::</span>

<span class="sd">            Enzyme overhang</span>

<span class="sd">            EcoRI  -4     --GAATTC--        --G       AATTC--</span>
<span class="sd">                            ||||||     --&gt;    |           |</span>
<span class="sd">                          --CTTAAG--        --CTTAA       G--</span>

<span class="sd">            KpnI    4     --GGTACC--        --GGTAC       C--</span>
<span class="sd">                            ||||||     --&gt;    |           |</span>
<span class="sd">                          --CCATGG--        --C       CATGG--</span>

<span class="sd">            SmaI    0     --CCCGGG--        --CCC       GGG--</span>
<span class="sd">                            ||||||     --&gt;    |||       |||</span>
<span class="sd">                          --GGGCCC--        --GGG       CCC--</span>


<span class="sd">        &gt;&gt;&gt; from Bio.Restriction import EcoRI, KpnI, SmaI</span>
<span class="sd">        &gt;&gt;&gt; EcoRI.ovhg</span>
<span class="sd">        -4</span>
<span class="sd">        &gt;&gt;&gt; KpnI.ovhg</span>
<span class="sd">        4</span>
<span class="sd">        &gt;&gt;&gt; SmaI.ovhg</span>
<span class="sd">        0</span>

<span class="sd">        Returns False if:</span>

<span class="sd">        - Cut positions fall outside the sequence (could be moved to Biopython)</span>
<span class="sd">        TODO: example</span>

<span class="sd">        - Overhang is not double stranded</span>
<span class="sd">        TODO: example</span>

<span class="sd">        - Recognition site is not double stranded or is outside the sequence</span>
<span class="sd">        TODO: example</span>

<span class="sd">        - For enzymes that cut twice, it checks that at least one possibility is valid</span>
<span class="sd">        TODO: example</span>



<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cutsite : CutSiteType</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        bool</span>
<span class="sd">            True if cutsite can cut the DNA fragment.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="n">cutsite</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;cutsite is None&quot;</span>

        <span class="n">enz</span> <span class="o">=</span> <span class="n">cutsite</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">watson</span><span class="p">,</span> <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cut_parameters</span><span class="p">(</span><span class="n">cutsite</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

        <span class="c1"># The overhang is double stranded</span>
        <span class="n">overhang_dseq</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="n">watson</span><span class="p">:</span><span class="n">crick</span><span class="p">]</span> <span class="k">if</span> <span class="n">ovhg</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="bp">self</span><span class="p">[</span><span class="n">crick</span><span class="p">:</span><span class="n">watson</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">overhang_dseq</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">!=</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">overhang_dseq</span><span class="o">.</span><span class="n">watson_ovhg</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># The recognition site is double stranded and within the sequence</span>
        <span class="n">start_of_recognition_site</span> <span class="o">=</span> <span class="n">watson</span> <span class="o">-</span> <span class="n">enz</span><span class="o">.</span><span class="n">fst5</span>
        <span class="k">if</span> <span class="n">start_of_recognition_site</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">start_of_recognition_site</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">end_of_recognition_site</span> <span class="o">=</span> <span class="n">start_of_recognition_site</span> <span class="o">+</span> <span class="n">enz</span><span class="o">.</span><span class="n">size</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">end_of_recognition_site</span> <span class="o">%=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">recognition_site</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="n">start_of_recognition_site</span><span class="p">:</span><span class="n">end_of_recognition_site</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">recognition_site</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
            <span class="ow">or</span> <span class="n">recognition_site</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">!=</span> <span class="mi">0</span>
            <span class="ow">or</span> <span class="n">recognition_site</span><span class="o">.</span><span class="n">watson_ovhg</span> <span class="o">!=</span> <span class="mi">0</span>
        <span class="p">):</span>
            <span class="k">if</span> <span class="n">enz</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">enz</span><span class="o">.</span><span class="n">scd5</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># For enzymes that cut twice, this might be referring to the second one</span>
                <span class="n">start_of_recognition_site</span> <span class="o">=</span> <span class="n">watson</span> <span class="o">-</span> <span class="n">enz</span><span class="o">.</span><span class="n">scd5</span>
                <span class="k">if</span> <span class="n">start_of_recognition_site</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">start_of_recognition_site</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
                <span class="n">end_of_recognition_site</span> <span class="o">=</span> <span class="n">start_of_recognition_site</span> <span class="o">+</span> <span class="n">enz</span><span class="o">.</span><span class="n">size</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
                    <span class="n">end_of_recognition_site</span> <span class="o">%=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
                <span class="n">recognition_site</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span>
                    <span class="n">start_of_recognition_site</span><span class="p">:</span><span class="n">end_of_recognition_site</span>
                <span class="p">]</span>

                <span class="k">if</span> <span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="n">recognition_site</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
                    <span class="ow">or</span> <span class="n">recognition_site</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">!=</span> <span class="mi">0</span>
                    <span class="ow">or</span> <span class="n">recognition_site</span><span class="o">.</span><span class="n">watson_ovhg</span> <span class="o">!=</span> <span class="mi">0</span>
                <span class="p">):</span>
                    <span class="k">return</span> <span class="kc">False</span>

        <span class="k">return</span> <span class="kc">True</span></div>


<div class="viewcode-block" id="Dseq.get_cutsites">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.get_cutsites">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_cutsites</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="o">*</span><span class="n">enzymes</span><span class="p">:</span> <span class="n">EnzymesType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_List</span><span class="p">[</span><span class="n">CutSiteType</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a list of cutsites, represented represented as `((cut_watson, ovhg), enz)`:</span>

<span class="sd">        - `cut_watson` is a positive integer contained in `[0,len(seq))`, where `seq` is the sequence</span>
<span class="sd">          that will be cut. It represents the position of the cut on the watson strand, using the full</span>
<span class="sd">          sequence as a reference. By &quot;full sequence&quot; I mean the one you would get from `str(Dseq)`.</span>

<span class="sd">        - `ovhg` is the overhang left after the cut. It has the same meaning as `ovhg` in</span>
<span class="sd">          the `Bio.Restriction` enzyme objects, or pydna&#39;s `Dseq` property.</span>

<span class="sd">        - `enz` is the enzyme object. It&#39;s not necessary to perform the cut, but can be</span>
<span class="sd">           used to keep track of which enzyme was used.</span>

<span class="sd">        Cuts are only returned if the recognition site and overhang are on the double-strand</span>
<span class="sd">        part of the sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        enzymes : Union[_RestrictionBatch,list[_AbstractCut]]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list[tuple[tuple[int,int], _AbstractCut]]</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; from Bio.Restriction import EcoRI</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; seq = Dseq(&#39;AAGAATTCAAGAATTC&#39;)</span>
<span class="sd">        &gt;&gt;&gt; seq.get_cutsites(EcoRI)</span>
<span class="sd">        [((3, -4), EcoRI), ((11, -4), EcoRI)]</span>

<span class="sd">        `cut_watson` is defined with respect to the &quot;full sequence&quot;, not the</span>
<span class="sd">        watson strand:</span>

<span class="sd">        &gt;&gt;&gt; dseq = Dseq.from_full_sequence_and_overhangs(&#39;aaGAATTCaa&#39;, 1, 0)</span>
<span class="sd">        &gt;&gt;&gt; dseq</span>
<span class="sd">        Dseq(-10)</span>
<span class="sd">         aGAATTCaa</span>
<span class="sd">        ttCTTAAGtt</span>
<span class="sd">        &gt;&gt;&gt; dseq.get_cutsites([EcoRI])</span>
<span class="sd">        [((3, -4), EcoRI)]</span>

<span class="sd">        Cuts are only returned if the recognition site and overhang are on the double-strand</span>
<span class="sd">        part of the sequence.</span>

<span class="sd">        &gt;&gt;&gt; Dseq(&#39;GAATTC&#39;).get_cutsites([EcoRI])</span>
<span class="sd">        [((1, -4), EcoRI)]</span>
<span class="sd">        &gt;&gt;&gt; Dseq.from_full_sequence_and_overhangs(&#39;GAATTC&#39;, -1, 0).get_cutsites([EcoRI])</span>
<span class="sd">        []</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">enzymes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">enzymes</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">_RestrictionBatch</span><span class="p">):</span>
            <span class="c1"># argument is probably a RestrictionBatch</span>
            <span class="n">enzymes</span> <span class="o">=</span> <span class="p">[</span><span class="n">e</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">enzymes</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

        <span class="n">enzymes</span> <span class="o">=</span> <span class="n">_flatten</span><span class="p">(</span><span class="n">enzymes</span><span class="p">)</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">enzymes</span><span class="p">:</span>
            <span class="c1"># Positions of the cut on the watson strand. They are 1-based, so we subtract</span>
            <span class="c1"># 1 to get 0-based positions</span>
            <span class="n">cuts_watson</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="o">-</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">e</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="p">(</span><span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">))]</span>

            <span class="n">out</span> <span class="o">+=</span> <span class="p">[((</span><span class="n">w</span><span class="p">,</span> <span class="n">e</span><span class="o">.</span><span class="n">ovhg</span><span class="p">),</span> <span class="n">e</span><span class="p">)</span> <span class="k">for</span> <span class="n">w</span> <span class="ow">in</span> <span class="n">cuts_watson</span><span class="p">]</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">cutsite</span> <span class="k">for</span> <span class="n">cutsite</span> <span class="ow">in</span> <span class="n">out</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">cutsite_is_valid</span><span class="p">(</span><span class="n">cutsite</span><span class="p">)])</span></div>


<div class="viewcode-block" id="Dseq.left_end_position">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.left_end_position">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">left_end_position</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The index in the full sequence of the watson and crick start positions.</span>

<span class="sd">        full sequence (str(self)) for all three cases is AAA</span>
<span class="sd">        ::</span>

<span class="sd">            AAA              AA               AAT</span>
<span class="sd">             TT             TTT               TTT</span>
<span class="sd">            Returns (0, 1)  Returns (1, 0)    Returns (0, 0)</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span><span class="p">,</span> <span class="mi">0</span>
        <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span></div>


<div class="viewcode-block" id="Dseq.right_end_position">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.right_end_position">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">right_end_position</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;The index in the full sequence of the watson and crick end positions.</span>

<span class="sd">        full sequence (str(self)) for all three cases is AAA</span>

<span class="sd">        ```</span>
<span class="sd">        AAA               AA                   AAA</span>
<span class="sd">        TT                TTT                  TTT</span>
<span class="sd">        Returns (3, 2)    Returns (2, 3)       Returns (3, 3)</span>
<span class="sd">        ```</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson_ovhg</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson_ovhg</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson_ovhg</span></div>


<div class="viewcode-block" id="Dseq.get_ss_meltsites">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.get_ss_meltsites">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_ss_meltsites</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Single stranded DNA melt sites</span>

<span class="sd">        Two lists of 2-tuples of integers are returned. Each tuple</span>
<span class="sd">        (`((from, to))`) contains the start and end positions of a single</span>
<span class="sd">        stranded region, shorter or equal to `length`.</span>

<span class="sd">        In the example below, the middle 2 nt part is released from the</span>
<span class="sd">        molecule.</span>

<span class="sd">        ::</span>


<span class="sd">            tagaa ta gtatg</span>
<span class="sd">            ||||| || |||||  --&gt;   [(6,8)], []</span>
<span class="sd">            atcttcatccatac</span>

<span class="sd">            tagaagtaggtatg</span>
<span class="sd">            ||||| || |||||  --&gt;   [], [(6,8)]</span>
<span class="sd">            atctt at catac</span>




<span class="sd">        The output of this method is used in the `melt_ss_dna` method in order</span>
<span class="sd">        to determine the start and end positions of single stranded regions.</span>

<span class="sd">        See get_ds_meltsites for melting ds sequences.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaaqtaqgtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaa ta gtatg</span>
<span class="sd">        atcttcatccatac</span>
<span class="sd">        &gt;&gt;&gt; cutsites = ds.get_ss_meltsites(2)</span>
<span class="sd">        &gt;&gt;&gt; cutsites</span>
<span class="sd">        ([(6, 8)], [])</span>
<span class="sd">        &gt;&gt;&gt; ds[6:8]</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        ta</span>
<span class="sd">        at</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaaptapgtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaagtaggtatg</span>
<span class="sd">        atctt at catac</span>
<span class="sd">        &gt;&gt;&gt; cutsites = ds.get_ss_meltsites(2)</span>
<span class="sd">        &gt;&gt;&gt; cutsites</span>
<span class="sd">        ([], [(6, 8)])</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">regex</span> <span class="o">=</span> <span class="n">regex_ss_melt_factory</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">spacer</span> <span class="o">=</span> <span class="n">length</span>
            <span class="n">cutfrom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="o">-</span><span class="n">length</span><span class="p">:]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">length</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">spacer</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">cutfrom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span>

        <span class="n">watson_cuts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">crick_cuts</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">regex</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">cutfrom</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">lastgroup</span> <span class="o">==</span> <span class="s2">&quot;watson&quot;</span><span class="p">:</span>
                <span class="n">cut1</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">+</span> <span class="n">spacer</span>
                <span class="n">cut2</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="o">+</span> <span class="n">spacer</span>
                <span class="n">watson_cuts</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">cut1</span><span class="p">,</span> <span class="n">cut2</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">m</span><span class="o">.</span><span class="n">lastgroup</span> <span class="o">==</span> <span class="s2">&quot;crick&quot;</span>
                <span class="n">cut1</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">+</span> <span class="n">spacer</span>
                <span class="n">cut2</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="o">+</span> <span class="n">spacer</span>
                <span class="n">crick_cuts</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">cut1</span><span class="p">,</span> <span class="n">cut2</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">watson_cuts</span><span class="p">,</span> <span class="n">crick_cuts</span></div>


<div class="viewcode-block" id="Dseq.get_ds_meltsites">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.get_ds_meltsites">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_ds_meltsites</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">DseqType</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_List</span><span class="p">[</span><span class="n">CutSiteType</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Double stranded DNA melt sites</span>

<span class="sd">        DNA molecules can fall apart by melting if they have internal single</span>
<span class="sd">        stranded regions. In the example below, the molecule has two gaps</span>
<span class="sd">        on opposite sides, two nucleotides apart, which means that it hangs</span>
<span class="sd">        together by two basepairs.</span>

<span class="sd">        This molecule can melt into two separate 8 bp double stranded</span>
<span class="sd">        molecules, each with 3 nt 3&#39; overhangs a depicted below.</span>

<span class="sd">        ::</span>

<span class="sd">            tagaagta gtatg        tagaagta          gtatg</span>
<span class="sd">            ||||| || |||||  --&gt;   |||||             |||||</span>
<span class="sd">            atctt atccatac        atctt          atccatac</span>


<span class="sd">        A list of 2-tuples is returned. Each tuple (`((cut_watson, ovhg), None)`)</span>
<span class="sd">        contains cut position and the overhang value in the same format as</span>
<span class="sd">        returned by the get_cutsites method for restriction enzymes.</span>

<span class="sd">        Note that this function deals with melting that results in two double</span>
<span class="sd">        stranded DNA molecules.</span>

<span class="sd">        See get_ss_meltsites for melting of single stranded regions from</span>
<span class="sd">        molecules.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaaptaqgtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaagta gtatg</span>
<span class="sd">        atctt atccatac</span>
<span class="sd">        &gt;&gt;&gt; cutsite = ds.get_ds_meltsites(2)</span>
<span class="sd">        &gt;&gt;&gt; cutsite</span>
<span class="sd">        [((8, 2), None)]</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">length</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">tuple</span><span class="p">()</span>

        <span class="n">regex</span> <span class="o">=</span> <span class="n">regex_ds_melt_factory</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">spacer</span> <span class="o">=</span> <span class="n">length</span>
            <span class="n">cutfrom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="o">-</span><span class="n">length</span><span class="p">:]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">length</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">spacer</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">cutfrom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span>

        <span class="n">cuts</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">regex</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">cutfrom</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">lastgroup</span> <span class="o">==</span> <span class="s2">&quot;watson&quot;</span><span class="p">:</span>
                <span class="n">cut</span> <span class="o">=</span> <span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="o">-</span> <span class="n">spacer</span><span class="p">,</span> <span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="o">-</span> <span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()),</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">m</span><span class="o">.</span><span class="n">lastgroup</span> <span class="o">==</span> <span class="s2">&quot;crick&quot;</span>
                <span class="n">cut</span> <span class="o">=</span> <span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">-</span> <span class="n">spacer</span><span class="p">,</span> <span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">-</span> <span class="n">m</span><span class="o">.</span><span class="n">end</span><span class="p">()),</span> <span class="kc">None</span>

            <span class="n">cuts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cut</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">cuts</span></div>


<div class="viewcode-block" id="Dseq.cast_to_ds_right">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.cast_to_ds_right">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cast_to_ds_right</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        NNNN               NNNNGATC</span>
<span class="sd">        ||||       --&gt;     ||||||||</span>
<span class="sd">        NNNNCTAG           NNNNCTAG</span>


<span class="sd">        NNNNGATC           NNNNGATC</span>
<span class="sd">        ||||       --&gt;     ||||||||</span>
<span class="sd">        NNNN               NNNNCTAG</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">p</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>

        <span class="n">ds_stuffer</span> <span class="o">=</span> <span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">sticky_right5</span> <span class="ow">or</span> <span class="n">p</span><span class="o">.</span><span class="n">sticky_right3</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
            <span class="n">dscode_to_full_sequence_table</span>
        <span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">sticky_left5</span> <span class="ow">or</span> <span class="n">p</span><span class="o">.</span><span class="n">sticky_left3</span><span class="p">)</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">middle</span> <span class="o">+</span> <span class="n">ds_stuffer</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.cast_to_ds_left">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.cast_to_ds_left">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cast_to_ds_left</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        GATCNNNN           GATCNNNN</span>
<span class="sd">            ||||   --&gt;     ||||||||</span>
<span class="sd">            NNNN           CTAGNNNN</span>

<span class="sd">            NNNN           GATCNNNN</span>
<span class="sd">            ||||   --&gt;     ||||||||</span>
<span class="sd">        CTAGNNNN           CTAGNNNN</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">p</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_parts</span><span class="p">()</span>

        <span class="n">ds_stuffer</span> <span class="o">=</span> <span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">sticky_left5</span> <span class="ow">or</span> <span class="n">p</span><span class="o">.</span><span class="n">sticky_left3</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
            <span class="n">dscode_to_full_sequence_table</span>
        <span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">ds_stuffer</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">middle</span> <span class="o">+</span> <span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">sticky_right5</span> <span class="ow">or</span> <span class="n">p</span><span class="o">.</span><span class="n">sticky_right3</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.get_cut_parameters">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.get_cut_parameters">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_cut_parameters</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">cut</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">CutSiteType</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span> <span class="n">is_left</span><span class="p">:</span> <span class="nb">bool</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;For a given cut expressed as ((cut_watson, ovhg), enz), returns</span>
<span class="sd">        a tuple (cut_watson, cut_crick, ovhg).</span>

<span class="sd">        - cut_watson: see get_cutsites docs</span>
<span class="sd">        - cut_crick: equivalent of cut_watson in the crick strand</span>
<span class="sd">        - ovhg: see get_cutsites docs</span>

<span class="sd">        The cut can be None if it represents the left or right end of the sequence.</span>
<span class="sd">        Then it will return the position of the watson and crick ends with respect</span>
<span class="sd">        to the &quot;full sequence&quot;. The `is_left` parameter is only used in this case.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">cut</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">watson</span><span class="p">,</span> <span class="n">ovhg</span> <span class="o">=</span> <span class="n">cut</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">crick</span> <span class="o">=</span> <span class="n">watson</span> <span class="o">-</span> <span class="n">ovhg</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
                <span class="n">crick</span> <span class="o">%=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">watson</span><span class="p">,</span> <span class="n">crick</span><span class="p">,</span> <span class="n">ovhg</span>

        <span class="k">assert</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">,</span> <span class="s2">&quot;Circular sequences should not have None cuts&quot;</span>

        <span class="k">if</span> <span class="n">is_left</span><span class="p">:</span>
            <span class="k">return</span> <span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">left_end_position</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">ovhg</span>
        <span class="c1"># In the right end, the overhang does not matter</span>
        <span class="k">return</span> <span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">right_end_position</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">watson_ovhg</span></div>


<div class="viewcode-block" id="Dseq.melt">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.melt">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">melt</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        TBD</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        length : TYPE</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        TYPE</span>
<span class="sd">            DESCRIPTION.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">length</span> <span class="ow">or</span> <span class="n">length</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">tuple</span><span class="p">()</span>

        <span class="c1"># First we need to get rid of single stranded sequences</span>
        <span class="n">new</span><span class="p">,</span> <span class="n">strands</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">melt_ss_dna</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>

        <span class="n">cutsites</span> <span class="o">=</span> <span class="n">new</span><span class="o">.</span><span class="n">get_ds_meltsites</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>

        <span class="n">cutsite_pairs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cutsite_pairs</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">new</span><span class="o">.</span><span class="n">apply_cut</span><span class="p">(</span><span class="o">*</span><span class="n">cutsite_pair</span><span class="p">)</span> <span class="k">for</span> <span class="n">cutsite_pair</span> <span class="ow">in</span> <span class="n">cutsite_pairs</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">([</span><span class="n">new</span><span class="p">])</span> <span class="k">if</span> <span class="n">strands</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">result</span> <span class="k">else</span> <span class="n">result</span>

        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">strands</span><span class="p">)</span> <span class="o">+</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.melt_ss_dna">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.melt_ss_dna">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">melt_ss_dna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="s2">&quot;Dseq&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="s2">&quot;Dseq&quot;</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Melt to separate single stranded DNA</span>

<span class="sd">        Single stranded DNA molecules shorter or equal to `length` shed from</span>
<span class="sd">        a double stranded DNA molecule without affecting the length of the</span>
<span class="sd">        remaining molecule.</span>

<span class="sd">        In the examples below, the middle 2 nt part is released from the</span>
<span class="sd">        molecule.</span>

<span class="sd">        ::</span>

<span class="sd">            tagaa ta gtatg        tagaa    gtatg          ta</span>
<span class="sd">            ||||| || |||||  --&gt;   |||||    |||||     +    ||</span>
<span class="sd">            atcttcatccatac        atcttcatccatac</span>

<span class="sd">            tagaagtaggtatg        tagaagtaggtatg</span>
<span class="sd">            ||||| || |||||  --&gt;   |||||    |||||     +    ||</span>
<span class="sd">            atctt at catac        atctt    catac          at</span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaaqtaqgtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaa ta gtatg</span>
<span class="sd">        atcttcatccatac</span>
<span class="sd">        &gt;&gt;&gt; new, strands  = ds.melt_ss_dna(2)</span>
<span class="sd">        &gt;&gt;&gt; new</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaa    gtatg</span>
<span class="sd">        atcttcatccatac</span>
<span class="sd">        &gt;&gt;&gt; strands[0]</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        ta</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaaptapgtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaagtaggtatg</span>
<span class="sd">        atctt at catac</span>
<span class="sd">        &gt;&gt;&gt; new, strands = ds.melt_ss_dna(2)</span>
<span class="sd">        &gt;&gt;&gt; new</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaagtaggtatg</span>
<span class="sd">        atctt    catac</span>
<span class="sd">        &gt;&gt;&gt; strands[0]</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        at</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">watsonnicks</span><span class="p">,</span> <span class="n">cricknicks</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_ss_meltsites</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>

        <span class="n">new</span><span class="p">,</span> <span class="n">strands</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shed_ss_dna</span><span class="p">(</span><span class="n">watsonnicks</span><span class="p">,</span> <span class="n">cricknicks</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new</span><span class="p">,</span> <span class="n">strands</span></div>


<div class="viewcode-block" id="Dseq.shed_ss_dna">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.shed_ss_dna">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">shed_ss_dna</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">watson_cutpairs</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">crick_cutpairs</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Separate parts of one of the DNA strands</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaagtaggtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaagtaggtatg</span>
<span class="sd">        atcttcatccatac</span>
<span class="sd">        &gt;&gt;&gt; new, strands = ds.shed_ss_dna([(6, 8)],[])</span>
<span class="sd">        &gt;&gt;&gt; new</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaag  ggtatg</span>
<span class="sd">        atcttcatccatac</span>
<span class="sd">        &gt;&gt;&gt; strands[0]</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        ta</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; new, strands = ds.shed_ss_dna([],[(6, 8)])</span>
<span class="sd">        &gt;&gt;&gt; new</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaagtaggtatg</span>
<span class="sd">        atcttc  ccatac</span>
<span class="sd">        &gt;&gt;&gt; strands[0]</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        at</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;tagaagtaggtatg&quot;)</span>
<span class="sd">        &gt;&gt;&gt; new, (strand1, strand2) = ds.shed_ss_dna([(6, 8), (9, 11)],[])</span>
<span class="sd">        &gt;&gt;&gt; new</span>
<span class="sd">        Dseq(-14)</span>
<span class="sd">        tagaag  g  atg</span>
<span class="sd">        atcttcatccatac</span>
<span class="sd">        &gt;&gt;&gt; strand1</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        ta</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; strand2</span>
<span class="sd">        Dseq(-2)</span>
<span class="sd">        gt</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">watson_cutpairs</span> <span class="o">=</span> <span class="n">watson_cutpairs</span> <span class="ow">or</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">crick_cutpairs</span> <span class="o">=</span> <span class="n">crick_cutpairs</span> <span class="ow">or</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">strands</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">new</span> <span class="o">=</span> <span class="nb">bytearray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">watson_cutpairs</span><span class="p">:</span>
            <span class="n">stuffer</span> <span class="o">=</span> <span class="n">new</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">y</span><span class="p">]</span>
            <span class="n">ss</span> <span class="o">=</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">quick</span><span class="p">(</span><span class="n">new</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">y</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_tail_table</span><span class="p">))</span>
            <span class="n">new</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">y</span><span class="p">]</span> <span class="o">=</span> <span class="n">stuffer</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_tail_table</span><span class="p">)</span>
            <span class="n">strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ss</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">crick_cutpairs</span><span class="p">:</span>
            <span class="n">stuffer</span> <span class="o">=</span> <span class="n">new</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">y</span><span class="p">]</span>
            <span class="n">ss</span> <span class="o">=</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">quick</span><span class="p">(</span><span class="n">stuffer</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_crick_tail_table</span><span class="p">))</span>
            <span class="n">new</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">y</span><span class="p">]</span> <span class="o">=</span> <span class="n">stuffer</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_tail_table</span><span class="p">)</span>
            <span class="n">strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ss</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">Dseq</span><span class="o">.</span><span class="n">quick</span><span class="p">(</span><span class="n">new</span><span class="p">),</span> <span class="n">strands</span></div>


<div class="viewcode-block" id="Dseq.apply_cut">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.apply_cut">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">apply_cut</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">left_cut</span><span class="p">:</span> <span class="n">CutSiteType</span><span class="p">,</span> <span class="n">right_cut</span><span class="p">:</span> <span class="n">CutSiteType</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;Dseq&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extracts a subfragment of the sequence between two cuts.</span>

<span class="sd">        For more detail see the documentation of get_cutsite_pairs.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        left_cut : Union[tuple[tuple[int,int], _AbstractCut], None]</span>
<span class="sd">        right_cut: Union[tuple[tuple[int,int], _AbstractCut], None]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Dseq</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from Bio.Restriction import EcoRI</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; dseq = Dseq(&#39;aaGAATTCaaGAATTCaa&#39;)</span>
<span class="sd">        &gt;&gt;&gt; cutsites = dseq.get_cutsites([EcoRI])</span>
<span class="sd">        &gt;&gt;&gt; cutsites</span>
<span class="sd">        [((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="sd">        &gt;&gt;&gt; p1, p2, p3 = dseq.get_cutsite_pairs(cutsites)</span>
<span class="sd">        &gt;&gt;&gt; p1</span>
<span class="sd">        (None, ((3, -4), EcoRI))</span>
<span class="sd">        &gt;&gt;&gt; dseq.apply_cut(*p1)</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        aaG</span>
<span class="sd">        ttCTTAA</span>
<span class="sd">        &gt;&gt;&gt; p2</span>
<span class="sd">        (((3, -4), EcoRI), ((11, -4), EcoRI))</span>
<span class="sd">        &gt;&gt;&gt; dseq.apply_cut(*p2)</span>
<span class="sd">        Dseq(-12)</span>
<span class="sd">        AATTCaaG</span>
<span class="sd">            GttCTTAA</span>
<span class="sd">        &gt;&gt;&gt; p3</span>
<span class="sd">        (((11, -4), EcoRI), None)</span>
<span class="sd">        &gt;&gt;&gt; dseq.apply_cut(*p3)</span>
<span class="sd">        Dseq(-7)</span>
<span class="sd">        AATTCaa</span>
<span class="sd">            Gtt</span>

<span class="sd">        &gt;&gt;&gt; dseq = Dseq(&#39;TTCaaGAA&#39;, circular=True)</span>
<span class="sd">        &gt;&gt;&gt; cutsites = dseq.get_cutsites([EcoRI])</span>
<span class="sd">        &gt;&gt;&gt; cutsites</span>
<span class="sd">        [((6, -4), EcoRI)]</span>
<span class="sd">        &gt;&gt;&gt; pair = dseq.get_cutsite_pairs(cutsites)[0]</span>
<span class="sd">        &gt;&gt;&gt; pair</span>
<span class="sd">        (((6, -4), EcoRI), ((6, -4), EcoRI))</span>
<span class="sd">        &gt;&gt;&gt; dseq.apply_cut(*pair)</span>
<span class="sd">        Dseq(-12)</span>
<span class="sd">        AATTCaaG</span>
<span class="sd">            GttCTTAA</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">_cuts_overlap</span><span class="p">(</span><span class="n">left_cut</span><span class="p">,</span> <span class="n">right_cut</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cuts by </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> overlap.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">left_cut</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">right_cut</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

        <span class="n">left_watson</span><span class="p">,</span> <span class="n">left_crick</span><span class="p">,</span> <span class="n">ovhg_left</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cut_parameters</span><span class="p">(</span><span class="n">left_cut</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">right_watson</span><span class="p">,</span> <span class="n">right_crick</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cut_parameters</span><span class="p">(</span><span class="n">right_cut</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">Dseq</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">[</span><span class="n">left_watson</span><span class="p">:</span><span class="n">right_watson</span><span class="p">]</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">),</span>
            <span class="bp">self</span><span class="p">[</span><span class="n">left_crick</span><span class="p">:</span><span class="n">right_crick</span><span class="p">]</span>
            <span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
            <span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">dscode_to_watson_table</span><span class="p">),</span>
            <span class="n">ovhg</span><span class="o">=</span><span class="n">ovhg_left</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="Dseq.get_cutsite_pairs">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.get_cutsite_pairs">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_cutsite_pairs</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">cutsites</span><span class="p">:</span> <span class="n">_List</span><span class="p">[</span><span class="n">CutSiteType</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_List</span><span class="p">[</span><span class="n">_Tuple</span><span class="p">[</span><span class="n">_Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">CutSiteType</span><span class="p">],</span> <span class="n">_Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">CutSiteType</span><span class="p">]]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns pairs of cutsites that render the edges of the resulting fragments.</span>

<span class="sd">        A fragment produced by restriction is represented by a tuple of length 2 that</span>
<span class="sd">        may contain cutsites or `None`:</span>

<span class="sd">            - Two cutsites: represents the extraction of a fragment between those two</span>
<span class="sd">              cutsites, in that orientation. To represent the opening of a circular</span>
<span class="sd">              molecule with a single cutsite, we put the same cutsite twice.</span>
<span class="sd">            - `None`, cutsite: represents the extraction of a fragment between the left</span>
<span class="sd">              edge of linear sequence and the cutsite.</span>
<span class="sd">            - cutsite, `None`: represents the extraction of a fragment between the cutsite</span>
<span class="sd">              and the right edge of a linear sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cutsites : list[tuple[tuple[int,int], _AbstractCut]]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list[tuple[tuple[tuple[int,int], _AbstractCut]|None],tuple[tuple[int,int], _AbstractCut]|None]</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; from Bio.Restriction import EcoRI</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; dseq = Dseq(&#39;aaGAATTCaaGAATTCaa&#39;)</span>
<span class="sd">        &gt;&gt;&gt; cutsites = dseq.get_cutsites([EcoRI])</span>
<span class="sd">        &gt;&gt;&gt; cutsites</span>
<span class="sd">        [((3, -4), EcoRI), ((11, -4), EcoRI)]</span>
<span class="sd">        &gt;&gt;&gt; dseq.get_cutsite_pairs(cutsites)</span>
<span class="sd">        [(None, ((3, -4), EcoRI)), (((3, -4), EcoRI), ((11, -4), EcoRI)), (((11, -4), EcoRI), None)]</span>

<span class="sd">        &gt;&gt;&gt; dseq = Dseq(&#39;TTCaaGAA&#39;, circular=True)</span>
<span class="sd">        &gt;&gt;&gt; cutsites = dseq.get_cutsites([EcoRI])</span>
<span class="sd">        &gt;&gt;&gt; cutsites</span>
<span class="sd">        [((6, -4), EcoRI)]</span>
<span class="sd">        &gt;&gt;&gt; dseq.get_cutsite_pairs(cutsites)</span>
<span class="sd">        [(((6, -4), EcoRI), ((6, -4), EcoRI))]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cutsites</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="n">cutsites</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="o">*</span><span class="n">cutsites</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Add the first cutsite at the end, for circular cuts</span>
            <span class="n">cutsites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cutsites</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">cutsites</span><span class="p">,</span> <span class="n">cutsites</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span></div>


<div class="viewcode-block" id="Dseq.get_parts">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Dseq.get_parts">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_parts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a DseqParts instance containing the parts (strings) of a dsDNA</span>
<span class="sd">        sequence. DseqParts instance field names:</span>

<span class="sd">        ::</span>

<span class="sd">             &quot;sticky_left5&quot;</span>
<span class="sd">             |</span>
<span class="sd">             |      &quot;sticky_right5&quot;</span>
<span class="sd">             |      |</span>
<span class="sd">            ---    ---</span>
<span class="sd">            GGGATCC</span>
<span class="sd">               TAGGTCA</span>
<span class="sd">               ----</span>
<span class="sd">                 |</span>
<span class="sd">                 &quot;middle&quot;</span>



<span class="sd">             &quot;sticky_left3&quot;</span>
<span class="sd">             |</span>
<span class="sd">             |      &quot;sticky_right3&quot;</span>
<span class="sd">             |      |</span>
<span class="sd">            ---    ---</span>
<span class="sd">               ATCCAGT</span>
<span class="sd">            CCCTAGG</span>
<span class="sd">               ----</span>
<span class="sd">                 |</span>
<span class="sd">                 &quot;middle&quot;</span>



<span class="sd">               &quot;single_watson&quot; (only an upper strand)</span>
<span class="sd">               |</span>
<span class="sd">            -------</span>
<span class="sd">            ATCCAGT</span>
<span class="sd">            |||||||</span>



<span class="sd">               &quot;single_crick&quot; (only a lower strand)</span>
<span class="sd">               |</span>
<span class="sd">            -------</span>

<span class="sd">            |||||||</span>
<span class="sd">            CCCTAGG</span>


<span class="sd">        Up to seven groups (0..6) are captured, but some are mutually exclusive</span>
<span class="sd">        which means that one of them is an empty string:</span>

<span class="sd">        0 or 1, not both, a DNA fragment has either 5&#39; or 3&#39; sticky end.</span>

<span class="sd">        2 or 5 or 6, a DNA molecule has a ds region or is single stranded.</span>

<span class="sd">        3 or 4, not both, either 5&#39; or 3&#39; sticky end.</span>

<span class="sd">        Note that internal single stranded regions are not identified and will</span>
<span class="sd">        be contained in the middle part if they are present.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; from pydna.dseq import Dseq</span>
<span class="sd">        &gt;&gt;&gt; ds = Dseq(&quot;PPPATCFQZ&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ds</span>
<span class="sd">        Dseq(-9)</span>
<span class="sd">        GGGATC</span>
<span class="sd">           TAGTCA</span>
<span class="sd">        &gt;&gt;&gt; parts = ds.get_parts()</span>
<span class="sd">        &gt;&gt;&gt; parts</span>
<span class="sd">        DseqParts(sticky_left5=&#39;PPP&#39;, sticky_left3=&#39;&#39;,</span>
<span class="sd">                  middle=&#39;ATC&#39;,</span>
<span class="sd">                  sticky_right3=&#39;&#39;, sticky_right5=&#39;FQZ&#39;,</span>
<span class="sd">                  single_watson=&#39;&#39;, single_crick=&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; Dseq(parts.sticky_left5)</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        GGG</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; Dseq(parts.middle)</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        ATC</span>
<span class="sd">        TAG</span>
<span class="sd">        &gt;&gt;&gt; Dseq(parts.sticky_right5)</span>
<span class="sd">        Dseq(-3)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        TCA</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        datastring : str</span>
<span class="sd">            A string with dscode.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        namedtuple</span>
<span class="sd">            Seven string fields describing the DNA molecule.</span>
<span class="sd">            fragment(sticky_left5=&#39;&#39;, sticky_left3=&#39;&#39;,</span>
<span class="sd">                     middle=&#39;&#39;,</span>
<span class="sd">                     sticky_right3=&#39;&#39;, sticky_right5=&#39;&#39;,</span>
<span class="sd">                     single_watson=&#39;&#39;, single_crick=&#39;&#39;)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">get_parts</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">))</span></div>
</div>

</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Bjrn F. Johansson
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2024, Bjrn F. Johansson.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>