
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pydna.genbank &#8212; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=b530ca2f" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eea9577"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pydna/genbank';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="icon" href="../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pydna 6.0.0a24.post17+b7b559bd66 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/pydna-group/pydna" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../index.html">
                    Pydna
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../getting_started.html">Getting started</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq.html">Representing sequences in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq_Features.html">Working with Features using the Dseqrecord class</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Importing_Seqs.html">Importing and viewing sequence files in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Restrict_Ligate_Cloning.html">Restriction and Ligation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/PCR.html">PCR</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../notebooks/primer_design.html">Primer design</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Gibson.html">Gibson Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/CRISPR.html">CRISPR-Cas9</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/history.html">Cloning history / Cloning strategy</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../example_gallery.html">Example gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Restriction.html">Plasmid Restriction/Ligation Cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Gibson.html">Gibson Assembly from the original publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_CRISPR.html">Designing a CRISPR-Cas9 toolbox for genome engineering.</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../how_pydna_works.html">How pydna works</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/new_assemblies.html">How assemblies work</a></li>


</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../reference/pydna.html">pydna package</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/pydna-group/pydna" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pydna.genbank</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1"># Copyright 2013-2023 by Björn Johansson.  All rights reserved.</span>
<span class="c1"># This code is part of the Python-dna distribution and governed by its</span>
<span class="c1"># license.  Please see the LICENSE.txt file that should have been included</span>
<span class="c1"># as part of this package.</span>
<span class="sd">&quot;&quot;&quot;This module provides a class for downloading sequences from genbank</span>
<span class="sd">called Genbank and an function that does the same thing called genbank.</span>

<span class="sd">The function can be used if the environmental variable **pydna_email** has</span>
<span class="sd">been set to a valid email address. The easiest way to do this permanantly is to edit the</span>
<span class="sd">`pydna.ini` file. See the documentation of :func:`pydna.open_config_folder`&quot;&quot;&quot;</span>

<span class="c1"># from pydna.utils import memorize as _memorize</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.genbankrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">GenbankRecord</span> <span class="k">as</span> <span class="n">_GenbankRecord</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.readers</span><span class="w"> </span><span class="kn">import</span> <span class="n">read</span> <span class="k">as</span> <span class="n">_read</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Entrez</span> <span class="k">as</span> <span class="n">_Entrez</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Literal</span> <span class="k">as</span> <span class="n">_Literal</span><span class="p">,</span> <span class="n">Optional</span> <span class="k">as</span> <span class="n">_Optional</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_os</span>

<span class="c1"># import logging as _logging</span>

<span class="c1"># _module_logger = _logging.getLogger(&quot;pydna.&quot; + __name__)</span>


<span class="c1"># TODO http://httpbin.org/ use for testing?</span>


<div class="viewcode-block" id="Genbank">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Genbank">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Genbank</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class to facilitate download from genbank. It is easier and</span>
<span class="sd">    quicker to use the :func:`pydna.genbank.genbank` function directly.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    users_email : string</span>
<span class="sd">        Has to be a valid email address. You should always tell</span>
<span class="sd">        Genbanks who you are, so that they can contact you.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; from pydna.genbank import Genbank</span>
<span class="sd">    &gt;&gt;&gt; gb=Genbank(&quot;bjornjobb@gmail.com&quot;)</span>
<span class="sd">    &gt;&gt;&gt; rec = gb.nucleotide(&quot;LP002422.1&quot;)   # &lt;- entry from genbank</span>
<span class="sd">    &gt;&gt;&gt; print(len(rec))</span>
<span class="sd">    1</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">users_email</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">tool</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;pydna&quot;</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">_re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span>
            <span class="sa">r</span><span class="s2">&quot;[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}&quot;</span><span class="p">,</span> <span class="n">users_email</span><span class="p">,</span> <span class="n">_re</span><span class="o">.</span><span class="n">IGNORECASE</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;email address </span><span class="si">{}</span><span class="s2"> is not valid.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">users_email</span><span class="p">))</span>

        <span class="c1"># _module_logger.info(&quot;#### Genbank ititiation ####&quot;)</span>
        <span class="c1"># _module_logger.info(&quot;Genbank initiated with email: %s&quot;, users_email)</span>
        <span class="c1"># _module_logger.info(&quot;Genbank initiated with tool : %s&quot;, tool)</span>

        <span class="k">if</span> <span class="n">users_email</span> <span class="o">==</span> <span class="s2">&quot;someone@example.com&quot;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;you have to set your email address in order to download from Genbank&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="n">users_email</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tool</span> <span class="o">=</span> <span class="n">tool</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;This method returns a short representation containing the email used to initiate.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;GenbankConnection(</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">email</span><span class="p">)</span>

    <span class="c1"># @_memorize(&quot;pydna.genbank.Genbank.nucleotide&quot;)</span>
<div class="viewcode-block" id="Genbank.nucleotide">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.Genbank.nucleotide">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">nucleotide</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">item</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">seq_start</span><span class="p">:</span> <span class="n">_Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">seq_stop</span><span class="p">:</span> <span class="n">_Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">strand</span><span class="p">:</span> <span class="n">_Literal</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_GenbankRecord</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;This method downloads a genbank nuclotide record from genbank. This method is</span>
<span class="sd">        cached by default. This can be controlled by editing the **pydna_cached_funcs** environment</span>
<span class="sd">        variable. The best way to do this permanently is to edit the edit the</span>
<span class="sd">        `pydna.ini` file. See the documentation of :func:`pydna.open_config_folder`</span>

<span class="sd">        Item is a string containing one genbank accession number</span>
<span class="sd">        for a nucleotide file. Genbank nucleotide accession numbers have this format:</span>

<span class="sd">        | A12345   = 1 letter  + 5 numerals</span>
<span class="sd">        | AB123456 = 2 letters + 6 numerals</span>

<span class="sd">        The accession number is sometimes followed by a point and version number</span>

<span class="sd">        | BK006936.2</span>

<span class="sd">        Item can also contain optional interval information in the following formats:</span>

<span class="sd">        | BK006936.2 REGION: complement(613900..615202)</span>
<span class="sd">        | NM_005546 REGION: 1..100</span>
<span class="sd">        | NM_005546 REGION: complement(1..100)</span>
<span class="sd">        | 21614549:1-100</span>
<span class="sd">        | 21614549:c100-1</span>
<span class="sd">        | 21614549 1-100</span>
<span class="sd">        | 21614549 c100-1</span>

<span class="sd">        It is useful to set an interval for large genbank records to limit the download time.</span>
<span class="sd">        The items above containing interval information and can be obtained directly by</span>
<span class="sd">        looking up an entry in Genbank and setting the `Change region shown` on the</span>
<span class="sd">        upper right side of the page. The `ACCESSION` line of the displayed Genbank</span>
<span class="sd">        file will have the formatting shown.</span>

<span class="sd">        Alternatively, seq_start and seq_stop can be set explicitly to the sequence intervals to be</span>
<span class="sd">        downloaded.</span>

<span class="sd">        If strand is 2. &quot;c&quot;, &quot;C&quot;, &quot;crick&quot;, &quot;Crick&quot;, &quot;antisense&quot;,&quot;Antisense&quot;,</span>
<span class="sd">        &quot;2&quot;, 2, &quot;-&quot; or &quot;-1&quot;, the antisense (Crick) strand is returned, otherwise</span>
<span class="sd">        the sense (Watson) strand is returned.</span>

<span class="sd">        Result is returned as a :class:`pydna.genbankrecord.GenbankRecord` object.</span>

<span class="sd">        References</span>
<span class="sd">        ----------</span>

<span class="sd">        .. [#]   http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm</span>
<span class="sd">        .. [#]   http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(REGION:\s(?P&lt;start&gt;\d+)\.\.(?P&lt;stop&gt;\d+))&quot;</span><span class="p">,</span> <span class="n">item</span><span class="p">)),</span>
            <span class="p">(</span>
                <span class="mi">2</span><span class="p">,</span>
                <span class="n">_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
                    <span class="sa">r</span><span class="s2">&quot;(REGION: complement\((?P&lt;start&gt;\d+)\.\.(?P&lt;stop&gt;\d+)\))&quot;</span><span class="p">,</span> <span class="n">item</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(:|\s)(?P&lt;start&gt;\d+)-(?P&lt;stop&gt;\d+)&quot;</span><span class="p">,</span> <span class="n">item</span><span class="p">)),</span>
            <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(:|\s)c(?P&lt;start&gt;\d+)-(?P&lt;stop&gt;\d+)&quot;</span><span class="p">,</span> <span class="n">item</span><span class="p">)),</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">strand_</span><span class="p">,</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">matches</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                <span class="n">seq_start</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s2">&quot;start&quot;</span><span class="p">)</span>
                <span class="n">seq_stop</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s2">&quot;stop&quot;</span><span class="p">)</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">item</span><span class="p">[:</span> <span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()]</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="n">strand_</span>
                <span class="k">break</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;c&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;crick&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;antisense&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">}[</span>
                    <span class="n">strand</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">KeyError</span><span class="p">,</span> <span class="ne">AttributeError</span><span class="p">):</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="c1"># _module_logger.info(&quot;#### Genbank download ####&quot;)</span>
        <span class="c1"># _module_logger.info(&quot;item  %s&quot;, item)</span>
        <span class="c1"># _module_logger.info(&quot;start %s&quot;, seq_start)</span>
        <span class="c1"># _module_logger.info(&quot;stop  %s&quot;, seq_stop)</span>

        <span class="c1"># _module_logger.info(&quot;strand  %s&quot;, str(strand))</span>

        <span class="n">_Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">email</span>
        <span class="n">_Entrez</span><span class="o">.</span><span class="n">tool</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tool</span>

        <span class="c1"># _module_logger.info(&quot;Entrez.email  %s&quot;, self.email)</span>
        <span class="n">text</span> <span class="o">=</span> <span class="n">_Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span>
            <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nuccore&quot;</span><span class="p">,</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">item</span><span class="p">,</span>
            <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gbwithparts&quot;</span><span class="p">,</span>
            <span class="n">seq_start</span><span class="o">=</span><span class="n">seq_start</span><span class="p">,</span>
            <span class="n">seq_stop</span><span class="o">=</span><span class="n">seq_stop</span><span class="p">,</span>
            <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span>
            <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

        <span class="c1"># _module_logger.info(&quot;text[:160]  %s&quot;, text[:160])</span>

        <span class="k">return</span> <span class="n">_GenbankRecord</span><span class="p">(</span>
            <span class="n">_read</span><span class="p">(</span><span class="n">text</span><span class="p">),</span> <span class="n">item</span><span class="o">=</span><span class="n">item</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">seq_start</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="n">seq_stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span>
        <span class="p">)</span></div>
</div>



<div class="viewcode-block" id="genbank">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.genbank">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">genbank</span><span class="p">(</span><span class="n">accession</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;CS570233.1&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_GenbankRecord</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download a genbank nuclotide record.</span>

<span class="sd">    This function takes the same paramenters as the</span>
<span class="sd">    :func:pydna.genbank.Genbank.nucleotide method. The email address stored</span>
<span class="sd">    in the `pydna_email` environment variable is used. The easiest way set</span>
<span class="sd">    this permanantly is to edit the `pydna.ini` file.</span>
<span class="sd">    See the documentation of :func:`pydna.open_config_folder`</span>

<span class="sd">    if no accession is given, a `very short Genbank</span>
<span class="sd">    entry &lt;https://www.ncbi.nlm.nih.gov/nuccore/CS570233.1&gt;`_</span>
<span class="sd">    is used as an example (see below). This can be useful for testing the</span>
<span class="sd">    connection to Genbank.</span>

<span class="sd">    Please note that this result is also cached by default by settings in</span>
<span class="sd">    the pydna.ini file.</span>
<span class="sd">    See the documentation of :func:`pydna.open_config_folder`</span>

<span class="sd">    ::</span>

<span class="sd">        LOCUS       CS570233                  14 bp    DNA     linear   PAT 18-MAY-2007</span>
<span class="sd">        DEFINITION  Sequence 6 from Patent WO2007025016.</span>
<span class="sd">        ACCESSION   CS570233</span>
<span class="sd">        VERSION     CS570233.1</span>
<span class="sd">        KEYWORDS    .</span>
<span class="sd">        SOURCE      synthetic construct</span>
<span class="sd">          ORGANISM  synthetic construct</span>
<span class="sd">                    other sequences; artificial sequences.</span>
<span class="sd">        REFERENCE   1</span>
<span class="sd">          AUTHORS   Shaw,R.W. and Cottenoir,M.</span>
<span class="sd">          TITLE     Inhibition of metallo-beta-lactamase by double-stranded dna</span>
<span class="sd">          JOURNAL   Patent: WO 2007025016-A1 6 01-MAR-2007;</span>
<span class="sd">                    Texas Tech University System (US)</span>
<span class="sd">        FEATURES             Location/Qualifiers</span>
<span class="sd">             source          1..14</span>
<span class="sd">                             /organism=&quot;synthetic construct&quot;</span>
<span class="sd">                             /mol_type=&quot;unassigned DNA&quot;</span>
<span class="sd">                             /db_xref=&quot;taxon:32630&quot;</span>
<span class="sd">                             /note=&quot;This is a 14bp aptamer inhibitor.&quot;</span>
<span class="sd">        ORIGIN</span>
<span class="sd">                1 atgttcctac atga</span>
<span class="sd">        //</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">email</span> <span class="o">=</span> <span class="n">_os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;pydna_email&quot;</span><span class="p">)</span>
    <span class="c1"># _module_logger.info(&quot;#### genbank function called ####&quot;)</span>
    <span class="c1"># _module_logger.info(&quot;email      %s&quot;, email)</span>
    <span class="c1"># _module_logger.info(&quot;accession  %s&quot;, email)</span>
    <span class="n">gb</span> <span class="o">=</span> <span class="n">Genbank</span><span class="p">(</span><span class="n">email</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gb</span><span class="o">.</span><span class="n">nucleotide</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Björn F. Johansson
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Björn F. Johansson.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>