
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pydna.parsers &#8212; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=b530ca2f" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eea9577"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pydna/parsers';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="icon" href="../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pydna 6.0.0a24.post17+b7b559bd66 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/pydna-group/pydna" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../index.html">
                    Pydna
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../getting_started.html">Getting started</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq.html">Representing sequences in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq_Features.html">Working with Features using the Dseqrecord class</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Importing_Seqs.html">Importing and viewing sequence files in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Restrict_Ligate_Cloning.html">Restriction and Ligation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/PCR.html">PCR</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../notebooks/primer_design.html">Primer design</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Gibson.html">Gibson Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/CRISPR.html">CRISPR-Cas9</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/history.html">Cloning history / Cloning strategy</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../example_gallery.html">Example gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Restriction.html">Plasmid Restriction/Ligation Cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Gibson.html">Gibson Assembly from the original publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_CRISPR.html">Designing a CRISPR-Cas9 toolbox for genome engineering.</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../how_pydna_works.html">How pydna works</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/new_assemblies.html">How assemblies work</a></li>


</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../reference/pydna.html">pydna package</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/pydna-group/pydna" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pydna.parsers</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1"># Copyright 2013-2023 by Björn Johansson.  All rights reserved.</span>
<span class="c1"># This code is part of the Python-dna distribution and governed by its</span>
<span class="c1"># license.  Please see the LICENSE.txt file that should have been included</span>
<span class="c1"># as part of this package.</span>

<span class="sd">&quot;&quot;&quot;Provides two functions, parse and parse_primers&quot;&quot;&quot;</span>

<span class="c1"># import os as _os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_io</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">textwrap</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_textwrap</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqIO</span> <span class="k">as</span> <span class="n">_SeqIO</span>

<span class="c1"># from pydna.genbankfile import GenbankFile as _GenbankFile</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.dseqrecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dseqrecord</span> <span class="k">as</span> <span class="n">_Dseqrecord</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span> <span class="k">as</span> <span class="n">_SeqRecord</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.opencloning_models</span><span class="w"> </span><span class="kn">import</span> <span class="n">UploadedFileSource</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.primer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Primer</span> <span class="k">as</span> <span class="n">_Primer</span>

<span class="c1"># from pydna.amplify import pcr as _pcr</span>
<span class="c1"># from copy import deepcopy as _deepcopy</span>
<span class="c1"># from Bio.SeqFeature import SeqFeature as _SeqFeature</span>
<span class="c1"># import xml.etree.ElementTree as _et</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">itertools</span><span class="w"> </span><span class="kn">import</span> <span class="n">pairwise</span> <span class="k">as</span> <span class="n">_pairwise</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_pairwise</span><span class="p">(</span><span class="n">iterable</span><span class="p">):</span>
        <span class="c1"># pairwise(&#39;ABCDEFG&#39;) → AB BC CD DE EF FG</span>
        <span class="n">iterator</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">iterator</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">iterator</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">b</span>


<span class="c1"># &quot;^&gt;.+?^(?=$|LOCUS|ID|&gt;|\#)|^(?:LOCUS|ID).+?^//&quot;</span>
<span class="c1"># &quot;(?:^&gt;.+\n^(?:^[^&gt;]+?)(?=\n\n|&gt;|^LOCUS|ID))|(?:(?:^LOCUS|ID)(?:(?:.|\n)+?)^//)&quot;</span>

<span class="c1"># gb_fasta_embl_regex = r&quot;(?:&gt;.+\n^(?:^[^&gt;]+?)(?=\n\n|&gt;|LOCUS|ID))|(?:(?:LOCUS|ID)(?:(?:.|\n)+?)^//)&quot;</span>

<span class="n">gb_fasta_embl_regex</span> <span class="o">=</span> <span class="p">(</span>
    <span class="sa">r</span><span class="s2">&quot;(?:^&gt;.+\n^(?:^[^&gt;]+?)(?=\n\n|&gt;|^LOCUS|^ID))|(?:(?:^LOCUS|^ID)(?:(?:.|\n)+?)^//)&quot;</span>
<span class="p">)</span>

<span class="c1"># The gb_fasta_embl_regex is meant to be able to extract sequences from</span>
<span class="c1"># text where sequences are mixed with other contents as well</span>
<span class="c1"># use https://regex101.com to get an idea how it works.</span>


<div class="viewcode-block" id="extract_from_text">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.extract_from_text">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_from_text</span><span class="p">(</span><span class="n">text</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">_textwrap</span><span class="o">.</span><span class="n">dedent</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">text</span><span class="p">))</span>
    <span class="n">mos</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">_re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">gb_fasta_embl_regex</span><span class="p">,</span> <span class="n">data</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">_re</span><span class="o">.</span><span class="n">MULTILINE</span><span class="p">))</span>

    <span class="k">class</span><span class="w"> </span><span class="nc">Fakemo</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">start</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">end</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">return</span> <span class="mi">0</span>

    <span class="n">mofirst</span> <span class="o">=</span> <span class="n">molast</span> <span class="o">=</span> <span class="n">Fakemo</span><span class="p">()</span>

    <span class="n">gaps</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">mo1</span><span class="p">,</span> <span class="n">mo2</span> <span class="ow">in</span> <span class="n">_pairwise</span><span class="p">([</span><span class="n">mofirst</span><span class="p">]</span> <span class="o">+</span> <span class="n">mos</span> <span class="o">+</span> <span class="p">[</span><span class="n">molast</span><span class="p">]):</span>
        <span class="n">gaps</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">mo1</span><span class="o">.</span><span class="n">end</span><span class="p">()</span> <span class="p">:</span> <span class="n">mo2</span><span class="o">.</span><span class="n">start</span><span class="p">()])</span>

    <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">mo</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">mo</span> <span class="ow">in</span> <span class="n">mos</span><span class="p">),</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">gaps</span><span class="p">)</span></div>



<div class="viewcode-block" id="embl_gb_fasta">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.embl_gb_fasta">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">embl_gb_fasta</span><span class="p">(</span><span class="n">text</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse embl, genbank or fasta format from text.</span>

<span class="sd">    Returns list of Bio.SeqRecord.SeqRecord</span>

<span class="sd">    annotations[&quot;molecule_type&quot;]</span>
<span class="sd">    annotations[&quot;topology&quot;]</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chunks</span><span class="p">,</span> <span class="n">gaps</span> <span class="o">=</span> <span class="n">extract_from_text</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>
    <span class="n">result_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># topology = &quot;linear&quot;</span>

    <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">chunks</span><span class="p">:</span>
        <span class="n">handle</span> <span class="o">=</span> <span class="n">_io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
        <span class="c1"># circular = False</span>
        <span class="n">first_line</span> <span class="o">=</span> <span class="n">chunk</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">parsed</span> <span class="o">=</span> <span class="n">_SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;embl&quot;</span><span class="p">)</span>
            <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;pydna_parse_sequence_file_format&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;embl&quot;</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="n">handle</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">parsed</span> <span class="o">=</span> <span class="n">_SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
                <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;pydna_parse_sequence_file_format&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;genbank&quot;</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="n">handle</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">parsed</span> <span class="o">=</span> <span class="n">_SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta-blast&quot;</span><span class="p">)</span>
                    <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;pydna_parse_sequence_file_format&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;fasta&quot;</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># hack to pick up molecule_type from FASTA header line</span>
                    <span class="k">if</span> <span class="s2">&quot;protein&quot;</span> <span class="ow">in</span> <span class="n">first_line</span><span class="p">:</span>
                        <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;protein&quot;</span>
                        <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;DNA&quot;</span>
            <span class="c1"># else:</span>
            <span class="c1">#     if _re.match(r&quot;LOCUS\s+(\S+)\s+(\S+)\s+aa&quot;, &quot; &quot;.join(first_line)):</span>
            <span class="c1">#         parsed.annotations[&quot;molecule_type&quot;] = &quot;protein&quot;</span>
            <span class="c1">#         parsed.annotations[&quot;topology&quot;] = &quot;linear&quot;</span>
        <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="c1"># hack to pick up topology from FASTA and malformed gb files</span>
        <span class="n">first_line</span> <span class="o">=</span> <span class="n">chunk</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">first_line</span><span class="p">:</span>
            <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;circular&quot;</span>
        <span class="c1"># assert parsed.annotations.get(&quot;topology&quot;), &quot;topology must be set&quot;</span>
        <span class="k">assert</span> <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">),</span> <span class="s2">&quot;molecule_type  must be set&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">parsed</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span>
        <span class="n">result_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">result_list</span><span class="p">)</span></div>



<div class="viewcode-block" id="parse">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.parse">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">parse</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">ds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">_Dseqrecord</span> <span class="o">|</span> <span class="n">_SeqRecord</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return *all* DNA sequences found in data.</span>

<span class="sd">    If no sequences are found, an empty list is returned. This is a greedy</span>
<span class="sd">    function, use carefully.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data : string or iterable</span>
<span class="sd">        The data parameter is a string containing:</span>

<span class="sd">        1. an absolute path to a local file.</span>
<span class="sd">           The file will be read in text</span>
<span class="sd">           mode and parsed for EMBL, FASTA</span>
<span class="sd">           and Genbank sequences. Can be</span>
<span class="sd">           a string or a Path object.</span>

<span class="sd">        2. a string containing one or more</span>
<span class="sd">           sequences in EMBL, GENBANK,</span>
<span class="sd">           or FASTA format. Mixed formats</span>
<span class="sd">           are allowed.</span>

<span class="sd">        3. data can be a list or other iterable where the elements are 1 or 2</span>

<span class="sd">    ds : bool</span>
<span class="sd">        If True double stranded :class:`Dseqrecord` objects are returned.</span>
<span class="sd">        If False single stranded :class:`Bio.SeqRecord` [#]_ objects are</span>
<span class="sd">        returned.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list</span>
<span class="sd">        contains Dseqrecord or SeqRecord objects</span>

<span class="sd">    References</span>
<span class="sd">    ----------</span>
<span class="sd">    .. [#] http://biopython.org/wiki/SeqRecord</span>

<span class="sd">    See Also</span>
<span class="sd">    --------</span>
<span class="sd">    read</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># a string is an iterable datatype but on Python2.x</span>
    <span class="c1"># it doesn&#39;t have an __iter__ method.</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s2">&quot;__iter__&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">)):</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">(</span><span class="n">data</span><span class="p">,)</span>

    <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># item is a path to a utf-8 encoded text file?</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">raw</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
            <span class="c1"># item was not a path, add sequences parsed from item</span>
            <span class="n">raw</span> <span class="o">=</span> <span class="n">item</span>
            <span class="n">path</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># item was a readable text file, seqences are parsed from the file</span>
            <span class="n">path</span> <span class="o">=</span> <span class="n">item</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">newsequences</span> <span class="o">=</span> <span class="n">embl_gb_fasta</span><span class="p">(</span><span class="n">raw</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">newsequences</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ds</span> <span class="ow">and</span> <span class="n">path</span><span class="p">:</span>
                    <span class="kn">from</span><span class="w"> </span><span class="nn">pydna.opencloning_models</span><span class="w"> </span><span class="kn">import</span> <span class="n">UploadedFileSource</span>

                    <span class="n">result</span> <span class="o">=</span> <span class="n">_Dseqrecord</span><span class="o">.</span><span class="n">from_SeqRecord</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="n">UploadedFileSource</span><span class="p">(</span>
                        <span class="n">file_name</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">),</span>  <span class="c1"># we use str to handle PosixPath</span>
                        <span class="n">sequence_file_format</span><span class="o">=</span><span class="n">s</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span>
                            <span class="s2">&quot;pydna_parse_sequence_file_format&quot;</span>
                        <span class="p">],</span>
                        <span class="n">index_in_file</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
                    <span class="c1"># sequences.append(_GenbankFile.from_SeqRecord(s, path=path))</span>
                <span class="k">elif</span> <span class="n">ds</span><span class="p">:</span>
                    <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_Dseqrecord</span><span class="o">.</span><span class="n">from_SeqRecord</span><span class="p">(</span><span class="n">s</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sequences</span></div>



<div class="viewcode-block" id="parse_primers">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.parse_primers">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_primers</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">_Primer</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">parse</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">ds</span><span class="o">=</span><span class="kc">False</span><span class="p">)]</span></div>



<div class="viewcode-block" id="parse_snapgene">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.parse_snapgene">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_snapgene</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">_Dseqrecord</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse a SnapGene file and return a Dseqrecord object.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : str</span>
<span class="sd">        The path to the SnapGene file to parse.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Dseqrecord</span>
<span class="sd">        The parsed SnapGene file as a Dseqrecord object.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">parsed_seq</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">_SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s2">&quot;snapgene&quot;</span><span class="p">))</span>
        <span class="n">circular</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;topology&quot;</span> <span class="ow">in</span> <span class="n">parsed_seq</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="ow">and</span> <span class="n">parsed_seq</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span>
        <span class="p">)</span>

        <span class="n">source</span> <span class="o">=</span> <span class="n">UploadedFileSource</span><span class="p">(</span>
            <span class="n">file_name</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">file_path</span><span class="p">),</span>
            <span class="n">sequence_file_format</span><span class="o">=</span><span class="s2">&quot;snapgene&quot;</span><span class="p">,</span>
            <span class="n">index_in_file</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">_Dseqrecord</span><span class="p">(</span><span class="n">parsed_seq</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="n">circular</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">)]</span></div>

</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Björn F. Johansson
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Björn F. Johansson.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>