
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pydna.utils &#8212; pydna 6.0.0a24.post17+b7b559bd66 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=b530ca2f" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eea9577"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pydna/utils';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="icon" href="../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">pydna 6.0.0a24.post17+b7b559bd66 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/pydna-group/pydna" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../index.html">
                    Pydna
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../getting_started.html">Getting started</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq.html">Representing sequences in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Dseq_Features.html">Working with Features using the Dseqrecord class</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Importing_Seqs.html">Importing and viewing sequence files in pydna</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Restrict_Ligate_Cloning.html">Restriction and Ligation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/PCR.html">PCR</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../notebooks/primer_design.html">Primer design</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Gibson.html">Gibson Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/CRISPR.html">CRISPR-Cas9</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/history.html">Cloning history / Cloning strategy</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../example_gallery.html">Example gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Restriction.html">Plasmid Restriction/Ligation Cloning</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_Gibson.html">Gibson Assembly from the original publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/Example_CRISPR.html">Designing a CRISPR-Cas9 toolbox for genome engineering.</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../how_pydna_works.html">How pydna works</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../notebooks/new_assemblies.html">How assemblies work</a></li>


</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../reference/pydna.html">pydna package</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/pydna-group/pydna" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pydna.utils</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1"># Copyright 2013-2023 by Björn Johansson.  All rights reserved.</span>
<span class="c1"># This code is part of the Python-dna distribution and governed by its</span>
<span class="c1"># license.  Please see the LICENSE.txt file that should have been included</span>
<span class="c1"># as part of this package.</span>
<span class="sd">&quot;&quot;&quot;Miscellaneous functions.&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Data.IUPACData</span><span class="w"> </span><span class="kn">import</span> <span class="n">ambiguous_dna_complement</span> <span class="k">as</span> <span class="n">_ambiguous_dna_complement</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">_maketrans</span>

<span class="c1"># import shelve as _shelve</span>
<span class="c1"># import os as _os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_re</span>

<span class="c1"># import logging as _logging</span>
<span class="c1"># import base64 as _base64</span>
<span class="c1"># import pickle as _pickle</span>
<span class="c1"># import hashlib as _hashlib</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">keyword</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_keyword</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_collections</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_itertools</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">copy</span><span class="w"> </span><span class="kn">import</span> <span class="n">deepcopy</span> <span class="k">as</span> <span class="n">_deepcopy</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">_subprocess</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bisect</span><span class="w"> </span><span class="kn">import</span> <span class="n">bisect</span> <span class="k">as</span> <span class="n">_bisect</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">math</span><span class="w"> </span><span class="kn">import</span> <span class="n">ceil</span> <span class="k">as</span> <span class="n">_ceil</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.codon</span><span class="w"> </span><span class="kn">import</span> <span class="n">weights</span> <span class="k">as</span> <span class="n">_weights</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pydna.codon</span><span class="w"> </span><span class="kn">import</span> <span class="n">rare_codons</span> <span class="k">as</span> <span class="n">_rare_codons</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqFeature</span><span class="w"> </span><span class="kn">import</span> <span class="n">SimpleLocation</span> <span class="k">as</span> <span class="n">_sl</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqFeature</span><span class="w"> </span><span class="kn">import</span> <span class="n">CompoundLocation</span> <span class="k">as</span> <span class="n">_cl</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqFeature</span><span class="w"> </span><span class="kn">import</span> <span class="n">Location</span> <span class="k">as</span> <span class="n">_Location</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Union</span> <span class="k">as</span> <span class="n">_Union</span><span class="p">,</span> <span class="n">TypeVar</span> <span class="k">as</span> <span class="n">_TypeVar</span><span class="p">,</span> <span class="n">List</span> <span class="k">as</span> <span class="n">_List</span>

<span class="c1"># For functions that take str or bytes as input and return str or bytes as output, matching the input type</span>
<span class="n">StrOrBytes</span> <span class="o">=</span> <span class="n">_TypeVar</span><span class="p">(</span><span class="s2">&quot;StrOrBytes&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">)</span>

<span class="c1"># _module_logger = _logging.getLogger(&quot;pydna.&quot; + __name__)</span>
<span class="n">_ambiguous_dna_complement</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;U&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">})</span>
<span class="n">_complement_table</span> <span class="o">=</span> <span class="n">_maketrans</span><span class="p">(</span><span class="n">_ambiguous_dna_complement</span><span class="p">)</span>


<div class="viewcode-block" id="three_frame_orfs">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.three_frame_orfs">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">three_frame_orfs</span><span class="p">(</span>
    <span class="n">dna</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">limit</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
    <span class="n">startcodons</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">,),</span>
    <span class="n">stopcodons</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;TAG&quot;</span><span class="p">,</span> <span class="s2">&quot;TAA&quot;</span><span class="p">,</span> <span class="s2">&quot;TGA&quot;</span><span class="p">),</span>
    <span class="c1"># startcodons: tuple[str, ...] = (&quot;ATG&quot;,),</span>
    <span class="c1"># stopcodons: tuple[str, ...] = (&quot;TAG&quot;, &quot;TAA&quot;, &quot;TGA&quot;),</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overlapping orfs in three frames.&quot;&quot;&quot;</span>
    <span class="c1"># breakpoint()</span>
    <span class="n">limit</span> <span class="o">=</span> <span class="n">_ceil</span><span class="p">(</span><span class="n">limit</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="n">dna</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

    <span class="n">orfs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">frame</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>

        <span class="n">codons</span> <span class="o">=</span> <span class="p">[</span><span class="n">dna</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">frame</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dna</span><span class="p">),</span> <span class="mi">3</span><span class="p">)]</span>

        <span class="n">startdindices</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cd</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">codons</span><span class="p">)</span> <span class="k">if</span> <span class="n">cd</span> <span class="ow">in</span> <span class="n">startcodons</span><span class="p">]</span>
        <span class="n">stopdindices</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cd</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">codons</span><span class="p">)</span> <span class="k">if</span> <span class="n">cd</span> <span class="ow">in</span> <span class="n">stopcodons</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">startindex</span> <span class="ow">in</span> <span class="n">startdindices</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">stopindex</span> <span class="o">=</span> <span class="n">stopdindices</span><span class="p">[</span><span class="n">_bisect</span><span class="p">(</span><span class="n">stopdindices</span><span class="p">,</span> <span class="n">startindex</span><span class="p">)]</span>
            <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">stopindex</span> <span class="o">-</span> <span class="n">startindex</span> <span class="o">&gt;=</span> <span class="n">limit</span><span class="p">:</span>
                    <span class="n">orfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="p">(</span><span class="n">frame</span><span class="p">,</span> <span class="n">startindex</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">frame</span><span class="p">,</span> <span class="p">(</span><span class="n">stopindex</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="n">frame</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="c1"># print(stopindex, startindex, limit)</span>
    <span class="k">return</span> <span class="n">orfs</span></div>



<div class="viewcode-block" id="shift_location">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.shift_location">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">shift_location</span><span class="p">(</span><span class="n">original_location</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">lim</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="n">newparts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">original_location</span><span class="o">.</span><span class="n">strand</span>
    <span class="k">if</span> <span class="n">lim</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">min</span><span class="p">(</span><span class="n">original_location</span><span class="p">)</span> <span class="o">+</span> <span class="n">shift</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Shift moves location below zero, use a `lim` to loop around if sequence is circular.&quot;</span>
            <span class="p">)</span>
        <span class="n">lim</span> <span class="o">=</span> <span class="n">_sys</span><span class="o">.</span><span class="n">maxsize</span>

    <span class="k">for</span> <span class="n">part</span> <span class="ow">in</span> <span class="n">original_location</span><span class="o">.</span><span class="n">parts</span><span class="p">:</span>
        <span class="n">new_start</span> <span class="o">=</span> <span class="p">(</span><span class="n">part</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="n">shift</span><span class="p">)</span> <span class="o">%</span> <span class="n">lim</span>
        <span class="n">new_end</span> <span class="o">=</span> <span class="p">(</span><span class="n">part</span><span class="o">.</span><span class="n">end</span> <span class="o">+</span> <span class="n">shift</span><span class="p">)</span> <span class="o">%</span> <span class="n">lim</span> <span class="ow">or</span> <span class="n">lim</span>
        <span class="n">old_start</span><span class="p">,</span> <span class="n">old_end</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">newparts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">newparts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">newparts</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># The &quot;join with old&quot; cases are for features with multiple parts</span>
        <span class="c1"># in which consecutive parts do not have any bases between them.</span>
        <span class="c1"># This type of feature is generated to represent a feature that</span>
        <span class="c1"># spans the origin of a circular sequence. See more details in</span>
        <span class="c1"># https://github.com/pydna-group/pydna/issues/195</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">part</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">newparts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_sl</span><span class="p">(</span><span class="n">new_start</span><span class="p">,</span> <span class="n">new_start</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
            <span class="k">continue</span>
        <span class="c1"># Join with old, case 1</span>
        <span class="k">elif</span> <span class="n">strand</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span> <span class="ow">and</span> <span class="n">old_end</span> <span class="o">==</span> <span class="n">new_start</span><span class="p">:</span>
            <span class="n">part</span> <span class="o">=</span> <span class="n">newparts</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="n">part</span><span class="o">.</span><span class="n">_end</span> <span class="o">=</span> <span class="n">new_end</span>
            <span class="n">new_start</span> <span class="o">=</span> <span class="n">part</span><span class="o">.</span><span class="n">start</span>
        <span class="c1"># Join with old, case 2</span>
        <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="ow">and</span> <span class="n">old_start</span> <span class="o">==</span> <span class="n">new_end</span><span class="p">:</span>
            <span class="n">part</span> <span class="o">=</span> <span class="n">newparts</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="n">part</span><span class="o">.</span><span class="n">_start</span> <span class="o">=</span> <span class="n">new_start</span>
            <span class="n">new_end</span> <span class="o">=</span> <span class="n">part</span><span class="o">.</span><span class="n">end</span>
        <span class="k">if</span> <span class="n">new_start</span> <span class="o">&lt;</span> <span class="n">new_end</span><span class="p">:</span>
            <span class="n">newparts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_sl</span><span class="p">(</span><span class="n">new_start</span><span class="p">,</span> <span class="n">new_end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">parttuple</span> <span class="o">=</span> <span class="p">(</span><span class="n">_sl</span><span class="p">(</span><span class="n">new_start</span><span class="p">,</span> <span class="n">lim</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="n">_sl</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">new_end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
            <span class="n">newparts</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">parttuple</span> <span class="k">if</span> <span class="n">strand</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">parttuple</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">newloc</span> <span class="o">=</span> <span class="n">_cl</span><span class="p">(</span><span class="n">newparts</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="n">newloc</span><span class="p">,</span> <span class="o">*</span><span class="n">n</span> <span class="o">=</span> <span class="n">newparts</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">newloc</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">original_location</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">newloc</span></div>



<span class="c1"># def shift_feature(feature, shift, lim):</span>
<span class="c1">#     &quot;&quot;&quot;Return a new feature with shifted location.&quot;&quot;&quot;</span>
<span class="c1">#     # TODO: Missing tests</span>
<span class="c1">#     new_location = shift_location(feature.location, shift, lim)</span>
<span class="c1">#     new_feature = _deepcopy(feature)</span>
<span class="c1">#     new_feature.location = new_location</span>
<span class="c1">#     return new_feature</span>


<div class="viewcode-block" id="shift_feature">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.shift_feature">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">shift_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">lim</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a new feature with shifted location.&quot;&quot;&quot;</span>
    <span class="c1"># TODO: Missing tests</span>
    <span class="n">new_location</span> <span class="o">=</span> <span class="n">shift_location</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">lim</span><span class="p">)</span>
    <span class="n">new_feature</span> <span class="o">=</span> <span class="n">_deepcopy</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
    <span class="n">new_feature</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">new_location</span>
    <span class="k">return</span> <span class="n">new_feature</span></div>



<span class="c1"># def smallest_rotation(s):</span>
<span class="c1">#     &quot;&quot;&quot;Smallest rotation of a string.</span>

<span class="c1">#     Algorithm described in Pierre Duval, Jean. 1983. Factorizing Words</span>
<span class="c1">#     over an Ordered Alphabet. Journal of Algorithms &amp; Computational Technology</span>
<span class="c1">#     4 (4) (December 1): 363–381. and Algorithms on strings and sequences based</span>
<span class="c1">#     on Lyndon words, David Eppstein 2011.</span>
<span class="c1">#     https://gist.github.com/dvberkel/1950267</span>

<span class="c1">#     Examples</span>
<span class="c1">#     --------</span>
<span class="c1">#     &gt;&gt;&gt; from pydna.utils import smallest_rotation</span>
<span class="c1">#     &gt;&gt;&gt; smallest_rotation(&quot;taaa&quot;)</span>
<span class="c1">#     &#39;aaat&#39;</span>

<span class="c1">#     &quot;&quot;&quot;</span>
<span class="c1">#     prev, rep = None, 0</span>
<span class="c1">#     ds = _array(&quot;u&quot;, 2 * s)</span>
<span class="c1">#     lens, lends = len(s), len(ds)</span>
<span class="c1">#     old = 0</span>
<span class="c1">#     k = 0</span>
<span class="c1">#     w = &quot;&quot;</span>
<span class="c1">#     while k &lt; lends:</span>
<span class="c1">#         i, j = k, k + 1</span>
<span class="c1">#         while j &lt; lends and ds[i] &lt;= ds[j]:</span>
<span class="c1">#             i = (ds[i] == ds[j]) and i + 1 or k</span>
<span class="c1">#             j += 1</span>
<span class="c1">#         while k &lt; i + 1:</span>
<span class="c1">#             k += j - i</span>
<span class="c1">#             prev = w</span>
<span class="c1">#             w = ds[old:k]</span>
<span class="c1">#             old = k</span>
<span class="c1">#             if w == prev:</span>
<span class="c1">#                 rep += 1</span>
<span class="c1">#             else:</span>
<span class="c1">#                 prev, rep = w, 1</span>
<span class="c1">#             if len(w) * rep == lens:</span>
<span class="c1">#                 return &quot;&quot;.join(w * rep)</span>


<div class="viewcode-block" id="smallest_rotation">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.smallest_rotation">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">smallest_rotation</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Smallest rotation of a string.</span>

<span class="sd">    Algorithm described in Pierre Duval, Jean. 1983. Factorizing Words</span>
<span class="sd">    over an Ordered Alphabet. Journal of Algorithms &amp; Computational Technology</span>
<span class="sd">    4 (4) (December 1): 363–381. and Algorithms on strings and sequences based</span>
<span class="sd">    on Lyndon words, David Eppstein 2011.</span>
<span class="sd">    https://gist.github.com/dvberkel/1950267</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; from pydna.utils import smallest_rotation</span>
<span class="sd">    &gt;&gt;&gt; smallest_rotation(&quot;taaa&quot;)</span>
<span class="sd">    &#39;aaat&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">pydivsufsort</span><span class="w"> </span><span class="kn">import</span> <span class="n">min_rotation</span>

    <span class="n">k</span> <span class="o">=</span> <span class="n">min_rotation</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="s2">&quot;ascii&quot;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">s</span><span class="p">[</span><span class="n">k</span><span class="p">:]</span> <span class="o">+</span> <span class="n">s</span><span class="p">[:</span><span class="n">k</span><span class="p">]</span></div>



<div class="viewcode-block" id="cai">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.cai">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">cai</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">organism</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;sce&quot;</span><span class="p">,</span> <span class="n">weights</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="n">_weights</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">cai2</span><span class="w"> </span><span class="kn">import</span> <span class="n">CAI</span> <span class="k">as</span> <span class="n">_CAI</span>

    <span class="k">return</span> <span class="nb">round</span><span class="p">(</span><span class="n">_CAI</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">[</span><span class="n">organism</span><span class="p">]),</span> <span class="mi">3</span><span class="p">)</span></div>



<div class="viewcode-block" id="rarecodons">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.rarecodons">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rarecodons</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;sce&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="n">rare</span> <span class="o">=</span> <span class="n">_rare_codons</span><span class="p">[</span><span class="n">organism</span><span class="p">]</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="n">slices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">):</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">i</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="n">i</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="mi">3</span>
        <span class="n">trip</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">x</span><span class="p">:</span><span class="n">y</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">trip</span> <span class="ow">in</span> <span class="n">rare</span><span class="p">:</span>
            <span class="n">slices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">slices</span></div>



<div class="viewcode-block" id="express">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.express">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">express</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;sce&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.</span>

<span class="sd">    **NOT IMPLEMENTED YET**</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># x = _PrettyTable([&quot;cds&quot;, &quot;len&quot;, &quot;cai&quot;, &quot;gc&quot;, &quot;sta&quot;, &quot;stp&quot;, &quot;n-end&quot;] + _rare_codons[organism] + [&quot;rare&quot;])</span>
    <span class="c1"># val = []</span>

    <span class="c1"># val.append(f&quot;{self._data.upper().decode(&#39;ASCII&#39;)[:3]}...&quot; f&quot;{self._data.upper().decode(&#39;ASCII&#39;)[-3:]}&quot;)</span>
    <span class="c1"># val.append(len(self) / 3)</span>
    <span class="c1"># val.append(cai(organism))</span>
    <span class="c1"># val.append(gc())</span>
    <span class="c1"># val.append(startcodon())</span>
    <span class="c1"># val.append(stopcodon())</span>
    <span class="c1"># val.append(_n_end[organism].get(_seq3(self[3:6].translate())))</span>
    <span class="c1"># s = self._data.upper().decode(&quot;ASCII&quot;)</span>
    <span class="c1"># trps = [s[i * 3 : i * 3 + 3] for i in range(0, len(s) // 3)]</span>
    <span class="c1"># tot = 0</span>
    <span class="c1"># for cdn in _rare_codons[organism]:</span>
    <span class="c1">#     cnt = trps.count(cdn)</span>
    <span class="c1">#     tot += cnt</span>
    <span class="c1">#     val.append(cnt)</span>
    <span class="c1"># val.append(round(tot / len(trps), 3))</span>
    <span class="c1"># x.add_row(val)</span>
    <span class="c1"># return x</span>
    <span class="k">raise</span> <span class="ne">NotImplementedError</span></div>



<div class="viewcode-block" id="open_folder">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.open_folder">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">open_folder</span><span class="p">(</span><span class="n">pth</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">_sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">==</span> <span class="s2">&quot;win32&quot;</span><span class="p">:</span>
        <span class="n">_subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="n">pth</span><span class="p">],</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">_sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">==</span> <span class="s2">&quot;darwin&quot;</span><span class="p">:</span>
        <span class="n">_subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s2">&quot;open&quot;</span><span class="p">,</span> <span class="n">pth</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">_subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s2">&quot;xdg-open&quot;</span><span class="p">,</span> <span class="n">pth</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;no cache to open.&quot;</span></div>



<div class="viewcode-block" id="rc">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.rc">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">rc</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="n">StrOrBytes</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StrOrBytes</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reverse complement.</span>

<span class="sd">    accepts mixed DNA/RNA</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">sequence</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">_complement_table</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>



<div class="viewcode-block" id="complement">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.complement">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Complement.</span>

<span class="sd">    accepts mixed DNA/RNA</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">sequence</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">_complement_table</span><span class="p">)</span></div>



<span class="c1"># def memorize(filename):</span>
<span class="c1">#     &quot;&quot;&quot;Cache functions and classes.</span>

<span class="c1">#     see pydna.download</span>
<span class="c1">#     &quot;&quot;&quot;</span>

<span class="c1">#     def decorator(f):</span>
<span class="c1">#         def wrappee(*args, **kwargs):</span>
<span class="c1">#             _module_logger.info(&quot;#### memorizer ####&quot;)</span>
<span class="c1">#             _module_logger.info(&quot;cache filename                   = %s&quot;, filename)</span>
<span class="c1">#             _module_logger.info(</span>
<span class="c1">#                 &quot;os.environ[&#39;pydna_cached_funcs&#39;] = %s&quot;,</span>
<span class="c1">#                 _os.getenv(&quot;pydna_cached_funcs&quot;, &quot;&quot;),</span>
<span class="c1">#             )</span>
<span class="c1">#             if filename not in _os.getenv(&quot;pydna_cached_funcs&quot;, &quot;&quot;):</span>
<span class="c1">#                 _module_logger.info(&quot;cache filename not among cached functions, made it new!&quot;)</span>
<span class="c1">#                 return f(*args, **kwargs)</span>
<span class="c1">#             key = _base64.urlsafe_b64encode(_hashlib.sha1(_pickle.dumps((args, kwargs))).digest()).decode(&quot;ascii&quot;)</span>
<span class="c1">#             _module_logger.info(&quot;key = %s&quot;, key)</span>
<span class="c1">#             cache = _shelve.open(</span>
<span class="c1">#                 _os.path.join(_os.environ[&quot;pydna_data_dir&quot;], identifier_from_string(filename)),</span>
<span class="c1">#                 writeback=False,</span>
<span class="c1">#             )</span>
<span class="c1">#             try:</span>
<span class="c1">#                 result = cache[key]</span>
<span class="c1">#             except KeyError:</span>
<span class="c1">#                 _module_logger.info(</span>
<span class="c1">#                     &quot;no result for key %s in shelve %s&quot;,</span>
<span class="c1">#                     key,</span>
<span class="c1">#                     identifier_from_string(filename),</span>
<span class="c1">#                 )</span>
<span class="c1">#                 result = f(*args, **kwargs)</span>
<span class="c1">#                 _module_logger.info(&quot;made it new!&quot;)</span>
<span class="c1">#                 cache[key] = result</span>
<span class="c1">#                 _module_logger.info(&quot;saved result under key %s&quot;, key)</span>
<span class="c1">#             else:</span>
<span class="c1">#                 _module_logger.info(&quot;found %s in cache&quot;, key)</span>
<span class="c1">#             cache.close()</span>
<span class="c1">#             return result</span>

<span class="c1">#         return wrappee</span>

<span class="c1">#     return decorator</span>


<div class="viewcode-block" id="identifier_from_string">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.identifier_from_string">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">identifier_from_string</span><span class="p">(</span><span class="n">s</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a valid python identifier.</span>

<span class="sd">    based on the argument s or an empty string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">_re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
    <span class="n">s</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">_re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;[^0-9a-zA-Z_]&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">s</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">isidentifier</span><span class="p">()</span> <span class="ow">or</span> <span class="n">_keyword</span><span class="o">.</span><span class="n">iskeyword</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;_</span><span class="si">{s}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">s</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="ow">or</span> <span class="n">s</span><span class="o">.</span><span class="n">isidentifier</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">s</span></div>



<div class="viewcode-block" id="flatten">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.flatten">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">flatten</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_List</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flattens an iterable of iterables.</span>

<span class="sd">    Down to str, bytes, bytearray or any of the pydna or Biopython seq objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">args</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">args</span><span class="p">:</span>
        <span class="n">top</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="n">top</span><span class="p">,</span> <span class="n">_collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Iterable</span><span class="p">)</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">top</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">,</span> <span class="nb">bytearray</span><span class="p">))</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">top</span><span class="p">,</span> <span class="s2">&quot;reverse_complement&quot;</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="n">args</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">top</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">top</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">output</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>



<div class="viewcode-block" id="seq31">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.seq31">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">seq31</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Turn a three letter code protein sequence into one with one letter code.</span>

<span class="sd">    The single input argument &#39;seq&#39; should be a protein sequence using single</span>
<span class="sd">    letter codes, as a python string.</span>

<span class="sd">    This function returns the amino acid sequence as a string using the one</span>
<span class="sd">    letter amino acid codes. Output follows the IUPAC standard (including</span>
<span class="sd">    ambiguous characters B for &quot;Asx&quot;, J for &quot;Xle&quot; and X for &quot;Xaa&quot;, and also U</span>
<span class="sd">    for &quot;Sel&quot; and O for &quot;Pyl&quot;) plus &quot;Ter&quot; for a terminator given as an</span>
<span class="sd">    asterisk.</span>

<span class="sd">    Any unknown</span>
<span class="sd">    character (including possible gap characters), is changed into &#39;Xaa&#39;.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; from Bio.SeqUtils import seq3</span>
<span class="sd">    &gt;&gt;&gt; seq3(&quot;MAIVMGRWKGAR*&quot;)</span>
<span class="sd">    &#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;</span>
<span class="sd">    &gt;&gt;&gt; from pydna.utils import seq31</span>
<span class="sd">    &gt;&gt;&gt; seq31(&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;)</span>
<span class="sd">    &#39;M  A  I  V  M  G  R  W  K  G  A  R  *&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">threecode</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Ala&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Asx&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Cys&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Asp&quot;</span><span class="p">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Glu&quot;</span><span class="p">:</span> <span class="s2">&quot;E&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Phe&quot;</span><span class="p">:</span> <span class="s2">&quot;F&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Gly&quot;</span><span class="p">:</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span>
        <span class="s2">&quot;His&quot;</span><span class="p">:</span> <span class="s2">&quot;H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Ile&quot;</span><span class="p">:</span> <span class="s2">&quot;I&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Lys&quot;</span><span class="p">:</span> <span class="s2">&quot;K&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Leu&quot;</span><span class="p">:</span> <span class="s2">&quot;L&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Met&quot;</span><span class="p">:</span> <span class="s2">&quot;M&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Asn&quot;</span><span class="p">:</span> <span class="s2">&quot;N&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Pro&quot;</span><span class="p">:</span> <span class="s2">&quot;P&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Gln&quot;</span><span class="p">:</span> <span class="s2">&quot;Q&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Arg&quot;</span><span class="p">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Ser&quot;</span><span class="p">:</span> <span class="s2">&quot;S&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Thr&quot;</span><span class="p">:</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Val&quot;</span><span class="p">:</span> <span class="s2">&quot;V&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Trp&quot;</span><span class="p">:</span> <span class="s2">&quot;W&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Tyr&quot;</span><span class="p">:</span> <span class="s2">&quot;Y&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Glx&quot;</span><span class="p">:</span> <span class="s2">&quot;Z&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Xaa&quot;</span><span class="p">:</span> <span class="s2">&quot;X&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Ter&quot;</span><span class="p">:</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Sel&quot;</span><span class="p">:</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Pyl&quot;</span><span class="p">:</span> <span class="s2">&quot;O&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Xle&quot;</span><span class="p">:</span> <span class="s2">&quot;J&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">nr_of_codons</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="mi">3</span> <span class="p">:</span> <span class="n">i</span> <span class="o">*</span> <span class="mi">3</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">title</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nr_of_codons</span><span class="p">)]</span>
    <span class="n">padding</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">2</span>
    <span class="k">return</span> <span class="n">padding</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">threecode</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="s2">&quot;X&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">])</span></div>



<div class="viewcode-block" id="randomRNA">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.randomRNA">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">randomRNA</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">maxlength</span> <span class="ow">and</span> <span class="n">maxlength</span> <span class="o">&gt;</span> <span class="n">length</span><span class="p">:</span>
        <span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">triangular</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)))</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s2">&quot;GAUC&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">)])</span></div>



<div class="viewcode-block" id="randomDNA">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.randomDNA">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">randomDNA</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">maxlength</span> <span class="ow">and</span> <span class="n">maxlength</span> <span class="o">&gt;</span> <span class="n">length</span><span class="p">:</span>
        <span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">triangular</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)))</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s2">&quot;GATC&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">)])</span></div>



<div class="viewcode-block" id="randomORF">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.randomORF">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">randomORF</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="n">length</span> <span class="o">-=</span> <span class="mi">2</span>
    <span class="k">if</span> <span class="n">maxlength</span> <span class="ow">and</span> <span class="n">maxlength</span> <span class="o">&gt;</span> <span class="n">length</span><span class="p">:</span>
        <span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">triangular</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)))</span>

    <span class="n">cdns</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;TTT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TTC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TTA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TTG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TCT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TCC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TCA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TCG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TAT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TAC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TGT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TGC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TGG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CTT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CTC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CTA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CTG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CCT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CCC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CCA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CCG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CAT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CAC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CAA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CAG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CGT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CGC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CGA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CGG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ATT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ATC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ATA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ATG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ACT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ACC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ACA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ACG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AAT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AAC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AAA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AAG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AGT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AGC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AGA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AGG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GTT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GTC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GTA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GTG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GCT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GCC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GCA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GCG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GAT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GAC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GAA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GAG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GGT&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GGC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GGA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GGG&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">starts</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">,)</span>
    <span class="n">stops</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;TAA&quot;</span><span class="p">,</span> <span class="s2">&quot;TAG&quot;</span><span class="p">,</span> <span class="s2">&quot;TGA&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">starts</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">cdns</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">)])</span>
        <span class="o">+</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">stops</span><span class="p">)</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="randomprot">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.randomprot">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">randomprot</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;docstring.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">maxlength</span> <span class="ow">and</span> <span class="n">maxlength</span> <span class="o">&gt;</span> <span class="n">length</span><span class="p">:</span>
        <span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">triangular</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">maxlength</span><span class="p">)))</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s2">&quot;ACDEFGHIKLMNPQRSTVWY&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">length</span><span class="p">)])</span></div>



<div class="viewcode-block" id="eq">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.eq">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">eq</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare two or more DNA sequences for equality.</span>

<span class="sd">    Compares two or more DNA sequences for equality i.e. if they</span>
<span class="sd">    represent the same double stranded DNA molecule.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    args : iterable</span>
<span class="sd">        iterable containing sequences</span>
<span class="sd">        args can be strings, Biopython Seq or SeqRecord, Dseqrecord</span>
<span class="sd">        or dsDNA objects.</span>
<span class="sd">    circular : bool, optional</span>
<span class="sd">        Consider all molecules circular or linear</span>
<span class="sd">    linear : bool, optional</span>
<span class="sd">        Consider all molecules circular or linear</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    eq : bool</span>
<span class="sd">        Returns True or False</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Compares two or more DNA sequences for equality i.e. if they</span>
<span class="sd">    represent the same DNA molecule.</span>

<span class="sd">    Two linear sequences are considiered equal if either:</span>

<span class="sd">    1. They have the same sequence (case insensitive)</span>
<span class="sd">    2. One sequence is the reverse complement of the other</span>

<span class="sd">    Two circular sequences are considered equal if they are circular</span>
<span class="sd">    permutations meaning that they have the same length and:</span>

<span class="sd">    1. One sequence can be found in the concatenation of the other sequence with itself.</span>
<span class="sd">    2. The reverse complement of one sequence can be found in the concatenation of the other sequence with itself.</span>

<span class="sd">    The topology for the comparison can be set using one of the keywords</span>
<span class="sd">    linear or circular to True or False.</span>

<span class="sd">    If circular or linear is not set, it will be deduced from the topology of</span>
<span class="sd">    each sequence for sequences that have a linear or circular attribute</span>
<span class="sd">    (like Dseq and Dseqrecord).</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord</span>
<span class="sd">    &gt;&gt;&gt; from pydna.utils import eq</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;aaa&quot;,&quot;AAA&quot;)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;aaa&quot;,&quot;AAA&quot;,&quot;TTT&quot;)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;aaa&quot;,&quot;AAA&quot;,&quot;TTT&quot;,&quot;tTt&quot;)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;aaa&quot;,&quot;AAA&quot;,&quot;TTT&quot;,&quot;tTt&quot;, linear=True)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;Taaa&quot;,&quot;aTaa&quot;, linear = True)</span>
<span class="sd">    False</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;Taaa&quot;,&quot;aTaa&quot;, circular = True)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; a=Dseqrecord(&quot;Taaa&quot;)</span>
<span class="sd">    &gt;&gt;&gt; b=Dseqrecord(&quot;aTaa&quot;)</span>
<span class="sd">    &gt;&gt;&gt; eq(a,b)</span>
<span class="sd">    False</span>
<span class="sd">    &gt;&gt;&gt; eq(a,b,circular=True)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; a=a.looped()</span>
<span class="sd">    &gt;&gt;&gt; b=b.looped()</span>
<span class="sd">    &gt;&gt;&gt; eq(a,b)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(a,b,circular=False)</span>
<span class="sd">    False</span>
<span class="sd">    &gt;&gt;&gt; eq(a,b,linear=True)</span>
<span class="sd">    False</span>
<span class="sd">    &gt;&gt;&gt; eq(a,b,linear=False)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;ggatcc&quot;,&quot;GGATCC&quot;)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;ggatcca&quot;,&quot;GGATCCa&quot;)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; eq(&quot;ggatcca&quot;,&quot;tGGATCC&quot;)</span>
<span class="sd">    True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">flatten</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>  <span class="c1"># flatten</span>

    <span class="n">topology</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="s2">&quot;linear&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;linear&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">topology</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;linear&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="n">topology</span> <span class="o">=</span> <span class="s2">&quot;circular&quot;</span>
    <span class="k">elif</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">topology</span> <span class="o">=</span> <span class="s2">&quot;circular&quot;</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="n">topology</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">topology</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
            <span class="p">[</span><span class="n">arg</span><span class="o">.</span><span class="n">circular</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="s2">&quot;circular&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">args</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">topology</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;sequences have different topologies&quot;</span><span class="p">)</span>
        <span class="n">topology</span> <span class="o">=</span> <span class="n">topology</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">topology</span> <span class="ow">in</span> <span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
            <span class="n">topology</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>
        <span class="k">elif</span> <span class="n">topology</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">topology</span> <span class="o">=</span> <span class="s2">&quot;circular&quot;</span>

    <span class="n">args</span> <span class="o">=</span> <span class="p">[</span><span class="n">arg</span><span class="o">.</span><span class="n">seq</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="s2">&quot;seq&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="n">arg</span> <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">args</span><span class="p">]</span>
    <span class="n">args_string_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">arg</span><span class="o">.</span><span class="n">watson</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="s2">&quot;watson&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="nb">str</span><span class="p">(</span><span class="n">arg</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">args</span>
    <span class="p">]</span>

    <span class="n">length</span> <span class="o">=</span> <span class="nb">set</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">args_string_list</span><span class="p">))</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">length</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="n">same</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">topology</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">:</span>
        <span class="c1"># force circular comparison of all given sequences</span>
        <span class="k">for</span> <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span> <span class="ow">in</span> <span class="n">_itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">args_string_list</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">s1</span> <span class="ow">in</span> <span class="n">s2</span> <span class="o">+</span> <span class="n">s2</span> <span class="ow">or</span> <span class="n">rc</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">s2</span> <span class="o">+</span> <span class="n">s2</span><span class="p">):</span>
                <span class="n">same</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">elif</span> <span class="n">topology</span> <span class="o">==</span> <span class="s2">&quot;linear&quot;</span><span class="p">:</span>
        <span class="c1"># force linear comparison of all given sequences</span>
        <span class="k">for</span> <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span> <span class="ow">in</span> <span class="n">_itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">args_string_list</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">s1</span> <span class="o">==</span> <span class="n">s2</span> <span class="ow">or</span> <span class="n">s1</span> <span class="o">==</span> <span class="n">rc</span><span class="p">(</span><span class="n">s2</span><span class="p">)):</span>
                <span class="n">same</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">return</span> <span class="n">same</span></div>



<span class="c1"># def cuts_overlap(left_cut, right_cut, seq_len):</span>
<span class="c1">#     # Special cases:</span>
<span class="c1">#     if left_cut is None or right_cut is None or left_cut == right_cut:</span>
<span class="c1">#         return False</span>

<span class="c1">#     # This block of code would not be necessary if the cuts were</span>
<span class="c1">#     # initially represented like this</span>
<span class="c1">#     (left_watson, left_ovhg), _ = left_cut</span>
<span class="c1">#     (right_watson, right_ovhg), _ = right_cut</span>
<span class="c1">#     # Position of the cut on the crick strands on the left and right</span>
<span class="c1">#     left_crick = left_watson - left_ovhg</span>
<span class="c1">#     right_crick = right_watson - right_ovhg</span>
<span class="c1">#     if left_crick &gt;= seq_len:</span>
<span class="c1">#         left_crick -= seq_len</span>
<span class="c1">#         left_watson -= seq_len</span>
<span class="c1">#     if right_crick &gt;= seq_len:</span>
<span class="c1">#         right_crick -= seq_len</span>
<span class="c1">#         right_watson -= seq_len</span>

<span class="c1">#     # Convert into ranges x and y and see if ranges overlap</span>
<span class="c1">#     x = sorted([left_watson, left_crick])</span>
<span class="c1">#     y = sorted([right_watson, right_crick])</span>
<span class="c1">#     return (x[1] &gt; y[0]) != (y[1] &lt; x[0])</span>


<span class="c1"># def location_boundaries(loc: _Union[_sl, _cl]):</span>
<span class="c1">#     if loc.strand == -1:</span>
<span class="c1">#         return loc.parts[-1].start, loc.parts[0].end</span>
<span class="c1">#     else:</span>
<span class="c1">#         return loc.parts[0].start, loc.parts[-1].end</span>


<div class="viewcode-block" id="cuts_overlap">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.cuts_overlap">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">cuts_overlap</span><span class="p">(</span><span class="n">left_cut</span><span class="p">,</span> <span class="n">right_cut</span><span class="p">,</span> <span class="n">seq_len</span><span class="p">):</span>
    <span class="c1"># Special cases:</span>
    <span class="k">if</span> <span class="n">left_cut</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">right_cut</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">left_cut</span> <span class="o">==</span> <span class="n">right_cut</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># This block of code would not be necessary if the cuts were</span>
    <span class="c1"># initially represented like this</span>
    <span class="p">(</span><span class="n">left_watson</span><span class="p">,</span> <span class="n">left_ovhg</span><span class="p">),</span> <span class="n">_</span> <span class="o">=</span> <span class="n">left_cut</span>
    <span class="p">(</span><span class="n">right_watson</span><span class="p">,</span> <span class="n">right_ovhg</span><span class="p">),</span> <span class="n">_</span> <span class="o">=</span> <span class="n">right_cut</span>
    <span class="c1"># Position of the cut on the crick strands on the left and right</span>
    <span class="n">left_crick</span> <span class="o">=</span> <span class="n">left_watson</span> <span class="o">-</span> <span class="n">left_ovhg</span>
    <span class="n">right_crick</span> <span class="o">=</span> <span class="n">right_watson</span> <span class="o">-</span> <span class="n">right_ovhg</span>
    <span class="k">if</span> <span class="n">left_crick</span> <span class="o">&gt;=</span> <span class="n">seq_len</span><span class="p">:</span>
        <span class="n">left_crick</span> <span class="o">-=</span> <span class="n">seq_len</span>
        <span class="n">left_watson</span> <span class="o">-=</span> <span class="n">seq_len</span>
    <span class="k">if</span> <span class="n">right_crick</span> <span class="o">&gt;=</span> <span class="n">seq_len</span><span class="p">:</span>
        <span class="n">right_crick</span> <span class="o">-=</span> <span class="n">seq_len</span>
        <span class="n">right_watson</span> <span class="o">-=</span> <span class="n">seq_len</span>

    <span class="c1"># Convert into ranges x and y and see if ranges overlap</span>
    <span class="n">x</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">left_watson</span><span class="p">,</span> <span class="n">left_crick</span><span class="p">])</span>
    <span class="n">y</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">right_watson</span><span class="p">,</span> <span class="n">right_crick</span><span class="p">])</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">y</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">!=</span> <span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span></div>



<div class="viewcode-block" id="location_boundaries">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.location_boundaries">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">location_boundaries</span><span class="p">(</span><span class="n">loc</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_sl</span><span class="p">,</span> <span class="n">_cl</span><span class="p">]):</span>
    <span class="k">if</span> <span class="n">loc</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">loc</span><span class="o">.</span><span class="n">parts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">end</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">loc</span><span class="o">.</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">parts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">end</span></div>



<div class="viewcode-block" id="locations_overlap">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.locations_overlap">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">locations_overlap</span><span class="p">(</span><span class="n">loc1</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_sl</span><span class="p">,</span> <span class="n">_cl</span><span class="p">],</span> <span class="n">loc2</span><span class="p">:</span> <span class="n">_Union</span><span class="p">[</span><span class="n">_sl</span><span class="p">,</span> <span class="n">_cl</span><span class="p">],</span> <span class="n">seq_len</span><span class="p">):</span>
    <span class="n">start1</span><span class="p">,</span> <span class="n">end1</span> <span class="o">=</span> <span class="n">location_boundaries</span><span class="p">(</span><span class="n">loc1</span><span class="p">)</span>
    <span class="n">start2</span><span class="p">,</span> <span class="n">end2</span> <span class="o">=</span> <span class="n">location_boundaries</span><span class="p">(</span><span class="n">loc2</span><span class="p">)</span>

    <span class="n">boundaries1</span> <span class="o">=</span> <span class="p">[(</span><span class="n">start1</span><span class="p">,</span> <span class="n">end1</span><span class="p">)]</span>
    <span class="n">boundaries2</span> <span class="o">=</span> <span class="p">[(</span><span class="n">start2</span><span class="p">,</span> <span class="n">end2</span><span class="p">)]</span>

    <span class="k">if</span> <span class="n">start1</span> <span class="o">&gt;</span> <span class="n">end1</span><span class="p">:</span>
        <span class="n">boundaries1</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="n">start1</span><span class="p">,</span> <span class="n">end1</span> <span class="o">+</span> <span class="n">seq_len</span><span class="p">],</span>
            <span class="p">[</span><span class="n">start1</span> <span class="o">-</span> <span class="n">seq_len</span><span class="p">,</span> <span class="n">end1</span><span class="p">],</span>
        <span class="p">]</span>
    <span class="k">if</span> <span class="n">start2</span> <span class="o">&gt;</span> <span class="n">end2</span><span class="p">:</span>
        <span class="n">boundaries2</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="n">start2</span><span class="p">,</span> <span class="n">end2</span> <span class="o">+</span> <span class="n">seq_len</span><span class="p">],</span>
            <span class="p">[</span><span class="n">start2</span> <span class="o">-</span> <span class="n">seq_len</span><span class="p">,</span> <span class="n">end2</span><span class="p">],</span>
        <span class="p">]</span>

    <span class="k">for</span> <span class="n">b1</span><span class="p">,</span> <span class="n">b2</span> <span class="ow">in</span> <span class="n">_itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">boundaries1</span><span class="p">,</span> <span class="n">boundaries2</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">b1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">b2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span> <span class="n">b1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">b2</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">return</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="sum_is_sticky">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.sum_is_sticky">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">sum_is_sticky</span><span class="p">(</span>
    <span class="n">three_prime_end</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">five_prime_end</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">partial</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the overlap length if the 3&#39; end of seq1 and 5&#39; end of seq2 ends are sticky and compatible for ligation.</span>
<span class="sd">    Return 0 if they are not compatible.&quot;&quot;&quot;</span>
    <span class="n">type_seq1</span><span class="p">,</span> <span class="n">sticky_seq1</span> <span class="o">=</span> <span class="n">three_prime_end</span>
    <span class="n">type_seq2</span><span class="p">,</span> <span class="n">sticky_seq2</span> <span class="o">=</span> <span class="n">five_prime_end</span>

    <span class="k">if</span> <span class="p">(</span>
        <span class="s2">&quot;blunt&quot;</span> <span class="o">!=</span> <span class="n">type_seq2</span>
        <span class="ow">and</span> <span class="n">type_seq2</span> <span class="o">==</span> <span class="n">type_seq1</span>
        <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky_seq2</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky_seq1</span><span class="p">))</span>
    <span class="p">):</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">sticky_seq1</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">partial</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="n">type_seq1</span> <span class="o">!=</span> <span class="n">type_seq2</span> <span class="ow">or</span> <span class="n">type_seq2</span> <span class="o">==</span> <span class="s2">&quot;blunt&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">0</span>
    <span class="k">elif</span> <span class="n">type_seq2</span> <span class="o">==</span> <span class="s2">&quot;5&#39;&quot;</span><span class="p">:</span>
        <span class="n">sticky_seq1</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky_seq1</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">type_seq2</span> <span class="o">==</span> <span class="s2">&quot;3&#39;&quot;</span><span class="p">:</span>
        <span class="n">sticky_seq2</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky_seq2</span><span class="p">))</span>

    <span class="n">ovhg_len</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sticky_seq1</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">sticky_seq2</span><span class="p">))</span>
    <span class="c1"># [::-1] to try the longest overhangs first</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ovhg_len</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">sticky_seq1</span><span class="p">[</span><span class="o">-</span><span class="n">i</span><span class="p">:]</span> <span class="o">==</span> <span class="n">sticky_seq2</span><span class="p">[:</span><span class="n">i</span><span class="p">]:</span>
            <span class="k">return</span> <span class="n">i</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">0</span></div>



<div class="viewcode-block" id="limit_iterator">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.limit_iterator">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">limit_iterator</span><span class="p">(</span><span class="n">iterator</span><span class="p">,</span> <span class="n">limit</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Call the function with an iterator to raise an error if the number of items is greater than the limit.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">iterator</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">limit</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Too many possible paths (more than </span><span class="si">{</span><span class="n">limit</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">yield</span> <span class="n">x</span></div>



<div class="viewcode-block" id="create_location">
<a class="viewcode-back" href="../../reference/pydna.html#pydna.all.create_location">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_location</span><span class="p">(</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">lim</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">strand</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">_Location</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a location object from a start and end position.</span>
<span class="sd">    If the end position is less than the start position, the location is circular. It handles negative positions.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    start : int</span>
<span class="sd">        The start position of the location.</span>
<span class="sd">    end : int</span>
<span class="sd">        The end position of the location.</span>
<span class="sd">    lim : int</span>
<span class="sd">        The length of the sequence.</span>
<span class="sd">    strand : int, optional</span>
<span class="sd">        The strand of the location. None, 1 or -1.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    location : Location</span>
<span class="sd">        The location object. Can be a SimpleLocation or a CompoundLocation if the feature spans the origin of</span>
<span class="sd">        a circular sequence.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; from pydna.utils import create_location</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(0, 5, 10,-1))</span>
<span class="sd">    &#39;[0:5](-)&#39;</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(0, 5, 10,+1))</span>
<span class="sd">    &#39;[0:5](+)&#39;</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(0, 5, 10))</span>
<span class="sd">    &#39;[0:5]&#39;</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(8, 2, 10))</span>
<span class="sd">    &#39;join{[8:10], [0:2]}&#39;</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(8, 2, 10,-1))</span>
<span class="sd">    &#39;join{[0:2](-), [8:10](-)}&#39;</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(-2, 2, 10))</span>
<span class="sd">    &#39;join{[8:10], [0:2]}&#39;</span>

<span class="sd">    Note this special case, 0 is the same as len(seq)</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(5, 0, 10))</span>
<span class="sd">    &#39;[5:10]&#39;</span>

<span class="sd">    Note the special case where if start and end are the same,</span>
<span class="sd">    the location spans the entire sequence (it&#39;s not empty).</span>
<span class="sd">    &gt;&gt;&gt; str(create_location(5, 5, 10))</span>
<span class="sd">    &#39;join{[5:10], [0:5]}&#39;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">while</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">start</span> <span class="o">+=</span> <span class="n">lim</span>
    <span class="k">while</span> <span class="n">end</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">end</span> <span class="o">+=</span> <span class="n">lim</span>
    <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">start</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">_sl</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">shift_location</span><span class="p">(</span><span class="n">_sl</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">+</span> <span class="n">lim</span><span class="p">,</span> <span class="n">strand</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="n">lim</span><span class="p">)</span></div>

</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Björn F. Johansson
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Björn F. Johansson.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>